<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns:mv="http://macVmlSchemaUri" xmlns="http://www.w3.org/TR/REC-html40">

<head>
<meta name=Title content=Manual>
<meta name=Keywords content="">
<meta http-equiv=Content-Type content="text/html; charset=windows-1252">
<meta name=ProgId content=Word.Document>
<meta name=Generator content="Microsoft Word 14">
<meta name=Originator content="Microsoft Word 14">
<link rel=File-List href="t_coffee_tutorial_files/filelist.xml">
<link rel=Edit-Time-Data href="t_coffee_tutorial_files/editdata.mso">
<!--[if !mso]>
<style>
v\:* {behavior:url(#default#VML);}
o\:* {behavior:url(#default#VML);}
w\:* {behavior:url(#default#VML);}
.shape {behavior:url(#default#VML);}
</style>
<![endif]-->
<title>Manual</title>
<!--[if gte mso 9]><xml>
 <o:DocumentProperties>
  <o:Author>Notredame</o:Author>
  <o:Template>\Documents and Settings\notredame\Application Data\Microsoft\Templates\t_coffee_doc.dot</o:Template>
  <o:LastAuthor>Notredame Cédric</o:LastAuthor>
  <o:Revision>2</o:Revision>
  <o:TotalTime>2</o:TotalTime>
  <o:LastPrinted>2006-06-18T16:16:00Z</o:LastPrinted>
  <o:Created>2015-08-05T16:39:00Z</o:Created>
  <o:LastSaved>2015-08-05T16:39:00Z</o:LastSaved>
  <o:Pages>3</o:Pages>
  <o:Words>23542</o:Words>
  <o:Characters>134195</o:Characters>
  <o:Company>SIB</o:Company>
  <o:Lines>1118</o:Lines>
  <o:Paragraphs>314</o:Paragraphs>
  <o:CharactersWithSpaces>157423</o:CharactersWithSpaces>
  <o:Version>14.0</o:Version>
 </o:DocumentProperties>
 <o:OfficeDocumentSettings>
  <o:PixelsPerInch>96</o:PixelsPerInch>
  <o:TargetScreenSize>800x600</o:TargetScreenSize>
 </o:OfficeDocumentSettings>
</xml><![endif]-->
<link rel=themeData href="t_coffee_tutorial_files/themedata.xml">
<!--[if gte mso 9]><xml>
 <w:WordDocument>
  <w:ActiveWritingStyle Lang="EN-US" VendorID="64" DLLVersion="131078"
   NLCheck="1">1</w:ActiveWritingStyle>
  <w:ActiveWritingStyle Lang="EN-GB" VendorID="64" DLLVersion="131078"
   NLCheck="1">1</w:ActiveWritingStyle>
  <w:ActiveWritingStyle Lang="FR" VendorID="64" DLLVersion="131078" NLCheck="1">1</w:ActiveWritingStyle>
  <w:ActiveWritingStyle Lang="FR-CH" VendorID="64" DLLVersion="131078"
   NLCheck="1">1</w:ActiveWritingStyle>
  <w:SpellingState>Clean</w:SpellingState>
  <w:GrammarState>Clean</w:GrammarState>
  <w:TrackMoves>false</w:TrackMoves>
  <w:TrackFormatting/>
  <w:HyphenationZone>21</w:HyphenationZone>
  <w:ValidateAgainstSchemas/>
  <w:SaveIfXMLInvalid>false</w:SaveIfXMLInvalid>
  <w:IgnoreMixedContent>false</w:IgnoreMixedContent>
  <w:AlwaysShowPlaceholderText>false</w:AlwaysShowPlaceholderText>
  <w:DoNotPromoteQF/>
  <w:LidThemeOther>EN-US</w:LidThemeOther>
  <w:LidThemeAsian>JA</w:LidThemeAsian>
  <w:LidThemeComplexScript>X-NONE</w:LidThemeComplexScript>
  <w:Compatibility>
   <w:BreakWrappedTables/>
   <w:SnapToGridInCell/>
   <w:WrapTextWithPunct/>
   <w:UseAsianBreakRules/>
   <w:UseWord2002TableStyleRules/>
   <w:DontUseIndentAsNumberingTabStop/>
   <w:FELineBreak11/>
   <w:WW11IndentRules/>
   <w:DontAutofitConstrainedTables/>
   <w:AutofitLikeWW11/>
   <w:HangulWidthLikeWW11/>
   <w:UseNormalStyleForList/>
   <w:DontVertAlignCellWithSp/>
   <w:DontBreakConstrainedForcedTables/>
   <w:DontVertAlignInTxbx/>
   <w:Word11KerningPairs/>
   <w:CachedColBalance/>
  </w:Compatibility>
  <m:mathPr>
   <m:mathFont m:val="Cambria Math"/>
   <m:brkBin m:val="before"/>
   <m:brkBinSub m:val="&#45;-"/>
   <m:smallFrac m:val="off"/>
   <m:dispDef/>
   <m:lMargin m:val="0"/>
   <m:rMargin m:val="0"/>
   <m:defJc m:val="centerGroup"/>
   <m:wrapIndent m:val="1440"/>
   <m:intLim m:val="subSup"/>
   <m:naryLim m:val="undOvr"/>
  </m:mathPr></w:WordDocument>
</xml><![endif]--><!--[if gte mso 9]><xml>
 <w:LatentStyles DefLockedState="false" DefUnhideWhenUsed="false"
  DefSemiHidden="false" DefQFormat="false" LatentStyleCount="276">
  <w:LsdException Locked="false" QFormat="true" Name="Normal"/>
  <w:LsdException Locked="false" QFormat="true" Name="heading 1"/>
  <w:LsdException Locked="false" QFormat="true" Name="heading 2"/>
  <w:LsdException Locked="false" QFormat="true" Name="heading 3"/>
  <w:LsdException Locked="false" QFormat="true" Name="heading 4"/>
  <w:LsdException Locked="false" QFormat="true" Name="heading 5"/>
  <w:LsdException Locked="false" QFormat="true" Name="heading 6"/>
  <w:LsdException Locked="false" QFormat="true" Name="heading 7"/>
  <w:LsdException Locked="false" QFormat="true" Name="heading 8"/>
  <w:LsdException Locked="false" QFormat="true" Name="heading 9"/>
  <w:LsdException Locked="false" QFormat="true" Name="caption"/>
  <w:LsdException Locked="false" QFormat="true" Name="Title"/>
  <w:LsdException Locked="false" QFormat="true" Name="Subtitle"/>
  <w:LsdException Locked="false" QFormat="true" Name="Strong"/>
  <w:LsdException Locked="false" QFormat="true" Name="Emphasis"/>
  <w:LsdException Locked="false" Priority="99" SemiHidden="true"
   UnhideWhenUsed="true" Name="Note Level 1"/>
  <w:LsdException Locked="false" Priority="99" SemiHidden="true"
   UnhideWhenUsed="true" Name="Note Level 2"/>
  <w:LsdException Locked="false" Priority="99" SemiHidden="true"
   UnhideWhenUsed="true" Name="Note Level 3"/>
  <w:LsdException Locked="false" Priority="99" SemiHidden="true"
   UnhideWhenUsed="true" Name="Note Level 4"/>
  <w:LsdException Locked="false" Priority="99" SemiHidden="true"
   UnhideWhenUsed="true" Name="Note Level 5"/>
  <w:LsdException Locked="false" Priority="99" SemiHidden="true"
   UnhideWhenUsed="true" Name="Note Level 6"/>
  <w:LsdException Locked="false" Priority="99" SemiHidden="true"
   UnhideWhenUsed="true" Name="Note Level 7"/>
  <w:LsdException Locked="false" Priority="99" SemiHidden="true"
   UnhideWhenUsed="true" Name="Note Level 8"/>
  <w:LsdException Locked="false" Priority="99" SemiHidden="true"
   UnhideWhenUsed="true" Name="Note Level 9"/>
  <w:LsdException Locked="false" Priority="99" SemiHidden="true"
   Name="Placeholder Text"/>
  <w:LsdException Locked="false" Priority="1" QFormat="true" Name="No Spacing"/>
  <w:LsdException Locked="false" Priority="60" Name="Light Shading"/>
  <w:LsdException Locked="false" Priority="61" Name="Light List"/>
  <w:LsdException Locked="false" Priority="62" Name="Light Grid"/>
  <w:LsdException Locked="false" Priority="63" Name="Medium Shading 1"/>
  <w:LsdException Locked="false" Priority="64" Name="Medium Shading 2"/>
  <w:LsdException Locked="false" Priority="65" Name="Medium List 1"/>
  <w:LsdException Locked="false" Priority="66" Name="Medium List 2"/>
  <w:LsdException Locked="false" Priority="67" Name="Medium Grid 1"/>
  <w:LsdException Locked="false" Priority="68" Name="Medium Grid 2"/>
  <w:LsdException Locked="false" Priority="69" Name="Medium Grid 3"/>
  <w:LsdException Locked="false" Priority="70" Name="Dark List"/>
  <w:LsdException Locked="false" Priority="71" Name="Colorful Shading"/>
  <w:LsdException Locked="false" Priority="72" Name="Colorful List"/>
  <w:LsdException Locked="false" Priority="73" Name="Colorful Grid"/>
  <w:LsdException Locked="false" Priority="60" Name="Light Shading Accent 1"/>
  <w:LsdException Locked="false" Priority="61" Name="Light List Accent 1"/>
  <w:LsdException Locked="false" Priority="62" Name="Light Grid Accent 1"/>
  <w:LsdException Locked="false" Priority="63" Name="Medium Shading 1 Accent 1"/>
  <w:LsdException Locked="false" Priority="64" Name="Medium Shading 2 Accent 1"/>
  <w:LsdException Locked="false" Priority="65" Name="Medium List 1 Accent 1"/>
  <w:LsdException Locked="false" Priority="99" SemiHidden="true" Name="Revision"/>
  <w:LsdException Locked="false" Priority="34" QFormat="true"
   Name="List Paragraph"/>
  <w:LsdException Locked="false" Priority="29" QFormat="true" Name="Quote"/>
  <w:LsdException Locked="false" Priority="30" QFormat="true"
   Name="Intense Quote"/>
  <w:LsdException Locked="false" Priority="66" Name="Medium List 2 Accent 1"/>
  <w:LsdException Locked="false" Priority="67" Name="Medium Grid 1 Accent 1"/>
  <w:LsdException Locked="false" Priority="68" Name="Medium Grid 2 Accent 1"/>
  <w:LsdException Locked="false" Priority="69" Name="Medium Grid 3 Accent 1"/>
  <w:LsdException Locked="false" Priority="70" Name="Dark List Accent 1"/>
  <w:LsdException Locked="false" Priority="71" Name="Colorful Shading Accent 1"/>
  <w:LsdException Locked="false" Priority="72" Name="Colorful List Accent 1"/>
  <w:LsdException Locked="false" Priority="73" Name="Colorful Grid Accent 1"/>
  <w:LsdException Locked="false" Priority="60" Name="Light Shading Accent 2"/>
  <w:LsdException Locked="false" Priority="61" Name="Light List Accent 2"/>
  <w:LsdException Locked="false" Priority="62" Name="Light Grid Accent 2"/>
  <w:LsdException Locked="false" Priority="63" Name="Medium Shading 1 Accent 2"/>
  <w:LsdException Locked="false" Priority="64" Name="Medium Shading 2 Accent 2"/>
  <w:LsdException Locked="false" Priority="65" Name="Medium List 1 Accent 2"/>
  <w:LsdException Locked="false" Priority="66" Name="Medium List 2 Accent 2"/>
  <w:LsdException Locked="false" Priority="67" Name="Medium Grid 1 Accent 2"/>
  <w:LsdException Locked="false" Priority="68" Name="Medium Grid 2 Accent 2"/>
  <w:LsdException Locked="false" Priority="69" Name="Medium Grid 3 Accent 2"/>
  <w:LsdException Locked="false" Priority="70" Name="Dark List Accent 2"/>
  <w:LsdException Locked="false" Priority="71" Name="Colorful Shading Accent 2"/>
  <w:LsdException Locked="false" Priority="72" Name="Colorful List Accent 2"/>
  <w:LsdException Locked="false" Priority="73" Name="Colorful Grid Accent 2"/>
  <w:LsdException Locked="false" Priority="60" Name="Light Shading Accent 3"/>
  <w:LsdException Locked="false" Priority="61" Name="Light List Accent 3"/>
  <w:LsdException Locked="false" Priority="62" Name="Light Grid Accent 3"/>
  <w:LsdException Locked="false" Priority="63" Name="Medium Shading 1 Accent 3"/>
  <w:LsdException Locked="false" Priority="64" Name="Medium Shading 2 Accent 3"/>
  <w:LsdException Locked="false" Priority="65" Name="Medium List 1 Accent 3"/>
  <w:LsdException Locked="false" Priority="66" Name="Medium List 2 Accent 3"/>
  <w:LsdException Locked="false" Priority="67" Name="Medium Grid 1 Accent 3"/>
  <w:LsdException Locked="false" Priority="68" Name="Medium Grid 2 Accent 3"/>
  <w:LsdException Locked="false" Priority="69" Name="Medium Grid 3 Accent 3"/>
  <w:LsdException Locked="false" Priority="70" Name="Dark List Accent 3"/>
  <w:LsdException Locked="false" Priority="71" Name="Colorful Shading Accent 3"/>
  <w:LsdException Locked="false" Priority="72" Name="Colorful List Accent 3"/>
  <w:LsdException Locked="false" Priority="73" Name="Colorful Grid Accent 3"/>
  <w:LsdException Locked="false" Priority="60" Name="Light Shading Accent 4"/>
  <w:LsdException Locked="false" Priority="61" Name="Light List Accent 4"/>
  <w:LsdException Locked="false" Priority="62" Name="Light Grid Accent 4"/>
  <w:LsdException Locked="false" Priority="63" Name="Medium Shading 1 Accent 4"/>
  <w:LsdException Locked="false" Priority="64" Name="Medium Shading 2 Accent 4"/>
  <w:LsdException Locked="false" Priority="65" Name="Medium List 1 Accent 4"/>
  <w:LsdException Locked="false" Priority="66" Name="Medium List 2 Accent 4"/>
  <w:LsdException Locked="false" Priority="67" Name="Medium Grid 1 Accent 4"/>
  <w:LsdException Locked="false" Priority="68" Name="Medium Grid 2 Accent 4"/>
  <w:LsdException Locked="false" Priority="69" Name="Medium Grid 3 Accent 4"/>
  <w:LsdException Locked="false" Priority="70" Name="Dark List Accent 4"/>
  <w:LsdException Locked="false" Priority="71" Name="Colorful Shading Accent 4"/>
  <w:LsdException Locked="false" Priority="72" Name="Colorful List Accent 4"/>
  <w:LsdException Locked="false" Priority="73" Name="Colorful Grid Accent 4"/>
  <w:LsdException Locked="false" Priority="60" Name="Light Shading Accent 5"/>
  <w:LsdException Locked="false" Priority="61" Name="Light List Accent 5"/>
  <w:LsdException Locked="false" Priority="62" Name="Light Grid Accent 5"/>
  <w:LsdException Locked="false" Priority="63" Name="Medium Shading 1 Accent 5"/>
  <w:LsdException Locked="false" Priority="64" Name="Medium Shading 2 Accent 5"/>
  <w:LsdException Locked="false" Priority="65" Name="Medium List 1 Accent 5"/>
  <w:LsdException Locked="false" Priority="66" Name="Medium List 2 Accent 5"/>
  <w:LsdException Locked="false" Priority="67" Name="Medium Grid 1 Accent 5"/>
  <w:LsdException Locked="false" Priority="68" Name="Medium Grid 2 Accent 5"/>
  <w:LsdException Locked="false" Priority="69" Name="Medium Grid 3 Accent 5"/>
  <w:LsdException Locked="false" Priority="70" Name="Dark List Accent 5"/>
  <w:LsdException Locked="false" Priority="71" Name="Colorful Shading Accent 5"/>
  <w:LsdException Locked="false" Priority="72" Name="Colorful List Accent 5"/>
  <w:LsdException Locked="false" Priority="73" Name="Colorful Grid Accent 5"/>
  <w:LsdException Locked="false" Priority="60" Name="Light Shading Accent 6"/>
  <w:LsdException Locked="false" Priority="61" Name="Light List Accent 6"/>
  <w:LsdException Locked="false" Priority="62" Name="Light Grid Accent 6"/>
  <w:LsdException Locked="false" Priority="63" Name="Medium Shading 1 Accent 6"/>
  <w:LsdException Locked="false" Priority="64" Name="Medium Shading 2 Accent 6"/>
  <w:LsdException Locked="false" Priority="65" Name="Medium List 1 Accent 6"/>
  <w:LsdException Locked="false" Priority="66" Name="Medium List 2 Accent 6"/>
  <w:LsdException Locked="false" Priority="67" Name="Medium Grid 1 Accent 6"/>
  <w:LsdException Locked="false" Priority="68" Name="Medium Grid 2 Accent 6"/>
  <w:LsdException Locked="false" Priority="69" Name="Medium Grid 3 Accent 6"/>
  <w:LsdException Locked="false" Priority="70" Name="Dark List Accent 6"/>
  <w:LsdException Locked="false" Priority="71" Name="Colorful Shading Accent 6"/>
  <w:LsdException Locked="false" Priority="72" Name="Colorful List Accent 6"/>
  <w:LsdException Locked="false" Priority="73" Name="Colorful Grid Accent 6"/>
  <w:LsdException Locked="false" Priority="19" QFormat="true"
   Name="Subtle Emphasis"/>
  <w:LsdException Locked="false" Priority="21" QFormat="true"
   Name="Intense Emphasis"/>
  <w:LsdException Locked="false" Priority="31" QFormat="true"
   Name="Subtle Reference"/>
  <w:LsdException Locked="false" Priority="32" QFormat="true"
   Name="Intense Reference"/>
  <w:LsdException Locked="false" Priority="33" QFormat="true" Name="Book Title"/>
  <w:LsdException Locked="false" Priority="37" SemiHidden="true"
   UnhideWhenUsed="true" Name="Bibliography"/>
  <w:LsdException Locked="false" Priority="39" SemiHidden="true"
   UnhideWhenUsed="true" QFormat="true" Name="TOC Heading"/>
 </w:LatentStyles>
</xml><![endif]-->
<style>
<!--
 /* Font Definitions */
@font-face
	{font-family:Arial;
	panose-1:2 11 6 4 2 2 2 2 2 4;
	mso-font-charset:0;
	mso-generic-font-family:auto;
	mso-font-pitch:variable;
	mso-font-signature:-536859905 -1073711037 9 0 511 0;}
@font-face
	{font-family:"Courier New";
	panose-1:2 7 3 9 2 2 5 2 4 4;
	mso-font-charset:0;
	mso-generic-font-family:auto;
	mso-font-pitch:variable;
	mso-font-signature:-536859905 -1073711037 9 0 511 0;}
@font-face
	{font-family:Wingdings;
	panose-1:5 0 0 0 0 0 0 0 0 0;
	mso-font-charset:2;
	mso-generic-font-family:auto;
	mso-font-pitch:variable;
	mso-font-signature:0 268435456 0 0 -2147483648 0;}
@font-face
	{font-family:Wingdings;
	panose-1:5 0 0 0 0 0 0 0 0 0;
	mso-font-charset:2;
	mso-generic-font-family:auto;
	mso-font-pitch:variable;
	mso-font-signature:0 268435456 0 0 -2147483648 0;}
@font-face
	{font-family:"Arial Black";
	panose-1:2 11 10 4 2 1 2 2 2 4;
	mso-font-charset:0;
	mso-generic-font-family:auto;
	mso-font-pitch:variable;
	mso-font-signature:3 0 0 0 1 0;}
@font-face
	{font-family:"Comic Sans MS";
	panose-1:3 15 7 2 3 3 2 2 2 4;
	mso-font-charset:0;
	mso-generic-font-family:auto;
	mso-font-pitch:variable;
	mso-font-signature:3 0 0 0 1 0;}
@font-face
	{font-family:Garamond;
	panose-1:2 2 4 4 3 3 1 1 8 3;
	mso-font-charset:0;
	mso-generic-font-family:auto;
	mso-font-pitch:variable;
	mso-font-signature:3 0 0 0 1 0;}
@font-face
	{font-family:"MS Mincho";
	mso-font-alt:"MS Mincho";
	mso-font-charset:128;
	mso-generic-font-family:modern;
	mso-font-pitch:fixed;
	mso-font-signature:-536870145 1791491579 18 0 131231 0;}
@font-face
	{font-family:starbats;
	mso-font-charset:2;
	mso-generic-font-family:auto;
	mso-font-pitch:variable;
	mso-font-signature:0 0 0 0 0 0;}
 /* Style Definitions */
p.MsoNormal, li.MsoNormal, div.MsoNormal
	{mso-style-unhide:no;
	mso-style-qformat:yes;
	mso-style-parent:"";
	margin:0cm;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:8.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
h1
	{mso-style-unhide:no;
	mso-style-qformat:yes;
	mso-style-link:"Heading 1 Char";
	mso-style-next:"Body Text";
	margin-top:12.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	mso-pagination:widow-orphan;
	page-break-after:avoid;
	mso-outline-level:1;
	font-size:16.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Arial Black";
	color:gray;
	letter-spacing:-1.25pt;
	mso-font-kerning:14.0pt;
	mso-ansi-language:EN-GB;
	font-weight:normal;}
h2
	{mso-style-update:auto;
	mso-style-unhide:no;
	mso-style-qformat:yes;
	mso-style-parent:"Heading 1";
	mso-style-next:"Body Text";
	margin-top:12.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	mso-line-height-alt:12.0pt;
	mso-pagination:widow-orphan;
	page-break-after:avoid;
	mso-outline-level:2;
	font-size:16.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Arial Black";
	color:gray;
	letter-spacing:-.5pt;
	mso-font-kerning:14.0pt;
	mso-ansi-language:EN-GB;
	font-weight:normal;}
h3
	{mso-style-update:auto;
	mso-style-unhide:no;
	mso-style-qformat:yes;
	mso-style-next:MyNormal;
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:17.0pt;
	text-align:justify;
	text-indent:0cm;
	mso-pagination:none;
	page-break-after:avoid;
	mso-outline-level:3;
	mso-list:l10 level3 lfo3;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:14.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Arial Black";
	font-weight:normal;}
h4
	{mso-style-update:auto;
	mso-style-unhide:no;
	mso-style-qformat:yes;
	mso-style-parent:MyNormal;
	mso-style-next:Normal;
	margin-top:2.0pt;
	margin-right:0cm;
	margin-bottom:2.0pt;
	margin-left:1.0cm;
	mso-pagination:none;
	page-break-after:avoid;
	mso-outline-level:4;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:11.0pt;
	font-family:"Times New Roman";
	mso-bidi-font-family:"Courier New";
	letter-spacing:1.5pt;
	mso-bidi-font-weight:normal;
	font-style:italic;
	mso-bidi-font-style:normal;}
h5
	{mso-style-unhide:no;
	mso-style-qformat:yes;
	mso-style-next:"Body Text";
	margin-top:2.0pt;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:0cm;
	mso-pagination:widow-orphan;
	page-break-after:avoid;
	mso-outline-level:5;
	mso-element:frame;
	mso-element-frame-width:90.0pt;
	mso-element-wrap:around;
	mso-element-anchor-vertical:paragraph;
	mso-element-anchor-horizontal:page;
	mso-element-left:60.05pt;
	mso-element-top:.05pt;
	mso-height-rule:exactly;
	font-size:9.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Arial Black";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;
	font-weight:normal;}
h6
	{mso-style-unhide:no;
	mso-style-qformat:yes;
	mso-style-next:"Body Text";
	margin:0cm;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	page-break-after:avoid;
	mso-outline-level:6;
	mso-element:frame;
	mso-element-frame-width:90.0pt;
	mso-element-wrap:around;
	mso-element-anchor-vertical:paragraph;
	mso-element-anchor-horizontal:page;
	mso-element-left:60.05pt;
	mso-element-top:.05pt;
	mso-height-rule:exactly;
	font-size:8.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Garamond;
	mso-ansi-language:EN-GB;
	font-weight:normal;}
p.MsoHeading7, li.MsoHeading7, div.MsoHeading7
	{mso-style-unhide:no;
	mso-style-qformat:yes;
	mso-style-next:"Body Text";
	margin-top:3.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:0cm;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	mso-outline-level:7;
	background:#F2F2F2;
	mso-shading:windowtext;
	mso-pattern:gray-5 auto;
	border:none;
	mso-border-alt:solid white .75pt;
	padding:0cm;
	mso-padding-alt:12.0pt 12.0pt 12.0pt 12.0pt;
	mso-element:frame;
	mso-element-frame-width:189.0pt;
	mso-element-frame-hspace:12.0pt;
	mso-element-wrap:around;
	mso-element-anchor-vertical:paragraph;
	mso-element-anchor-horizontal:page;
	mso-element-left:74.45pt;
	mso-element-top:.05pt;
	mso-height-rule:exactly;
	font-size:14.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;
	font-style:italic;
	mso-bidi-font-style:normal;}
p.MsoHeading8, li.MsoHeading8, div.MsoHeading8
	{mso-style-unhide:no;
	mso-style-qformat:yes;
	mso-style-next:"Body Text";
	margin-top:3.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:0cm;
	margin-bottom:.0001pt;
	text-align:center;
	line-height:16.0pt;
	mso-line-height-rule:exactly;
	mso-pagination:widow-orphan;
	page-break-after:avoid;
	mso-outline-level:8;
	border:none;
	mso-border-top-alt:solid windowtext 3.0pt;
	mso-border-bottom-alt:solid windowtext .75pt;
	padding:0cm;
	mso-padding-alt:0cm 0cm 0cm 0cm;
	mso-element:frame;
	mso-element-frame-width:93.0pt;
	mso-element-wrap:around;
	mso-element-anchor-vertical:paragraph;
	mso-element-anchor-horizontal:page;
	mso-element-left:60.05pt;
	mso-element-top:.05pt;
	mso-height-rule:exactly;
	font-size:7.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Arial Black";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	text-transform:uppercase;
	letter-spacing:3.0pt;
	mso-ansi-language:EN-GB;}
p.MsoHeading9, li.MsoHeading9, div.MsoHeading9
	{mso-style-unhide:no;
	mso-style-qformat:yes;
	mso-style-next:"Body Text";
	margin-top:4.0pt;
	margin-right:0cm;
	margin-bottom:3.0pt;
	margin-left:0cm;
	mso-pagination:widow-orphan;
	page-break-after:avoid;
	mso-outline-level:9;
	font-size:8.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	mso-font-kerning:14.0pt;
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-weight:normal;
	font-style:italic;
	mso-bidi-font-style:normal;}
p.MsoIndex1, li.MsoIndex1, div.MsoIndex1
	{mso-style-update:auto;
	mso-style-noshow:yes;
	mso-style-unhide:no;
	mso-style-next:Normal;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:8.0pt;
	margin-bottom:.0001pt;
	text-indent:-8.0pt;
	mso-pagination:widow-orphan;
	font-size:8.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.MsoToc1, li.MsoToc1, div.MsoToc1
	{mso-style-noshow:yes;
	mso-style-unhide:no;
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:0cm;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:14.0pt;
	mso-bidi-font-size:12.0pt;
	font-family:"Arial Black";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	color:black;
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-style:italic;}
p.MsoToc2, li.MsoToc2, div.MsoToc2
	{mso-style-noshow:yes;
	mso-style-unhide:no;
	mso-style-parent:"TOC 1";
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:0cm;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:11.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	color:black;
	mso-ansi-language:EN-GB;
	font-weight:bold;
	font-style:italic;
	mso-bidi-font-style:normal;}
p.MsoToc3, li.MsoToc3, div.MsoToc3
	{mso-style-noshow:yes;
	mso-style-unhide:no;
	mso-style-next:Normal;
	margin:0cm;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.MsoToc4, li.MsoToc4, div.MsoToc4
	{mso-style-noshow:yes;
	mso-style-unhide:no;
	mso-style-next:Normal;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:17.0pt;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-style:italic;
	mso-bidi-font-style:normal;}
p.MsoHeader, li.MsoHeader, div.MsoHeader
	{mso-style-unhide:no;
	margin:0cm;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan lines-together;
	tab-stops:center 216.0pt right 432.0pt;
	font-size:7.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Arial Black";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	text-transform:uppercase;
	letter-spacing:3.0pt;
	mso-ansi-language:EN-GB;}
p.MsoFooter, li.MsoFooter, div.MsoFooter
	{mso-style-unhide:no;
	margin:0cm;
	margin-bottom:.0001pt;
	text-align:center;
	mso-pagination:widow-orphan lines-together;
	tab-stops:center 216.0pt right 432.0pt;
	border:none;
	mso-border-top-alt:solid windowtext .75pt;
	padding:0cm;
	mso-padding-alt:3.0pt 0cm 0cm 0cm;
	font-size:8.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Arial Black";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.MsoCaption, li.MsoCaption, div.MsoCaption
	{mso-style-unhide:no;
	mso-style-qformat:yes;
	mso-style-next:"Body Text";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:0cm;
	mso-pagination:widow-orphan;
	font-size:8.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
span.MsoCommentReference
	{mso-style-noshow:yes;
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:8.0pt;}
span.MsoLineNumber
	{mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:9.0pt;
	font-family:Arial;
	mso-ascii-font-family:Arial;
	mso-hansi-font-family:Arial;
	mso-bidi-font-family:Arial;}
span.MsoPageNumber
	{mso-style-unhide:no;
	mso-style-parent:"";
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.MsoList, li.MsoList, div.MsoList
	{mso-style-unhide:no;
	mso-style-parent:"Body Text";
	mso-style-link:"List Char";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:18.0pt;
	text-align:justify;
	mso-pagination:widow-orphan;
	tab-stops:36.0pt;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoListBullet, li.MsoListBullet, div.MsoListBullet
	{mso-style-unhide:no;
	mso-style-parent:List;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:18.0pt;
	text-align:justify;
	text-indent:0cm;
	mso-pagination:widow-orphan;
	mso-list:l23 level1 lfo5;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoListNumber, li.MsoListNumber, div.MsoListNumber
	{mso-style-unhide:no;
	mso-style-parent:List;
	margin-top:0cm;
	margin-right:18.0pt;
	margin-bottom:12.0pt;
	margin-left:36.0pt;
	text-align:justify;
	text-indent:-18.0pt;
	mso-pagination:widow-orphan;
	mso-list:l8 level1 lfo7;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoList2, li.MsoList2, div.MsoList2
	{mso-style-unhide:no;
	mso-style-parent:List;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:54.0pt;
	text-align:justify;
	mso-pagination:widow-orphan;
	tab-stops:54.0pt;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoList3, li.MsoList3, div.MsoList3
	{mso-style-unhide:no;
	mso-style-parent:List;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:72.0pt;
	text-align:justify;
	mso-pagination:widow-orphan;
	tab-stops:72.0pt;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoList4, li.MsoList4, div.MsoList4
	{mso-style-unhide:no;
	mso-style-parent:List;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:90.0pt;
	text-align:justify;
	mso-pagination:widow-orphan;
	tab-stops:90.0pt;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoList5, li.MsoList5, div.MsoList5
	{mso-style-unhide:no;
	mso-style-parent:List;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:108.0pt;
	text-align:justify;
	mso-pagination:widow-orphan;
	tab-stops:108.0pt;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoListBullet2, li.MsoListBullet2, div.MsoListBullet2
	{mso-style-unhide:no;
	mso-style-parent:"List Bullet";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:54.0pt;
	text-align:justify;
	text-indent:0cm;
	mso-pagination:widow-orphan;
	mso-list:l7 level1 lfo8;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoListBullet3, li.MsoListBullet3, div.MsoListBullet3
	{mso-style-unhide:no;
	mso-style-parent:"List Bullet";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:72.0pt;
	text-align:justify;
	text-indent:0cm;
	mso-pagination:widow-orphan;
	mso-list:l6 level1 lfo9;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoListBullet4, li.MsoListBullet4, div.MsoListBullet4
	{mso-style-unhide:no;
	mso-style-parent:"List Bullet";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:90.0pt;
	text-align:justify;
	text-indent:0cm;
	mso-pagination:widow-orphan;
	mso-list:l5 level1 lfo10;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoListBullet5, li.MsoListBullet5, div.MsoListBullet5
	{mso-style-unhide:no;
	margin-top:0cm;
	margin-right:18.0pt;
	margin-bottom:0cm;
	margin-left:18.0pt;
	margin-bottom:.0001pt;
	text-indent:-18.0pt;
	line-height:16.0pt;
	mso-line-height-rule:exactly;
	mso-pagination:widow-orphan;
	mso-list:l21 level1 lfo12;
	tab-stops:list 18.0pt;
	border:none;
	mso-border-bottom-alt:solid windowtext .75pt;
	padding:0cm;
	mso-padding-alt:0cm 0cm 0cm 0cm;
	mso-element:frame;
	mso-element-frame-width:93.0pt;
	mso-element-wrap:around;
	mso-element-anchor-vertical:paragraph;
	mso-element-anchor-horizontal:page;
	mso-element-left:60.05pt;
	mso-element-top:.05pt;
	mso-height-rule:exactly;
	font-size:9.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.MsoListNumber2, li.MsoListNumber2, div.MsoListNumber2
	{mso-style-unhide:no;
	mso-style-parent:"List Number";
	margin-top:0cm;
	margin-right:18.0pt;
	margin-bottom:12.0pt;
	margin-left:54.0pt;
	text-align:justify;
	text-indent:-18.0pt;
	mso-pagination:widow-orphan;
	mso-list:l3 level1 lfo14;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoListNumber3, li.MsoListNumber3, div.MsoListNumber3
	{mso-style-unhide:no;
	mso-style-parent:"List Number";
	margin-top:0cm;
	margin-right:18.0pt;
	margin-bottom:12.0pt;
	margin-left:72.0pt;
	text-align:justify;
	text-indent:-18.0pt;
	mso-pagination:widow-orphan;
	mso-list:l2 level1 lfo15;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoListNumber4, li.MsoListNumber4, div.MsoListNumber4
	{mso-style-unhide:no;
	mso-style-parent:"List Number";
	margin-top:0cm;
	margin-right:18.0pt;
	margin-bottom:12.0pt;
	margin-left:90.0pt;
	text-align:justify;
	text-indent:-18.0pt;
	mso-pagination:widow-orphan;
	mso-list:l1 level1 lfo16;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoListNumber5, li.MsoListNumber5, div.MsoListNumber5
	{mso-style-unhide:no;
	mso-style-parent:"List Number";
	margin-top:0cm;
	margin-right:18.0pt;
	margin-bottom:12.0pt;
	margin-left:108.0pt;
	text-align:justify;
	text-indent:-18.0pt;
	mso-pagination:widow-orphan;
	mso-list:l0 level1 lfo17;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoTitle, li.MsoTitle, div.MsoTitle
	{mso-style-unhide:no;
	mso-style-qformat:yes;
	mso-style-parent:"Heading Base";
	margin-top:5.0pt;
	margin-right:0cm;
	margin-bottom:180.0pt;
	margin-left:0cm;
	text-align:center;
	line-height:30.0pt;
	mso-line-height-rule:exactly;
	mso-pagination:widow-orphan;
	page-break-after:avoid;
	border:none;
	mso-border-bottom-alt:solid gray .75pt;
	padding:0cm;
	mso-padding-alt:0cm 0cm 14.0pt 0cm;
	font-size:24.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Arial Black";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	color:gray;
	letter-spacing:-1.75pt;
	mso-font-kerning:14.0pt;
	mso-ansi-language:EN-GB;}
p.MsoBodyText, li.MsoBodyText, div.MsoBodyText
	{mso-style-update:auto;
	mso-style-unhide:no;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:0cm;
	text-align:justify;
	mso-pagination:widow-orphan;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoBodyTextIndent, li.MsoBodyTextIndent, div.MsoBodyTextIndent
	{mso-style-unhide:no;
	mso-style-parent:"Body Text";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:0cm;
	text-align:justify;
	text-indent:18.0pt;
	mso-pagination:widow-orphan;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoListContinue, li.MsoListContinue, div.MsoListContinue
	{mso-style-unhide:no;
	mso-style-parent:List;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:8.0pt;
	margin-left:18.0pt;
	text-align:justify;
	mso-pagination:widow-orphan;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoListContinue2, li.MsoListContinue2, div.MsoListContinue2
	{mso-style-unhide:no;
	mso-style-parent:"List Continue";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:8.0pt;
	margin-left:54.0pt;
	text-align:justify;
	mso-pagination:widow-orphan;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoListContinue3, li.MsoListContinue3, div.MsoListContinue3
	{mso-style-unhide:no;
	mso-style-parent:"List Continue";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:8.0pt;
	margin-left:72.0pt;
	text-align:justify;
	mso-pagination:widow-orphan;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoListContinue4, li.MsoListContinue4, div.MsoListContinue4
	{mso-style-unhide:no;
	mso-style-parent:"List Continue";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:8.0pt;
	margin-left:90.0pt;
	text-align:justify;
	mso-pagination:widow-orphan;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoListContinue5, li.MsoListContinue5, div.MsoListContinue5
	{mso-style-unhide:no;
	mso-style-parent:"List Continue";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:8.0pt;
	margin-left:108.0pt;
	text-align:justify;
	mso-pagination:widow-orphan;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MsoSubtitle, li.MsoSubtitle, div.MsoSubtitle
	{mso-style-unhide:no;
	mso-style-qformat:yes;
	mso-style-parent:Title;
	mso-style-next:"Body Text";
	margin-top:97.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:0cm;
	margin-bottom:.0001pt;
	text-align:center;
	line-height:10.0pt;
	mso-pagination:widow-orphan;
	page-break-after:avoid;
	border:none;
	mso-border-bottom-alt:solid gray .75pt;
	padding:0cm;
	mso-padding-alt:0cm 0cm 14.0pt 0cm;
	font-size:9.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	color:gray;
	text-transform:uppercase;
	letter-spacing:1.5pt;
	mso-font-kerning:14.0pt;
	mso-ansi-language:EN-GB;
	mso-bidi-font-weight:bold;}
p.MsoDate, li.MsoDate, div.MsoDate
	{mso-style-unhide:no;
	mso-style-parent:"Body Text";
	margin-top:24.0pt;
	margin-right:0cm;
	margin-bottom:8.0pt;
	margin-left:0cm;
	text-align:center;
	mso-pagination:widow-orphan;
	font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
a:link, span.MsoHyperlink
	{mso-style-unhide:no;
	mso-style-parent:"";
	color:blue;
	text-decoration:underline;
	text-underline:single;}
a:visited, span.MsoHyperlinkFollowed
	{mso-style-unhide:no;
	mso-style-parent:"";
	color:purple;
	text-decoration:underline;
	text-underline:single;}
em
	{mso-style-unhide:no;
	mso-style-qformat:yes;
	mso-style-parent:"";
	mso-ansi-font-size:9.0pt;
	font-family:"Arial Black";
	mso-ascii-font-family:"Arial Black";
	mso-hansi-font-family:"Arial Black";
	font-style:normal;}
p.MsoPlainText, li.MsoPlainText, div.MsoPlainText
	{mso-style-unhide:no;
	margin:0cm;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:FR;
	mso-fareast-language:FR;}
pre
	{mso-style-unhide:no;
	margin:0cm;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"MS Mincho";
	mso-ansi-language:FR;
	mso-fareast-language:JA;}
span.Heading1Char
	{mso-style-name:"Heading 1 Char";
	mso-style-unhide:no;
	mso-style-locked:yes;
	mso-style-parent:"";
	mso-style-link:"Heading 1";
	mso-ansi-font-size:16.0pt;
	font-family:"Arial Black";
	mso-ascii-font-family:"Arial Black";
	mso-hansi-font-family:"Arial Black";
	color:gray;
	letter-spacing:-1.25pt;
	mso-font-kerning:14.0pt;
	mso-ansi-language:EN-US;
	mso-fareast-language:EN-US;
	mso-bidi-language:AR-SA;}
p.MyNormal, li.MyNormal, div.MyNormal
	{mso-style-name:MyNormal;
	mso-style-update:auto;
	mso-style-unhide:no;
	mso-style-link:"MyNormal Char";
	margin-top:2.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:42.55pt;
	text-align:justify;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:11.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Courier New";}
span.MyNormalChar
	{mso-style-name:"MyNormal Char";
	mso-style-unhide:no;
	mso-style-locked:yes;
	mso-style-parent:"";
	mso-style-link:MyNormal;
	mso-ansi-font-size:11.0pt;
	font-family:"Courier New";
	mso-bidi-font-family:"Courier New";
	mso-ansi-language:EN-US;
	mso-fareast-language:EN-US;
	mso-bidi-language:AR-SA;}
p.examplefileverysmall, li.examplefileverysmall, div.examplefileverysmall
	{mso-style-name:example_file_very_small;
	mso-style-unhide:no;
	mso-style-parent:example_file;
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	mso-add-space:auto;
	text-align:justify;
	mso-pagination:lines-together;
	page-break-after:avoid;
	mso-hyphenate:none;
	background:#E6E6E6;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	border:none;
	mso-border-alt:solid windowtext .5pt;
	padding:0cm;
	mso-padding-alt:1.0pt 4.0pt 1.0pt 4.0pt;
	font-size:8.0pt;
	mso-bidi-font-size:9.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.examplefileverysmallCxSpFirst, li.examplefileverysmallCxSpFirst, div.examplefileverysmallCxSpFirst
	{mso-style-name:example_file_very_smallCxSpFirst;
	mso-style-unhide:no;
	mso-style-parent:example_file;
	mso-style-type:export-only;
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:0cm;
	margin-bottom:.0001pt;
	mso-add-space:auto;
	text-align:justify;
	mso-pagination:lines-together;
	page-break-after:avoid;
	mso-hyphenate:none;
	background:#E6E6E6;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	border:none;
	mso-border-alt:solid windowtext .5pt;
	padding:0cm;
	mso-padding-alt:1.0pt 4.0pt 1.0pt 4.0pt;
	font-size:8.0pt;
	mso-bidi-font-size:9.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.examplefileverysmallCxSpMiddle, li.examplefileverysmallCxSpMiddle, div.examplefileverysmallCxSpMiddle
	{mso-style-name:example_file_very_smallCxSpMiddle;
	mso-style-unhide:no;
	mso-style-parent:example_file;
	mso-style-type:export-only;
	margin:0cm;
	margin-bottom:.0001pt;
	mso-add-space:auto;
	text-align:justify;
	mso-pagination:lines-together;
	page-break-after:avoid;
	mso-hyphenate:none;
	background:#E6E6E6;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	border:none;
	mso-border-alt:solid windowtext .5pt;
	padding:0cm;
	mso-padding-alt:1.0pt 4.0pt 1.0pt 4.0pt;
	font-size:8.0pt;
	mso-bidi-font-size:9.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.examplefileverysmallCxSpLast, li.examplefileverysmallCxSpLast, div.examplefileverysmallCxSpLast
	{mso-style-name:example_file_very_smallCxSpLast;
	mso-style-unhide:no;
	mso-style-parent:example_file;
	mso-style-type:export-only;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	mso-add-space:auto;
	text-align:justify;
	mso-pagination:lines-together;
	page-break-after:avoid;
	mso-hyphenate:none;
	background:#E6E6E6;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	border:none;
	mso-border-alt:solid windowtext .5pt;
	padding:0cm;
	mso-padding-alt:1.0pt 4.0pt 1.0pt 4.0pt;
	font-size:8.0pt;
	mso-bidi-font-size:9.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.examplefile, li.examplefile, div.examplefile
	{mso-style-name:example_file;
	mso-style-update:auto;
	mso-style-unhide:no;
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	mso-add-space:auto;
	text-align:justify;
	mso-pagination:lines-together;
	page-break-after:avoid;
	mso-hyphenate:none;
	background:#E6E6E6;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	border:none;
	mso-border-alt:solid windowtext .5pt;
	padding:0cm;
	mso-padding-alt:1.0pt 4.0pt 1.0pt 4.0pt;
	font-size:9.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.examplefileCxSpFirst, li.examplefileCxSpFirst, div.examplefileCxSpFirst
	{mso-style-name:example_fileCxSpFirst;
	mso-style-update:auto;
	mso-style-unhide:no;
	mso-style-type:export-only;
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:0cm;
	margin-bottom:.0001pt;
	mso-add-space:auto;
	text-align:justify;
	mso-pagination:lines-together;
	page-break-after:avoid;
	mso-hyphenate:none;
	background:#E6E6E6;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	border:none;
	mso-border-alt:solid windowtext .5pt;
	padding:0cm;
	mso-padding-alt:1.0pt 4.0pt 1.0pt 4.0pt;
	font-size:9.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.examplefileCxSpMiddle, li.examplefileCxSpMiddle, div.examplefileCxSpMiddle
	{mso-style-name:example_fileCxSpMiddle;
	mso-style-update:auto;
	mso-style-unhide:no;
	mso-style-type:export-only;
	margin:0cm;
	margin-bottom:.0001pt;
	mso-add-space:auto;
	text-align:justify;
	mso-pagination:lines-together;
	page-break-after:avoid;
	mso-hyphenate:none;
	background:#E6E6E6;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	border:none;
	mso-border-alt:solid windowtext .5pt;
	padding:0cm;
	mso-padding-alt:1.0pt 4.0pt 1.0pt 4.0pt;
	font-size:9.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.examplefileCxSpLast, li.examplefileCxSpLast, div.examplefileCxSpLast
	{mso-style-name:example_fileCxSpLast;
	mso-style-update:auto;
	mso-style-unhide:no;
	mso-style-type:export-only;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	mso-add-space:auto;
	text-align:justify;
	mso-pagination:lines-together;
	page-break-after:avoid;
	mso-hyphenate:none;
	background:#E6E6E6;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	border:none;
	mso-border-alt:solid windowtext .5pt;
	padding:0cm;
	mso-padding-alt:1.0pt 4.0pt 1.0pt 4.0pt;
	font-size:9.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.FAQ, li.FAQ, div.FAQ
	{mso-style-name:FAQ;
	mso-style-update:auto;
	mso-style-unhide:no;
	mso-style-parent:"Heading 3";
	margin-top:12.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:1.0cm;
	text-align:justify;
	mso-pagination:none;
	page-break-after:avoid;
	mso-outline-level:3;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:14.0pt;
	font-family:"Comic Sans MS";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
span.ListChar
	{mso-style-name:"List Char";
	mso-style-unhide:no;
	mso-style-locked:yes;
	mso-style-parent:"";
	mso-style-link:List;
	mso-ansi-font-size:12.0pt;
	font-family:Garamond;
	mso-ascii-font-family:Garamond;
	mso-hansi-font-family:Garamond;
	letter-spacing:-.25pt;
	mso-ansi-language:EN-US;
	mso-fareast-language:EN-US;
	mso-bidi-language:AR-SA;}
span.TextbodyChar
	{mso-style-name:"Text body Char";
	mso-style-unhide:no;
	mso-style-locked:yes;
	mso-style-parent:"";
	mso-style-link:"Text body";
	mso-ansi-font-size:12.0pt;
	mso-ansi-language:EN-US;
	mso-fareast-language:EN-US;
	mso-bidi-language:AR-SA;}
p.Textbody, li.Textbody, div.Textbody
	{mso-style-name:"Text body";
	mso-style-unhide:no;
	mso-style-link:"Text body Char";
	margin:0cm;
	margin-bottom:.0001pt;
	text-align:justify;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.HeadingBase, li.HeadingBase, div.HeadingBase
	{mso-style-name:"Heading Base";
	mso-style-unhide:no;
	mso-style-next:"Body Text";
	margin-top:12.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	mso-pagination:widow-orphan;
	page-break-after:avoid;
	font-size:18.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Arial;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	mso-font-kerning:14.0pt;
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.BlockQuotation, li.BlockQuotation, div.BlockQuotation
	{mso-style-name:"Block Quotation";
	mso-style-unhide:no;
	mso-style-next:"Body Text";
	margin-top:0cm;
	margin-right:30.0pt;
	margin-bottom:12.0pt;
	margin-left:30.0pt;
	text-align:justify;
	mso-pagination:widow-orphan;
	background:#484848;
	mso-shading:windowtext;
	mso-pattern:gray-10 gray;
	border:none;
	mso-border-alt:solid white .75pt;
	padding:0cm;
	mso-padding-alt:12.0pt 12.0pt 12.0pt 12.0pt;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.BlockQuotationFirst, li.BlockQuotationFirst, div.BlockQuotationFirst
	{mso-style-name:"Block Quotation First";
	mso-style-unhide:no;
	mso-style-next:"Block Quotation";
	margin-top:0cm;
	margin-right:24.0pt;
	margin-bottom:0cm;
	margin-left:24.0pt;
	margin-bottom:.0001pt;
	text-indent:3.0pt;
	mso-pagination:widow-orphan lines-together;
	background:#E5E5E5;
	mso-shading:windowtext;
	mso-pattern:gray-10 auto;
	border:none;
	mso-border-top-alt:solid white .75pt;
	mso-border-left-alt:solid white .75pt;
	mso-border-right-alt:solid white .75pt;
	padding:0cm;
	mso-padding-alt:6.0pt 6.0pt 0cm 6.0pt;
	font-size:10.5pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Arial Black";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	letter-spacing:-.5pt;
	mso-ansi-language:EN-GB;}
p.BlockQuotationLast, li.BlockQuotationLast, div.BlockQuotationLast
	{mso-style-name:"Block Quotation Last";
	mso-style-unhide:no;
	mso-style-parent:"Block Quotation";
	mso-style-next:"Body Text";
	margin-top:0cm;
	margin-right:36.0pt;
	margin-bottom:12.0pt;
	margin-left:36.0pt;
	mso-pagination:widow-orphan lines-together;
	font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-style:italic;
	mso-bidi-font-style:normal;}
p.BodyTextKeep, li.BodyTextKeep, div.BodyTextKeep
	{mso-style-name:"Body Text Keep";
	mso-style-unhide:no;
	mso-style-parent:"Body Text";
	mso-style-next:"Body Text";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:0cm;
	text-align:justify;
	mso-pagination:widow-orphan;
	page-break-after:avoid;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.ChapterLabel, li.ChapterLabel, div.ChapterLabel
	{mso-style-name:"Chapter Label";
	mso-style-unhide:no;
	mso-style-next:"Body Text";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:0cm;
	mso-pagination:widow-orphan;
	page-break-after:avoid;
	border:none;
	mso-border-bottom-alt:solid windowtext .75pt;
	padding:0cm;
	mso-padding-alt:0cm 0cm 3.0pt 0cm;
	font-size:7.5pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Arial Black";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	text-transform:uppercase;
	letter-spacing:3.5pt;
	mso-font-kerning:14.0pt;
	mso-ansi-language:EN-GB;}
p.ChapterSubtitle, li.ChapterSubtitle, div.ChapterSubtitle
	{mso-style-name:"Chapter Subtitle";
	mso-style-unhide:no;
	mso-style-next:"Body Text";
	margin-top:0cm;
	margin-right:90.0pt;
	margin-bottom:18.0pt;
	margin-left:0cm;
	mso-line-height-alt:12.0pt;
	mso-pagination:widow-orphan lines-together;
	page-break-after:avoid;
	font-size:14.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	letter-spacing:-1.0pt;
	mso-font-kerning:14.0pt;
	mso-ansi-language:EN-GB;
	font-style:italic;
	mso-bidi-font-style:normal;}
p.ChapterTitle, li.ChapterTitle, div.ChapterTitle
	{mso-style-name:"Chapter Title";
	mso-style-update:auto;
	mso-style-unhide:no;
	mso-style-link:"Chapter Title Char";
	mso-style-next:"Chapter Subtitle";
	margin-top:30.0pt;
	margin-right:4.25pt;
	margin-bottom:50.0pt;
	margin-left:0cm;
	text-align:center;
	mso-line-height-alt:22.0pt;
	page-break-before:always;
	mso-pagination:widow-orphan lines-together;
	page-break-after:avoid;
	background:#CCCCCC;
	border:none;
	mso-border-alt:solid windowtext 3.0pt;
	padding:0cm;
	mso-padding-alt:1.0pt 4.0pt 1.0pt 4.0pt;
	font-size:36.0pt;
	mso-bidi-font-size:48.0pt;
	font-family:"Arial Black";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	color:white;
	letter-spacing:-1.75pt;
	mso-font-kerning:14.0pt;
	mso-ansi-language:EN-GB;}
span.ChapterTitleChar
	{mso-style-name:"Chapter Title Char";
	mso-style-unhide:no;
	mso-style-locked:yes;
	mso-style-parent:"";
	mso-style-link:"Chapter Title";
	mso-ansi-font-size:36.0pt;
	mso-bidi-font-size:48.0pt;
	font-family:"Arial Black";
	mso-ascii-font-family:"Arial Black";
	mso-hansi-font-family:"Arial Black";
	color:white;
	letter-spacing:-1.75pt;
	mso-font-kerning:14.0pt;
	mso-ansi-language:EN-GB;
	mso-fareast-language:EN-US;
	mso-bidi-language:AR-SA;}
p.CompanyName, li.CompanyName, div.CompanyName
	{mso-style-name:"Company Name";
	mso-style-unhide:no;
	mso-style-next:Normal;
	margin-top:21.0pt;
	margin-right:0cm;
	margin-bottom:3.0pt;
	margin-left:0cm;
	line-height:16.0pt;
	mso-line-height-rule:exactly;
	mso-pagination:widow-orphan;
	font-size:19.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	text-transform:uppercase;
	mso-font-kerning:18.0pt;
	mso-ansi-language:EN-GB;}
p.DocumentLabel, li.DocumentLabel, div.DocumentLabel
	{mso-style-name:"Document Label";
	mso-style-unhide:no;
	margin-top:12.0pt;
	margin-right:0cm;
	margin-bottom:18.0pt;
	margin-left:0cm;
	mso-pagination:widow-orphan;
	page-break-after:avoid;
	font-size:18.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	mso-font-kerning:14.0pt;
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.FooterEven, li.FooterEven, div.FooterEven
	{mso-style-name:"Footer Even";
	mso-style-unhide:no;
	mso-style-parent:Footer;
	margin:0cm;
	margin-bottom:.0001pt;
	text-align:center;
	mso-pagination:widow-orphan lines-together;
	tab-stops:center 216.0pt right 432.0pt;
	border:none;
	mso-border-top-alt:solid windowtext .75pt;
	padding:0cm;
	mso-padding-alt:3.0pt 0cm 0cm 0cm;
	font-size:8.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Arial Black";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.FooterFirst, li.FooterFirst, div.FooterFirst
	{mso-style-name:"Footer First";
	mso-style-unhide:no;
	mso-style-parent:Footer;
	margin:0cm;
	margin-bottom:.0001pt;
	text-align:center;
	mso-pagination:widow-orphan lines-together;
	tab-stops:center 216.0pt;
	font-size:8.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Arial Black";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	letter-spacing:-.5pt;
	mso-ansi-language:EN-GB;}
p.FooterOdd, li.FooterOdd, div.FooterOdd
	{mso-style-name:"Footer Odd";
	mso-style-unhide:no;
	mso-style-parent:Footer;
	margin:0cm;
	margin-bottom:.0001pt;
	text-align:center;
	mso-pagination:widow-orphan lines-together;
	tab-stops:right 0cm center 216.0pt right 432.0pt;
	border:none;
	mso-border-top-alt:solid windowtext .75pt;
	padding:0cm;
	mso-padding-alt:3.0pt 0cm 0cm 0cm;
	font-size:8.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Arial Black";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.FootnoteBase, li.FootnoteBase, div.FootnoteBase
	{mso-style-name:"Footnote Base";
	mso-style-unhide:no;
	margin-top:12.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:0cm;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:9.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.HeaderBase, li.HeaderBase, div.HeaderBase
	{mso-style-name:"Header Base";
	mso-style-unhide:no;
	margin:0cm;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan lines-together;
	tab-stops:center 216.0pt right 432.0pt;
	font-size:8.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.HeaderEven, li.HeaderEven, div.HeaderEven
	{mso-style-name:"Header Even";
	mso-style-unhide:no;
	mso-style-parent:Header;
	margin:0cm;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan lines-together;
	tab-stops:center 216.0pt right 432.0pt;
	font-size:7.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Arial Black";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	text-transform:uppercase;
	letter-spacing:3.0pt;
	mso-ansi-language:EN-GB;}
p.HeaderFirst, li.HeaderFirst, div.HeaderFirst
	{mso-style-name:"Header First";
	mso-style-unhide:no;
	mso-style-parent:Header;
	margin:0cm;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan lines-together;
	tab-stops:center 216.0pt;
	font-size:7.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	text-transform:uppercase;
	letter-spacing:3.0pt;
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.HeaderOdd, li.HeaderOdd, div.HeaderOdd
	{mso-style-name:"Header Odd";
	mso-style-unhide:no;
	mso-style-parent:Header;
	margin:0cm;
	margin-bottom:.0001pt;
	text-align:right;
	mso-pagination:widow-orphan lines-together;
	tab-stops:right 0cm center 216.0pt right 432.0pt;
	font-size:7.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Arial Black";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	text-transform:uppercase;
	letter-spacing:3.0pt;
	mso-ansi-language:EN-GB;}
p.Icon1, li.Icon1, div.Icon1
	{mso-style-name:"Icon 1";
	mso-style-unhide:no;
	margin-top:3.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:0cm;
	margin-bottom:.0001pt;
	text-align:center;
	line-height:72.0pt;
	mso-line-height-rule:exactly;
	mso-pagination:widow-orphan;
	background:#E5E5E5;
	mso-shading:windowtext;
	mso-pattern:gray-10 auto;
	mso-element:frame;
	mso-element-frame-width:72.0pt;
	mso-element-frame-hspace:9.35pt;
	mso-element-wrap:around;
	mso-element-anchor-vertical:paragraph;
	mso-element-anchor-horizontal:margin;
	mso-element-top:.05pt;
	mso-height-rule:exactly;
	font-size:80.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Wingdings;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	color:white;
	letter-spacing:-.5pt;
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.IndexBase, li.IndexBase, div.IndexBase
	{mso-style-name:"Index Base";
	mso-style-unhide:no;
	margin:0cm;
	margin-bottom:.0001pt;
	line-height:12.0pt;
	mso-pagination:widow-orphan;
	tab-stops:right 198.0pt;
	font-size:9.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.ListBulletFirst, li.ListBulletFirst, div.ListBulletFirst
	{mso-style-name:"List Bullet First";
	mso-style-unhide:no;
	mso-style-parent:"List Bullet";
	mso-style-next:"List Bullet";
	margin-top:4.0pt;
	margin-right:0cm;
	margin-bottom:8.0pt;
	margin-left:18.0pt;
	text-indent:0cm;
	mso-pagination:widow-orphan;
	mso-list:l23 level1 lfo5;
	font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.ListBulletLast, li.ListBulletLast, div.ListBulletLast
	{mso-style-name:"List Bullet Last";
	mso-style-unhide:no;
	mso-style-parent:"List Bullet";
	mso-style-next:"Body Text";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:18.0pt;
	text-indent:0cm;
	mso-pagination:widow-orphan;
	mso-list:l23 level1 lfo5;
	font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.ListFirst, li.ListFirst, div.ListFirst
	{mso-style-name:"List First";
	mso-style-unhide:no;
	mso-style-parent:List;
	mso-style-next:List;
	margin-top:4.0pt;
	margin-right:0cm;
	margin-bottom:4.0pt;
	margin-left:36.0pt;
	text-indent:-18.0pt;
	mso-pagination:widow-orphan;
	tab-stops:36.0pt;
	font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.ListLast, li.ListLast, div.ListLast
	{mso-style-name:"List Last";
	mso-style-unhide:no;
	mso-style-parent:List;
	mso-style-next:"Body Text";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:36.0pt;
	text-indent:-18.0pt;
	mso-pagination:widow-orphan;
	tab-stops:36.0pt;
	font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.ListNumberFirst, li.ListNumberFirst, div.ListNumberFirst
	{mso-style-name:"List Number First";
	mso-style-unhide:no;
	mso-style-parent:"List Number";
	mso-style-next:"List Number";
	margin-top:4.0pt;
	margin-right:0cm;
	margin-bottom:8.0pt;
	margin-left:36.0pt;
	text-indent:-18.0pt;
	mso-pagination:widow-orphan;
	mso-list:l8 level1 lfo7;
	font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.ListNumberLast, li.ListNumberLast, div.ListNumberLast
	{mso-style-name:"List Number Last";
	mso-style-unhide:no;
	mso-style-parent:"List Number";
	mso-style-next:"Body Text";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:36.0pt;
	text-indent:-18.0pt;
	mso-pagination:widow-orphan;
	mso-list:l8 level1 lfo7;
	font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.PartLabel, li.PartLabel, div.PartLabel
	{mso-style-name:"Part Label";
	mso-style-unhide:no;
	mso-style-link:"Part Label Char";
	mso-style-next:Normal;
	margin-top:16.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:0cm;
	margin-bottom:.0001pt;
	text-align:center;
	line-height:78.0pt;
	mso-line-height-rule:exactly;
	mso-pagination:widow-orphan;
	background:#CCCCCC;
	mso-shading:windowtext;
	mso-pattern:gray-20 auto;
	mso-element:frame;
	mso-element-frame-width:102.25pt;
	mso-element-frame-hspace:9.35pt;
	mso-element-frame-vspace:9.35pt;
	mso-element-wrap:no-wrap-beside;
	mso-element-anchor-vertical:page;
	mso-element-anchor-horizontal:margin;
	mso-element-left:right;
	mso-element-top:48.3pt;
	mso-height-rule:exactly;
	font-size:98.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Arial Black";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	color:white;
	mso-ansi-language:EN-GB;}
span.PartLabelChar
	{mso-style-name:"Part Label Char";
	mso-style-unhide:no;
	mso-style-locked:yes;
	mso-style-parent:"";
	mso-style-link:"Part Label";
	mso-ansi-font-size:98.0pt;
	font-family:"Arial Black";
	mso-ascii-font-family:"Arial Black";
	mso-hansi-font-family:"Arial Black";
	color:white;
	mso-ansi-language:EN-US;
	mso-fareast-language:EN-US;
	mso-bidi-language:AR-SA;}
p.PartSubtitle, li.PartSubtitle, div.PartSubtitle
	{mso-style-name:"Part Subtitle";
	mso-style-unhide:no;
	mso-style-next:"Body Text";
	margin-top:18.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	text-align:center;
	mso-pagination:widow-orphan;
	page-break-after:avoid;
	font-size:16.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Arial;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	mso-font-kerning:14.0pt;
	mso-ansi-language:EN-GB;
	font-style:italic;
	mso-bidi-font-style:normal;}
p.PartTitle, li.PartTitle, div.PartTitle
	{mso-style-name:"Part Title";
	mso-style-unhide:no;
	mso-style-next:"Part Label";
	margin:0cm;
	margin-bottom:.0001pt;
	text-align:center;
	line-height:24.0pt;
	mso-line-height-rule:exactly;
	page-break-before:always;
	mso-pagination:widow-orphan;
	page-break-after:avoid;
	background:#CCCCCC;
	mso-shading:windowtext;
	mso-pattern:gray-20 auto;
	mso-element:frame;
	mso-element-frame-width:102.25pt;
	mso-element-frame-hspace:9.35pt;
	mso-element-frame-vspace:9.35pt;
	mso-element-wrap:no-wrap-beside;
	mso-element-anchor-vertical:page;
	mso-element-anchor-horizontal:margin;
	mso-element-left:right;
	mso-element-top:48.3pt;
	mso-height-rule:exactly;
	font-size:18.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Arial Black";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	letter-spacing:-2.5pt;
	mso-ansi-language:EN-GB;}
p.Picture, li.Picture, div.Picture
	{mso-style-name:Picture;
	mso-style-unhide:no;
	mso-style-parent:"Body Text";
	mso-style-next:Caption;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:0cm;
	text-align:justify;
	mso-pagination:widow-orphan;
	page-break-after:avoid;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.ReturnAddress, li.ReturnAddress, div.ReturnAddress
	{mso-style-name:"Return Address";
	mso-style-unhide:no;
	margin:0cm;
	margin-bottom:.0001pt;
	text-align:center;
	mso-pagination:widow-orphan;
	font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	letter-spacing:-.15pt;
	mso-ansi-language:EN-GB;}
p.SectionHeading, li.SectionHeading, div.SectionHeading
	{mso-style-name:"Section Heading";
	mso-style-unhide:no;
	mso-style-next:"Body Text";
	margin:0cm;
	margin-bottom:.0001pt;
	line-height:32.0pt;
	mso-pagination:widow-orphan;
	font-size:7.5pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Arial Black";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	text-transform:uppercase;
	letter-spacing:3.0pt;
	mso-ansi-language:EN-GB;}
p.SectionLabel, li.SectionLabel, div.SectionLabel
	{mso-style-name:"Section Label";
	mso-style-unhide:no;
	mso-style-next:Normal;
	margin-top:102.0pt;
	margin-right:0cm;
	margin-bottom:18.0pt;
	margin-left:0cm;
	line-height:24.0pt;
	mso-pagination:widow-orphan;
	font-size:24.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Arial Black";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	color:gray;
	letter-spacing:-1.75pt;
	mso-ansi-language:EN-GB;}
p.SubtitleCover, li.SubtitleCover, div.SubtitleCover
	{mso-style-name:"Subtitle Cover";
	mso-style-unhide:no;
	mso-style-next:Normal;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:264.0pt;
	margin-left:0cm;
	line-height:24.0pt;
	mso-line-height-rule:exactly;
	mso-pagination:widow-orphan;
	page-break-after:avoid;
	border:none;
	mso-border-top-alt:solid windowtext .75pt;
	padding:0cm;
	mso-padding-alt:1.0pt 0cm 0cm 0cm;
	font-size:22.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	letter-spacing:-.75pt;
	mso-font-kerning:14.0pt;
	mso-ansi-language:EN-GB;}
p.TitleCover, li.TitleCover, div.TitleCover
	{mso-style-name:"Title Cover";
	mso-style-unhide:no;
	mso-style-parent:"Heading Base";
	mso-style-next:"Subtitle Cover";
	margin-top:0cm;
	margin-right:30.0pt;
	margin-bottom:0cm;
	margin-left:30.0pt;
	margin-bottom:.0001pt;
	text-align:right;
	line-height:72.0pt;
	mso-line-height-rule:exactly;
	mso-pagination:widow-orphan;
	background:#E5E5E5;
	mso-shading:windowtext;
	mso-pattern:gray-10 auto;
	border:none;
	mso-border-alt:solid white .75pt;
	padding:0cm;
	mso-padding-alt:31.0pt 31.0pt 31.0pt 31.0pt;
	font-size:72.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	letter-spacing:-3.5pt;
	mso-font-kerning:14.0pt;
	mso-ansi-language:EN-GB;}
p.TOCBase, li.TOCBase, div.TOCBase
	{mso-style-name:"TOC Base";
	mso-style-unhide:no;
	mso-style-parent:"TOC 2";
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:0cm;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:11.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	color:black;
	mso-ansi-language:EN-GB;
	font-weight:bold;
	font-style:italic;
	mso-bidi-font-style:normal;}
span.Lead-inEmphasis
	{mso-style-name:"Lead-in Emphasis";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:11.0pt;
	text-transform:uppercase;}
span.Superscript
	{mso-style-name:Superscript;
	mso-style-unhide:no;
	mso-style-parent:"";
	position:relative;
	top:0pt;
	mso-text-raise:0pt;
	vertical-align:super;}
p.Heading, li.Heading, div.Heading
	{mso-style-name:Heading;
	mso-style-unhide:no;
	mso-style-next:Normal;
	margin-top:12.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	mso-pagination:none;
	page-break-after:avoid;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:14.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Helvetica;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
span.WW-Absatz-Standardschriftart
	{mso-style-name:WW-Absatz-Standardschriftart;
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.FootnoteSymbol
	{mso-style-name:"Footnote Symbol";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	vertical-align:super;}
span.WW-Absatz-Standardschriftart1
	{mso-style-name:WW-Absatz-Standardschriftart1;
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-FootnoteSymbol
	{mso-style-name:"WW-Footnote Symbol";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	vertical-align:super;}
span.WW-Absatz-Standardschriftart11
	{mso-style-name:WW-Absatz-Standardschriftart11;
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-FootnoteSymbol1
	{mso-style-name:"WW-Footnote Symbol1";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	vertical-align:super;}
span.WW-Absatz-Standardschriftart111
	{mso-style-name:WW-Absatz-Standardschriftart111;
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-FootnoteSymbol11
	{mso-style-name:"WW-Footnote Symbol11";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	vertical-align:super;}
span.WW-Absatz-Standardschriftart1111
	{mso-style-name:WW-Absatz-Standardschriftart1111;
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-FootnoteSymbol111
	{mso-style-name:"WW-Footnote Symbol111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	vertical-align:super;}
span.WW-Absatz-Standardschriftart11111
	{mso-style-name:WW-Absatz-Standardschriftart11111;
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-FootnoteSymbol1111
	{mso-style-name:"WW-Footnote Symbol1111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	vertical-align:super;}
span.WW-Absatz-Standardschriftart111111
	{mso-style-name:WW-Absatz-Standardschriftart111111;
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-FootnoteSymbol11111
	{mso-style-name:"WW-Footnote Symbol11111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	vertical-align:super;}
span.WW-Absatz-Standardschriftart1111111
	{mso-style-name:WW-Absatz-Standardschriftart1111111;
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-FootnoteSymbol111111
	{mso-style-name:"WW-Footnote Symbol111111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	vertical-align:super;}
span.WW-Absatz-Standardschriftart11111111
	{mso-style-name:WW-Absatz-Standardschriftart11111111;
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-FootnoteSymbol1111111
	{mso-style-name:"WW-Footnote Symbol1111111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	vertical-align:super;}
span.WW-Absatz-Standardschriftart111111111
	{mso-style-name:WW-Absatz-Standardschriftart111111111;
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-FootnoteSymbol11111111
	{mso-style-name:"WW-Footnote Symbol11111111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	vertical-align:super;}
span.EndnoteSymbol
	{mso-style-name:"Endnote Symbol";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-EndnoteSymbol
	{mso-style-name:"WW-Endnote Symbol";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-EndnoteSymbol1
	{mso-style-name:"WW-Endnote Symbol1";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-EndnoteSymbol11
	{mso-style-name:"WW-Endnote Symbol11";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-EndnoteSymbol111
	{mso-style-name:"WW-Endnote Symbol111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-EndnoteSymbol1111
	{mso-style-name:"WW-Endnote Symbol1111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-EndnoteSymbol11111
	{mso-style-name:"WW-Endnote Symbol11111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-EndnoteSymbol111111
	{mso-style-name:"WW-Endnote Symbol111111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-EndnoteSymbol1111111
	{mso-style-name:"WW-Endnote Symbol1111111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.Internetlink
	{mso-style-name:"Internet link";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	color:navy;
	text-decoration:underline;
	text-underline:single;}
span.WW-Internetlink
	{mso-style-name:"WW-Internet link";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	color:navy;
	text-decoration:underline;
	text-underline:single;}
span.WW-Internetlink1
	{mso-style-name:"WW-Internet link1";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	color:navy;
	text-decoration:underline;
	text-underline:single;}
span.WW-Internetlink11
	{mso-style-name:"WW-Internet link11";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	color:navy;
	text-decoration:underline;
	text-underline:single;}
span.WW-Internetlink111
	{mso-style-name:"WW-Internet link111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	color:navy;
	text-decoration:underline;
	text-underline:single;}
span.WW-Internetlink1111
	{mso-style-name:"WW-Internet link1111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	color:navy;
	text-decoration:underline;
	text-underline:single;}
span.WW-Internetlink11111
	{mso-style-name:"WW-Internet link11111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	color:navy;
	text-decoration:underline;
	text-underline:single;}
span.WW-Internetlink111111
	{mso-style-name:"WW-Internet link111111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	color:navy;
	text-decoration:underline;
	text-underline:single;}
span.WW-Internetlink1111111
	{mso-style-name:"WW-Internet link1111111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	color:navy;
	text-decoration:underline;
	text-underline:single;}
span.VisitedInternetLink
	{mso-style-name:"Visited Internet Link";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	color:maroon;
	text-decoration:underline;
	text-underline:single;}
span.WW-VisitedInternetLink
	{mso-style-name:"WW-Visited Internet Link";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	color:maroon;
	text-decoration:underline;
	text-underline:single;}
span.WW-VisitedInternetLink1
	{mso-style-name:"WW-Visited Internet Link1";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	color:maroon;
	text-decoration:underline;
	text-underline:single;}
span.WW-VisitedInternetLink11
	{mso-style-name:"WW-Visited Internet Link11";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	color:maroon;
	text-decoration:underline;
	text-underline:single;}
span.WW-VisitedInternetLink111
	{mso-style-name:"WW-Visited Internet Link111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	color:maroon;
	text-decoration:underline;
	text-underline:single;}
span.WW-VisitedInternetLink1111
	{mso-style-name:"WW-Visited Internet Link1111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	color:maroon;
	text-decoration:underline;
	text-underline:single;}
span.WW-VisitedInternetLink11111
	{mso-style-name:"WW-Visited Internet Link11111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	color:maroon;
	text-decoration:underline;
	text-underline:single;}
span.WW-VisitedInternetLink111111
	{mso-style-name:"WW-Visited Internet Link111111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	color:maroon;
	text-decoration:underline;
	text-underline:single;}
span.WW-VisitedInternetLink1111111
	{mso-style-name:"WW-Visited Internet Link1111111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;
	color:maroon;
	text-decoration:underline;
	text-underline:single;}
span.NumberingSymbols
	{mso-style-name:"Numbering Symbols";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-NumberingSymbols
	{mso-style-name:"WW-Numbering Symbols";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-NumberingSymbols1
	{mso-style-name:"WW-Numbering Symbols1";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-NumberingSymbols11
	{mso-style-name:"WW-Numbering Symbols11";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-NumberingSymbols111
	{mso-style-name:"WW-Numbering Symbols111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-NumberingSymbols1111
	{mso-style-name:"WW-Numbering Symbols1111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.WW-NumberingSymbols11111
	{mso-style-name:"WW-Numbering Symbols11111";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:12.0pt;}
span.BulletSymbols
	{mso-style-name:"Bullet Symbols";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:9.0pt;
	font-family:starbats;
	mso-ascii-font-family:starbats;
	mso-hansi-font-family:starbats;}
span.WW-BulletSymbols
	{mso-style-name:"WW-Bullet Symbols";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:9.0pt;
	font-family:starbats;
	mso-ascii-font-family:starbats;
	mso-hansi-font-family:starbats;}
span.WW-BulletSymbols1
	{mso-style-name:"WW-Bullet Symbols1";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:9.0pt;
	font-family:starbats;
	mso-ascii-font-family:starbats;
	mso-hansi-font-family:starbats;}
span.WW-BulletSymbols11
	{mso-style-name:"WW-Bullet Symbols11";
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-ansi-font-size:9.0pt;
	font-family:starbats;
	mso-ascii-font-family:starbats;
	mso-hansi-font-family:starbats;}
p.Index, li.Index, div.Index
	{mso-style-name:Index;
	mso-style-unhide:no;
	margin:0cm;
	margin-bottom:.0001pt;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.ContentsHeading, li.ContentsHeading, div.ContentsHeading
	{mso-style-name:"Contents Heading";
	mso-style-unhide:no;
	margin-top:12.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	mso-pagination:none;
	page-break-after:avoid;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:16.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Helvetica;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.Tableofcontents6, li.Tableofcontents6, div.Tableofcontents6
	{mso-style-name:"Table of contents 6";
	mso-style-unhide:no;
	mso-style-parent:Index;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:70.75pt;
	margin-bottom:.0001pt;
	mso-pagination:none;
	mso-hyphenate:none;
	tab-stops:right dotted 432.0pt;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.Tableofcontents7, li.Tableofcontents7, div.Tableofcontents7
	{mso-style-name:"Table of contents 7";
	mso-style-unhide:no;
	mso-style-parent:Index;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:84.9pt;
	margin-bottom:.0001pt;
	mso-pagination:none;
	mso-hyphenate:none;
	tab-stops:right dotted 432.0pt;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.Tableofcontents8, li.Tableofcontents8, div.Tableofcontents8
	{mso-style-name:"Table of contents 8";
	mso-style-unhide:no;
	mso-style-parent:Index;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:99.05pt;
	margin-bottom:.0001pt;
	mso-pagination:none;
	mso-hyphenate:none;
	tab-stops:right dotted 432.0pt;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.Tableofcontents9, li.Tableofcontents9, div.Tableofcontents9
	{mso-style-name:"Table of contents 9";
	mso-style-unhide:no;
	mso-style-parent:Index;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:113.2pt;
	margin-bottom:.0001pt;
	mso-pagination:none;
	mso-hyphenate:none;
	tab-stops:right dotted 432.0pt;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.Tableofcontents10, li.Tableofcontents10, div.Tableofcontents10
	{mso-style-name:"Table of contents 10";
	mso-style-unhide:no;
	mso-style-parent:Index;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:127.35pt;
	margin-bottom:.0001pt;
	mso-pagination:none;
	mso-hyphenate:none;
	tab-stops:right dotted 432.0pt;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.Textbodyindent, li.Textbodyindent, div.Textbodyindent
	{mso-style-name:"Text body indent";
	mso-style-unhide:no;
	mso-style-parent:"Text body";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:14.15pt;
	margin-bottom:.0001pt;
	text-align:justify;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.ListContents, li.ListContents, div.ListContents
	{mso-style-name:"List Contents";
	mso-style-unhide:no;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:28.3pt;
	margin-bottom:.0001pt;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.FAQexample, li.FAQexample, div.FAQexample
	{mso-style-name:FAQ_example;
	mso-style-unhide:no;
	margin-top:12.0pt;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:1.0cm;
	mso-add-space:auto;
	line-height:200%;
	mso-pagination:widow-orphan lines-together;
	page-break-after:avoid;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.FAQexampleCxSpFirst, li.FAQexampleCxSpFirst, div.FAQexampleCxSpFirst
	{mso-style-name:FAQ_exampleCxSpFirst;
	mso-style-unhide:no;
	mso-style-type:export-only;
	margin-top:12.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:1.0cm;
	margin-bottom:.0001pt;
	mso-add-space:auto;
	line-height:200%;
	mso-pagination:widow-orphan lines-together;
	page-break-after:avoid;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.FAQexampleCxSpMiddle, li.FAQexampleCxSpMiddle, div.FAQexampleCxSpMiddle
	{mso-style-name:FAQ_exampleCxSpMiddle;
	mso-style-unhide:no;
	mso-style-type:export-only;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:1.0cm;
	margin-bottom:.0001pt;
	mso-add-space:auto;
	line-height:200%;
	mso-pagination:widow-orphan lines-together;
	page-break-after:avoid;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.FAQexampleCxSpLast, li.FAQexampleCxSpLast, div.FAQexampleCxSpLast
	{mso-style-name:FAQ_exampleCxSpLast;
	mso-style-unhide:no;
	mso-style-type:export-only;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:1.0cm;
	mso-add-space:auto;
	line-height:200%;
	mso-pagination:widow-orphan lines-together;
	page-break-after:avoid;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.Firstlineindent, li.Firstlineindent, div.Firstlineindent
	{mso-style-name:"First line indent";
	mso-style-unhide:no;
	mso-style-parent:"Text body";
	margin:0cm;
	margin-bottom:.0001pt;
	text-align:justify;
	text-indent:14.15pt;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.Hangingindent, li.Hangingindent, div.Hangingindent
	{mso-style-name:"Hanging indent";
	mso-style-unhide:no;
	mso-style-parent:"Text body";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:1.0cm;
	margin-bottom:.0001pt;
	text-align:justify;
	text-indent:-14.15pt;
	mso-pagination:none;
	mso-hyphenate:none;
	tab-stops:1.0cm;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.Numbering3, li.Numbering3, div.Numbering3
	{mso-style-name:"Numbering 3";
	mso-style-unhide:no;
	mso-style-parent:List;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:42.45pt;
	text-align:justify;
	text-indent:-14.15pt;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.TableContents, li.TableContents, div.TableContents
	{mso-style-name:"Table Contents";
	mso-style-unhide:no;
	mso-style-parent:"Text body";
	margin:0cm;
	margin-bottom:.0001pt;
	text-align:justify;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.TableHeading, li.TableHeading, div.TableHeading
	{mso-style-name:"Table Heading";
	mso-style-unhide:no;
	mso-style-parent:"Table Contents";
	margin:0cm;
	margin-bottom:.0001pt;
	text-align:center;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-weight:normal;
	font-style:italic;
	mso-bidi-font-style:normal;}
p.author, li.author, div.author
	{mso-style-name:author;
	mso-style-unhide:no;
	margin:0cm;
	margin-bottom:.0001pt;
	text-align:justify;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.numberedlist, li.numberedlist, div.numberedlist
	{mso-style-name:"numbered list";
	mso-style-update:auto;
	mso-style-unhide:no;
	mso-style-parent:"Heading 4";
	margin-top:3.0pt;
	margin-right:0cm;
	margin-bottom:3.0pt;
	margin-left:0cm;
	mso-pagination:none;
	page-break-after:avoid;
	mso-outline-level:4;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	mso-bidi-font-size:12.0pt;
	font-family:"Arial Black";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Courier New";
	text-transform:uppercase;
	letter-spacing:1.0pt;}
p.Noteboxtitle, li.Noteboxtitle, div.Noteboxtitle
	{mso-style-name:"Notebox\.title";
	mso-style-unhide:no;
	margin:0cm;
	margin-bottom:.0001pt;
	text-align:center;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:18.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Arial;
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;}
p.noteboxexample, li.noteboxexample, div.noteboxexample
	{mso-style-name:"notebox\.example";
	mso-style-unhide:no;
	margin:0cm;
	margin-bottom:.0001pt;
	text-indent:36.0pt;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;}
p.noteboxbody, li.noteboxbody, div.noteboxbody
	{mso-style-name:"notebox\.body";
	mso-style-unhide:no;
	margin:0cm;
	margin-bottom:.0001pt;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-style:italic;}
p.Note, li.Note, div.Note
	{mso-style-name:Note;
	mso-style-update:auto;
	mso-style-unhide:no;
	mso-style-link:"Note Char";
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:42.55pt;
	text-align:justify;
	mso-pagination:none;
	mso-hyphenate:none;
	tab-stops:1.0cm;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	mso-bidi-font-size:8.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
span.NoteChar
	{mso-style-name:"Note Char";
	mso-style-unhide:no;
	mso-style-locked:yes;
	mso-style-parent:"";
	mso-style-link:Note;
	mso-bidi-font-size:8.0pt;
	mso-ansi-language:EN-GB;
	mso-fareast-language:EN-US;
	mso-bidi-language:AR-SA;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.Table, li.Table, div.Table
	{mso-style-name:Table;
	mso-style-update:auto;
	mso-style-unhide:no;
	mso-style-parent:Index;
	margin:0cm;
	margin-bottom:.0001pt;
	mso-pagination:lines-together;
	page-break-after:avoid;
	mso-hyphenate:none;
	tab-stops:48.75pt;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.flag, li.flag, div.flag
	{mso-style-name:flag;
	mso-style-unhide:no;
	mso-style-parent:"Heading 4";
	margin-top:12.0pt;
	margin-right:0cm;
	margin-bottom:3.0pt;
	margin-left:1.0cm;
	mso-pagination:none;
	page-break-after:avoid;
	mso-outline-level:4;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Courier New";
	text-transform:uppercase;
	letter-spacing:1.0pt;}
p.flagusage, li.flagusage, div.flagusage
	{mso-style-name:"flag usage";
	mso-style-update:auto;
	mso-style-unhide:no;
	mso-style-parent:"Heading 5";
	margin-top:3.0pt;
	margin-right:0cm;
	margin-bottom:3.0pt;
	margin-left:42.55pt;
	mso-pagination:none;
	mso-outline-level:5;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:13.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:FR;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.flagdefault, li.flagdefault, div.flagdefault
	{mso-style-name:"flag default";
	mso-style-unhide:no;
	mso-style-parent:"Heading 5";
	margin-top:3.0pt;
	margin-right:0cm;
	margin-bottom:3.0pt;
	margin-left:18.0pt;
	mso-pagination:none;
	mso-outline-level:5;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:13.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;}
p.flagdescription, li.flagdescription, div.flagdescription
	{mso-style-name:"flag description";
	mso-style-update:auto;
	mso-style-unhide:no;
	mso-style-link:"flag description Char";
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:42.55pt;
	margin-bottom:.0001pt;
	text-align:justify;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
span.flagdescriptionChar
	{mso-style-name:"flag description Char";
	mso-style-unhide:no;
	mso-style-locked:yes;
	mso-style-parent:"";
	mso-style-link:"flag description";
	mso-ansi-language:EN-US;
	mso-fareast-language:EN-US;
	mso-bidi-language:AR-SA;}
p.flagNote, li.flagNote, div.flagNote
	{mso-style-name:"flag Note";
	mso-style-unhide:no;
	margin:0cm;
	margin-bottom:.0001pt;
	text-align:justify;
	text-indent:.05pt;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;}
p.flagdescriptionid1, li.flagdescriptionid1, div.flagdescriptionid1
	{mso-style-name:"flag description id1";
	mso-style-unhide:no;
	mso-style-parent:"flag description";
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:70.9pt;
	margin-bottom:.0001pt;
	text-align:justify;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.output, li.output, div.output
	{mso-style-name:output;
	mso-style-unhide:no;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:18.0pt;
	margin-bottom:.0001pt;
	text-align:justify;
	text-indent:.05pt;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:8.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Courier;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.instruction, li.instruction, div.instruction
	{mso-style-name:instruction;
	mso-style-update:auto;
	mso-style-unhide:no;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:42.55pt;
	margin-bottom:.0001pt;
	line-height:200%;
	mso-pagination:widow-orphan lines-together;
	page-break-after:avoid;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.Format, li.Format, div.Format
	{mso-style-name:Format;
	mso-style-unhide:no;
	margin-top:12.0pt;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:1.0cm;
	mso-add-space:auto;
	line-height:200%;
	mso-pagination:widow-orphan lines-together;
	page-break-after:avoid;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.FormatCxSpFirst, li.FormatCxSpFirst, div.FormatCxSpFirst
	{mso-style-name:FormatCxSpFirst;
	mso-style-unhide:no;
	mso-style-type:export-only;
	margin-top:12.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:1.0cm;
	margin-bottom:.0001pt;
	mso-add-space:auto;
	line-height:200%;
	mso-pagination:widow-orphan lines-together;
	page-break-after:avoid;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.FormatCxSpMiddle, li.FormatCxSpMiddle, div.FormatCxSpMiddle
	{mso-style-name:FormatCxSpMiddle;
	mso-style-unhide:no;
	mso-style-type:export-only;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:1.0cm;
	margin-bottom:.0001pt;
	mso-add-space:auto;
	line-height:200%;
	mso-pagination:widow-orphan lines-together;
	page-break-after:avoid;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.FormatCxSpLast, li.FormatCxSpLast, div.FormatCxSpLast
	{mso-style-name:FormatCxSpLast;
	mso-style-unhide:no;
	mso-style-type:export-only;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:1.0cm;
	mso-add-space:auto;
	line-height:200%;
	mso-pagination:widow-orphan lines-together;
	page-break-after:avoid;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.ListinFlag, li.ListinFlag, div.ListinFlag
	{mso-style-name:"List in Flag";
	mso-style-unhide:no;
	mso-style-parent:List;
	margin:0cm;
	margin-bottom:.0001pt;
	text-align:justify;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.example2, li.example2, div.example2
	{mso-style-name:example2;
	mso-style-unhide:no;
	margin-top:12.0pt;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:2.0cm;
	mso-add-space:auto;
	line-height:200%;
	mso-pagination:widow-orphan lines-together;
	page-break-after:avoid;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.example2CxSpFirst, li.example2CxSpFirst, div.example2CxSpFirst
	{mso-style-name:example2CxSpFirst;
	mso-style-unhide:no;
	mso-style-type:export-only;
	margin-top:12.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:2.0cm;
	margin-bottom:.0001pt;
	mso-add-space:auto;
	line-height:200%;
	mso-pagination:widow-orphan lines-together;
	page-break-after:avoid;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.example2CxSpMiddle, li.example2CxSpMiddle, div.example2CxSpMiddle
	{mso-style-name:example2CxSpMiddle;
	mso-style-unhide:no;
	mso-style-type:export-only;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:2.0cm;
	margin-bottom:.0001pt;
	mso-add-space:auto;
	line-height:200%;
	mso-pagination:widow-orphan lines-together;
	page-break-after:avoid;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.example2CxSpLast, li.example2CxSpLast, div.example2CxSpLast
	{mso-style-name:example2CxSpLast;
	mso-style-unhide:no;
	mso-style-type:export-only;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:2.0cm;
	mso-add-space:auto;
	line-height:200%;
	mso-pagination:widow-orphan lines-together;
	page-break-after:avoid;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.StyleNote, li.StyleNote, div.StyleNote
	{mso-style-name:"Style Note";
	mso-style-unhide:no;
	mso-style-parent:Note;
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:-35.45pt;
	text-align:justify;
	mso-pagination:none;
	mso-hyphenate:none;
	tab-stops:1.0cm;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	mso-bidi-font-size:8.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;}
p.Warning, li.Warning, div.Warning
	{mso-style-name:Warning;
	mso-style-update:auto;
	mso-style-unhide:no;
	mso-style-link:"Warning Char";
	mso-margin-top-alt:auto;
	margin-right:0cm;
	mso-margin-bottom-alt:auto;
	margin-left:0cm;
	mso-pagination:widow-orphan;
	background:#F7CBB7;
	border:none;
	mso-border-alt:solid windowtext 1.5pt;
	padding:0cm;
	mso-padding-alt:0cm 0cm 0cm 0cm;
	font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
span.WarningChar
	{mso-style-name:"Warning Char";
	mso-style-unhide:no;
	mso-style-locked:yes;
	mso-style-parent:"";
	mso-style-link:Warning;
	mso-bidi-font-size:8.0pt;
	font-family:Garamond;
	mso-ascii-font-family:Garamond;
	mso-hansi-font-family:Garamond;
	mso-ansi-language:EN-US;
	mso-fareast-language:EN-US;
	mso-bidi-language:AR-SA;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.Heaing3, li.Heaing3, div.Heaing3
	{mso-style-name:"Heaing 3";
	mso-style-unhide:no;
	mso-style-parent:Note;
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:42.55pt;
	text-align:justify;
	mso-pagination:none;
	mso-hyphenate:none;
	tab-stops:1.0cm;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	mso-bidi-font-size:8.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.NormalJustified, li.NormalJustified, div.NormalJustified
	{mso-style-name:"Normal + Justified";
	mso-style-unhide:no;
	mso-style-link:"Normal + Justified Char";
	margin:0cm;
	margin-bottom:.0001pt;
	text-align:justify;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
span.NormalJustifiedChar
	{mso-style-name:"Normal + Justified Char";
	mso-style-unhide:no;
	mso-style-locked:yes;
	mso-style-parent:"";
	mso-style-link:"Normal + Justified";
	mso-ansi-font-size:12.0pt;
	mso-ansi-language:EN-GB;
	mso-fareast-language:EN-US;
	mso-bidi-language:AR-SA;}
p.MyNormalID, li.MyNormalID, div.MyNormalID
	{mso-style-name:MyNormalID;
	mso-style-update:auto;
	mso-style-unhide:no;
	mso-style-parent:"List Bullet";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:31.2pt;
	margin-bottom:.0001pt;
	text-indent:0cm;
	mso-pagination:none;
	mso-list:l8 level1 lfo7;
	mso-hyphenate:none;
	tab-stops:list 18.0pt;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.MyExample, li.MyExample, div.MyExample
	{mso-style-name:MyExample;
	mso-style-update:auto;
	mso-style-unhide:no;
	mso-margin-top-alt:auto;
	margin-right:0cm;
	mso-margin-bottom-alt:auto;
	margin-left:2.0cm;
	mso-pagination:lines-together;
	page-break-after:avoid;
	mso-hyphenate:none;
	background:#FFFFCC;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	border:none;
	mso-border-top-alt:solid black .25pt;
	mso-border-bottom-alt:solid black .25pt;
	padding:0cm;
	mso-padding-alt:7.0pt 0cm 10.0pt 0cm;
	font-size:9.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	color:black;
	mso-ansi-language:ZU;
	font-weight:bold;
	mso-bidi-font-weight:normal;
	mso-no-proof:yes;}
p.Style1, li.Style1, div.Style1
	{mso-style-name:Style1;
	mso-style-update:auto;
	mso-style-unhide:no;
	margin:0cm;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:8.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.StyleHeading6Bold, li.StyleHeading6Bold, div.StyleHeading6Bold
	{mso-style-name:"Style Heading 6 + Bold";
	mso-style-update:auto;
	mso-style-unhide:no;
	mso-style-link:"Style Heading 6 + Bold Char";
	margin:0cm;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:8.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:Garamond;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;}
span.StyleHeading6BoldChar
	{mso-style-name:"Style Heading 6 + Bold Char";
	mso-style-unhide:no;
	mso-style-locked:yes;
	mso-style-parent:"";
	mso-style-link:"Style Heading 6 + Bold";
	mso-ansi-font-size:8.0pt;
	font-family:Garamond;
	mso-ascii-font-family:Garamond;
	mso-hansi-font-family:Garamond;
	mso-ansi-language:EN-US;
	mso-fareast-language:EN-US;
	mso-bidi-language:AR-SA;
	font-weight:bold;}
p.MyNormaBoldl, li.MyNormaBoldl, div.MyNormaBoldl
	{mso-style-name:MyNormaBoldl;
	mso-style-unhide:no;
	mso-style-parent:MyNormal;
	margin-top:2.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:42.55pt;
	text-align:justify;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:11.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Courier New";}
p.Mylist, li.Mylist, div.Mylist
	{mso-style-name:Mylist;
	mso-style-unhide:no;
	mso-style-parent:MyNormal;
	margin-top:2.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:128.7pt;
	text-align:justify;
	text-indent:-72.0pt;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:11.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Courier New";
	font-style:italic;
	mso-bidi-font-style:normal;}
p.MyNormalBold, li.MyNormalBold, div.MyNormalBold
	{mso-style-name:MyNormalBold;
	mso-style-unhide:no;
	mso-style-parent:MyNormalID;
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:31.2pt;
	text-indent:0cm;
	mso-pagination:lines-together;
	mso-list:l8 level1 lfo7;
	tab-stops:list 18.0pt;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.StyleflagdescriptionBoldItalic, li.StyleflagdescriptionBoldItalic, div.StyleflagdescriptionBoldItalic
	{mso-style-name:"Style flag description + Bold Italic";
	mso-style-update:auto;
	mso-style-unhide:no;
	mso-style-parent:"flag description";
	mso-style-link:"Style flag description + Bold Italic Char";
	margin-top:3.0pt;
	margin-right:0cm;
	margin-bottom:3.0pt;
	margin-left:42.55pt;
	text-align:justify;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;
	font-style:italic;}
span.StyleflagdescriptionBoldItalicChar
	{mso-style-name:"Style flag description + Bold Italic Char";
	mso-style-unhide:no;
	mso-style-locked:yes;
	mso-style-parent:"";
	mso-style-link:"Style flag description + Bold Italic";
	mso-ansi-language:EN-US;
	mso-fareast-language:EN-US;
	mso-bidi-language:AR-SA;
	font-weight:bold;
	font-style:italic;}
p.Formula, li.Formula, div.Formula
	{mso-style-name:Formula;
	mso-style-unhide:no;
	mso-style-parent:"flag description";
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:42.55pt;
	margin-bottom:.0001pt;
	text-align:justify;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.examplefilesmall, li.examplefilesmall, div.examplefilesmall
	{mso-style-name:example_file_small;
	mso-style-unhide:no;
	mso-style-parent:example_file;
	margin-top:6.0pt;
	margin-right:1.0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	mso-add-space:auto;
	text-align:justify;
	mso-pagination:lines-together;
	page-break-after:avoid;
	mso-hyphenate:none;
	background:#E6E6E6;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	border:none;
	mso-border-alt:solid windowtext .5pt;
	padding:0cm;
	mso-padding-alt:1.0pt 4.0pt 1.0pt 4.0pt;
	font-size:8.0pt;
	mso-bidi-font-size:9.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.examplefilesmallCxSpFirst, li.examplefilesmallCxSpFirst, div.examplefilesmallCxSpFirst
	{mso-style-name:example_file_smallCxSpFirst;
	mso-style-unhide:no;
	mso-style-parent:example_file;
	mso-style-type:export-only;
	margin-top:6.0pt;
	margin-right:1.0cm;
	margin-bottom:0cm;
	margin-left:0cm;
	margin-bottom:.0001pt;
	mso-add-space:auto;
	text-align:justify;
	mso-pagination:lines-together;
	page-break-after:avoid;
	mso-hyphenate:none;
	background:#E6E6E6;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	border:none;
	mso-border-alt:solid windowtext .5pt;
	padding:0cm;
	mso-padding-alt:1.0pt 4.0pt 1.0pt 4.0pt;
	font-size:8.0pt;
	mso-bidi-font-size:9.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.examplefilesmallCxSpMiddle, li.examplefilesmallCxSpMiddle, div.examplefilesmallCxSpMiddle
	{mso-style-name:example_file_smallCxSpMiddle;
	mso-style-unhide:no;
	mso-style-parent:example_file;
	mso-style-type:export-only;
	margin-top:0cm;
	margin-right:1.0cm;
	margin-bottom:0cm;
	margin-left:0cm;
	margin-bottom:.0001pt;
	mso-add-space:auto;
	text-align:justify;
	mso-pagination:lines-together;
	page-break-after:avoid;
	mso-hyphenate:none;
	background:#E6E6E6;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	border:none;
	mso-border-alt:solid windowtext .5pt;
	padding:0cm;
	mso-padding-alt:1.0pt 4.0pt 1.0pt 4.0pt;
	font-size:8.0pt;
	mso-bidi-font-size:9.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.examplefilesmallCxSpLast, li.examplefilesmallCxSpLast, div.examplefilesmallCxSpLast
	{mso-style-name:example_file_smallCxSpLast;
	mso-style-unhide:no;
	mso-style-parent:example_file;
	mso-style-type:export-only;
	margin-top:0cm;
	margin-right:1.0cm;
	margin-bottom:6.0pt;
	margin-left:0cm;
	mso-add-space:auto;
	text-align:justify;
	mso-pagination:lines-together;
	page-break-after:avoid;
	mso-hyphenate:none;
	background:#E6E6E6;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	border:none;
	mso-border-alt:solid windowtext .5pt;
	padding:0cm;
	mso-padding-alt:1.0pt 4.0pt 1.0pt 4.0pt;
	font-size:8.0pt;
	mso-bidi-font-size:9.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;}
p.StylenoteboxbodyBoldNotItalicBoxSinglesolidlineAut, li.StylenoteboxbodyBoldNotItalicBoxSinglesolidlineAut, div.StylenoteboxbodyBoldNotItalicBoxSinglesolidlineAut
	{mso-style-name:"Style notebox\.body + Bold Not Italic Box\: \(Single solid line Aut\.\.\.";
	mso-style-unhide:no;
	mso-style-parent:"notebox\.body";
	margin:0cm;
	margin-bottom:.0001pt;
	mso-pagination:none;
	mso-hyphenate:none;
	background:#FFFF89;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	border:none;
	mso-border-alt:solid windowtext 1.5pt;
	padding:0cm;
	mso-padding-alt:1.0pt 4.0pt 1.0pt 4.0pt;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;}
p.MyNormalJustified, li.MyNormalJustified, div.MyNormalJustified
	{mso-style-name:MyNormalJustified;
	mso-style-unhide:no;
	mso-style-parent:MyNormal;
	margin-top:2.0pt;
	margin-right:0cm;
	margin-bottom:6.0pt;
	margin-left:42.55pt;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:11.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Courier New";}
p.flagheading2, li.flagheading2, div.flagheading2
	{mso-style-name:"flag heading 2";
	mso-style-update:auto;
	mso-style-unhide:no;
	mso-style-parent:"flag description";
	margin-top:6.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:45.0pt;
	margin-bottom:.0001pt;
	text-align:justify;
	mso-pagination:none;
	mso-hyphenate:none;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	mso-ansi-language:EN-GB;
	font-weight:bold;
	mso-bidi-font-weight:normal;
	font-style:italic;
	mso-bidi-font-style:normal;
	text-decoration:underline;
	text-underline:single;}
p.example1, li.example1, div.example1
	{mso-style-name:example1;
	mso-style-unhide:no;
	mso-style-parent:example2;
	margin-top:12.0pt;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:2.0cm;
	mso-add-space:auto;
	line-height:200%;
	mso-pagination:widow-orphan lines-together;
	page-break-after:avoid;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.example1CxSpFirst, li.example1CxSpFirst, div.example1CxSpFirst
	{mso-style-name:example1CxSpFirst;
	mso-style-unhide:no;
	mso-style-parent:example2;
	mso-style-type:export-only;
	margin-top:12.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:2.0cm;
	margin-bottom:.0001pt;
	mso-add-space:auto;
	line-height:200%;
	mso-pagination:widow-orphan lines-together;
	page-break-after:avoid;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.example1CxSpMiddle, li.example1CxSpMiddle, div.example1CxSpMiddle
	{mso-style-name:example1CxSpMiddle;
	mso-style-unhide:no;
	mso-style-parent:example2;
	mso-style-type:export-only;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:2.0cm;
	margin-bottom:.0001pt;
	mso-add-space:auto;
	line-height:200%;
	mso-pagination:widow-orphan lines-together;
	page-break-after:avoid;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.example1CxSpLast, li.example1CxSpLast, div.example1CxSpLast
	{mso-style-name:example1CxSpLast;
	mso-style-unhide:no;
	mso-style-parent:example2;
	mso-style-type:export-only;
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:2.0cm;
	mso-add-space:auto;
	line-height:200%;
	mso-pagination:widow-orphan lines-together;
	page-break-after:avoid;
	mso-layout-grid-align:none;
	punctuation-wrap:simple;
	text-autospace:none;
	font-size:10.0pt;
	font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	font-weight:bold;
	mso-bidi-font-weight:normal;}
p.Headin3, li.Headin3, div.Headin3
	{mso-style-name:Headin3;
	mso-style-unhide:no;
	mso-style-parent:"Body Text";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:0cm;
	text-align:justify;
	mso-pagination:widow-orphan;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
p.MyNormal0, li.MyNormal0, div.MyNormal0
	{mso-style-name:"My Normal";
	mso-style-unhide:no;
	mso-style-parent:"Body Text";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:12.0pt;
	margin-left:0cm;
	text-align:justify;
	mso-pagination:widow-orphan;
	font-size:12.0pt;
	mso-bidi-font-size:10.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";
	letter-spacing:-.25pt;
	mso-ansi-language:EN-GB;}
span.SpellE
	{mso-style-name:"";
	mso-spl-e:yes;}
span.GramE
	{mso-style-name:"";
	mso-gram-e:yes;}
 /* Page Definitions */
@page
	{mso-footnote-separator:url(":t_coffee_tutorial_files:header.htm") fs;
	mso-footnote-continuation-separator:url(":t_coffee_tutorial_files:header.htm") fcs;
	mso-endnote-separator:url(":t_coffee_tutorial_files:header.htm") es;
	mso-endnote-continuation-separator:url(":t_coffee_tutorial_files:header.htm") ecs;}
@page WordSection1
	{size:21.0cm 842.0pt;
	margin:90.0pt 60.0pt 90.0pt 60.0pt;
	mso-header-margin:48.0pt;
	mso-footer-margin:48.0pt;
	mso-page-numbers:1;
	mso-header:url(":t_coffee_tutorial_files:header.htm") h1;
	mso-footer:url(":t_coffee_tutorial_files:header.htm") f1;
	mso-first-header:url(":t_coffee_tutorial_files:header.htm") fh1;
	mso-paper-source:0;}
div.WordSection1
	{page:WordSection1;}
@page WordSection2
	{size:21.0cm 842.0pt;
	margin:72.0pt 90.0pt 72.0pt 90.0pt;
	mso-header-margin:36.0pt;
	mso-footer-margin:36.0pt;
	mso-header:url(":t_coffee_tutorial_files:header.htm") h2;
	mso-footer:url(":t_coffee_tutorial_files:header.htm") f2;
	mso-first-header:url(":t_coffee_tutorial_files:header.htm") fh2;
	mso-paper-source:0;}
div.WordSection2
	{page:WordSection2;}
 /* List Definitions */
@list l0
	{mso-list-id:-132;
	mso-list-type:simple;
	mso-list-template-ids:243462170;}
@list l0:level1
	{mso-level-style-link:"List Number 2";
	mso-level-tab-stop:90.0pt;
	mso-level-number-position:left;
	margin-left:90.0pt;
	text-indent:-18.0pt;}
@list l1
	{mso-list-id:-131;
	mso-list-type:simple;
	mso-list-template-ids:579796312;}
@list l1:level1
	{mso-level-style-link:"List Bullet 5";
	mso-level-tab-stop:72.0pt;
	mso-level-number-position:left;
	margin-left:72.0pt;
	text-indent:-18.0pt;}
@list l2
	{mso-list-id:-130;
	mso-list-type:simple;
	mso-list-template-ids:-525153058;}
@list l2:level1
	{mso-level-style-link:"List Bullet 4";
	mso-level-tab-stop:54.0pt;
	mso-level-number-position:left;
	margin-left:54.0pt;
	text-indent:-18.0pt;}
@list l3
	{mso-list-id:-129;
	mso-list-type:simple;
	mso-list-template-ids:1970563230;}
@list l3:level1
	{mso-level-style-link:"List Bullet 3";
	mso-level-tab-stop:36.0pt;
	mso-level-number-position:left;
	text-indent:-18.0pt;}
@list l4
	{mso-list-id:-128;
	mso-list-type:simple;
	mso-list-template-ids:-118980482;}
@list l4:level1
	{mso-level-number-format:bullet;
	mso-level-style-link:"List Bullet 2";
	mso-level-text:\F0B7;
	mso-level-tab-stop:90.0pt;
	mso-level-number-position:left;
	margin-left:90.0pt;
	text-indent:-18.0pt;
	font-family:Symbol;}
@list l5
	{mso-list-id:-127;
	mso-list-type:simple;
	mso-list-template-ids:-1465100272;}
@list l5:level1
	{mso-level-number-format:bullet;
	mso-level-style-link:"List 5";
	mso-level-text:\F0B7;
	mso-level-tab-stop:72.0pt;
	mso-level-number-position:left;
	margin-left:72.0pt;
	text-indent:-18.0pt;
	font-family:Symbol;}
@list l6
	{mso-list-id:-126;
	mso-list-type:simple;
	mso-list-template-ids:-368430812;}
@list l6:level1
	{mso-level-number-format:bullet;
	mso-level-style-link:"List 4";
	mso-level-text:\F0B7;
	mso-level-tab-stop:54.0pt;
	mso-level-number-position:left;
	margin-left:54.0pt;
	text-indent:-18.0pt;
	font-family:Symbol;}
@list l7
	{mso-list-id:-125;
	mso-list-type:simple;
	mso-list-template-ids:-706941758;}
@list l7:level1
	{mso-level-number-format:bullet;
	mso-level-style-link:"List 3";
	mso-level-text:\F0B7;
	mso-level-tab-stop:36.0pt;
	mso-level-number-position:left;
	text-indent:-18.0pt;
	font-family:Symbol;}
@list l8
	{mso-list-id:-120;
	mso-list-type:simple;
	mso-list-template-ids:1576020400;}
@list l8:level1
	{mso-level-style-link:"Plain Text";
	mso-level-tab-stop:18.0pt;
	mso-level-number-position:left;
	margin-left:18.0pt;
	text-indent:-18.0pt;}
@list l9
	{mso-list-id:-119;
	mso-list-type:simple;
	mso-list-template-ids:96769246;}
@list l9:level1
	{mso-level-number-format:bullet;
	mso-level-style-link:"List Char";
	mso-level-text:\F0B7;
	mso-level-tab-stop:18.0pt;
	mso-level-number-position:left;
	margin-left:18.0pt;
	text-indent:-18.0pt;
	font-family:Symbol;}
@list l10
	{mso-list-id:-5;
	mso-list-template-ids:-1557078836;}
@list l10:level1
	{mso-level-number-format:none;
	mso-level-suffix:none;
	mso-level-text:"";
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	margin-left:0cm;
	text-indent:0cm;}
@list l10:level2
	{mso-level-number-format:none;
	mso-level-text:"";
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	mso-level-legacy:yes;
	mso-level-legacy-indent:0cm;
	mso-level-legacy-space:0cm;
	margin-left:0cm;
	text-indent:0cm;}
@list l10:level3
	{mso-level-number-format:none;
	mso-level-style-link:"Heading 3";
	mso-level-text:"";
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	mso-level-legacy:yes;
	mso-level-legacy-indent:0cm;
	mso-level-legacy-space:0cm;
	margin-left:0cm;
	text-indent:0cm;}
@list l10:level4
	{mso-level-number-format:none;
	mso-level-text:"";
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	mso-level-legacy:yes;
	mso-level-legacy-indent:0cm;
	mso-level-legacy-space:0cm;
	margin-left:0cm;
	text-indent:0cm;
	mso-ansi-language:EN-GB;}
@list l10:level5
	{mso-level-start-at:0;
	mso-level-number-format:none;
	mso-level-text:"";
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	margin-left:0cm;
	text-indent:0cm;}
@list l10:level6
	{mso-level-start-at:0;
	mso-level-number-format:none;
	mso-level-text:"";
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	margin-left:0cm;
	text-indent:0cm;}
@list l10:level7
	{mso-level-start-at:0;
	mso-level-number-format:none;
	mso-level-text:"";
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	margin-left:0cm;
	text-indent:0cm;}
@list l10:level8
	{mso-level-start-at:0;
	mso-level-number-format:none;
	mso-level-text:"";
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	margin-left:0cm;
	text-indent:0cm;}
@list l10:level9
	{mso-level-start-at:0;
	mso-level-number-format:none;
	mso-level-text:"";
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	margin-left:0cm;
	text-indent:0cm;}
@list l11
	{mso-list-id:-2;
	mso-list-type:simple;
	mso-list-template-ids:-1;}
@list l11:level1
	{mso-level-start-at:0;
	mso-level-text:*;
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	margin-left:0cm;
	text-indent:0cm;}
@list l12
	{mso-list-id:31855817;
	mso-list-type:hybrid;
	mso-list-template-ids:1137623166 -1 -1 -1 -1 -1 -1 -1 -1 -1;}
@list l12:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0A7;
	mso-level-tab-stop:108.0pt;
	mso-level-number-position:left;
	margin-left:108.0pt;
	text-indent:-18.0pt;
	font-family:Wingdings;}
@list l12:level2
	{mso-level-number-format:bullet;
	mso-level-text:o;
	mso-level-tab-stop:144.0pt;
	mso-level-number-position:left;
	margin-left:144.0pt;
	text-indent:-18.0pt;
	font-family:"Courier New";
	mso-bidi-font-family:"Times New Roman";}
@list l12:level3
	{mso-level-number-format:bullet;
	mso-level-text:\F0A7;
	mso-level-tab-stop:180.0pt;
	mso-level-number-position:left;
	margin-left:180.0pt;
	text-indent:-18.0pt;
	font-family:Wingdings;}
@list l12:level4
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:216.0pt;
	mso-level-number-position:left;
	margin-left:216.0pt;
	text-indent:-18.0pt;
	font-family:Symbol;}
@list l12:level5
	{mso-level-number-format:bullet;
	mso-level-text:o;
	mso-level-tab-stop:252.0pt;
	mso-level-number-position:left;
	margin-left:252.0pt;
	text-indent:-18.0pt;
	font-family:"Courier New";
	mso-bidi-font-family:"Times New Roman";}
@list l12:level6
	{mso-level-number-format:bullet;
	mso-level-text:\F0A7;
	mso-level-tab-stop:288.0pt;
	mso-level-number-position:left;
	margin-left:288.0pt;
	text-indent:-18.0pt;
	font-family:Wingdings;}
@list l12:level7
	{mso-level-number-format:bullet;
	mso-level-text:\F0B7;
	mso-level-tab-stop:324.0pt;
	mso-level-number-position:left;
	margin-left:324.0pt;
	text-indent:-18.0pt;
	font-family:Symbol;}
@list l12:level8
	{mso-level-number-format:bullet;
	mso-level-text:o;
	mso-level-tab-stop:360.0pt;
	mso-level-number-position:left;
	margin-left:360.0pt;
	text-indent:-18.0pt;
	font-family:"Courier New";
	mso-bidi-font-family:"Times New Roman";}
@list l12:level9
	{mso-level-number-format:bullet;
	mso-level-text:\F0A7;
	mso-level-tab-stop:396.0pt;
	mso-level-number-position:left;
	margin-left:396.0pt;
	text-indent:-18.0pt;
	font-family:Wingdings;}
@list l13
	{mso-list-id:142738275;
	mso-list-type:hybrid;
	mso-list-template-ids:1833586492 -1819087664 67895321 67895323 67895311 67895321 67895323 67895311 67895321 67895323;}
@list l13:level1
	{mso-level-text:%1-;
	mso-level-tab-stop:108.3pt;
	mso-level-number-position:left;
	margin-left:108.3pt;
	text-indent:-37.5pt;}
@list l13:level2
	{mso-level-number-format:alpha-lower;
	mso-level-tab-stop:124.8pt;
	mso-level-number-position:left;
	margin-left:124.8pt;
	text-indent:-18.0pt;}
@list l13:level3
	{mso-level-number-format:roman-lower;
	mso-level-tab-stop:160.8pt;
	mso-level-number-position:right;
	margin-left:160.8pt;
	text-indent:-9.0pt;}
@list l13:level4
	{mso-level-tab-stop:196.8pt;
	mso-level-number-position:left;
	margin-left:196.8pt;
	text-indent:-18.0pt;}
@list l13:level5
	{mso-level-number-format:alpha-lower;
	mso-level-tab-stop:232.8pt;
	mso-level-number-position:left;
	margin-left:232.8pt;
	text-indent:-18.0pt;}
@list l13:level6
	{mso-level-number-format:roman-lower;
	mso-level-tab-stop:268.8pt;
	mso-level-number-position:right;
	margin-left:268.8pt;
	text-indent:-9.0pt;}
@list l13:level7
	{mso-level-tab-stop:304.8pt;
	mso-level-number-position:left;
	margin-left:304.8pt;
	text-indent:-18.0pt;}
@list l13:level8
	{mso-level-number-format:alpha-lower;
	mso-level-tab-stop:340.8pt;
	mso-level-number-position:left;
	margin-left:340.8pt;
	text-indent:-18.0pt;}
@list l13:level9
	{mso-level-number-format:roman-lower;
	mso-level-tab-stop:376.8pt;
	mso-level-number-position:right;
	margin-left:376.8pt;
	text-indent:-9.0pt;}
@list l14
	{mso-list-id:429860257;
	mso-list-type:simple;
	mso-list-template-ids:-1193904264;}
@list l14:level1
	{mso-level-tab-stop:none;
	mso-level-number-position:left;
	mso-level-legacy:yes;
	mso-level-legacy-indent:18.0pt;
	mso-level-legacy-space:0cm;
	text-indent:-18.0pt;
	mso-ansi-font-size:10.0pt;
	mso-ascii-font-family:"Arial Black";
	mso-hansi-font-family:"Arial Black";}
@list l15
	{mso-list-id:497695395;
	mso-list-type:hybrid;
	mso-list-template-ids:-1444274642 -1181029460 67895321 67895323 67895311 67895321 67895323 67895311 67895321 67895323;}
@list l15:level1
	{mso-level-text:%1-;
	mso-level-tab-stop:53.25pt;
	mso-level-number-position:left;
	margin-left:53.25pt;
	text-indent:-18.0pt;}
@list l15:level2
	{mso-level-number-format:alpha-lower;
	mso-level-tab-stop:89.25pt;
	mso-level-number-position:left;
	margin-left:89.25pt;
	text-indent:-18.0pt;}
@list l15:level3
	{mso-level-number-format:roman-lower;
	mso-level-tab-stop:125.25pt;
	mso-level-number-position:right;
	margin-left:125.25pt;
	text-indent:-9.0pt;}
@list l15:level4
	{mso-level-tab-stop:161.25pt;
	mso-level-number-position:left;
	margin-left:161.25pt;
	text-indent:-18.0pt;}
@list l15:level5
	{mso-level-number-format:alpha-lower;
	mso-level-tab-stop:197.25pt;
	mso-level-number-position:left;
	margin-left:197.25pt;
	text-indent:-18.0pt;}
@list l15:level6
	{mso-level-number-format:roman-lower;
	mso-level-tab-stop:233.25pt;
	mso-level-number-position:right;
	margin-left:233.25pt;
	text-indent:-9.0pt;}
@list l15:level7
	{mso-level-tab-stop:269.25pt;
	mso-level-number-position:left;
	margin-left:269.25pt;
	text-indent:-18.0pt;}
@list l15:level8
	{mso-level-number-format:alpha-lower;
	mso-level-tab-stop:305.25pt;
	mso-level-number-position:left;
	margin-left:305.25pt;
	text-indent:-18.0pt;}
@list l15:level9
	{mso-level-number-format:roman-lower;
	mso-level-tab-stop:341.25pt;
	mso-level-number-position:right;
	margin-left:341.25pt;
	text-indent:-9.0pt;}
@list l16
	{mso-list-id:671949508;
	mso-list-type:hybrid;
	mso-list-template-ids:1439100602 1099232202 67895321 67895323 67895311 67895321 67895323 67895311 67895321 67895323;}
@list l16:level1
	{mso-level-text:%1-;
	mso-level-tab-stop:53.4pt;
	mso-level-number-position:left;
	margin-left:53.4pt;
	text-indent:-18.0pt;}
@list l16:level2
	{mso-level-number-format:alpha-lower;
	mso-level-tab-stop:89.4pt;
	mso-level-number-position:left;
	margin-left:89.4pt;
	text-indent:-18.0pt;}
@list l16:level3
	{mso-level-number-format:roman-lower;
	mso-level-tab-stop:125.4pt;
	mso-level-number-position:right;
	margin-left:125.4pt;
	text-indent:-9.0pt;}
@list l16:level4
	{mso-level-tab-stop:161.4pt;
	mso-level-number-position:left;
	margin-left:161.4pt;
	text-indent:-18.0pt;}
@list l16:level5
	{mso-level-number-format:alpha-lower;
	mso-level-tab-stop:197.4pt;
	mso-level-number-position:left;
	margin-left:197.4pt;
	text-indent:-18.0pt;}
@list l16:level6
	{mso-level-number-format:roman-lower;
	mso-level-tab-stop:233.4pt;
	mso-level-number-position:right;
	margin-left:233.4pt;
	text-indent:-9.0pt;}
@list l16:level7
	{mso-level-tab-stop:269.4pt;
	mso-level-number-position:left;
	margin-left:269.4pt;
	text-indent:-18.0pt;}
@list l16:level8
	{mso-level-number-format:alpha-lower;
	mso-level-tab-stop:305.4pt;
	mso-level-number-position:left;
	margin-left:305.4pt;
	text-indent:-18.0pt;}
@list l16:level9
	{mso-level-number-format:roman-lower;
	mso-level-tab-stop:341.4pt;
	mso-level-number-position:right;
	margin-left:341.4pt;
	text-indent:-9.0pt;}
@list l17
	{mso-list-id:795760782;
	mso-list-type:hybrid;
	mso-list-template-ids:338989358 -1181029460 67895321 67895323 67895311 67895321 67895323 67895311 67895321 67895323;}
@list l17:level1
	{mso-level-text:%1-;
	mso-level-tab-stop:95.8pt;
	mso-level-number-position:left;
	margin-left:95.8pt;
	text-indent:-18.0pt;}
@list l17:level2
	{mso-level-number-format:alpha-lower;
	mso-level-tab-stop:114.55pt;
	mso-level-number-position:left;
	margin-left:114.55pt;
	text-indent:-18.0pt;}
@list l17:level3
	{mso-level-number-format:roman-lower;
	mso-level-tab-stop:150.55pt;
	mso-level-number-position:right;
	margin-left:150.55pt;
	text-indent:-9.0pt;}
@list l17:level4
	{mso-level-tab-stop:186.55pt;
	mso-level-number-position:left;
	margin-left:186.55pt;
	text-indent:-18.0pt;}
@list l17:level5
	{mso-level-number-format:alpha-lower;
	mso-level-tab-stop:222.55pt;
	mso-level-number-position:left;
	margin-left:222.55pt;
	text-indent:-18.0pt;}
@list l17:level6
	{mso-level-number-format:roman-lower;
	mso-level-tab-stop:258.55pt;
	mso-level-number-position:right;
	margin-left:258.55pt;
	text-indent:-9.0pt;}
@list l17:level7
	{mso-level-tab-stop:294.55pt;
	mso-level-number-position:left;
	margin-left:294.55pt;
	text-indent:-18.0pt;}
@list l17:level8
	{mso-level-number-format:alpha-lower;
	mso-level-tab-stop:330.55pt;
	mso-level-number-position:left;
	margin-left:330.55pt;
	text-indent:-18.0pt;}
@list l17:level9
	{mso-level-number-format:roman-lower;
	mso-level-tab-stop:366.55pt;
	mso-level-number-position:right;
	margin-left:366.55pt;
	text-indent:-9.0pt;}
@list l18
	{mso-list-id:873344605;
	mso-list-template-ids:1828727762;}
@list l18:level1
	{mso-level-number-format:bullet;
	mso-level-text:\F0A7;
	mso-level-tab-stop:18.0pt;
	mso-level-number-position:left;
	margin-left:18.0pt;
	text-indent:-18.0pt;
	font-family:Wingdings;}
@list l18:level2
	{mso-level-number-format:none;
	mso-level-text:"";
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	mso-level-legacy:yes;
	mso-level-legacy-indent:0cm;
	mso-level-legacy-space:0cm;
	margin-left:0cm;
	text-indent:0cm;}
@list l18:level3
	{mso-level-number-format:none;
	mso-level-text:"";
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	mso-level-legacy:yes;
	mso-level-legacy-indent:0cm;
	mso-level-legacy-space:0cm;
	margin-left:0cm;
	text-indent:0cm;}
@list l18:level4
	{mso-level-number-format:none;
	mso-level-text:"";
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	mso-level-legacy:yes;
	mso-level-legacy-indent:0cm;
	mso-level-legacy-space:0cm;
	margin-left:0cm;
	text-indent:0cm;}
@list l18:level5
	{mso-level-start-at:0;
	mso-level-number-format:none;
	mso-level-text:"";
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	margin-left:0cm;
	text-indent:0cm;}
@list l18:level6
	{mso-level-start-at:0;
	mso-level-number-format:none;
	mso-level-text:"";
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	margin-left:0cm;
	text-indent:0cm;}
@list l18:level7
	{mso-level-start-at:0;
	mso-level-number-format:none;
	mso-level-text:"";
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	margin-left:0cm;
	text-indent:0cm;}
@list l18:level8
	{mso-level-start-at:0;
	mso-level-number-format:none;
	mso-level-text:"";
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	margin-left:0cm;
	text-indent:0cm;}
@list l18:level9
	{mso-level-start-at:0;
	mso-level-number-format:none;
	mso-level-text:"";
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	margin-left:0cm;
	text-indent:0cm;}
@list l19
	{mso-list-id:1130830429;
	mso-list-type:hybrid;
	mso-list-template-ids:539937826 -1181029460 67895321 67895323 67895311 67895321 67895323 67895311 67895321 67895323;}
@list l19:level1
	{mso-level-text:%1-;
	mso-level-tab-stop:53.25pt;
	mso-level-number-position:left;
	margin-left:53.25pt;
	text-indent:-18.0pt;}
@list l19:level2
	{mso-level-number-format:alpha-lower;
	mso-level-tab-stop:89.25pt;
	mso-level-number-position:left;
	margin-left:89.25pt;
	text-indent:-18.0pt;}
@list l19:level3
	{mso-level-number-format:roman-lower;
	mso-level-tab-stop:125.25pt;
	mso-level-number-position:right;
	margin-left:125.25pt;
	text-indent:-9.0pt;}
@list l19:level4
	{mso-level-tab-stop:161.25pt;
	mso-level-number-position:left;
	margin-left:161.25pt;
	text-indent:-18.0pt;}
@list l19:level5
	{mso-level-number-format:alpha-lower;
	mso-level-tab-stop:197.25pt;
	mso-level-number-position:left;
	margin-left:197.25pt;
	text-indent:-18.0pt;}
@list l19:level6
	{mso-level-number-format:roman-lower;
	mso-level-tab-stop:233.25pt;
	mso-level-number-position:right;
	margin-left:233.25pt;
	text-indent:-9.0pt;}
@list l19:level7
	{mso-level-tab-stop:269.25pt;
	mso-level-number-position:left;
	margin-left:269.25pt;
	text-indent:-18.0pt;}
@list l19:level8
	{mso-level-number-format:alpha-lower;
	mso-level-tab-stop:305.25pt;
	mso-level-number-position:left;
	margin-left:305.25pt;
	text-indent:-18.0pt;}
@list l19:level9
	{mso-level-number-format:roman-lower;
	mso-level-tab-stop:341.25pt;
	mso-level-number-position:right;
	margin-left:341.25pt;
	text-indent:-9.0pt;}
@list l20
	{mso-list-id:1264142537;
	mso-list-type:hybrid;
	mso-list-template-ids:657350672 67895311 67895321 67895323 67895311 67895321 67895323 67895311 67895321 67895323;}
@list l20:level1
	{mso-level-tab-stop:108.0pt;
	mso-level-number-position:left;
	margin-left:108.0pt;
	text-indent:-18.0pt;}
@list l20:level2
	{mso-level-number-format:alpha-lower;
	mso-level-tab-stop:144.0pt;
	mso-level-number-position:left;
	margin-left:144.0pt;
	text-indent:-18.0pt;}
@list l20:level3
	{mso-level-number-format:roman-lower;
	mso-level-tab-stop:180.0pt;
	mso-level-number-position:right;
	margin-left:180.0pt;
	text-indent:-9.0pt;}
@list l20:level4
	{mso-level-tab-stop:216.0pt;
	mso-level-number-position:left;
	margin-left:216.0pt;
	text-indent:-18.0pt;}
@list l20:level5
	{mso-level-number-format:alpha-lower;
	mso-level-tab-stop:252.0pt;
	mso-level-number-position:left;
	margin-left:252.0pt;
	text-indent:-18.0pt;}
@list l20:level6
	{mso-level-number-format:roman-lower;
	mso-level-tab-stop:288.0pt;
	mso-level-number-position:right;
	margin-left:288.0pt;
	text-indent:-9.0pt;}
@list l20:level7
	{mso-level-tab-stop:324.0pt;
	mso-level-number-position:left;
	margin-left:324.0pt;
	text-indent:-18.0pt;}
@list l20:level8
	{mso-level-number-format:alpha-lower;
	mso-level-tab-stop:360.0pt;
	mso-level-number-position:left;
	margin-left:360.0pt;
	text-indent:-18.0pt;}
@list l20:level9
	{mso-level-number-format:roman-lower;
	mso-level-tab-stop:396.0pt;
	mso-level-number-position:right;
	margin-left:396.0pt;
	text-indent:-9.0pt;}
@list l21
	{mso-list-id:1280378370;
	mso-list-type:simple;
	mso-list-template-ids:-247183560;}
@list l21:level1
	{mso-level-number-format:bullet;
	mso-level-style-link:"List Bullet 2";
	mso-level-text:\F0A7;
	mso-level-tab-stop:18.0pt;
	mso-level-number-position:left;
	margin-left:18.0pt;
	text-indent:-18.0pt;
	font-family:Wingdings;}
@list l22
	{mso-list-id:1469779017;
	mso-list-type:simple;
	mso-list-template-ids:-1193904264;}
@list l22:level1
	{mso-level-tab-stop:none;
	mso-level-number-position:left;
	mso-level-legacy:yes;
	mso-level-legacy-indent:18.0pt;
	mso-level-legacy-space:0cm;
	text-indent:-18.0pt;
	mso-ansi-font-size:10.0pt;
	mso-ascii-font-family:"Arial Black";
	mso-hansi-font-family:"Arial Black";}
@list l23
	{mso-list-id:1598251408;
	mso-list-type:simple;
	mso-list-template-ids:-674336410;}
@list l23:level1
	{mso-level-number-format:bullet;
	mso-level-style-link:"List Char";
	mso-level-text:\F0A7;
	mso-level-tab-stop:18.0pt;
	mso-level-number-position:left;
	margin-left:18.0pt;
	text-indent:-18.0pt;
	font-family:Wingdings;}
@list l24
	{mso-list-id:1649675689;
	mso-list-type:simple;
	mso-list-template-ids:-1193904264;}
@list l24:level1
	{mso-level-tab-stop:none;
	mso-level-number-position:left;
	mso-level-legacy:yes;
	mso-level-legacy-indent:18.0pt;
	mso-level-legacy-space:0cm;
	text-indent:-18.0pt;
	mso-ansi-font-size:10.0pt;
	mso-ascii-font-family:"Arial Black";
	mso-hansi-font-family:"Arial Black";}
@list l11:level1 lfo19
	{mso-level-number-format:bullet;
	mso-level-numbering:continue;
	mso-level-text:\F031;
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	mso-level-legacy:yes;
	mso-level-legacy-indent:18.0pt;
	mso-level-legacy-space:0cm;
	margin-left:18.0pt;
	text-indent:-18.0pt;
	mso-ansi-font-size:11.0pt;
	mso-ascii-font-family:Wingdings;
	mso-hansi-font-family:Wingdings;}
@list l11:level1 lfo20
	{mso-level-number-format:bullet;
	mso-level-numbering:continue;
	mso-level-text:\F021;
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	mso-level-legacy:yes;
	mso-level-legacy-indent:18.0pt;
	mso-level-legacy-space:0cm;
	margin-left:18.0pt;
	text-indent:-18.0pt;
	mso-ansi-font-size:12.0pt;
	mso-ascii-font-family:Wingdings;
	mso-hansi-font-family:Wingdings;}
@list l11:level1 lfo21
	{mso-level-number-format:bullet;
	mso-level-numbering:continue;
	mso-level-text:\F03A;
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	mso-level-legacy:yes;
	mso-level-legacy-indent:18.0pt;
	mso-level-legacy-space:0cm;
	margin-left:18.0pt;
	text-indent:-18.0pt;
	mso-ansi-font-size:11.0pt;
	mso-ascii-font-family:Wingdings;
	mso-hansi-font-family:Wingdings;}
@list l11:level1 lfo22
	{mso-level-number-format:bullet;
	mso-level-numbering:continue;
	mso-level-text:\F026;
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	mso-level-legacy:yes;
	mso-level-legacy-indent:18.0pt;
	mso-level-legacy-space:0cm;
	margin-left:18.0pt;
	text-indent:-18.0pt;
	mso-ansi-font-size:10.0pt;
	mso-ascii-font-family:Wingdings;
	mso-hansi-font-family:Wingdings;}
@list l11:level1 lfo23
	{mso-level-number-format:bullet;
	mso-level-numbering:continue;
	mso-level-text:\F0B7;
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	mso-level-legacy:yes;
	mso-level-legacy-indent:18.0pt;
	mso-level-legacy-space:0cm;
	margin-left:0cm;
	text-indent:-18.0pt;
	mso-ansi-font-size:14.0pt;
	mso-ascii-font-family:Symbol;
	mso-hansi-font-family:Symbol;}
@list l11:level1 lfo24
	{mso-level-number-format:bullet;
	mso-level-numbering:continue;
	mso-level-text:\F0B7;
	mso-level-tab-stop:none;
	mso-level-number-position:left;
	mso-level-legacy:yes;
	mso-level-legacy-indent:18.0pt;
	mso-level-legacy-space:0cm;
	margin-left:0cm;
	text-indent:-18.0pt;
	mso-ansi-font-size:14.0pt;
	mso-ascii-font-family:Symbol;
	mso-hansi-font-family:Symbol;}
ol
	{margin-bottom:0cm;}
ul
	{margin-bottom:0cm;}
-->
</style>
<!--[if gte mso 10]>
<style>
 /* Style Definitions */
table.MsoNormalTable
	{mso-style-name:"Table Normal";
	mso-tstyle-rowband-size:0;
	mso-tstyle-colband-size:0;
	mso-style-noshow:yes;
	mso-style-unhide:no;
	mso-style-parent:"";
	mso-padding-alt:0cm 5.4pt 0cm 5.4pt;
	mso-para-margin:0cm;
	mso-para-margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:10.0pt;
	font-family:"Times New Roman";}
table.MsoTableColumns1
	{mso-style-name:"Table Columns 1";
	mso-tstyle-rowband-size:0;
	mso-tstyle-colband-size:1;
	mso-style-unhide:no;
	border:solid black 1.5pt;
	mso-padding-alt:0cm 5.4pt 0cm 5.4pt;
	mso-para-margin:0cm;
	mso-para-margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:10.0pt;
	font-family:"Times New Roman";
	font-weight:bold;}
table.MsoTableColumns1FirstRow
	{mso-style-name:"Table Columns 1";
	mso-table-condition:first-row;
	mso-style-unhide:no;
	mso-tstyle-border-bottom:2.25pt double black;
	mso-tstyle-diagonal-down:none;
	mso-tstyle-diagonal-up:none;
	mso-ansi-font-weight:normal;
	mso-bidi-font-weight:normal;}
table.MsoTableColumns1LastRow
	{mso-style-name:"Table Columns 1";
	mso-table-condition:last-row;
	mso-style-unhide:no;
	mso-tstyle-diagonal-down:none;
	mso-tstyle-diagonal-up:none;
	mso-ansi-font-weight:normal;
	mso-bidi-font-weight:normal;}
table.MsoTableColumns1FirstCol
	{mso-style-name:"Table Columns 1";
	mso-table-condition:first-column;
	mso-style-unhide:no;
	mso-tstyle-diagonal-down:none;
	mso-tstyle-diagonal-up:none;
	mso-ansi-font-weight:normal;
	mso-bidi-font-weight:normal;}
table.MsoTableColumns1LastCol
	{mso-style-name:"Table Columns 1";
	mso-table-condition:last-column;
	mso-style-unhide:no;
	mso-tstyle-diagonal-down:none;
	mso-tstyle-diagonal-up:none;
	mso-ansi-font-weight:normal;
	mso-bidi-font-weight:normal;}
table.MsoTableColumns1OddColumn
	{mso-style-name:"Table Columns 1";
	mso-table-condition:odd-column;
	mso-style-unhide:no;
	mso-tstyle-shading:white;
	mso-tstyle-pattern:gray-25 black;
	color:windowtext;}
table.MsoTableColumns1EvenColumn
	{mso-style-name:"Table Columns 1";
	mso-table-condition:even-column;
	mso-style-unhide:no;
	mso-tstyle-shading:white;
	mso-tstyle-pattern:gray-25 yellow;
	color:windowtext;}
table.MsoTableColumns1NECell
	{mso-style-name:"Table Columns 1";
	mso-table-condition:ne-cell;
	mso-style-unhide:no;
	mso-tstyle-diagonal-down:none;
	mso-tstyle-diagonal-up:none;
	mso-ansi-font-weight:bold;
	mso-bidi-font-weight:bold;}
table.MsoTableColumns1SWCell
	{mso-style-name:"Table Columns 1";
	mso-table-condition:sw-cell;
	mso-style-unhide:no;
	mso-tstyle-diagonal-down:none;
	mso-tstyle-diagonal-up:none;
	mso-ansi-font-weight:bold;
	mso-bidi-font-weight:bold;}
table.MsoTableColumns3
	{mso-style-name:"Table Columns 3";
	mso-tstyle-rowband-size:0;
	mso-tstyle-colband-size:1;
	mso-style-unhide:no;
	border:solid navy 1.0pt;
	mso-border-alt:solid navy .75pt;
	mso-padding-alt:0cm 5.4pt 0cm 5.4pt;
	mso-border-insidev:.75pt solid navy;
	mso-para-margin:0cm;
	mso-para-margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:10.0pt;
	font-family:"Times New Roman";
	font-weight:bold;}
table.MsoTableColumns3FirstRow
	{mso-style-name:"Table Columns 3";
	mso-table-condition:first-row;
	mso-style-unhide:no;
	mso-tstyle-shading:white;
	mso-tstyle-pattern:solid navy;
	mso-tstyle-diagonal-down:none;
	mso-tstyle-diagonal-up:none;
	color:white;}
table.MsoTableColumns3LastRow
	{mso-style-name:"Table Columns 3";
	mso-table-condition:last-row;
	mso-style-unhide:no;
	mso-tstyle-border-top:.75pt solid navy;
	mso-tstyle-diagonal-down:none;
	mso-tstyle-diagonal-up:none;
	mso-ansi-font-weight:normal;
	mso-bidi-font-weight:normal;}
table.MsoTableColumns3FirstCol
	{mso-style-name:"Table Columns 3";
	mso-table-condition:first-column;
	mso-style-unhide:no;
	mso-tstyle-diagonal-down:none;
	mso-tstyle-diagonal-up:none;
	mso-ansi-font-weight:normal;
	mso-bidi-font-weight:normal;}
table.MsoTableColumns3LastCol
	{mso-style-name:"Table Columns 3";
	mso-table-condition:last-column;
	mso-style-unhide:no;
	mso-tstyle-diagonal-down:none;
	mso-tstyle-diagonal-up:none;
	mso-ansi-font-weight:normal;
	mso-bidi-font-weight:normal;}
table.MsoTableColumns3OddColumn
	{mso-style-name:"Table Columns 3";
	mso-table-condition:odd-column;
	mso-style-unhide:no;
	mso-tstyle-shading:white;
	mso-tstyle-pattern:solid silver;
	color:windowtext;}
table.MsoTableColumns3EvenColumn
	{mso-style-name:"Table Columns 3";
	mso-table-condition:even-column;
	mso-style-unhide:no;
	mso-tstyle-shading:white;
	mso-tstyle-pattern:gray-10 black;
	color:windowtext;}
table.MsoTableColumns3NECell
	{mso-style-name:"Table Columns 3";
	mso-table-condition:ne-cell;
	mso-style-unhide:no;
	mso-tstyle-diagonal-down:none;
	mso-tstyle-diagonal-up:none;
	mso-ansi-font-weight:bold;
	mso-bidi-font-weight:bold;}
table.MsoTableElegant
	{mso-style-name:"Table Elegant";
	mso-tstyle-rowband-size:0;
	mso-tstyle-colband-size:0;
	mso-style-unhide:no;
	border:double black 2.25pt;
	mso-padding-alt:0cm 5.4pt 0cm 5.4pt;
	mso-border-insideh:.75pt solid black;
	mso-border-insidev:.75pt solid black;
	mso-para-margin:0cm;
	mso-para-margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:10.0pt;
	font-family:"Times New Roman";}
table.MsoTableElegantFirstRow
	{mso-style-name:"Table Elegant";
	mso-table-condition:first-row;
	mso-style-unhide:no;
	mso-tstyle-diagonal-down:none;
	mso-tstyle-diagonal-up:none;
	color:windowtext;
	text-transform:uppercase;}
table.MsoTableGrid
	{mso-style-name:"Table Grid";
	mso-tstyle-rowband-size:0;
	mso-tstyle-colband-size:0;
	mso-style-unhide:no;
	border:solid windowtext 1.0pt;
	mso-border-alt:solid windowtext .5pt;
	mso-padding-alt:0cm 5.4pt 0cm 5.4pt;
	mso-border-insideh:.5pt solid windowtext;
	mso-border-insidev:.5pt solid windowtext;
	mso-para-margin:0cm;
	mso-para-margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:10.0pt;
	font-family:"Times New Roman";}
</style>
<![endif]--><!--[if gte mso 9]><xml>
 <o:shapedefaults v:ext="edit" spidmax="2050"/>
</xml><![endif]--><!--[if gte mso 9]><xml>
 <o:shapelayout v:ext="edit">
  <o:idmap v:ext="edit" data="1"/>
 </o:shapelayout></xml><![endif]-->
</head>

<body lang=EN-US link=blue vlink=purple style='tab-interval:35.4pt'>

<div class=WordSection1>

<div style='mso-element:frame;mso-element-wrap:no-wrap-beside;mso-height-rule:
exactly'>

<table cellspacing=0 cellpadding=0 hspace=0 vspace=0 width=112 align=right>
 <tr>
  <td valign=top align=left style='padding-top:9.35pt;padding-right:9.35pt;
  padding-bottom:9.35pt;padding-left:9.35pt'>
  <p class=PartTitle><span class=GramE><span lang=EN-GB>.T</span></span><span
  lang=EN-GB>-coffee Tutorial</span></p>
  </td>
 </tr>
</table>

</div>

<![if !supportTextWrap]><br clear=ALL>
<![endif]>

<div style='mso-element:para-border-div;border:solid white 1.0pt;mso-border-alt:
solid white .75pt;padding:31.0pt 31.0pt 31.0pt 31.0pt;background:#E5E5E5;
mso-shading:windowtext;mso-pattern:gray-10 auto'>

<p class=TitleCover style='margin:0cm;margin-bottom:.0001pt;background:#E5E5E5;
mso-shading:windowtext;mso-pattern:gray-10 auto'><a name="_Toc235985429"><span
lang=EN-GB>T-Coffee Cheat Sheet </span></a></p>

</div>

<h1><span style='mso-bookmark:_Toc235985429'><span lang=EN-GB>Proteins</span></span></h1>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6'>

<p class=examplefileCxSpFirst><span lang=EN-GB>Mode<span style='mso-tab-count:
3'>                 </span>Command</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>============================================================================</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>Very <span class=GramE>Fast</span><span
style='mso-tab-count:2'>            </span>t_coffee sample_aln1.fasta -mode <span
class=SpellE>quickaln</span></span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB><span
style="mso-spacerun:yes">                </span><span style='mso-tab-count:
1'>     </span><span class=GramE>lower</span> -<span class=SpellE>ndiag</span>
if the sequences are very similar </span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>Regular<span style='mso-tab-count:
2'>              </span>t_coffee sample_aln1.fasta</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB><span
style="mso-spacerun:yes">    </span><span style='mso-tab-count:3'>                 </span><span
class=GramE>use</span> the output.html to estimate the MSA accuracy</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>Very Accurate<span
style='mso-tab-count:1'> </span> <span style='mso-tab-count:1'>      </span>t_coffee
sample_aln1.fasta -mode accurate</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB><span
style="mso-spacerun:yes">          </span><span style='mso-tab-count:1'>    </span><span
style='mso-tab-count:1'>       </span><span class=GramE>slow</span>, combines
structures, sequences and profiles</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>M-Coffee<span style='mso-tab-count:
2'>             </span>t_coffee sample_aln1.fasta -mode <span class=SpellE>mcoffee</span></span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB><span style='mso-tab-count:
3'>                     </span><span class=GramE>combines</span> most of the
existing MSA packages</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>3D-Coffee<span
style='mso-tab-count:2'>            </span>t_coffee sample_aln1.fasta -mode 3dcoffee</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB><span style='mso-tab-count:
3'>                     </span><span class=GramE>uses</span> the structure of
your sequences if named with PDBID</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>Expresso<span style='mso-tab-count:
2'>             </span>t_coffee sample_aln1.fasta -mode <span class=SpellE>expresso</span></span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB><span style='mso-tab-count:
3'>                     </span><span class=GramE>finds</span> structures
homologous to your sequences</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>PSI-Coffee<span
style='mso-tab-count:2'>           </span>t_coffee sample_aln1.fasta -mode
psicoffee</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB><span style='mso-tab-count:
3'>                     </span><span class=GramE>enriches</span> your sequence
with profile information</span></p>

<p class=examplefileCxSpLast><span lang=EN-GB>----------------------------------------------------------------------------</span></p>

</div>

<h1><a name="_Toc235985430"><span lang=EN-GB>DNA</span></a></h1>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6'>

<p class=examplefile><span lang=EN-GB>R-Coffee<span style='mso-tab-count:2'>             </span>t_coffee
<span class=SpellE>three_cdna.fasta</span> -mode <span class=SpellE>cdna</span></span></p>

</div>

<h1><a name="_Toc235985431"><span lang=EN-GB>RNA</span></a></h1>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6'>

<p class=examplefileCxSpFirst><span lang=EN-GB>Mode<span style='mso-tab-count:
3'>                 </span>Command</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>============================================================================</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>R-Coffee<span style='mso-tab-count:
2'>             </span>t_coffee sample_rnaseq1.fasta -mode <span class=SpellE>rcoffee</span></span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB><span
style="mso-spacerun:yes">                </span><span style='mso-tab-count:
1'>     </span><span class=GramE>use</span> the predicted secondary structure
of your sequences</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>RM-Coffee<span
style='mso-tab-count:2'>            </span>t_coffee sample_rnaseq1.fasta -mode <span
class=SpellE>rmcoffee</span></span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB><span
style="mso-spacerun:yes">                </span><span style='mso-tab-count:
1'>     </span><span class=GramE>use</span> M-Coffee + secondary structure
prediction</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>R-Coffee <span class=SpellE>Consan</span><span
style='mso-tab-count:1'>      </span>t_coffee sample_rnaseq1.fasta -mode <span
class=SpellE>rcoffee_consan</span></span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB><span
style="mso-spacerun:yes">   </span><span style='mso-tab-count:1'>    </span><span
style="mso-spacerun:yes">   </span><span style='mso-tab-count:1'>    </span><span
style="mso-spacerun:yes"> </span><span style='mso-tab-count:1'>      </span><span
class=GramE>use</span> <span class=SpellE>rcoffee</span> to combine <span
class=SpellE>consan</span> alignments. Accurate and Slow</span></p>

<p class=examplefileCxSpLast><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

</div>

<h1><a name="_Toc235985432"><span lang=EN-GB>Memory Problems</span></a></h1>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6'>

<p class=examplefile><span class=GramE><span lang=EN-GB>memory</span></span><span
lang=EN-GB><span style='mso-tab-count:2'>        </span><span style='mso-tab-count:
1'>       </span>t_coffee sample_aln1.fasta -mode memory</span></p>

</div>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<div style='mso-element:para-border-div;border:solid white 1.0pt;mso-border-alt:
solid white .75pt;padding:31.0pt 31.0pt 31.0pt 31.0pt;background:#E5E5E5;
mso-shading:windowtext;mso-pattern:gray-10 auto'>

<p class=TitleCover style='margin:0cm;margin-bottom:.0001pt;background:#E5E5E5;
mso-shading:windowtext;mso-pattern:gray-10 auto'><a name="_Toc235985434"><span
lang=EN-GB>T-Coffee:</span></a></p>

<p class=TitleCover style='margin:0cm;margin-bottom:.0001pt;background:#E5E5E5;
mso-shading:windowtext;mso-pattern:gray-10 auto'><span style='mso-bookmark:
_Toc235985434'><span lang=EN-GB>Tutorial and FAQ</span></span></p>

</div>

<div style='mso-element:para-border-div;border:none;border-top:solid windowtext 1.0pt;
mso-border-top-alt:solid windowtext .75pt;padding:1.0pt 0cm 0cm 0cm'>

<p class=SubtitleCover><span style='mso-bookmark:_Toc235985434'><span
lang=EN-GB><o:p>&nbsp;</o:p></span></span></p>

</div>

<h1><span style='mso-bookmark:_Toc235985434'><span lang=EN-GB>Before You Start</span></span></h1>

<h2><span style='mso-bookmark:_Toc235985434'><span lang=EN-GB>Foreword</span></span></h2>

<p class=MsoBodyText><span lang=EN-GB>A lot of the stuff presented here
emanates form two summer schools that were tentatively called the &quot;<span
class=SpellE>Prosite</span> Workshops&quot; and were held in Marseille, in 2001
and 2002. These workshops were mostly an excuse to go rambling and swimming in
the <span class=SpellE>callanques</span>. Yet, when we got tired of lazing in
the sun, we eventually did a bit of work to chill out. Most of our experiments
were revolving around the development of sequence analysis tools. Many of the
most advanced ideas in T-Coffee were launched during these fruitful sessions.
Participants included Phillip Bucher, Laurent <span class=SpellE>Falquet</span>,
Marco <span class=SpellE>Pagni</span>, <span class=SpellE>Alexandre</span> <span
class=SpellE>Gattiker</span>, Nicolas <span class=SpellE>Hulo</span>, Christian
Siegfried, Anne-<span class=SpellE>Lise</span> <span class=SpellE>Veuthey</span>,
<span class=SpellE>Virginie</span> <span class=SpellE>Leseau</span>, Lorenzo <span
class=SpellE>Ceruti</span> and Cedric Notredame.</span></p>

<p class=MsoBodyText><span lang=EN-GB>This Document contains two main sections.
The first one is a tutorial, where we go from simple things to more complicated
and show you how to use all the subtleties of T-Coffee. We have tried to put as
many of these functionalities on the web (www.tcoffee.org) but if you need to
do something special and highly reproducible, the Command Line is the only way.
<span style="mso-spacerun:yes"> </span></span></p>

<h2><a name="_Toc235985435"><span lang=EN-GB>Pre-Requisite</span></a></h2>

<p class=MsoBodyText><span lang=EN-GB>This tutorial relies on the assumption
that you have installed T-Coffee, version 6.18 or higher.</span></p>

<p class=MsoBodyText><span lang=EN-GB>T-Coffee is a freeware open source
running on all Unix-like platforms, including MAC-<span class=SpellE>osX</span>
and Cygwin. All the relevant information for installing T-Coffee is contained
in the Technical Documentation (tcoffee_technical.doc in the doc directory.) </span></p>

<p class=MsoBodyText><span lang=EN-GB>T-Coffee cannot run on the Microsoft Windows
shell. If you need to run T -Coffee on windows, start by installing <span
class=SpellE>cygwin</span> (www.cygwin.com). Cygwin is a freeware open source
that makes it possible to run a <span class=SpellE>unix</span>-like command
line on your Microsoft Windows PC without having to reboot. Cygwin is free of
charge and very easy to install. Yet, as the first installation requires
downloading substantial amounts of data, you should make sure you have access
to a <span class=GramE>broad-band</span> connection.</span></p>

<p class=MsoBodyText><span lang=EN-GB>In the course of this tutorial, we expect
you to use a <span class=SpellE>unix</span>-like command line shell. If you
work on Cygwin, this means clicking on the <span class=SpellE>cygwin</span>
icon and typing commands in the window that appears. If you don't want to
bother with command line stuff, try using the online <span class=SpellE>tcoffee</span>
webserver at: <b style='mso-bidi-font-weight:normal'><u>www.tcoffee.org<o:p></o:p></u></b></span></p>

<h2><a name="_Toc235985436"><span lang=EN-GB>Getting the Example Files of the
Tutorial</span></a></h2>

<p class=MsoBodyText><span lang=EN-GB>We encourage you to try all the following
examples with your own sequences/structures. If you want to try with ours, you
can get the material from the example directory of the distribution. If you do
not know where this file leaves or if you do not have access to it, the
simplest thing to do is to:</span></p>

<p class=MsoBodyText style='margin-left:53.25pt;text-indent:-18.0pt;mso-list:
l19 level1 lfo36;tab-stops:list 53.25pt'><![if !supportLists]><span lang=EN-GB><span
style='mso-list:Ignore'>1-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><span class=GramE><span lang=EN-GB>download</span></span><span
lang=EN-GB> T-Coffee's latest version from www.tcoffee.org (Follow the link to
the T-Coffee Home Page)</span></p>

<p class=MsoBodyText style='margin-left:53.25pt;text-indent:-18.0pt;mso-list:
l19 level1 lfo36;tab-stops:list 53.25pt'><![if !supportLists]><span lang=EN-GB><span
style='mso-list:Ignore'>2-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><span lang=EN-GB>Download the latest
distribution</span></p>

<p class=MsoBodyText style='margin-left:53.25pt;text-indent:-18.0pt;mso-list:
l19 level1 lfo36;tab-stops:list 53.25pt'><![if !supportLists]><span lang=EN-GB><span
style='mso-list:Ignore'>3-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><span class=SpellE><span class=GramE><span
lang=EN-GB>gunzip</span></span></span><span lang=EN-GB> &lt;<span class=SpellE>distrib</span>&gt;.tar.gz</span></p>

<p class=MsoBodyText style='margin-left:53.25pt;text-indent:-18.0pt;mso-list:
l19 level1 lfo36;tab-stops:list 53.25pt'><![if !supportLists]><span lang=EN-GB><span
style='mso-list:Ignore'>4-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><span class=GramE><span lang=EN-GB>tar</span></span><span
lang=EN-GB> -<span class=SpellE>xvf</span> &lt;<span class=SpellE>distrib</span>&gt;.tar</span></p>

<p class=MsoBodyText style='margin-left:53.25pt;text-indent:-18.0pt;mso-list:
l19 level1 lfo36;tab-stops:list 53.25pt'><![if !supportLists]><span lang=EN-GB><span
style='mso-list:Ignore'>5-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]><span class=GramE><span lang=EN-GB>go</span></span><span
lang=EN-GB> into &lt;<span class=SpellE>distrib</span>&gt;/example</span></p>

<p class=MsoBodyText><span lang=EN-GB>This is all you need to do to run ALL the
examples provided in this tutorial.</span></p>

<p class=MyNormal><a name="_Toc135066740"><o:p>&nbsp;</o:p></a></p>

<h1><span style='mso-bookmark:_Toc135066740'><a name="_Toc235985437"><span
lang=EN-GB>What Is <br>
T-COFFEE<br>
?</span></a></span></h1>

<h2><a name="_Toc235985438"></a><a name="_Toc135066741"></a><a
name="_Toc102494660"></a><a name="_Toc102494281"></a><a name="_Toc101065191"><span
style='mso-bookmark:_Toc102494281'><span style='mso-bookmark:_Toc102494660'><span
style='mso-bookmark:_Toc135066741'><span style='mso-bookmark:_Toc235985438'><span
lang=EN-GB>What is T-Coffee?</span></span></span></span></span></a></h2>

<p class=MyNormal>Before going deep into the core of the matter, here are a few
words to quickly explain some of the things T-Coffee will do for you.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985439"></a><a
name="_Toc135066742"></a><a name="_Toc102494661"></a><a name="_Toc102494282"></a><a
name="_Toc101065192"><span style='mso-bookmark:_Toc102494282'><span
style='mso-bookmark:_Toc102494661'><span style='mso-bookmark:_Toc135066742'><span
style='mso-bookmark:_Toc235985439'>What does it do?</span></span></span></span></a></h3>

<p class=MyNormal>T-Coffee is a multiple sequence alignment program: given a
set of sequences previously gathered using database search programs like BLAST,
FASTA or Smith and Waterman, T-Coffee will produce a multiple sequence
alignment. <b style='mso-bidi-font-weight:normal'>To use T-Coffee you must
already have your sequences ready</b>.</p>

<p class=MyNormal>T-Coffee can also be used to compare alignments, reformat
them or evaluate them using structural information, it is the mode known as
seq_reformat.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985440"></a><a
name="_Toc135066743"></a><a name="_Toc102494662"></a><a name="_Toc102494283"></a><a
name="_Toc101065193"><span style='mso-bookmark:_Toc102494283'><span
style='mso-bookmark:_Toc102494662'><span style='mso-bookmark:_Toc135066743'><span
style='mso-bookmark:_Toc235985440'>What can it align?</span></span></span></span></a></h3>

<p class=MyNormal>T-Coffee will align nucleic and protein sequences alike. It
will be able to use structural information for protein sequences with a known
structure or the RNA sequences. On a new PC mid of the range, T-Coffee will
align up to a 100 sequences, about 1000 amino acid long. </p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985441"></a><a
name="_Toc135066744"></a><a name="_Toc102494663"></a><a name="_Toc102494284"></a><a
name="_Toc101065194"><span style='mso-bookmark:_Toc102494284'><span
style='mso-bookmark:_Toc102494663'><span style='mso-bookmark:_Toc135066744'><span
style='mso-bookmark:_Toc235985441'>How can I use it?</span></span></span></span></a></h3>

<p class=MyNormal>T-Coffee is not an interactive program. It runs from your
UNIX or Linux command line and you must provide it with the correct parameters.
If you do not like typing commands, here is the simplest available mode where
T-Coffee only needs the name of the sequence file:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU><span
style='mso-tab-count:2'>          </span>PROMPT: t_coffee sample_seq1.fasta </span></p>

</div>

<p class=MyNormal>Installing and using T-Coffee requires a minimum acquaintance
with the Linux/Unix operating system. If you feel this is beyond your computer
skills, we suggest you use one of the available online servers.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985442">Is
There an Online Server</a></h3>

<p class=MyNormal>Yes, at www.tcoffee.org</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985443"></a><a
name="_Toc135066745"></a><a name="_Toc102494664"></a><a name="_Toc102494285"></a><a
name="_Toc101065195"><span style='mso-bookmark:_Toc102494285'><span
style='mso-bookmark:_Toc102494664'><span style='mso-bookmark:_Toc135066745'><span
style='mso-bookmark:_Toc235985443'>Is T-Coffee different from ClustalW?</span></span></span></span></a></h3>

<p class=MyNormal>According to several benchmarks, T-Coffee appears to be more
accurate than ClustalW. Yet, this increased accuracy comes at a price: T-Coffee
is slower than Clustal (about N times fro N Sequences).</p>

<p class=MyNormal>If you are familiar with ClustalW, or if you run a ClustalW
server, you will find that we have made some efforts to ensure as much
compatibility as possible between ClustalW and T-COFFEE. Whenever it was
relevant, we have kept the flag names and the flag syntax of ClustalW. Yet, you
will find that T-Coffee also has many extra possibilities </p>

<p class=MyNormal>If you want to align closely related sequences, T-Coffee can
also be used in a fast mode, much faster than ClustalW, and about as accurate:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU><span
style='mso-tab-count:2'>          </span>PROMPT: t_coffee sample_seq1.fasta -mode
quickaln </span></p>

</div>

<p class=MyNormal>This mode works by only considering the best diagonals
between two sequences. By default all the diagonals having a substitution score
&gt;0 are considered, but you can lower this by specifying:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU><span
style='mso-tab-count:2'>          </span>PROMPT: t_coffee sample_seq1.fasta -mode
quickaln -ndiag=10</span></p>

</div>

<p class=MyNormal>That will only consider the top 10 diagonals. This will be
very useful if you have long and very similar sequences to align (DNA for
instance).</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985444">Is
T-Coffee very accurate?</a></h3>

<p class=MyNormal>T-Coffee combines methods, and can be made as accurate (and
hopefully more) as the methods it combines. If you need a very accurate
alignment (and you have the full package installed with SOAP);</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU><span
style='mso-tab-count:1'>   </span>PROMPT: t_coffee sample_seq1.fasta -mode
accurate</span></p>

</div>

<p class=MyNormal>If you cannot run this job, go to the first section of the
technical manual (Installing BLAST for T-Coffee). You don't necessary need to
install BLAST locally but you must have access to a remote server (EBI or
NCBI).</p>

<p class=MyNormal>This mode is very slow but also very accurate. On average
this mode is about 10 % more accurate than normal aligners on sequences less
than 30% similar. If you want something faster:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;text-indent:14.1pt;background:#FFFFCC'><span
lang=ZU>t_coffee sample_seq1.fasta</span></p>

</div>

<p class=MyNormal>This is the normal mode. It is one of the most accurate of
its kind, roughly like <span class=SpellE>Probcons</span>. </p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span></p>

<h2><a name="_Toc235985445"></a><a name="_Toc135066746"></a><a
name="_Toc102494665"></a><a name="_Toc102494286"></a><a name="_Toc101065196"><span
style='mso-bookmark:_Toc102494286'><span style='mso-bookmark:_Toc102494665'><span
style='mso-bookmark:_Toc135066746'><span style='mso-bookmark:_Toc235985445'><span
lang=EN-GB>What T-Coffee Can and </span></span></span></span></span></a><span
class=GramE><span style='mso-bookmark:_Toc101065196'><span style='mso-bookmark:
_Toc102494286'><span style='mso-bookmark:_Toc102494665'><span style='mso-bookmark:
_Toc135066746'><span style='mso-bookmark:_Toc235985445'><span lang=EN-GB>Cannot</span></span></span></span></span></span></span><span
style='mso-bookmark:_Toc101065196'><span style='mso-bookmark:_Toc102494286'><span
style='mso-bookmark:_Toc102494665'><span style='mso-bookmark:_Toc135066746'><span
style='mso-bookmark:_Toc235985445'><span lang=EN-GB> do for you </span></span></span></span></span></span><span
style='mso-bookmark:_Toc135066746'><span style='mso-bookmark:_Toc235985445'><span
lang=EN-GB></span></span></span></h2>

<div style='mso-element:para-border-div;border:solid windowtext 1.5pt;
padding:0cm 0cm 0cm 0cm;background:#F7CBB7'>

<p class=Warning><span lang=EN-GB>IMPORTANT: All the files mentioned here (<span
class=SpellE>sample_seq</span><span class=GramE>..</span>.) can be found in the
example directory of the distribution.</span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985446"></a><a
name="_Toc135066747"></a><a name="_Toc102494666"></a><a name="_Toc102494287"></a><a
name="_Toc101065197"><span style='mso-bookmark:_Toc102494287'><span
style='mso-bookmark:_Toc102494666'><span style='mso-bookmark:_Toc135066747'><span
style='mso-bookmark:_Toc235985446'>(NOT) Fetching Sequences</span></span></span></span></a>
</h3>

<p class=MyNormal>T-Coffee will <span style='mso-bidi-font-weight:bold'>NOT</span>
fetch sequences for you: you must select the sequences you want to align before
hand. We suggest you use any BLAST server and format your sequences in FASTA so
that T-COFFEE can use them easily. The <span class=SpellE>expasy</span> BLAST
server (www.expasy.ch) provides a nice interface for integrating database
searches.</p>

<p class=MyNormal>Yet, the new modes of </p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985447"></a><a
name="_Toc135066748"></a><a name="_Toc102494667"></a><a name="_Toc102494288"></a><a
name="_Toc101065198"><span style='mso-bookmark:_Toc102494288'><span
style='mso-bookmark:_Toc102494667'><span style='mso-bookmark:_Toc135066748'><span
style='mso-bookmark:_Toc235985447'>Aligning Sequences</span></span></span></span></a></h3>

<p class=MyNormal>T-Coffee will compute (or at least try to compute!) accurate
multiple alignments of DNA, RNA or Protein sequences.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985448"></a><a
name="_Toc135066749"></a><a name="_Toc102494668"></a><a name="_Toc102494289"></a><a
name="_Toc101065199"><span style='mso-bookmark:_Toc102494289'><span
style='mso-bookmark:_Toc102494668'><span style='mso-bookmark:_Toc135066749'><span
style='mso-bookmark:_Toc235985448'>Combining Alignments</span></span></span></span></a></h3>

<p class=MyNormal>T-Coffee allows you to combine results obtained with several
alignment methods. For instance if you have an alignment coming from ClustalW,
an other alignment coming from <span class=SpellE>Dialign</span>, and a
structural alignment of some of your sequences, T-Coffee will combine all that
information and produce a new multiple sequence alignment having the best
agreement with all these methods (see the FAQ for more details)</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee &#8211;aln=sproteases_small.cw_aln, sproteases_small.muscle,
sproteases_small.tc_aln &#8211;outfile=combined_aln.aln <a name="_Toc101065200"></a></span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><span style='mso-bookmark:
_Toc101065200'><a name="_Toc235985449"></a><a name="_Toc135066750"></a><a
name="_Toc102494669"></a><a name="_Toc102494290"><span style='mso-bookmark:
_Toc102494669'><span style='mso-bookmark:_Toc135066750'><span style='mso-bookmark:
_Toc235985449'>Evaluating Alignments</span></span></span></a></span></h3>

<p class=MyNormal>You can use T-Coffee to measure the reliability of your
Multiple Sequence alignment. If you want to find out about that, read the FAQ
or the documentation for the -output flag.</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=FR
style='mso-ansi-language:FR'>PROMPT: </span><span lang=ZU>t_coffee
&#8211;infile=sproteases_small.aln &#8211;special_mode=evaluate </span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985450"></a><a
name="_Toc135066751"></a><a name="_Toc102494670"></a><a name="_Toc102494291"></a><a
name="_Toc101065201"><span style='mso-bookmark:_Toc102494291'><span
style='mso-bookmark:_Toc102494670'><span style='mso-bookmark:_Toc135066751'><span
style='mso-bookmark:_Toc235985450'>Combining Sequences and Structures</span></span></span></span></a></h3>

<p class=MyNormal>One of the latest improvements of T-Coffee is to let you
combine sequences and structures, so that your alignments are of higher
quality. You need to have sap package installed to fully benefit of this
facility. If you have the EBI BLAST client installed (see installation
procedure), you can run the following:</p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee 3d.fasta &#8211;special_mode=expresso</span></p>

</div>

<p class=MyNormal>BLAST will identify the best PDB target for each <span
class=GramE>sequences</span>, and T-Coffee will use sap (or any other
structural package) to align your structures and your sequences. If you do not
have BLAST installed, or if you want to specify the templates yourself, you can
use 3D-Coffee:</p>

<p class=MyNormal style='margin-left:0cm'><o:p>&nbsp;</o:p></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee 3d.fasta &#8211;special_mode=3dcoffee</span></p>

</div>

<p class=MyNormal>In this case, the sequences must be names according to their
PDB targets. <span style="mso-spacerun:yes"> </span>All these <span
class=GramE>network based</span> operations are carried out using <span
class=SpellE>wget</span>. If <span class=SpellE>wget</span> is not installed on
your system, you can get it for free from (www.wget.org). To make sure <span
class=SpellE>wget</span> is installed on your system, type</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
which wget <a name="_Toc101065202"></a></span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><span style='mso-bookmark:
_Toc101065202'><a name="_Toc235985451"></a><a name="_Toc135066752"></a><a
name="_Toc102494671"></a><a name="_Toc102494292"><span style='mso-bookmark:
_Toc102494671'><span style='mso-bookmark:_Toc135066752'><span style='mso-bookmark:
_Toc235985451'>Identifying Occurrences of a Motif: </span></span></span></a><span
class=SpellE><span style='mso-bookmark:_Toc102494292'><span style='mso-bookmark:
_Toc102494671'><span style='mso-bookmark:_Toc135066752'><span style='mso-bookmark:
_Toc235985451'>Mocca</span></span></span></span></span><span style='mso-bookmark:
_Toc235985451'></span><span style='mso-bookmark:_Toc135066752'></span><span
style='mso-bookmark:_Toc102494671'></span><span style='mso-bookmark:_Toc102494292'></span></span></h3>

<p class=MyNormal><span class=SpellE>Mocca</span> is a special mode of T-Coffee
that allows you to extract a series of repeats from a single sequence or a set
of sequences. In other words, if you know the coordinates of one copy of a
repeat, you can extract all the other occurrences. If you want to use <span
class=SpellE>Mocca</span>, simply type:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee &#8211;other_pg mocca sample_seq1.fasta </span></p>

</div>

<p class=MyNormal>The program needs some time to compute a library and it will
then prompt you with an interactive menu. Follow the instructions.</p>

<h2><a name="_Toc235985452"></a><a name="_Toc135066753"></a><a
name="_Toc102494672"></a><a name="_Toc102494293"></a><a name="_Toc101065203"><span
style='mso-bookmark:_Toc102494293'><span style='mso-bookmark:_Toc102494672'><span
style='mso-bookmark:_Toc135066753'><span style='mso-bookmark:_Toc235985452'><span
lang=EN-GB>How Does T-Coffee works</span></span></span></span></span></a></h2>

<p class=MyNormal>If you only want to make a standard multiple alignments, you
may skip these explanations. But if you want to do more sophisticated things,
these few indications may help before you start reading the doc and the papers.</p>

<p class=MyNormal>When you run T-Coffee, the first thing it does is to compute
a library. The library is a list of pairs of residues that <i>could</i> be
aligned. It is like a Xmas list: you can ask anything you fancy, but it is down
to Santa to assemble a collection of Toys that won't get him stuck at the
airport, while going through the metal detector.</p>

<p class=MyNormal>Given a standard library, it is not possible to have all the
residues aligned at the same time because all the lines of the library may not
agree. For instance, line 1 may say </p>

<p class=example2>Residue 1 of <span class=SpellE>seq</span> A with Residue 5
of <span class=SpellE>seq</span> B,</p>

<p class=MyNormal><span class=GramE>and</span> line 100 may say </p>

<p class=example2>Residue 1 of <span class=SpellE>seq</span> A with Residue 29
of <span class=SpellE>seq</span> B, </p>

<p class=MyNormal>Each of these constraints comes with a weight and in the end,
the T-Coffee algorithm tries to generate the multiple alignment that contains
constraints whose sum of weights yields the highest score. In other words, it
tries to make happy as many constraints as possible (replace the word
constraint with, friends, family members, collaborators and you will know
exactly what we mean).</p>

<p class=MyNormal>You can generate this list of constraints however you like.
You may even provide it yourself, forcing important residues to be aligned by
giving them high weights (see the FAQ). For your convenience, T-Coffee can
generate (this is the default) its own list by making all the possible global
pairwise alignments, and the 10 best local alignments associated with each pair
of sequences. Each pair of residues observed aligned in these pairwise
alignments becomes a line in the library.</p>

<p class=MyNormal>Yet be aware that nothing forces you to use this library and
that you could build it using other methods (see the FAQ). In protein language,
T-COFEE is synonymous for freedom, the freedom of being aligned however you
fancy <span class=GramE>( I</span> was a Tryptophan in some previous life). </p>

<p class=MsoBodyText><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<p class=MsoBodyText><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<p class=MsoBodyText><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<p class=MsoBodyText><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<p class=MsoBodyText><span lang=EN-GB><span style='mso-tab-count:1'>            </span></span></p>

<h1><a name="_Toc235985453"><span lang=EN-GB>Preparing Your Data:<br>
Reformatting and Trimming With seq_reformat</span></a></h1>

<p class=MsoBodyText><span lang=EN-GB>Nothing is more frustrating than
downloading important data and realizing you need to format it *before* using
it. In general, you should avoid manual reformatting: it is by essence
inconsistent and will get you into trouble. It will also get you depressed when
you will realize that you have spend the whole day adding carriage return to
each line in your files.</span></p>

<h2><a name="_Toc235985454"><span lang=EN-GB>Seq_reformat</span></a></h2>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985455">Accessing
the T-Coffee Reformatting Utility</a></h3>

<p class=MyNormal>T-Coffee comes along with a very powerful reformatting
utility named seq_reformat. You can use seq_reformat by invoking the t_coffee
shell.</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat </span></p>

</div>

<p class=MyNormal>This will output the online flag usage of seq_reformat.
Seq_reformat recognizes automatically the most common formats. You can use it
to:</p>

<p class=MyNormal>Reformat your sequences.</p>

<p class=MyNormal><span class=GramE>extract</span> sub-portions of alignments</p>

<p class=MyNormal>Extract sequences. </p>

<p class=MyNormal>In this section we give you a few examples of things you can
do with seq_reformat:</p>

<div style='mso-element:para-border-div;border:solid windowtext 1.5pt;
padding:0cm 0cm 0cm 0cm;background:#F7CBB7'>

<p class=Warning><span lang=EN-GB>Warning: after the flag -<span class=SpellE>other_pg</span>,
the T-Coffee flags are not any more recognized. It is like if you were using a
different programme </span></p>

</div>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985456"><span
lang=EN-GB style='mso-ansi-language:EN-GB'>An overview of seq_reformat</span></a><span
lang=EN-GB style='mso-ansi-language:EN-GB'><o:p></o:p></span></h3>

<p class=MyNormal style='margin-left:35.4pt'><span class=GramE><span
lang=EN-GB style='mso-ansi-language:EN-GB'>seq</span></span><span lang=EN-GB
style='mso-ansi-language:EN-GB'>_reformat is a reformatting utility. It reads
in via the -in and -in2 flags and outputs in whatever specified format via the
-output flag. In the meantime, you can use the flag '-action' to modify your
data, using any of the flag. If you want a complete list of things seq_reformat
can do for you, try:<o:p></o:p></span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat </span></p>

</div>

<h2><a name="_Toc235985457"><span lang=EN-GB>Reformatting your data</span></a></h2>

<p class=MyNormal style='margin-left:35.4pt'><span style='mso-tab-count:1'>            </span><span
lang=EN-GB style='mso-ansi-language:EN-GB'><o:p></o:p></span></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985458"><span
lang=EN-GB style='mso-ansi-language:EN-GB'>Changing MSA formats</span></a><span
lang=EN-GB style='mso-ansi-language:EN-GB'><o:p></o:p></span></h3>

<p class=MyNormal>It can be necessary to change from one MSA format to another.
If your sequences are in ClustalW format and you want to turn them into <span
class=SpellE>fasta</span>, while keeping the gaps, try</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.aln -output fasta_aln &gt;
sproteases_small.fasta_aln</span></p>

</div>

<p class=MyNormal>If you want to turn a <span class=SpellE>clustalw</span>
alignment into an alignment having the pileup format (MSF), try:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.aln -output msf &gt; sproteases_small.msf</span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985459">Dealing
with Non-automatically recognized formats</a></h3>

<p class=MyNormal><span style="mso-spacerun:yes">    </span>Format recognition
is not 100% full proof. Occasionally you will have to inform the program about
the nature of the file you are trying to reformat:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU><span
style="mso-spacerun:yes">    </span>-in_f msf_aln for intance</span></p>

</div>

<h2><a name="_Toc235985460"><span lang=EN-GB>Automated Sequence Edition</span></a></h2>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985461"><span
lang=EN-GB style='mso-ansi-language:EN-GB'>Removing the gaps from an alignment</span></a><span
lang=EN-GB style='mso-ansi-language:EN-GB'><o:p></o:p></span></h3>

<p class=MyNormal>If you want to recover your sequences from some pre-computed
alignment, you can try:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.aln -output fasta_seq &gt;
sproteases_small.fasta</span></p>

</div>

<p class=MyNormal>This will remove all the gaps.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985462"><span
lang=EN-GB style='mso-ansi-language:EN-GB'>Changing the case of your sequences</span></a><span
lang=EN-GB style='mso-ansi-language:EN-GB'><o:p></o:p></span></h3>

<p class=MyNormal>If you need to change the case of your sequences, you can use
more sophisticated functions embedded in seq_reformat. We call these modifiers,
and they are accessed via the -action flag. For instance, to write our
sequences in lower case:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.aln -action +lower -output
clustalw</span></p>

</div>

<p class=MyNormal>No prize for guessing that +upper will do exactly the opposite....</p>

<h3 style='mso-list:none'><a name="_Toc235985463">Changing the case of specific
residues</a></h3>

<p class=MyNormal>If you want to change the case of a specific residue, you can
use the flag: +<span class=SpellE>edit_residue</span> &lt;sequence&gt;
&lt;residue #&gt; &lt;<span class=SpellE>lower|upper|symbol</span>&gt;. If you
have more than one residue to color, you can put all the coordinates in a file,
(one coordinate per line). Spans are not yet supported.</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_aln1.aln -action +upper
+edit_residue hmgb_chite 10 lower</span></p>

</div>

<h3 style='mso-list:none'><a name="_Toc235985464">Changing the case depending
on the score</a></h3>

<p class=MyNormal>If you want to change the case depending on the score, you
must either evaluate your alignment, or provide cache (see next section for the
cache). If you want to evaluate on the fly, try:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_aln1.aln -in3 sample_aln1.aln
-action +upper +3evaluate idmat +lower '[5-9]'</span></p>

</div>

<p class=MyNormal>Will lower the case of every residue identical to more than
50% of the residue in its column.</p>

<h2><a name="_Toc235985465"><span lang=EN-GB>Protecting Important Sequence
Names</span></a></h2>

<p class=MsoBodyText><span lang=EN-GB>Few programs support long sequence names.
Sometimes, when going through some pipeline the names of your sequences can be
damaged (truncated or modified). To avoid this, seq_reformat contains a utility
that can automatically rename your sequences into a form that will be machine
friendly, while making it easy to return to the human friendly form.</span></p>

<p class=MsoBodyText><span lang=EN-GB>The first thing to do is to generate a
list of names that will be used in place of the long original name of the
sequences. For instance:</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_large.fasta -output code_name
&gt; sproteases_large.code_name</span></p>

</div>

<p class=MsoBodyText><span lang=EN-GB>Will create a file where each original name
is associated with a coded name (<span class=SpellE>Cxxxx</span>). You can then
use this file to either code or decode your dataset. For instance, the
following command:</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -code sproteases_large.code_name -in
sproteases_large.fasta &gt;sproteases_large.coded.fasta</span></p>

</div>

<p class=MsoBodyText><span lang=EN-GB>Will code all the names of the original
data. You can work with the file <span class=SpellE>sproteases_large.coded.fasta</span>,
and when you are done, you can de-code the names of your sequences using:</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -decode sproteases_large.code_name -in
sproteases_large.coded.fasta</span></p>

</div>

<h2><a name="_Toc235985466"><span lang=EN-GB>Colouring/Editing Residues in an
Alignment</span></a></h2>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985467">Coloring
specific types of residues</a></h3>

<p class=MyNormal>You can color all the residues of your sequences on the fly.
For instance, the following command:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_aln1.aln -in3 sample_aln1.aln
-action<span style="mso-spacerun:yes">  </span>+3convert a0 -output color_html &gt;
colored.html</span></p>

</div>

<p class=MyNormal><span class=GramE>will</span> color all the As in color 0
(blue). </p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985468">Coloring
a specific residue of a specific sequence</a></h3>

<p class=MyNormal>If you want to color a specific residue, you can use the
flag: +<span class=SpellE>color_residue</span> &lt;sequence&gt; &lt;residue
#&gt; &lt;color #&gt;. If you have more than one residue to color, you can put
all the coordinates in a file, (one coordinate per line). Spans are not yet
supported.</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_aln1.aln -action +color_residue hmgb_chite
10 1 -output color_html &gt; color.html</span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985469"><![if !supportLists]><![endif]>Coloring
according to the conservation</a></h3>

<p class=MyNormal>Use the +evaluate flag if you want to color your alignment
according to its conservation level</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_aln1.aln -in3 sample_aln1.aln
-action +3evaluate pam250mt- output color_html &gt; color.html</span></p>

</div>

<p class=MyNormal>You can also use the <span class=SpellE>boxshade</span> <span
class=SpellE>scoreing</span> scheme:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_aln1.aln -in3 sample_aln1.aln
-action +3evaluate boxshade -output color_html &gt; color.html</span></p>

</div>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<h2><a name="_Toc235985470"><span lang=EN-GB>Colouring/Editing residues in an
Alignment Using a Cache</span></a></h2>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985471">Overview</a></h3>

<p class=MyNormal>To color an alignment, two files are needed: the alignment (<span
class=SpellE>aln</span>) and the cache (cache). The cache is a file where
residues to be colored are declared along with the colors. Nine different
colors are currently supported. They are set by default but can be modified by
the user (see last changing default colors). The cache can either look like a
standard sequence or alignment file (see below) or like a standard T-Coffee
library (see next section). In this section we show you how to specifically
modify your original sequences to turn them into a cache.</p>

<p class=MyNormal>In the cache, the colors of each residue are declared with a
number between 0 and 9.<span style="mso-spacerun:yes">  </span>Undeclared
residues will appear without any color in the final alignment.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985472">Preparing
a Sequence or Alignment Cache</a></h3>

<p class=MyNormal>Let us consider the following file: </p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmallCxSpFirst><span lang=EN-GB>CLUSTAL FORMAT </span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>B<span
style="mso-spacerun:yes">               </span>CTGAGA-AGCCGC---CTGAGG--TCG</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>C<span
style="mso-spacerun:yes">               </span>TTAAGG-TCCAGA---TTGCGG--AGC</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>D<span
style="mso-spacerun:yes">               </span>CTTCGT-AGTCGT---TTAAGA--ca-</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>A<span
style="mso-spacerun:yes">               </span><span class=SpellE>CTCCGTgTCTAGGagtTTACGTggAGT</span></span></p>

<p class=examplefilesmallCxSpLast><span lang=EN-GB><span
style="mso-spacerun:yes">                 </span>*<span
style="mso-spacerun:yes">  </span>*<span style="mso-spacerun:yes">     
</span>*<span style="mso-spacerun:yes">     </span>*<span
style="mso-spacerun:yes">  </span>*<span style="mso-spacerun:yes">      </span></span></p>

</div>

<p class=MyNormal>The command</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in=sample_aln6.aln -output=clustalw_aln
-out=cache.aln -action +convert 'Aa1' '.--' +convert '#0'</span></p>

</div>

<p class=MyNormal style='text-indent:14.15pt'>The conversion will proceed as
follows:</p>

<p class=MyNormal style='margin-left:70.8pt;text-indent:35.4pt'>-<span
class=SpellE><span class=GramE>conv</span></span> indicates the filters for
character conversion:</p>

<p class=MyNormal><span style='mso-tab-count:2'>                     </span>- <span
class=GramE>will</span> remain - </p>

<p class=MyNormal><span style='mso-tab-count:2'>                     </span>A
and a will be turned into 1</p>

<p class=MyNormal><span style='mso-tab-count:2'>                     </span>All
the other symbols (#) will be turned into 0.</p>

<p class=MyNormal style='margin-left:106.2pt'>-<span class=GramE>action</span>
+convert, indicates the actions that must be carried out on the alignment
before it is output into cache.</p>

<p class=MsoBodyText><span lang=EN-GB>This command generates the following
alignment (called a cache):</span></p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmallCxSpFirst><span lang=EN-GB>CLUSTAL FORMAT for
SEQ_REFORMAT Version 1.00, CPU=0.00 sec, SCORE=0, <span class=SpellE>Nseq</span>=4,
Len=27</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>B<span
style="mso-spacerun:yes">               </span>000101-100000---000100--000</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>C<span
style="mso-spacerun:yes">            </span><span
style="mso-spacerun:yes">   </span>001100-000101---000000--100</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>D<span
style="mso-spacerun:yes">               </span>000000-100000---001101--01-</span></p>

<p class=examplefilesmallCxSpLast><span lang=EN-GB>A<span
style="mso-spacerun:yes">               </span>000000000010010000100000100</span></p>

</div>

<p class=MyNormal>Other <span class=GramE>alternative are</span> possible. For
instance, the following command:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in=sample_aln6.aln -output=fasta_seq
-out=cache.seq -action +convert 'Aa1' '.--' +convert '#0'</span></p>

</div>

<p class=MyNormal><span class=GramE>will</span> produce the following file <span
class=SpellE>cache_seq</span></p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmallCxSpFirst><span lang=EN-GB>&gt;B</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>000101100000000100000</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;C</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>001100000101000000100</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;D</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>00000010000000110101</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;A</span></p>

<p class=examplefilesmallCxSpLast><span lang=EN-GB>000000000010010000100000100</span></p>

</div>

<p class=MyNormal><span class=GramE>where</span> each residue has been replaced
with a number according to what was specified by conv. Note that it is not
necessary to replace EVERY residue with a code. For instance, the following
file would also be suitable as a cache:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in=sample_aln6.aln -output=fasta_seq
-out=cache -action +convert 'Aa1' '.--'</span></p>

</div>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmallCxSpFirst><span lang=EN-GB>&gt;B</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>CTG1G11GCCGCCTG1GGTCG</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;C</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>TT11GGTCC1G1TTGCGG1GC</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;D</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>CTTCGT1GTCGTTT11G1c1</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;A</span></p>

<p class=examplefilesmallCxSpLast><span lang=EN-GB>CTCCGTgTCT1GG1gtTT1CGTgg1GT</span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985473">Preparing
a Library Cache</a></h3>

<p class=MyNormal>The Library is a special format used by T-Coffee to declare
special relationships between pairs of residues. The cache library format can
also be used to declare the color of specific residues in an alignment. For
instance, the following file</p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmallCxSpFirst><span lang=EN-GB>! TC_LIB_FORMAT_01</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>4</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>A 27 <span class=SpellE>CTCCGTgTCTAGGagtTTACGTggAGT</span></span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>B 21 CTGAGAAGCCGCCTGAGGTCG</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>C 21 TTAAGGTCCAGATTGCGGAGC</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>D 20 <span class=SpellE>CTTCGTAGTCGTTTAAGAca</span></span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>#1 <span class=SpellE>1</span></span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB><span
style="mso-spacerun:yes">    </span>1<span style="mso-spacerun:yes">    
</span><span class=SpellE>1</span><span style="mso-spacerun:yes">   </span>3</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB><span
style="mso-spacerun:yes">    </span>4<span style="mso-spacerun:yes">    
</span><span class=SpellE>4</span><span style="mso-spacerun:yes">   </span>5</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>#3 <span class=SpellE>3</span></span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB><span
style="mso-spacerun:yes">    </span>6<span style="mso-spacerun:yes">    
</span><span class=SpellE>6</span><span style="mso-spacerun:yes">   </span>1</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB><span
style="mso-spacerun:yes">    </span>9<span style="mso-spacerun:yes">    
</span><span class=SpellE>9</span><span style="mso-spacerun:yes">   </span>4</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>! CPU 240</span></p>

<p class=examplefilesmallCxSpLast><span lang=EN-GB>! SEQ_1_TO_N</span></p>

</div>

<p class=MyNormal><span class=GramE>sample</span>_lib5.tc_lib declares that
residue 1 of sequence 3 will be receive color 6, while residue 20 of sequence 4
will receive color 20. Note that the sequence number and the residue index are
duplicated, owing to the recycling of this format from its original usage.</p>

<p class=MyNormal>It is also possible to use the BLOCK operator when defining
the library (c.f. technical doc, library format). For instance:</p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmallCxSpFirst><span lang=EN-GB>! TC_LIB_FORMAT_01</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>4</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>A 27 <span class=SpellE>CTCCGTgTCTAGGagtTTACGTggAGT</span></span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>B 21 CTGAGAAGCCGCCTGAGGTCG</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>C 21 TTAAGGTCCAGATTGCGGAGC</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>D 20 <span class=SpellE>CTTCGTAGTCGTTTAAGAca</span></span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>#1 <span class=SpellE>1</span></span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB><span
style="mso-spacerun:yes">    </span>+BLOCK+ 10 1<span
style="mso-spacerun:yes">     </span><span class=SpellE>1</span><span
style="mso-spacerun:yes">   </span><span style="mso-spacerun:yes"> </span>3</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB><span
style="mso-spacerun:yes">    </span>+BLOCK+ <span class=GramE>5 <span
style="mso-spacerun:yes"> </span>15</span><span style="mso-spacerun:yes">   
</span><span class=SpellE>15</span><span style="mso-spacerun:yes">   </span>5</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>#3 <span class=SpellE>3</span></span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB><span
style="mso-spacerun:yes">    </span>6<span style="mso-spacerun:yes">    
</span><span class=SpellE>6</span><span style="mso-spacerun:yes">   </span>1</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB><span
style="mso-spacerun:yes">    </span>9<span style="mso-spacerun:yes">    
</span><span class=SpellE>9</span><span style="mso-spacerun:yes">   </span>4</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>! CPU 240</span></p>

<p class=examplefilesmallCxSpLast><span lang=EN-GB>! SEQ_1_TO_N</span></p>

</div>

<p class=MyNormal>The number right after BLOCK indicates the block length (10).
The two next numbers (1 <span class=SpellE>1</span>) indicate the position of
the first element in the block. The last value is the color.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985474">Coloring
an Alignment using a cache</a></h3>

<p class=MyNormal>If you have a cache alignment or a cache library, you can use
it to color your alignment and either <span class=GramE>make</span> a post
script, html or PDF output. For instance, if you use the file <span
class=SpellE>cache.seq</span>:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU><span
style='mso-tab-count:1'>   </span>PROMPT: t_coffee -other_pg seq_reformat
-in=sample_aln6.aln -struc_in=sample_aln6.cache -struc_in_f number_fasta
-output=color_html -out=x.html</span></p>

</div>

<p class=MyNormal>This will produce a colored version readable with any
standard web browser, while:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU><span
style='mso-tab-count:1'>   </span>PROMPT: t_coffee -other_pg seq_reformat
-in=sample_aln6.aln -struc_in=sample_aln6.cache -struc_in_f number_fasta
-output=color_pdf -out=x.pdf</span></p>

</div>

<p class=MyNormal>This will produce a colored version readable with acrobat
reader. </p>

<div style='mso-element:para-border-div;border:solid windowtext 1.5pt;
padding:0cm 0cm 0cm 0cm;background:#F7CBB7'>

<p class=Warning><span lang=EN-GB>Warning: ps2pdf must be installed on your
system </span></p>

</div>

<p class=MyNormal>You can also use a cache library like the one shown above
(sample_lib5.tc_lib):</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in=sample_aln6.aln
-struc_in=sample_lib5.tc_lib -output=color_html -out=x.html</span></p>

</div>

<h2><a name="_Toc235985475"><span lang=EN-GB>Changing the default </span></a><span
class=SpellE><span style='mso-bookmark:_Toc235985475'><span lang=EN-GB>colors</span></span></span><span
style='mso-bookmark:_Toc235985475'></span></h2>

<p class=MyNormal>Colors are hard coded in the program, but if you wish, you
can change them, simply create a file named:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU><span
style='mso-tab-count:1'>   </span>seq_reformat.color</span></p>

</div>

<p class=MyNormal>That is used to declare the color values:</p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmallCxSpFirst><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>0 #FFAA00 1 0.2 0</span></p>

<p class=examplefilesmallCxSpLast><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

</div>

<p class=MyNormal>This indicates that the value 0 in the cache corresponds now
to #FFAA00 in html, and in RGB 1<span class=GramE>,<span
style="mso-spacerun:yes">  </span>0.2</span> and 0. The name of the file (<span
class=SpellE>seq_reformat.color</span>) is defined in: <span class=SpellE>programmes_define.h</span>,
COLOR_FILE. And can be changed before compilation. <b style='mso-bidi-font-weight:
normal'>By default, the file is searched in the current directory<o:p></o:p></b></p>

<h2><a name="_Toc235985476"><span lang=EN-GB>Evaluating an alignment and
producing a cache</span></a></h2>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985477">Evaluating
an alignment with T-Coffee</a></h3>

<p class=MyNormal>As suggested in a previous section, it is possible to
evaluate the accuracy of any alignment using a T-Coffee library. The simplest
way to do that is to compute a default library and evaluate the target
alignment against this library: </p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -infile sample_aln1.aln -mode evaluate </span></p>

</div>

<p class=MyNormal>This command will output a file named sample_aln1.score_asccii
that can then be used to either evaluate the local accuracy of the alignment or
automatically filter it using the seq_reformat utility.</p>

<p class=MyNormal>In some circumstances, you may also want to evaluate your
alignment against a pre-computed library. This can be easily achieved:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -infile sample_aln1.aln -out_lib sample_aln1.tc_lib -lib_only </span></p>

</div>

<p class=MsoBodyText><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -infile sample_aln1.aln -mode evaluate -lib sample_aln1.tc_lib</span></p>

</div>

<p class=MyNormal>When using this last command, the reference library will be
the one provided by the user. The local score thus reported is the CORE index. </p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985478">Evaluating
the level of conservation with a substitution matrix</a></h3>

<p class=MyNormal>It is possible to use seq_reformat in a similar way to infer
the local level of identity, either using an identity matrix or with any regular
matrix, in which case, every residue with a substitution score higher than 0 is
counted as an identity. This can be achieved as follows for identity measure:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_aln1.aln -action +evaluate idmat
-output score_ascii</span></p>

</div>

<p class=MyNormal>Or with the following for measuring similarity with a
blosum62</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_aln1.aln -action +evaluate
blosum62mt -output score_ascii</span></p>

</div>

<p class=MyNormal>Finally, it is also possible to display in color the
conservation levels:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU><span
style='mso-tab-count:1'>   </span>PROMPT: t_coffee -other_pg seq_reformat -in
sample_aln1.aln -action +evaluate blosum62mt -output score_html &gt; x.html</span></p>

</div>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MsoBodyText><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<h2><a name="_Toc235985479"><span lang=EN-GB>Selective Reformatting</span></a></h2>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985480">Removing
gapped columns</a></h3>

<p class=MyNormal>You can remove all the columns containing a certain
proportion of gaps. For instance:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_aln7.aln -action +rm_gap 50</span></p>

</div>

<p class=MyNormal>Will delete all the residues occurring in a column that
contains 50% or more gaps (use 1 to delete residues from columns having 1 gap
or more).</p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><o:p>&nbsp;</o:p></h3>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985481">Selectively
turn some residues to lower case</a></h3>

<p class=MyNormal>Consider the following alignment (sample_aln7.aln)</p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmallCxSpFirst><span lang=EN-GB>CLUSTAL FORMAT for T-COFFEE
Version_4.62 [http://www.tcoffee.org], CPU=0.04 sec, SCORE=0, <span
class=SpellE>Nseq</span>=4, Len=28</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>A<span
style="mso-spacerun:yes">               </span>CTCCGTGTCTAGGAGT-TTACGTGGAGT</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>B<span
style="mso-spacerun:yes">               </span>CTGAGA----AGCCGCCTGAGGTCG---</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>D<span
style="mso-spacerun:yes">               </span>CTTCGT----AGTCGT-TTAAGACA---</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>C<span
style="mso-spacerun:yes">               </span>-TTAAGGTCC---AGATTGCGGAGC---</span></p>

<p class=examplefilesmallCxSpLast><span lang=EN-GB><span
style="mso-spacerun:yes">                 </span><span class=GramE>* ..</span><span
style="mso-spacerun:yes">        </span><span class=GramE>.*</span><span
style="mso-spacerun:yes">  </span>* . *:</span></p>

</div>

<p class=MyNormal><span class=GramE>and</span> the following cache
(sample_aln7.cache_aln):</p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmallCxSpFirst><span lang=EN-GB>CLUSTAL FORMAT for T-COFFEE
Version_4.62 [http://www.tcoffee.org], CPU=0.04 sec, SCORE=0, <span
class=SpellE>Nseq</span>=4, Len=28</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>A<span
style="mso-spacerun:yes">               </span>3133212131022021-11032122021</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>B<span
style="mso-spacerun:yes">               </span>312020----023323312022132---</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>D<span
style="mso-spacerun:yes">               </span>311321----021321-11002030---</span></p>

<p class=examplefilesmallCxSpLast><span lang=EN-GB>C<span
style="mso-spacerun:yes">               </span>-110022133---020112322023---</span></p>

</div>

<p class=MsoBodyText><span lang=EN-GB style='font-size:8.0pt;mso-bidi-font-size:
10.0pt;font-family:"Courier New";letter-spacing:0pt'><o:p>&nbsp;</o:p></span></p>

<p class=MyNormal>You can turn to lower case all the residues having a score
between 1 and 2:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_aln7.aln -struc_in
sample_aln7.cache_aln -struc_in_f number_aln -action +lower '[1-2]'</span></p>

</div>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmallCxSpFirst><span lang=EN-GB>CLUSTAL FORMAT for T-COFFEE
Version_4.62 [http://www.tcoffee.org], CPU=0.05 sec, SCORE=0, <span
class=SpellE>Nseq</span>=4, Len=28</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>A<span
style="mso-spacerun:yes">               </span><span class=SpellE>CtCCgtgtCtAggAgt-ttACgtggAgt</span></span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>B<span
style="mso-spacerun:yes">               </span><span class=SpellE>CtgAgA</span>----<span
class=SpellE>AgCCgCCtgAggtCg</span>---</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>D<span
style="mso-spacerun:yes">               </span><span class=SpellE>CttCgt</span>----<span
class=SpellE>AgtCgt-ttAAgACA</span>---</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>C<span
style="mso-spacerun:yes">               </span>-<span class=SpellE>ttAAggtCC</span>---<span
class=SpellE>AgAttgCggAgC</span>---</span></p>

<p class=examplefilesmallCxSpLast><span lang=EN-GB><span
style="mso-spacerun:yes">                 </span><span class=GramE>* ..</span><span
style="mso-spacerun:yes">        </span><span class=GramE>.*</span><span
style="mso-spacerun:yes">  </span>* . *:</span></p>

</div>

<p class=MsoBodyText><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<div style='mso-element:para-border-div;border:solid windowtext 1.5pt;
padding:1.0pt 4.0pt 1.0pt 4.0pt;background:#FFFF89'>

<p class=StylenoteboxbodyBoldNotItalicBoxSinglesolidlineAut><span lang=EN-GB>Note:
<span class=GramE>that residues</span> not concerned will keep their original
case </span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985482">Selectively
modifying residues</a></h3>

<p class=MyNormal><span class=GramE>The range operator is supported by three
other important modifiers</span>:</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span><b
style='mso-bidi-font-weight:normal'>-<span class=GramE>upper</span>:</b> to
uppercase your residues</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span><b
style='mso-bidi-font-weight:normal'>-<span class=GramE>lower</span>:</b> to
lowercase your residues</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span><b
style='mso-bidi-font-weight:normal'>-<span class=SpellE><span class=GramE>switchcase</span></span>:</b>
to selectively toggle the case of your residues</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span><b
style='mso-bidi-font-weight:normal'>-<span class=GramE>keep</span>:</b> to only
keep the residues within the range</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span><b
style='mso-bidi-font-weight:normal'>-<span class=GramE>remove</span>:</b> to
remove the residues within the range</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span><b
style='mso-bidi-font-weight:normal'>-<span class=GramE>convert</span>:</b> to
only convert the residues within the range.</p>

<p class=MyNormal>For instance, to selectively turn all the G having a score
between 1 and 2, use:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_aln7.aln -struc_in
sample_aln7.cache_aln -struc_in_f number_aln -action +convert '[1-2]' CX</span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985483">Keeping
only the best portion of an alignment</a></h3>

<p class=MyNormal>To do this, you need an evaluation file that may have been
generated with T-Coffee, either running a de-novo alignment</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_seq1.fasta -output score_ascii, aln</span></p>

</div>

<p class=MyNormal>Or evaluating a pre-existing alignment</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_seq1.aln -action +evaluate
blosum62mt -output score_ascii</span></p>

</div>

<p class=MyNormal>This generates a <span class=SpellE>score_ascii</span> file
that you can then use to filter out the bad bits in your <span class=SpellE>alignement</span>:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_seq1.aln -struc_in sample_seq1.score_ascii
-struc_in_f number_aln<span style="mso-spacerun:yes">  </span>-action +keep '[8-9]'</span></p>

</div>

<p class=MyNormal>This command considers the individual score of each residue
to trigger the filtering. It is also possible to do this according to the whole
column. Simply add the &quot;+<span class=SpellE>use_cons</span>&quot; flag.</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_seq1.aln -struc_in sample_seq1.score_ascii
-struc_in_f number_aln<span style="mso-spacerun:yes">  </span>-action +use_cons
+keep '[8-9]'</span></p>

</div>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<h2><span lang=EN-GB><o:p>&nbsp;</o:p></span></h2>

<h2><a name="_Toc235985484"><span lang=EN-GB>Extracting Portions of Dataset</span></a></h2>

<p class=MsoBodyText><span lang=EN-GB>Extracting portions of a dataset is
something very frequently <span class=GramE>needed.</span> You may need to
extract all the sequences that contain the word human in their name, or you may
want all the sequences containing a simple motif. We show you here how to do a
couple of these things.</span></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985485"><span
lang=EN-GB style='mso-ansi-language:EN-GB'>Extracting The High Scoring Blocks</span></a><span
lang=EN-GB style='mso-ansi-language:EN-GB'><o:p></o:p></span></h3>

<p class=MyNormal><span lang=EN-GB style='mso-ansi-language:EN-GB'>It is
possible to use a <span class=SpellE>score_ascii</span> file <span class=GramE>(
as</span> produced in the previous section) in order to extract high scoring
portions of an alignment. For instance, the following command:<o:p></o:p></span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU><span
style='mso-tab-count:1'>   </span>PROMPT: t_coffee -other_pg seq_reformat -in
sample_aln1.aln -action +evaluate blosum62mt +use_cons +keep '[5-9]'</span></p>

</div>

<p class=MyNormal><span class=GramE><span lang=EN-GB style='mso-ansi-language:
EN-GB'>will</span></span><span lang=EN-GB style='mso-ansi-language:EN-GB'> keep
all the residues having a column conservation score between 5 and 9 <o:p></o:p></span></p>

<div style='mso-element:para-border-div;border:solid windowtext 1.5pt;
padding:1.0pt 4.0pt 1.0pt 4.0pt;background:#FFFF89'>

<p class=StylenoteboxbodyBoldNotItalicBoxSinglesolidlineAut><span lang=EN-GB>Note:
Don't forget the simple quotes! (') </span></p>

</div>

<p class=MyNormal><span lang=EN-GB style='mso-ansi-language:EN-GB'>It is also
possible to re-use pre-computed <span class=SpellE>score_ascii</span> files,
such as those obtained when computing a T-Coffee multiple alignment. For
instance, the following series of command will make it possible to extract the
positions having a consistency score between 6 and 9:<o:p></o:p></span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU><span
style='mso-tab-count:1'>   </span>PROMPT: t_coffee sample_aln1.fasta -output
score_ascii -outfile sample1.score_ascii</span></p>

</div>

<p class=MyNormal><span lang=EN-GB style='mso-ansi-language:EN-GB'><o:p>&nbsp;</o:p></span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU><span
style='mso-tab-count:1'>   </span>PROMPT: t_coffee -other_pg seq_reformat -in
sample_aln1.aln -struc_in sample1.score_ascii -struc_in_f number_aln -action
+use_cons +keep '[8-9]'</span></p>

</div>

<p class=MyNormal><span lang=EN-GB style='mso-ansi-language:EN-GB'><o:p>&nbsp;</o:p></span></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985486"><span
lang=EN-GB style='mso-ansi-language:EN-GB'>Extracting Sequences According to a
Pattern</span></a><span lang=EN-GB style='mso-ansi-language:EN-GB'><o:p></o:p></span></h3>

<p class=MyNormal>You can extract any sequence by requesting a specific pattern
to be found either in the name, the comment or the sequence. For instance, if
you want to extract all the sequences whose name contain the word HUMAN:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.aln -action +grep NAME
KEEP HUMAN -output clustalw</span></p>

</div>

<p class=MyNormal>The modifier is &quot;+<span class=SpellE>grep</span>&quot;.
NAME indicates that the extraction is made according to the sequences names,
and KEEP means that you will keep all the sequences containing the string
HUMAN. If you wanted to remove all the sequences whose name contains the word
HUMAN, you should have typed:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.aln -action +grep NAME
REMOVE HUMAN -output clustalw</span></p>

</div>

<p class=MyNormal>Note <span class=GramE>that<span style="mso-spacerun:yes"> 
</span>HUMAN</span> is case sensitive (Human, HUMAN and <span class=SpellE>hUman</span>
will not yield the same results). You can also select the sequences according
to some pattern found in their COMMENT section or directly in the sequence. For
instance</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.aln -action +grep COMMENT
KEEP sapiens -output clustalw</span></p>

</div>

<p class=MyNormal>Will keep all the sequences containing the word <b
style='mso-bidi-font-weight:normal'><i style='mso-bidi-font-style:normal'>sapiens</i></b>
in the comment section. Last but not least, you should know that the pattern <span
class=GramE>can</span> be any <span class=SpellE>perl</span> legal regular
expression (See www.comp.leeds.ac.uk/Perl/matching.html for some background on
regular expressions). For instance:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.aln -action +grep NAME
REMOVE '[ILM]K' -output clustalw</span></p>

</div>

<p class=MyNormal>Will extract all the sequences containing the pattern [<span
class=GramE>ILM]K</span>.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985487"><span
lang=EN-GB style='mso-ansi-language:EN-GB'>Extracting Sequences by Names</span></a><span
lang=EN-GB style='mso-ansi-language:EN-GB'><o:p></o:p></span></h3>

<p class=MyNormal><b style='mso-bidi-font-weight:normal'><i style='mso-bidi-font-style:
normal'>Extracting Two Sequences</i></b>: If you want to extract several
sequences, in order to make a subset. You can do the following:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.aln -action +extract_seq_list
'sp|P29786|TRY3_AEDAE' 'sp|P35037|TRY3_ANOGA' </span></p>

</div>

<div style='mso-element:para-border-div;border:solid windowtext 1.5pt;
padding:0cm 0cm 0cm 0cm;background:#F7CBB7'>

<p class=Warning><span lang=EN-GB>Note the single quotes ('). They are meant to
protect the name of your sequence and prevent the UNIX shell to interpret it
like an instruction.</span></p>

</div>

<p class=MyNormal><b style='mso-bidi-font-weight:normal'>Removing Columns of
Gaps. </b>Removing intermediate sequences results in columns of gaps appearing
here and there. Keeping them is convenient if some features are mapped on your alignment.
On the other hand, if you want to remove these columns you can use:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.aln -action +extract_seq_list
'sp|P29786|TRY3_AEDAE' 'sp|P35037|TRY3_ANOGA' +rm_gap</span></p>

</div>

<p class=MyNormal><b style='mso-bidi-font-weight:normal'><i style='mso-bidi-font-style:
normal'>Extracting Sub sequences</i></b>: You may want to extract portions of
your sequences. This is possible if you specify the coordinates after the
sequences name:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.aln -action +extract_seq
'sp|P29786|TRY3_AEDAE' 20 200 'sp|P35037|TRY3_ANOGA' 10 150 +rm_gap</span></p>

</div>

<p class=MyNormal><b style='mso-bidi-font-weight:normal'><i style='mso-bidi-font-style:
normal'>Keeping the original Sequence Names.</i></b> Note that your sequences
are now renamed according to the extraction coordinates. You can keep the
original names by using the +<span class=SpellE>keep_name</span> modifier:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.aln -action +keep_name
+extract_seq 'sp|P29786|TRY3_AEDAE' 20 200 'sp|P35037|TRY3_ANOGA' 10 150
+rm_gap</span></p>

</div>

<div style='mso-element:para-border-div;border:solid windowtext 1.5pt;
padding:0cm 0cm 0cm 0cm;background:#F7CBB7'>

<p class=Warning><span lang=EN-GB>Note: +<span class=SpellE>keep_name</span>
must come BEFORE +<span class=SpellE>extract_seq</span></span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985488"><span
lang=EN-GB style='mso-ansi-language:EN-GB'>Removing Sequences by Names</span></a><span
lang=EN-GB style='mso-ansi-language:EN-GB'><o:p></o:p></span></h3>

<p class=MyNormal><b style='mso-bidi-font-weight:normal'><i style='mso-bidi-font-style:
normal'>Removing Two Sequences</i></b>. If you want to remove several
sequences, use <span class=SpellE>rm_seq</span> instead of <span class=SpellE>keep_seq</span>.</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.aln -action +remove_seq
'sp|P29786|TRY3_AEDAE' 'sp|P35037|TRY3_ANOGA' </span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985489"><span
lang=EN-GB style='mso-ansi-language:EN-GB'>Extracting Blocks Within Alignment</span></a><span
lang=EN-GB style='mso-ansi-language:EN-GB'><o:p></o:p></span></h3>

<p class=MyNormal><b style='mso-bidi-font-weight:normal'><i style='mso-bidi-font-style:
normal'>Extracting a Block.</i></b> If you only want to keep one block in your
alignment, use </p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.aln -action +extract_block
cons 150 200</span></p>

</div>

<p class=MyNormal>In this command line, <span class=GramE><b style='mso-bidi-font-weight:
normal'><i style='mso-bidi-font-style:normal'>cons</i></b> indicates</span>
that you are counting the positions according to the consensus of the alignment
(i.e. the positions correspond to the columns # of the alignment). If you want
to extract your block relatively to a specific sequence, you should replace
cons with this sequence name. For instance:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.aln -action +extract_block
'sp|Q03238|GRAM_RAT' 10 200</span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985490"><span
lang=EN-GB style='mso-ansi-language:EN-GB'>Concatenating Alignments</span></a><span
lang=EN-GB style='mso-ansi-language:EN-GB'><o:p></o:p></span></h3>

<p class=MyNormal>If you have extracted several blocks and you now want to glue
them together, you can use the <span class=SpellE>cat_aln</span> function</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.aln -action +extract_block
cons 100 120 &gt; block1.aln </span></p>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.aln -action +extract_block
cons 150 200 &gt; block2.aln </span></p>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in block1.aln -in2 block2.aln -action +cat_aln
</span></p>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU><o:p>&nbsp;</o:p></span></p>

</div>

<div style='mso-element:para-border-div;border:solid windowtext 1.5pt;
padding:0cm 0cm 0cm 0cm;background:#F7CBB7'>

<p class=Warning><span lang=EN-GB>Note: The alignments do not need to have the
same number of sequences and the sequences do not need to come in the same
order.</span></p>

</div>

<h2><a name="_Toc235985491"><span lang=EN-GB>Analyzing your Multiple Sequence
Alignment</span></a></h2>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985492">Estimating
the diversity in your alignment</a></h3>

<p class=MyNormal>It is easy to measure the level of diversity within your
multiple sequence alignment. The following command:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_aln1.aln -output sim</span></p>

</div>

<p class=MyNormal>Will output all the pairwise identities, as well as the
average level of identity between each sequence and the others. You can sort
and <span class=SpellE>grep</span> in order to select the sequences you are
interested in.</p>

<p class=MsoBodyText><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<h2><a name="_Toc235985493"><span lang=EN-GB>Reducing and improving your
dataset</span></a></h2>

<p class=MsoBodyText><span lang=EN-GB>Large datasets are problematic because
they can be difficult to analyze. The problem is that when there are too many
sequences, MSA programs tend to become very slow and inaccurate. Furthermore,
you will find that large datasets are difficult to display and analyze. In
short, the best size for an MSA dataset is between 20 and 40 sequences. This
way you have enough sequences to <b style='mso-bidi-font-weight:normal'><i
style='mso-bidi-font-style:normal'>see</i></b> the effect of evolution, but at
the same time the dataset is small enough so that you can visualize your alignment
and <span class=SpellE>recompute</span> it as many times as needed.</span></p>

<div style='mso-element:para-border-div;border:solid windowtext 1.5pt;
padding:0cm 0cm 0cm 0cm;background:#F7CBB7'>

<p class=Warning><span lang=EN-GB>Note: If your sequence dataset is very large,
seq_reformat will compute the similarity matrix between your sequences once
only. It will then keep it in its cache and re-use it any time you re-use that
dataset. In short this means that it will take much longer to run the first
time.</span></p>

</div>

<p class=MsoBodyText><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985494">Extracting
the N most informative sequences</a></h3>

<p class=MyNormal>To be informative, a sequence must contain information the
other sequences do not contain. The N most informative sequences are the N sequences
that are as different as possible to one another, given the initial dataset.</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_large.fasta -action +trim
_seq_n10<span style="mso-spacerun:yes">  </span>-output fasta_seq</span></p>

</div>

<p class=MyNormal>The arguments to trim include <b style='mso-bidi-font-weight:
normal'><i style='mso-bidi-font-style:normal'>_<span class=SpellE>seq</span><span
class=GramE>_<span style='font-weight:normal;font-style:normal'> .</span></span></i></b>
It means your sequences are provided unaligned. If your sequences are already
aligned, you do not need to provide this parameter. It is <span class=SpellE>generaly</span>
more accurate to use unaligned sequences. </p>

<p class=MyNormal>The argument <b style='mso-bidi-font-weight:normal'><i
style='mso-bidi-font-style:normal'>_n10</i></b> means you want to extract the
10 most informative sequences. If you would rather extract the 20% most
informative sequences, use </p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_large.fasta -action +trim
_seq_N20<span style="mso-spacerun:yes">  </span>-output fasta_seq</span></p>

</div>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985495">Extracting
all the sequences less than X% identical</a></h3>

<p class=MyNormal>Removing the most similar sequences is often what people have
in mind when they talk about removing redundancy. You can do so using the trim
option. For instance, to generate a dataset where no pair of sequences has more
than 50% identity, use:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_large.fasta -action +trim
_seq_%%50_</span></p>

</div>

<h3 style='margin-left:0cm;mso-list:none'><o:p>&nbsp;</o:p></h3>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985496">Speeding
up the process</a></h3>

<p class=MyNormal>If you start form unaligned sequences, the removal of
redundancy can be slow. If your sequences have already been aligned using a
fast method, you can take advantage of this by replacing the _<span
class=SpellE>seq</span>_ with _<span class=SpellE>aln</span>_</p>

<p class=MyNormal>Note the difference of speed between these two <span
class=GramE>command</span> and the previous one:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in kinases.aln -action +trim _aln_%%50_</span></p>

</div>

<p class=MyNormal style='margin-left:0cm'><o:p>&nbsp;</o:p></p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>t_coffee
-other_pg seq_reformat -in kinases.fasta -action +trim _seq_%%50_</span></p>

</div>

<p class=MyNormal>Of course, using the MSA will mean that you rely on a more
approximate estimation of sequence similarity. </p>

<h3 style='margin-left:0cm;text-indent:17.0pt;mso-list:none'><a
name="_Toc235985497">Forcing Specific Sequences to be kept</a></h3>

<p class=MyNormal>Sometimes you want to trim while making sure specific important
sequences remain in your dataset. You can do so by providing trim with a <b
style='mso-bidi-font-weight:normal'>string. </b>Trim will keep all the
sequences whose name contains the string. For instance, if you want to force
trim to keep all the sequences that contain the word HUMAN, no matter how
similar they are to one another, you can run the following command:</p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_large.fasta -action +trim
_seq_%%50 HUMAN</span></p>

</div>

<p class=MyNormal>When you give this command, the program will first make sure
that all the HUMAN sequences are kept and it will then assemble your 50% dataset
while keeping the HUMAN sequences. Note that string is a <span class=SpellE>perl</span>
regular expression. </p>

<p class=MyNormal>By default, string causes all the sequences whose name it
matches to be kept. You can also make sure that sequences whose COMMENT or
SEQUENCE matches string are kept. For instance, the following line</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_large.fasta -action +trim
_seq_%%50_fCOMMENT '.apiens'</span></p>

</div>

<p class=MyNormal>Will cause all the sequences containing the regular
expression '<span class=GramE>.<span class=SpellE>apiens</span>'</span> in the
comment to be kept. The _f symbol before COMMENT stands for &quot;_field&quot;
If you want <span class=GramE>to<span style="mso-spacerun:yes">  </span>make</span>
a selection on the sequences:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_large.fasta -action +trim
_seq_%%50_fSEQ '[MLV][RK]'</span></p>

</div>

<p class=MyNormal>You can also specify the sequences you want to keep. To do
so, give a <span class=SpellE>fasta</span> file containing the name of these
sequences via the -in2 file</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:t_coffee
-other_pg seq_reformat -in sproteases_large.fasta -in2 sproteases_small.fasta
-action +trim _seq_%%40</span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985498">Identifying
and Removing </a><span class=SpellE><span style='mso-bookmark:_Toc235985498'>Outlayers</span></span><span
style='mso-bookmark:_Toc235985498'></span></h3>

<p class=MyNormal>Sequences that are too distantly related from the rest of the
set will sometimes have very negative effects on the overall alignment. To
prevent this, it is advisable not to use them. This can be done when trimming
the sequences. For instance, </p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_large.fasta<span
style="mso-spacerun:yes">  </span>-action +trim _seq_%%50_O40</span></p>

</div>

<p class=MyNormal>The symbol _O stands for <span class=SpellE>Outlayers</span>.
It will lead to the removal of all the sequences that have less than 40%
average accuracy with all the other sequences in the dataset.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985499">Chaining
Important Sequences</a></h3>

<p class=MyNormal>In order to align two distantly related sequences, most
multiple sequence alignment packages perform better when provided with many
intermediate sequences that make it possible to &quot;bridge&quot; your two
sequences. The modifier <b style='mso-bidi-font-weight:normal'><i
style='mso-bidi-font-style:normal'>+chain</i></b> makes it possible to extract
from a dataset a subset of intermediate sequences that chain the sequences you
are interested in.</p>

<p class=MyNormal>For instance, le us consider the two sequences:</p>

<p class=MyNormal><span style="mso-spacerun:yes"> </span><span class=GramE>sp</span>|P21844|MCPT5_MOUSE<span
style="mso-spacerun:yes">      </span>sp|P29786|TRY3_AEDAE</p>

<p class=MyNormal>These sequences have 26% identity. This is high enough to
make a case for a homology relationship between them, but this is too low to blindly
trust any pairwise alignment. With the names of the two sequences written in
the file <span class=SpellE>sproteases_pair.fasta</span>, run the following
command:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_large.fasta -in2
sproteases_pair.fasta -action +chain &gt; sproteases_chain.fasta</span></p>

</div>

<p class=MyNormal>This will generate a dataset of 21 sequences, <span
class=SpellE>whith</span> the following chain of similarity between your two
sequences:</p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmallCxSpFirst><span lang=EN-GB>N: 21 Lower: 40 <span
class=SpellE>Sim</span>: 25 DELTA: 15</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>#<span class=GramE>sp</span>|P21844|MCPT5_MOUSE
--&gt;93 --&gt;sp|P50339|MCPT3_RAT --&gt;85 --&gt;sp|P50341|MCPT2_MERUN
--&gt;72 --&gt;sp|P52195|MCPT1_PAPHA --&gt;98 --&gt;sp|P56435|MCPT1_MACFA
--&gt;97 --&gt;sp|P23946|MCPT1_HUMAN --&gt;8</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>1 --&gt;<span class=GramE>sp</span>|P21842|MCPT1_CANFA
--&gt;77 --&gt;sp|P79204|MCPT2_SHEEP --&gt;60 --&gt;sp|P21812|MCPT4_MOUSE
--&gt;90 --&gt;sp|P09650|MCPT1_RAT --&gt;83 --&gt;sp|P50340|MCPT1_MERUN
--&gt;73 --&gt;sp|P11034|MCPT1_MOUSE</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>--&gt;76 --&gt;<span
class=GramE>sp</span>|P00770|MCPT2_RAT --&gt;71 --&gt;sp|P97592|MCPT4_RAT
--&gt;66 --&gt;sp|Q00356|MCPTX_MOUSE --&gt;97 --&gt;sp|O35164|MCPT9_MOUSE
--&gt;61 --&gt;sp|P15119|MCPT2_MOUSE --&gt;50 --&gt;sp|Q06606|GRZ2_RAT -</span></p>

<p class=examplefilesmallCxSpLast><span lang=EN-GB>-&gt;54 --&gt;<span
class=GramE>sp</span>|P80931|MCT1A_SHEEP --&gt;40 --&gt;sp|Q90629|TRY3_CHICK
--&gt;41 --&gt;sp|P29786|TRY3_AEDAE</span></p>

</div>

<p class=MyNormal>This is probably the best way to generate a high quality
alignment of your two sequences when using a progressive method like ClustalW,
T-Coffee, Muscle or Mafft.</p>

<h2><a name="_Toc235985500"><span lang=EN-GB>Manipulating DNA sequences</span></a></h2>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985501">Translating
DNA sequences into Proteins</a></h3>

<p class=MyNormal>If your sequences are DNA coding sequences, it is always
safer to align them as proteins. Seq_reformat makes it easy for you to
translate your sequences:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small_dna.fasta -action
+translate -output fasta_seq</span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985502">Back-Translation
With the Bona-Fide DNA sequences</a></h3>

<p class=MyNormal>Once your sequences have been aligned, you may want to turn
your protein alignment back into a DNA alignment, either to do phylogeny, or
maybe in order to design PCR probes. To do so, use the following command:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small_dna.fasta -in2 sproteases_small.aln
-action +thread_dna_on_prot_aln -output clustalw</span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985503">Finding
the Bona-Fide Sequences for the Back-Translation</a></h3>

<p class=MyNormal>Use the online server Protogene, available from
www.tcoffee.org.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985504">Guessing
Your Back Translation</a> </h3>

<p class=MyNormal><span class=GramE>Back-translating</span> means turning a
protein sequence into a DNA sequence. If you do not have the original DNA
sequence, this operation will not be exact, owing to the fact that the genetic
code is degenerated. Yet, if a random-back translation is fine with you, you
can use the following command.</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small_dna.fasta -in2 sproteases_small.aln
-action +thread_dna_on_prot_aln<span style="mso-spacerun:yes">  </span>-output
clustalw</span></p>

</div>

<p class=MyNormal><span class=GramE>In<span style="mso-spacerun:yes"> 
</span>this</span> process, codons are chosen randomly. For instance, if an <span
class=GramE>amino-acid</span> has four codons, the back-translation process
will randomly select one of these. If you need more sophisticated
back-translations that take into account the codon bias, we suggest you use
more specific tools like: alpha.dmi.unict.it/~ctnyu/bbocushelp.html</p>

<h2><a name="_Toc235985505"><span lang=EN-GB>Fetching a Structure</span></a></h2>

<p class=MsoBodyText><span lang=EN-GB>There are many reasons why you may need a
structure. T-Coffee contains a powerful utility named <span class=SpellE><b
style='mso-bidi-font-weight:normal'><i style='mso-bidi-font-style:normal'>extract_from_pdb</i></b></span>
that makes it possible to fetch the PDB coordinates of a structure or its FASTA
sequence without requiring a local installation.</span></p>

<p class=MsoBodyText><span lang=EN-GB>By default, <span class=SpellE>extract_from_pdb</span>
will start looking for the structure in the current directory; it will then
look it up locally (PDB_DIR) and eventually try to fetch it from the web (via a
<span class=SpellE>wget</span> to www.rcsb.org). All these settings can be
customized using environment variables (see the last section).</span></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985506">Fetching
a PDB structure</a></h3>

<p class=MyNormal>If you want to fetch the chain E of the PDB structure 1PPG,
you can use:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg extract_from_pdb -infile 1PPGE</span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985507">Fetching
The Sequence of a PDB structure</a></h3>

<p class=MyNormal>To <span class=GramE>Fetch</span> the sequence, use:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg extract_from_pdb -infile 1PPGE -fasta</span></p>

</div>

<p class=MyNormal>Will fetch the <span class=SpellE>fasta</span> sequence.</p>

<p class=MsoBodyText><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<h3 style='mso-list:none'><a name="_Toc235985508">Adapting </a><span
class=SpellE><span style='mso-bookmark:_Toc235985508'>extract_from_pdb</span></span><span
style='mso-bookmark:_Toc235985508'> to your own environment</span></h3>

<p class=MyNormal>If you have the PDB installed locally, simply set the
variable PDB_DIR to the absolute location of the directory in which the PDB is
installed. The PDB can either be installed in its divided form or in its full
form.</p>

<p class=MyNormal>If the file you are looking for is neither in the current
directory nor in the local PDB version, <span class=SpellE>extract_from_pdb</span>
will try to fetch it from <span class=SpellE>rcsb</span>. If you do not want
this to happen, you should either set the environment variable NO_REMOTE_PDB_DIR
to 1 or use the -<span class=SpellE>no_remote_pdb_dir</span> flag:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>export
NO_REMOTE_PDB_FILE=1</span></p>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>or</span></p>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>t_coffee
-other_pg extract_from_pdb -infile 1PPGE -fasta -no_remote_pdb_file</span></p>

</div>

<h1><a name="_Toc235985509"><span lang=EN-GB>Manipulating RNA sequences with
seq_reformat</span></a></h1>

<h2><a name="_Toc235985510"><span lang=EN-GB>Producing a Stockholm output:
adding predicted secondary structures</span></a></h2>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985511">Producing
a consensus structure</a></h3>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MyNormal>Given <span class=GramE>an RNA</span> multiple sequence
alignment, it is possible to compute the <span class=SpellE>alifold</span>
(Vienna package) consensus secondary structure and output in in <span
class=SpellE>stockholm</span>:</p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_rnaseq2.aln -action +aln2alifold
-output stockholm_aln</span></p>

</div>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985512">Adding
a consensus structure to an alignment</a></h3>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_rnaseq2.aln -action +add_alifold
-output stockholm_aln</span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'>Adding a pre-computed
consensus structure to an alignment</h3>

<p class=MyNormal>The file sample_rnaseq2.aalifold contains the raw output of
the <span class=SpellE>alifold</span> program captured as follows:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>RNAalifold
&lt;sample_rnaseq2.aln<span style="mso-spacerun:yes">  </span>&gt;
sample_rnaseq2.alifold</span></p>

</div>

<p class=MyNormal>It is possible to add this secondary structure to an
alignment using:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_rnaseq2.aln -in2
sample_rnaseq2.alifold -input2 alifold -action +add_alifold -output
stockholm_aln</span></p>

</div>

<div style='mso-element:para-border-div;border:solid windowtext 1.5pt;
padding:0cm 0cm 0cm 0cm;background:#F7CBB7'>

<p class=Warning><span lang=EN-GB>WARNING:</span></p>

<p class=Warning><span lang=EN-GB>The <span class=SpellE>alifold</span>
structure and the alignment MUST be compatible. The function makes no attempt
to thread or align the structure. It merely <span class=GramE>stack</span> it
below the MSA.</span></p>

<p class=Warning><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

</div>

<p class=MyNormal><a name="_Toc235985513">It is also possible to stack
Stockholm formatted secondary structures:</a></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc235985513'><span lang=ZU>PROMPT: seq_reformat -in
sample_rnaseq2.aln -in2 sample_rnaseq2.cons.stk -action +add_alifold -output
stockholm_aln</span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc235985513'><o:p>&nbsp;</o:p></span></p>

<h2><span style='mso-bookmark:_Toc235985513'><span lang=EN-GB>Analyzing an <span
class=SpellE>RNAalifold</span> secondary structure prediction</span></span></h2>

<p class=MsoBodyText><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<p class=MyNormal><span class=GramE>the</span> following commands can either be
applied on a Stockholm or a standard MSA. In the second case (standard MSA) the
secondary structure will be automatically re-computed by <span class=SpellE>alifold</span>.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985514">Analyzing
matching columns</a></h3>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MyNormal><b><i>+<span class=GramE>alifold2cov</span>_stat</i></b> will
estimate the number of pairs of columns that are perfect Watson and Crick, those
that are neutral (including a GU) and those that include correlated mutations.
The <span class=SpellE>WCcomp</span> are the compensated mutations maintaining
WC base pairing</p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_rnaseq2.stk -action +alifold2analyze
stat</span></p>

</div>

<p class=MyNormal>Other arguments can given, to display the list of paired
positions and their status (compensated, Watson, etc)</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_rnaseq2.stk -action +alifold2analyze
list</span></p>

</div>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985515">Visualizing
compensatory mutations</a></h3>

<p class=MyNormal>The following command will output a <span class=GramE>color
coded</span> version of your alignment with matching columns indicated as
follows:</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span>I: Incompatible
pair (i.e. at least one pair is not WC)</p>

<p class=MyNormal style='text-indent:28.25pt'>N: pairs are Gus or WC</p>

<p class=MyNormal style='text-indent:28.25pt'>W: All <span class=GramE>pairs</span>
are <span class=GramE>Watson</span></p>

<p class=MyNormal style='text-indent:28.25pt'><span class=GramE>c</span> :
Compensatory mutations</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span>C: WC
compensatory mutations</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_rnaseq2.aln -action +alifold2analyze
aln</span></p>

</div>

<p class=MyNormal>It is possible to turn this output into a colored one using:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sample_rnaseq2.aln -action +alifold2analyze
color_htm</span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985516">Handling
gapped columns</a></h3>

<p class=MyNormal><span class=GramE>by</span> default gapped column are ignored
but they can be included by adding the tag <span class=SpellE>usegap</span></p>

<h2><a name="_Toc235985517"><span lang=EN-GB>Comparing alternative folds</span></a></h2>

<p class=MyNormal>The folds associated with alternative alignments can be
compared. This comparison involves counting how many identical pairs of
residues are predicted on each sequence in one fold and in the other. The folds
can either be provided via Stockholm alignments</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>t_coffee
-other_pg seq_reformat -in sample_rnaseq2.cw.stk -in2
sample_rnaseq2.tcoffee.stk -action +RNAfold_cmp</span></p>

</div>

<p class=MyNormal>The top of the output (@@lines) summarizes the results that
are displayed on the -in alignment. If the provided alignments do not have a
fold, this fold will be estimated with <span class=SpellE>alifold</span>.</p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<h1><a name="_Toc235985518"><span lang=EN-GB>Phylogenetic Trees Manipulation with
seq_reformat</span></a></h1>

<h2><a name="_Toc235985519"><span lang=EN-GB>Producing phylogenetic trees</span></a></h2>

<p class=MyNormal>Seq_reformat is NOT a phylogeny package, yet over the time it
has accumulated a few functions that make it possible to compute simple
phylogenetic trees, or similar types of clustering:</p>

<p class=MyNormal>Given a multiple sequence alignment, it is possible to
compute either a UPGM or an NJ tree:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>seq_reformat
-in &lt;aln&gt; -action +aln2tree -output newick </span></p>

</div>

<p class=MyNormal>Will use an identity matrix to compare your sequences and
will output an <span class=SpellE>unrooted</span> NJ tree in newick format. If
you want to produce a rooted UPGMA tree:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>seq_reformat
-in &lt;aln&gt; -action +aln2tree _TMODE_upgma -output newick </span></p>

</div>

<p class=MyNormal>If your data is not data sequence, but a matrix of 1 and Os
(i.e. SAR matrix for instance), you can use a different matrix to compute the
pairwise distances:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU><span
style='mso-tab-count:1'>   </span>seq_reformat -in &lt;aln&gt; -action
+aln2tree _MATRIX_sarmat -output newick </span></p>

</div>

<p class=MyNormal>All these parameters can be concatenated:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU><span
style='mso-tab-count:1'>   </span>seq_reformat -in &lt;aln&gt; -action
+aln2tree _TMODE_upgma_MATRIX_sarmat -output newick </span></p>

</div>

<p class=MyNormal>Bootstrap facilities will also be added at some point  <span
class=GramE>For</span> now we recommend you use <span class=SpellE>Phylip</span>
if you need some serious phylogeny</p>

<p class=MyNormal style='margin-left:0cm'><o:p>&nbsp;</o:p></p>

<h2><a name="_Toc235985520"><span lang=EN-GB>Comparing two phylogenetic trees</span></a></h2>

<p class=MyNormal>Consider the following file (sample_tree1.dnd)</p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmall><span style="mso-spacerun:yes"> </span><span
lang=EN-GB>(<span class=GramE>( A:0.50000</span>, C:0.50000):0.00000,(
D:0.00500, E:0.00500):0.99000, B:0.50000);</span></p>

</div>

<p class=MyNormal><span class=GramE>and</span> the file sample_tree3.dnd.</p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmall><span style="mso-spacerun:yes"> </span><span
lang=EN-GB>(<span class=GramE>( E:0.50000</span>, C:0.50000):0.00000,(
A:0.00500, B:0.00500):0.99000, D:0.50000);</span></p>

</div>

<p class=MyNormal>You can compare them using:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:6.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC;border:none;
mso-border-top-alt:solid black .25pt;mso-border-bottom-alt:solid black .25pt;
padding:0cm;mso-padding-alt:6.0pt 0cm 10.0pt 0cm'><span lang=ZU>seq_reformat
-in sample_tree2.dnd -in2 sample_tree3.dnd -action +tree_cmp -output newick</span></p>

</div>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmallCxSpFirst><span class=SpellE><span class=GramE><span
lang=EN-GB>tree</span></span><span lang=EN-GB>_cpm|T</span></span><span
lang=EN-GB>: 75 W: 71.43 L: 50.50</span></p>

<p class=examplefilesmallCxSpMiddle><span class=GramE><span lang=EN-GB>tree</span></span><span
lang=EN-GB>_cpm|8 Nodes in T1 with 5 Sequences</span></p>

<p class=examplefilesmallCxSpMiddle><span class=SpellE><span class=GramE><span
lang=EN-GB>tree</span></span><span lang=EN-GB>_cmp|T</span></span><span
lang=EN-GB>: ratio of identical nodes</span></p>

<p class=examplefilesmallCxSpMiddle><span class=SpellE><span class=GramE><span
lang=EN-GB>tree</span></span><span lang=EN-GB>_cmp|W</span></span><span
lang=EN-GB>: ratio of identical nodes weighted with the min <span class=SpellE>Nseq</span>
below node</span></p>

<p class=examplefilesmallCxSpMiddle><span class=SpellE><span class=GramE><span
lang=EN-GB>tree</span></span><span lang=EN-GB>_cmp|L</span></span><span
lang=EN-GB>: average branch length similarity</span></p>

<p class=examplefilesmallCxSpLast><span lang=EN-GB>(<span class=GramE>(
A:1.00000</span>, C:1.00000):-2.00000,( D:1.00000, E:1.00000):-2.00000,
B:1.00000);</span></p>

</div>

<p class=MyNormal>Please consider the following aspects when exploiting these
results:</p>

<p class=MyNormal style='text-indent:28.25pt'>-The comparison is made on the <span
class=SpellE>unrooted</span> trees</p>

<p class=MyNormal style='margin-left:70.8pt'>T: Fraction of the branches
conserved between the two trees. This is obtained by considering the split
induced by each branch and by checking whether that split is found in both
trees</p>

<p class=MyNormal style='margin-left:70.8pt'>W: Fraction of the branches
conserved between the two trees. Each branch is weighted with MIN the minimum
number of leaf on its left or right (Number leaf left, Number leaf Right)</p>

<p class=MyNormal style='text-indent:28.25pt'>L: Fraction of branch length
difference between the two considered trees.</p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MyNormal>The last portion of the output contains a tree where
distances have been replaced by the number of leaf under the considered node</p>

<p class=MyNormal style='text-indent:28.25pt'>Positive values (i.e. 2, 5)
indicate a node common to both trees and correspond to MIN.</p>

<p class=MyNormal style='text-indent:28.25pt'>Negative values indicate a node
found in tree1 but not in tree2</p>

<p class=MyNormal style='text-indent:28.25pt'>The higher this value, the deeper
the node.</p>

<p class=MyNormal>You can extract this tree for further usage by typing:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU><span
style='mso-tab-count:1'>   </span>cat outfile | grep -v &quot;tree_cmp&quot;</span></p>

</div>

<h2><a name="_Toc235985521"><span lang=EN-GB>Scanning Phylogenetic Trees</span></a></h2>

<p class=MyNormal>It is possible to scan an alignment and locally measure the
similarity between an estimated local tree and some reference tree provided
from an external source (or computed on the fly). The following command:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>seq_reformat
-in &lt;aln&gt; -in2 &lt;reftree&gt; -action +tree_scan _MODE_scan__W_10_ &gt;
ph_tree_scan.txt </span></p>

</div>

<p class=MyNormal>For each position of the alignment, W*2 blocks of size 2*1+1
up to W*2+<span class=GramE>1<span style="mso-spacerun:yes">  </span>will</span>
be extracted, for each of these block a tree will be estimated and the
similarity of that tree with the reference tree will be estimated with <span
class=SpellE>cmp_tree</span>. For each position, the tree giving the best fit
will be reported, along with the size of the block leading to that tree:</p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmallCxSpFirst><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>P: &lt;position&gt;<span
style="mso-spacerun:yes">  </span>&lt;block start&gt; &lt;<span class=SpellE>blck_end</span>&gt;
&lt;block score&gt; &lt;block Length&gt;</span></p>

<p class=examplefilesmallCxSpLast><span lang=EN-GB><span
style="mso-spacerun:yes"> </span></span></p>

</div>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<h2><a name="_Toc235985522"><span lang=EN-GB>Pruning Phylogenetic Trees</span></a></h2>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MyNormal>Pruning removes leaves from an existing tree and <span
class=SpellE>recomputes</span> distances so that no information is lost</p>

<p class=MyNormal>Consider the file sample_tree2.dnd:</p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmall><span style="mso-spacerun:yes"> </span><span
lang=EN-GB>(<span class=GramE>( A:0.50000</span>, C:0.50000):0.00000,(
D:0.00500, E:0.00500):0.99000, B:0.50000);</span></p>

</div>

<p class=MyNormal>And the file sample_seq8.seq</p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmallCxSpFirst><span lang=EN-GB>&gt;A</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;B</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;C</span></p>

<p class=examplefilesmallCxSpLast><span lang=EN-GB>&gt;D</span></p>

</div>

<div style='mso-element:para-border-div;border:solid windowtext 1.5pt;
padding:1.0pt 4.0pt 1.0pt 4.0pt;background:#FFFF89'>

<p class=StylenoteboxbodyBoldNotItalicBoxSinglesolidlineAut><span lang=EN-GB>Note:
Sample_seq8 is merely a FASTA file where sequences can be omitted. Sequences
can be omitted, but you can also leave them, at your entire convenience.</span></p>

</div>

<p class=MsoBodyText><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>seq_reformat
-in sample_tree2.dnd -in2 sample_seq8.seq -action +tree_prune -output newick</span></p>

</div>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmall><span lang=ZU><span style="mso-spacerun:yes"> </span></span><span
lang=EN-GB>(<span class=GramE>( A:0.50000</span>, C:0.50000):0.00000,
B:0.50000, D:0.99500);</span></p>

</div>

<p class=MsoBodyText><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<h2><span lang=EN-GB><o:p>&nbsp;</o:p></span></h2>

</div>

<span lang=EN-GB style='font-size:12.0pt;mso-bidi-font-size:10.0pt;font-family:
"Times New Roman";mso-fareast-font-family:"Times New Roman";letter-spacing:
-.25pt;mso-ansi-language:EN-GB;mso-fareast-language:EN-US;mso-bidi-language:
AR-SA'><br clear=all style='page-break-before:auto;mso-break-type:section-break'>
</span>

<div class=WordSection2>

<h1><a name="_Toc235985523"><span lang=EN-GB>Building Multiple Sequence
Alignments</span></a></h1>

<h2><a name="_Toc235985524"><span lang=EN-GB>How to generate The Alignment You
Need?</span></a></h2>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985525">What
is a Good Alignment?</a></h3>

<p class=MyNormal>This is a trick question. A good alignment is an alignment
that makes it possible to do good biology. If you want to reconstruct a
phylogeny, a good alignment will be an alignment leading to an accurate
reconstruction.</p>

<p class=MyNormal>In practice, the alignment community has become used to
measuring the accuracy of alignment methods using structures. Structures are
relatively easy to align correctly, even when the sequences have diverged quite
a lot. The most common usage is therefore to compare structure based alignments
with their <span class=GramE>sequence based</span> counterpart and to evaluate
the accuracy of the method using these criterions.</p>

<p class=MyNormal>Unfortunately it is not easy to establish <span class=GramE>structure
based</span> standards of truth. Several of these exist and they do not
necessarily agree. To summarize, the situation is as roughly as follows:</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span>Above 40%
identity (within the reference datasets), all the reference collections agree
with one another and all the established methods give roughly the same results.
These alignments can be trusted blindly.</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span>Below 40%
accuracy within the reference datasets, the reference collections stop agreeing
and the methods do not give consistent results. In this area of similarity it
is not necessarily easy to determine who is right and who is wrong, although
most studies seem to indicate that consistency based methods (T-Coffee,
Mafft-slow and ProbCons) have an edge over traditional methods.</p>

<p class=MyNormal>When dealing with distantly related sequences, the only way
to produce reliable alignments is to us structural information. T-Coffee
provides many facilities to do so in a seamless fashion. Several important
factors need to be taken into account when selecting an alignment method: </p>

<p class=MyNormal>-The best methods are not <b style='mso-bidi-font-weight:
normal'><i style='mso-bidi-font-style:normal'>always</i></b> doing best. Given
a difficult dataset, the best method is only more likely to deliver the best
alignment, but there is no guaranty it will do so. It is very much like betting
on the horse with the best odds.</p>

<p class=MyNormal>-Secondly, the difference in accuracy (as measured on
reference datasets<span class=GramE>)<span style="mso-spacerun:yes"> 
</span>between</span> all the available methods is not incredibly high. It is
unclear whether this is an artifact caused by the use of &quot;easy&quot;
reference alignments, or whether this is a reality. The only thing that can
change dramatically the accuracy of the alignment is the use of structural
information.</p>

<p class=MyNormal>Last, but not least, bear in mind that these methods have
only been evaluated by comparison with reference structure based sequence
alignments. This is merely one criterion among many. In theory, these methods
should be evaluated for their ability to produce alignments that lead to
accurate trees, good profiles or good models. Unfortunately, these evaluation
procedures do not yet exist.</p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985526">The
Main Methods and their Scope</a></h3>

<p class=MyNormal>There are many MSA packages around. The main ones are
ClustalW, Muscle, Mafft, T-Coffee and ProbCons. You can almost forget about the
other packages, as there is virtually nothing you could do with them that you
will not be able to do with these packages.</p>

<p class=MyNormal>These packages offer a complex trade-off between speed,
accuracy and versatility. </p>

<h4>ClustalW: everywhere you look</h4>

<p class=MyNormal>ClustalW is still the most widely used multiple sequence
alignment package. Yet things are gradually changing as recent tests have consistently
shown that ClustalW is neither the most accurate nor the fastest package around.
This being said, ClustalW is everywhere and if your sequences are similar
enough, it should deliver a fairly reasonable alignment.</p>

<h4>Mafft and Muscle: Aligning Many Sequences</h4>

<p class=MyNormal>If you have many sequences to align Muscle or Mafft are the
obvious choice. Mafft is often described as the fastest and the most efficient.
This is not entirely true. In its fast mode (<span style='font-size:8.5pt;
font-family:Arial;color:#111111'>FFT-NS-1)</span>, Mafft is similar to Muscle
and although it is fairly accurate it is about 5 points less accurate than the
consistency based packages (ProbCons and T-Coffee). In its most accurate mode (<span
style='font-size:8.5pt;font-family:Arial;color:#111111'>L-INS-<span
class=SpellE>i</span></span>) Mafft uses local alignments and consistency. It becomes
much more accurate but also slower, and more sensitive to the number of
sequences. </p>

<p class=MyNormal>The alignments generated using the fast modes of these
programs will be very suitable for several important applications such as:</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span>-Distance based
phylogenetic reconstruction (NJ trees)</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span>-Secondary
structure predictions</p>

<p class=MyNormal>However they may not be suitable for more refined application
such as</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span>-Profile
construction</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span>-Structure
Modeling</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span>-3D structure
prediction</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span>-Function
analysis</p>

<p class=MyNormal>In that case you may need to use more accurate methods</p>

<h4>T-Coffee and ProbCons: Slow and Accurate</h4>

<p class=MyNormal>T-Coffee works by first assembling a library and then by
turning this library into an alignment. The library is a list of potential
pairs of residues. All of them are not compatible and the job of the algorithm
is to make sure that as many possible constraints as possible find their way
into the final alignment. Each library line is a constraint and the purpose is
to assemble the alignment that accommodates the more all the constraints.</p>

<p class=MyNormal>It is very much like building a high school schedule, where
each teachers says something &quot;I need my Monday morning&quot;, &quot;I
can't come on Thursday afternoon&quot;, and so on. In the end you want a
schedule that makes everybody happy, if <span class=SpellE><span class=GramE>possible.The</span></span>
nice thing about the library is that it can be used as a media to combine as
many methods as one wishes. It is just a matter of generating the right
constraints with the right method and <span class=GramE>compile</span> them
into the library.</p>

<p class=MyNormal>ProbCons and Mafft (<span style='font-size:8.5pt;font-family:
Arial;color:#111111'>L-INS-<span class=SpellE>i</span></span>) <span
class=GramE>uses</span> a similar algorithm, but with a Bayesian twist in the
case of <span class=SpellE>Probcons</span>. In practice, however, <span
class=SpellE>probcons</span> and T-Coffee give very similar results and have
similar running time. Mafft is significantly faster. </p>

<p class=MyNormal>All these packages are ideal for the following applications:</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span>-Profile
reconstruction</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span>-Function
analysis</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span>-3D Prediction</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985527">Choosing
The Right Package</a></h3>

<p class=MyNormal>Each available package has something to go for it. It is just
a matter of knowing what you want to do. T-Coffee is probably the most
versatile, but it comes at a price and it is currently slower than many
alternative packages.</p>

<p class=MyNormal>In the rest of this tutorial we give some hints on how to
carry out each of these applications with T-Coffee.</p>

<table class=MsoNormalTable border=1 cellspacing=0 cellpadding=0 width=486
 style='width:486.0pt;margin-left:-48.6pt;border-collapse:collapse;border:none;
 mso-border-alt:double black 2.25pt;mso-yfti-tbllook:480;mso-padding-alt:0cm 5.4pt 0cm 5.4pt;
 mso-border-insideh:.75pt solid black;mso-border-insidev:.75pt solid black'>
 <tr style='mso-yfti-irow:0;mso-yfti-firstrow:yes'>
  <td width=157 valign=top style='width:157.25pt;border:double black 2.25pt;
  border-right:solid black 1.0pt;mso-border-alt:double black 2.25pt;mso-border-right-alt:
  solid black .75pt;background:#E6E6E6;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table><span style='text-transform:uppercase'><o:p>&nbsp;</o:p></span></p>
  </td>
  <td width=61 valign=top style='width:61.2pt;border-top:double black 2.25pt;
  border-left:none;border-bottom:double black 2.25pt;border-right:solid black 1.0pt;
  mso-border-left-alt:solid black .75pt;mso-border-top-alt:double 2.25pt;
  mso-border-left-alt:solid .75pt;mso-border-bottom-alt:double 2.25pt;
  mso-border-right-alt:solid .75pt;mso-border-color-alt:black;background:#E6E6E6;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table><span style='text-transform:uppercase'>Muscle<o:p></o:p></span></p>
  </td>
  <td width=56 valign=top style='width:56.0pt;border-top:double black 2.25pt;
  border-left:none;border-bottom:double black 2.25pt;border-right:solid black 1.0pt;
  mso-border-left-alt:solid black .75pt;mso-border-top-alt:double 2.25pt;
  mso-border-left-alt:solid .75pt;mso-border-bottom-alt:double 2.25pt;
  mso-border-right-alt:solid .75pt;mso-border-color-alt:black;background:#E6E6E6;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table><span style='text-transform:uppercase'>Mafft<o:p></o:p></span></p>
  </td>
  <td width=68 valign=top style='width:68.45pt;border-top:double black 2.25pt;
  border-left:none;border-bottom:double black 2.25pt;border-right:solid black 1.0pt;
  mso-border-left-alt:solid black .75pt;mso-border-top-alt:double 2.25pt;
  mso-border-left-alt:solid .75pt;mso-border-bottom-alt:double 2.25pt;
  mso-border-right-alt:solid .75pt;mso-border-color-alt:black;background:#E6E6E6;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table><span style='text-transform:uppercase'>ProbCons<o:p></o:p></span></p>
  </td>
  <td width=71 valign=top style='width:71.1pt;border-top:double black 2.25pt;
  border-left:none;border-bottom:double black 2.25pt;border-right:solid black 1.0pt;
  mso-border-left-alt:solid black .75pt;mso-border-top-alt:double 2.25pt;
  mso-border-left-alt:solid .75pt;mso-border-bottom-alt:double 2.25pt;
  mso-border-right-alt:solid .75pt;mso-border-color-alt:black;background:#E6E6E6;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table><span style='text-transform:uppercase'>T-Coffee<o:p></o:p></span></p>
  </td>
  <td width=72 valign=top style='width:72.0pt;border-top:double black 2.25pt;
  border-left:none;border-bottom:double black 2.25pt;border-right:solid black 1.0pt;
  mso-border-left-alt:solid black .75pt;mso-border-top-alt:double 2.25pt;
  mso-border-left-alt:solid .75pt;mso-border-bottom-alt:double 2.25pt;
  mso-border-right-alt:solid .75pt;mso-border-color-alt:black;background:#E6E6E6;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table><span style='text-transform:uppercase'>ClustalW<o:p></o:p></span></p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:1'>
  <td width=157 valign=top style='width:157.25pt;border-top:none;border-left:
  double black 2.25pt;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:double black 2.25pt;mso-border-top-alt:double 2.25pt;
  mso-border-left-alt:double 2.25pt;mso-border-bottom-alt:solid .75pt;
  mso-border-right-alt:solid .75pt;mso-border-color-alt:black;background:#E6E6E6;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>Accuracy</p>
  </td>
  <td width=61 valign=top style='width:61.2pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  double black 2.25pt;mso-border-left-alt:solid black .75pt;mso-border-alt:
  solid black .75pt;mso-border-top-alt:double black 2.25pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>++</p>
  </td>
  <td width=56 valign=top style='width:56.0pt;border-top:none;border-left:none;
  border-bottom:double black 2.25pt;border-right:solid black 1.0pt;mso-border-top-alt:
  double black 2.25pt;mso-border-left-alt:solid black .75pt;mso-border-top-alt:
  double 2.25pt;mso-border-left-alt:solid .75pt;mso-border-bottom-alt:double 2.25pt;
  mso-border-right-alt:solid .75pt;mso-border-color-alt:black;background:#FFFF99;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=68 valign=top style='width:68.45pt;border-top:none;border-left:
  none;border-bottom:double black 2.25pt;border-right:solid black 1.0pt;
  mso-border-top-alt:double black 2.25pt;mso-border-left-alt:solid black .75pt;
  mso-border-top-alt:double 2.25pt;mso-border-left-alt:solid .75pt;mso-border-bottom-alt:
  double 2.25pt;mso-border-right-alt:solid .75pt;mso-border-color-alt:black;
  background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=71 valign=top style='width:71.1pt;border-top:none;border-left:none;
  border-bottom:double black 2.25pt;border-right:solid black 1.0pt;mso-border-top-alt:
  double black 2.25pt;mso-border-left-alt:solid black .75pt;mso-border-top-alt:
  double 2.25pt;mso-border-left-alt:solid .75pt;mso-border-bottom-alt:double 2.25pt;
  mso-border-right-alt:solid .75pt;mso-border-color-alt:black;background:#FFFF99;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++<span style='mso-tab-count:1'>    </span></p>
  </td>
  <td width=72 valign=top style='width:72.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:double black 2.25pt;mso-border-top-alt:
  double black 2.25pt;mso-border-left-alt:solid black .75pt;mso-border-top-alt:
  double 2.25pt;mso-border-left-alt:solid .75pt;mso-border-bottom-alt:solid .75pt;
  mso-border-right-alt:double 2.25pt;mso-border-color-alt:black;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+</p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:2'>
  <td width=157 valign=top style='width:157.25pt;border-top:none;border-left:
  double black 2.25pt;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:double black 2.25pt;mso-border-top-alt:double 2.25pt;
  mso-border-left-alt:double 2.25pt;mso-border-bottom-alt:solid .75pt;
  mso-border-right-alt:solid .75pt;mso-border-color-alt:black;background:#E6E6E6;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>&lt;100 Seq.</p>
  </td>
  <td width=61 valign=top style='width:61.2pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  double black 2.25pt;mso-border-left-alt:solid black .75pt;mso-border-alt:
  solid black .75pt;mso-border-top-alt:double black 2.25pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>++</p>
  </td>
  <td width=56 valign=top style='width:56.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  double black 2.25pt;mso-border-left-alt:solid black .75pt;mso-border-alt:
  solid black .75pt;mso-border-top-alt:double black 2.25pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>++</p>
  </td>
  <td width=68 valign=top style='width:68.45pt;border-top:none;border-left:
  none;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:double black 2.25pt;mso-border-left-alt:solid black .75pt;
  mso-border-alt:solid black .75pt;mso-border-top-alt:double black 2.25pt;
  background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=71 valign=top style='width:71.1pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  double black 2.25pt;mso-border-left-alt:solid black .75pt;mso-border-alt:
  solid black .75pt;mso-border-top-alt:double black 2.25pt;background:#FFFF99;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=72 valign=top style='width:72.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:double black 2.25pt;mso-border-top-alt:
  double black 2.25pt;mso-border-left-alt:solid black .75pt;mso-border-top-alt:
  double 2.25pt;mso-border-left-alt:solid .75pt;mso-border-bottom-alt:solid .75pt;
  mso-border-right-alt:double 2.25pt;mso-border-color-alt:black;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+</p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:3'>
  <td width=157 valign=top style='width:157.25pt;border-top:none;border-left:
  double black 2.25pt;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-left-alt:double black 2.25pt;background:#E6E6E6;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>&gt;100 Seq.</p>
  </td>
  <td width=61 valign=top style='width:61.2pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=56 valign=top style='width:56.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=68 valign=top style='width:68.45pt;border-top:none;border-left:
  none;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-left-alt:solid black .75pt;
  mso-border-alt:solid black .75pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=71 valign=top style='width:71.1pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+</p>
  </td>
  <td width=72 valign=top style='width:72.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:double black 2.25pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-right-alt:double black 2.25pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+</p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:4'>
  <td width=157 valign=top style='width:157.25pt;border-top:none;border-left:
  double black 2.25pt;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-left-alt:double black 2.25pt;background:#E6E6E6;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>Remote Homologues</p>
  </td>
  <td width=61 valign=top style='width:61.2pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>++</p>
  </td>
  <td width=56 valign=top style='width:56.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=68 valign=top style='width:68.45pt;border-top:none;border-left:
  none;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-left-alt:solid black .75pt;
  mso-border-alt:solid black .75pt;background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=71 valign=top style='width:71.1pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=72 valign=top style='width:72.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:double black 2.25pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-right-alt:double black 2.25pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+</p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:5'>
  <td width=157 valign=top style='width:157.25pt;border-top:none;border-left:
  double black 2.25pt;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-left-alt:double black 2.25pt;background:#E6E6E6;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>MSA <span class=SpellE>vs</span> Seq.</p>
  </td>
  <td width=61 valign=top style='width:61.2pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=56 valign=top style='width:56.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=68 valign=top style='width:68.45pt;border-top:none;border-left:
  none;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-left-alt:solid black .75pt;
  mso-border-alt:solid black .75pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table><o:p>&nbsp;</o:p></p>
  </td>
  <td width=71 valign=top style='width:71.1pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=72 valign=top style='width:72.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:double black 2.25pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-right-alt:double black 2.25pt;background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:6'>
  <td width=157 valign=top style='width:157.25pt;border-top:none;border-left:
  double black 2.25pt;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-left-alt:double black 2.25pt;background:#E6E6E6;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>MSA <span class=SpellE>vs</span> MSA</p>
  </td>
  <td width=61 valign=top style='width:61.2pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=56 valign=top style='width:56.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=68 valign=top style='width:68.45pt;border-top:none;border-left:
  none;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-left-alt:solid black .75pt;
  mso-border-alt:solid black .75pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=71 valign=top style='width:71.1pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=72 valign=top style='width:72.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:double black 2.25pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-right-alt:double black 2.25pt;background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:7'>
  <td width=157 valign=top style='width:157.25pt;border-top:none;border-left:
  double black 2.25pt;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-left-alt:double black 2.25pt;background:#E6E6E6;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>&gt;2 MSAs</p>
  </td>
  <td width=61 valign=top style='width:61.2pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=56 valign=top style='width:56.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=68 valign=top style='width:68.45pt;border-top:none;border-left:
  none;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-left-alt:solid black .75pt;
  mso-border-alt:solid black .75pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=71 valign=top style='width:71.1pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=72 valign=top style='width:72.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:double black 2.25pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-right-alt:double black 2.25pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:8'>
  <td width=157 valign=top style='width:157.25pt;border-top:none;border-left:
  double black 2.25pt;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-left-alt:double black 2.25pt;background:#E6E6E6;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>Seq. <span class=SpellE>vs</span> <span class=SpellE>Struc</span>.</p>
  </td>
  <td width=61 valign=top style='width:61.2pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=56 valign=top style='width:56.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=68 valign=top style='width:68.45pt;border-top:none;border-left:
  none;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-left-alt:solid black .75pt;
  mso-border-alt:solid black .75pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=71 valign=top style='width:71.1pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=72 valign=top style='width:72.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:double black 2.25pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-right-alt:double black 2.25pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+</p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:9'>
  <td width=157 valign=top style='width:157.25pt;border-top:none;border-left:
  double black 2.25pt;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-left-alt:double black 2.25pt;background:#E6E6E6;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>Splicing Var.</p>
  </td>
  <td width=61 valign=top style='width:61.2pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=56 valign=top style='width:56.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=68 valign=top style='width:68.45pt;border-top:none;border-left:
  none;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-left-alt:solid black .75pt;
  mso-border-alt:solid black .75pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=71 valign=top style='width:71.1pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=72 valign=top style='width:72.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:double black 2.25pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-right-alt:double black 2.25pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:10'>
  <td width=157 valign=top style='width:157.25pt;border-top:none;border-left:
  double black 2.25pt;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-left-alt:double black 2.25pt;background:#E6E6E6;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>Reformat</p>
  </td>
  <td width=61 valign=top style='width:61.2pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=56 valign=top style='width:56.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=68 valign=top style='width:68.45pt;border-top:none;border-left:
  none;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-left-alt:solid black .75pt;
  mso-border-alt:solid black .75pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=71 valign=top style='width:71.1pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=72 valign=top style='width:72.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:double black 2.25pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-right-alt:double black 2.25pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>++</p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:11'>
  <td width=157 valign=top style='width:157.25pt;border-top:none;border-left:
  double black 2.25pt;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-left-alt:double black 2.25pt;background:#E6E6E6;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>Phylogeny</p>
  </td>
  <td width=61 valign=top style='width:61.2pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=56 valign=top style='width:56.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=68 valign=top style='width:68.45pt;border-top:none;border-left:
  none;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-left-alt:solid black .75pt;
  mso-border-alt:solid black .75pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=71 valign=top style='width:71.1pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+</p>
  </td>
  <td width=72 valign=top style='width:72.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:double black 2.25pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-right-alt:double black 2.25pt;background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>++</p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:12'>
  <td width=157 valign=top style='width:157.25pt;border-top:none;border-left:
  double black 2.25pt;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-left-alt:double black 2.25pt;background:#E6E6E6;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>Evaluation</p>
  </td>
  <td width=61 valign=top style='width:61.2pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=56 valign=top style='width:56.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
  <td width=68 valign=top style='width:68.45pt;border-top:none;border-left:
  none;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-left-alt:solid black .75pt;
  mso-border-alt:solid black .75pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+</p>
  </td>
  <td width=71 valign=top style='width:71.1pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=72 valign=top style='width:72.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:double black 2.25pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-right-alt:double black 2.25pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>-</p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:13;mso-yfti-lastrow:yes'>
  <td width=157 valign=top style='width:157.25pt;border-top:none;border-left:
  double black 2.25pt;border-bottom:double black 2.25pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-top-alt:solid .75pt;
  mso-border-left-alt:double 2.25pt;mso-border-bottom-alt:double 2.25pt;
  mso-border-right-alt:solid .75pt;mso-border-color-alt:black;background:#E6E6E6;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>Speed</p>
  </td>
  <td width=61 valign=top style='width:61.2pt;border-top:none;border-left:none;
  border-bottom:double black 2.25pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-bottom-alt:double black 2.25pt;background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=56 valign=top style='width:56.0pt;border-top:none;border-left:none;
  border-bottom:double black 2.25pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-bottom-alt:double black 2.25pt;background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=68 valign=top style='width:68.45pt;border-top:none;border-left:
  none;border-bottom:double black 2.25pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-left-alt:solid black .75pt;
  mso-border-alt:solid black .75pt;mso-border-bottom-alt:double black 2.25pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+</p>
  </td>
  <td width=71 valign=top style='width:71.1pt;border-top:none;border-left:none;
  border-bottom:double black 2.25pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-bottom-alt:double black 2.25pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+</p>
  </td>
  <td width=72 valign=top style='width:72.0pt;border-top:none;border-left:none;
  border-bottom:double black 2.25pt;border-right:double black 2.25pt;
  mso-border-top-alt:solid black .75pt;mso-border-left-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>++</p>
  </td>
 </tr>
</table>

<p class=MyNormal><b style='mso-bidi-font-weight:normal'>Table 1.</b> Relative possibilities
associated with the main packages (T-Coffee Tutorial, C. Notredame,
www.tcoffee.org). In any of the situations corresponding to each table line,
(+++) indicates that the method is the best suited, (++) indicates that the
method is not optimal but behaves reasonably well, (+) indicates that it is
possible but not recommended (-) indicates that the option is not available. </p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<table class=MsoNormalTable border=1 cellspacing=0 cellpadding=0 width=495
 style='width:495.0pt;margin-left:-57.6pt;border-collapse:collapse;border:none;
 mso-border-alt:double black 2.25pt;mso-yfti-tbllook:480;mso-padding-alt:0cm 5.4pt 0cm 5.4pt;
 mso-border-insideh:.75pt solid black;mso-border-insidev:.75pt solid black'>
 <tr style='mso-yfti-irow:0;mso-yfti-firstrow:yes'>
  <td width=166 valign=top style='width:166.25pt;border:double black 2.25pt;
  border-right:solid black 1.0pt;mso-border-alt:double black 2.25pt;mso-border-right-alt:
  solid black .75pt;background:#E6E6E6;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table><span style='text-transform:uppercase'><o:p>&nbsp;</o:p></span></p>
  </td>
  <td width=61 valign=top style='width:61.2pt;border-top:double black 2.25pt;
  border-left:none;border-bottom:double black 2.25pt;border-right:solid black 1.0pt;
  mso-border-left-alt:solid black .75pt;mso-border-top-alt:double 2.25pt;
  mso-border-left-alt:solid .75pt;mso-border-bottom-alt:double 2.25pt;
  mso-border-right-alt:solid .75pt;mso-border-color-alt:black;background:#E6E6E6;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table><span style='text-transform:uppercase'>Muscle<o:p></o:p></span></p>
  </td>
  <td width=56 valign=top style='width:56.0pt;border-top:double black 2.25pt;
  border-left:none;border-bottom:double black 2.25pt;border-right:solid black 1.0pt;
  mso-border-left-alt:solid black .75pt;mso-border-top-alt:double 2.25pt;
  mso-border-left-alt:solid .75pt;mso-border-bottom-alt:double 2.25pt;
  mso-border-right-alt:solid .75pt;mso-border-color-alt:black;background:#E6E6E6;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table><span style='text-transform:uppercase'>Mafft<o:p></o:p></span></p>
  </td>
  <td width=68 valign=top style='width:68.45pt;border-top:double black 2.25pt;
  border-left:none;border-bottom:double black 2.25pt;border-right:solid black 1.0pt;
  mso-border-left-alt:solid black .75pt;mso-border-top-alt:double 2.25pt;
  mso-border-left-alt:solid .75pt;mso-border-bottom-alt:double 2.25pt;
  mso-border-right-alt:solid .75pt;mso-border-color-alt:black;background:#E6E6E6;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table><span style='text-transform:uppercase'>ProbCons<o:p></o:p></span></p>
  </td>
  <td width=71 valign=top style='width:71.1pt;border-top:double black 2.25pt;
  border-left:none;border-bottom:double black 2.25pt;border-right:solid black 1.0pt;
  mso-border-left-alt:solid black .75pt;mso-border-top-alt:double 2.25pt;
  mso-border-left-alt:solid .75pt;mso-border-bottom-alt:double 2.25pt;
  mso-border-right-alt:solid .75pt;mso-border-color-alt:black;background:#E6E6E6;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table><span style='text-transform:uppercase'>T-Coffee<o:p></o:p></span></p>
  </td>
  <td width=72 valign=top style='width:72.0pt;border-top:double black 2.25pt;
  border-left:none;border-bottom:double black 2.25pt;border-right:solid black 1.0pt;
  mso-border-left-alt:solid black .75pt;mso-border-top-alt:double 2.25pt;
  mso-border-left-alt:solid .75pt;mso-border-bottom-alt:double 2.25pt;
  mso-border-right-alt:solid .75pt;mso-border-color-alt:black;background:#E6E6E6;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table><span style='text-transform:uppercase'>ClustalW<o:p></o:p></span></p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:1'>
  <td width=166 valign=top style='width:166.25pt;border-top:none;border-left:
  double black 2.25pt;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:double black 2.25pt;mso-border-top-alt:double 2.25pt;
  mso-border-left-alt:double 2.25pt;mso-border-bottom-alt:solid .75pt;
  mso-border-right-alt:solid .75pt;mso-border-color-alt:black;background:#E6E6E6;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>Dist Based Phylogeny</p>
  </td>
  <td width=61 valign=top style='width:61.2pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  double black 2.25pt;mso-border-left-alt:solid black .75pt;mso-border-alt:
  solid black .75pt;mso-border-top-alt:double black 2.25pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=56 valign=top style='width:56.0pt;border-top:none;border-left:none;
  border-bottom:double black 2.25pt;border-right:solid black 1.0pt;mso-border-top-alt:
  double black 2.25pt;mso-border-left-alt:solid black .75pt;mso-border-top-alt:
  double 2.25pt;mso-border-left-alt:solid .75pt;mso-border-bottom-alt:double 2.25pt;
  mso-border-right-alt:solid .75pt;mso-border-color-alt:black;background:#FFFF99;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=68 valign=top style='width:68.45pt;border-top:none;border-left:
  none;border-bottom:double black 2.25pt;border-right:solid black 1.0pt;
  mso-border-top-alt:double black 2.25pt;mso-border-left-alt:solid black .75pt;
  mso-border-top-alt:double 2.25pt;mso-border-left-alt:solid .75pt;mso-border-bottom-alt:
  double 2.25pt;mso-border-right-alt:solid .75pt;mso-border-color-alt:black;
  background:white;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>++</p>
  </td>
  <td width=71 valign=top style='width:71.1pt;border-top:none;border-left:none;
  border-bottom:double black 2.25pt;border-right:solid black 1.0pt;mso-border-top-alt:
  double black 2.25pt;mso-border-left-alt:solid black .75pt;mso-border-top-alt:
  double 2.25pt;mso-border-left-alt:solid .75pt;mso-border-bottom-alt:double 2.25pt;
  mso-border-right-alt:solid .75pt;mso-border-color-alt:black;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>++<span style='mso-tab-count:1'>     </span></p>
  </td>
  <td width=72 valign=top style='width:72.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:double black 2.25pt;mso-border-top-alt:
  double black 2.25pt;mso-border-left-alt:solid black .75pt;mso-border-top-alt:
  double 2.25pt;mso-border-left-alt:solid .75pt;mso-border-bottom-alt:solid .75pt;
  mso-border-right-alt:double 2.25pt;mso-border-color-alt:black;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>++</p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:2'>
  <td width=166 valign=top style='width:166.25pt;border-top:none;border-left:
  double black 2.25pt;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:double black 2.25pt;mso-border-top-alt:double 2.25pt;
  mso-border-left-alt:double 2.25pt;mso-border-bottom-alt:solid .75pt;
  mso-border-right-alt:solid .75pt;mso-border-color-alt:black;background:#E6E6E6;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>ML or MP Phylogeny</p>
  </td>
  <td width=61 valign=top style='width:61.2pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  double black 2.25pt;mso-border-left-alt:solid black .75pt;mso-border-alt:
  solid black .75pt;mso-border-top-alt:double black 2.25pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>++</p>
  </td>
  <td width=56 valign=top style='width:56.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  double black 2.25pt;mso-border-left-alt:solid black .75pt;mso-border-alt:
  solid black .75pt;mso-border-top-alt:double black 2.25pt;background:#FFFF99;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=68 valign=top style='width:68.45pt;border-top:none;border-left:
  none;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:double black 2.25pt;mso-border-left-alt:solid black .75pt;
  mso-border-alt:solid black .75pt;mso-border-top-alt:double black 2.25pt;
  background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=71 valign=top style='width:71.1pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  double black 2.25pt;mso-border-left-alt:solid black .75pt;mso-border-alt:
  solid black .75pt;mso-border-top-alt:double black 2.25pt;background:#FFFF99;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=72 valign=top style='width:72.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:double black 2.25pt;mso-border-top-alt:
  double black 2.25pt;mso-border-left-alt:solid black .75pt;mso-border-top-alt:
  double 2.25pt;mso-border-left-alt:solid .75pt;mso-border-bottom-alt:solid .75pt;
  mso-border-right-alt:double 2.25pt;mso-border-color-alt:black;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>++</p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:3'>
  <td width=166 valign=top style='width:166.25pt;border-top:none;border-left:
  double black 2.25pt;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-left-alt:double black 2.25pt;background:#E6E6E6;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>Profile Construction</p>
  </td>
  <td width=61 valign=top style='width:61.2pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>++</p>
  </td>
  <td width=56 valign=top style='width:56.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=68 valign=top style='width:68.45pt;border-top:none;border-left:
  none;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-left-alt:solid black .75pt;
  mso-border-alt:solid black .75pt;background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=71 valign=top style='width:71.1pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=72 valign=top style='width:72.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:double black 2.25pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-right-alt:double black 2.25pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>++</p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:4'>
  <td width=166 valign=top style='width:166.25pt;border-top:none;border-left:
  double black 2.25pt;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-left-alt:double black 2.25pt;background:#E6E6E6;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>3D Modeling</p>
  </td>
  <td width=61 valign=top style='width:61.2pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>++</p>
  </td>
  <td width=56 valign=top style='width:56.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>++</p>
  </td>
  <td width=68 valign=top style='width:68.45pt;border-top:none;border-left:
  none;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-left-alt:solid black .75pt;
  mso-border-alt:solid black .75pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>++</p>
  </td>
  <td width=71 valign=top style='width:71.1pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=72 valign=top style='width:72.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:double black 2.25pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-right-alt:double black 2.25pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+</p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:5;mso-yfti-lastrow:yes'>
  <td width=166 valign=top style='width:166.25pt;border-top:none;border-left:
  double black 2.25pt;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-left-alt:double black 2.25pt;background:#E6E6E6;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>Secondary Structure P</p>
  </td>
  <td width=61 valign=top style='width:61.2pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=56 valign=top style='width:56.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  background:#FFFF99;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>+++</p>
  </td>
  <td width=68 valign=top style='width:68.45pt;border-top:none;border-left:
  none;border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;
  mso-border-top-alt:solid black .75pt;mso-border-left-alt:solid black .75pt;
  mso-border-alt:solid black .75pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>++</p>
  </td>
  <td width=71 valign=top style='width:71.1pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:solid black 1.0pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>++</p>
  </td>
  <td width=72 valign=top style='width:72.0pt;border-top:none;border-left:none;
  border-bottom:solid black 1.0pt;border-right:double black 2.25pt;mso-border-top-alt:
  solid black .75pt;mso-border-left-alt:solid black .75pt;mso-border-alt:solid black .75pt;
  mso-border-right-alt:double black 2.25pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=Table>++</p>
  </td>
 </tr>
</table>

<p class=MyNormal><b style='mso-bidi-font-weight:normal'>Table 2.</b> Most
Suitable <span class=SpellE>Appplications</span> of each package (T-Coffee
Tutorial, C. Notredame, www.tcoffee.org). In any of the situations
corresponding to each table line, (+++) indicates that the method is the best
suited, (++) indicates that the method is not optimal but behaves reasonably
well, (+) indicates that it is possible but not recommended (-) indicates that
the option is not available. </p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<h2><a name="_Toc235985528"><span lang=EN-GB>Computing Multiple Sequence Alignments
With T-Coffee</span></a></h2>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985529">Computing
Very accurate (but slow) alignments with PSI-Coffee</a></h3>

<p class=MyNormal>PSI-Coffee builds a profile associated with each of your
input sequence and then makes a multiple profile alignment. If you do not have
any structure, it is the most accurate mode of T-Coffee.</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sproteases_small.fasta -mode psicoffee</span></p>

</div>

<p class=MyNormal>If you want to go further, and be even slower, you can use
the accurate mode that will combine profile and structural information</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sproteases_small.fasta -mode accurate</span></p>

</div>

<p class=MyNormal>It is probably one of the most accurate <span class=GramE>way</span>
of aligning sequences currently available.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985530">A
Simple Multiple Sequence Alignment</a></h3>

<p class=MyNormal>T-Coffee <span class=GramE>is meant to be run</span> like
ClustalW. This means you can use it like ClustalW for most simple applications.
For instance, the following instruction</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sproteases_small.fasta</span></p>

</div>

<p class=MyNormal>This instruction will compute a multiple sequence alignment
of your sequences, using the default mode of T-Coffee. It will output the
alignment on the screen and in a file named <span class=SpellE>sproteases_small.aln</span>.
This file contains your alignment in ClustalW format.</p>

<p class=MyNormal>The program will also output a file named <span class=SpellE>sproteases_small.dnd</span>
that contains the guide tree used to assemble the progressive alignment.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985531">Controlling
the Output Format</a></h3>

<p class=MyNormal>If you need to, you can also trigger different <span
class=SpellE>ouput</span> formats using the -output flag:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sproteases_small.fasta -output=clustalw,fasta_aln,msf</span></p>

</div>

<p class=MyNormal>You can specify as many formats as you want.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985532">Computing
a Phylogenetic tree</a></h3>

<p class=MyNormal>T-Coffee is not a phylogeny package. Yet, it has some limited
abilities to turn your MSA into a phylogenetic tree. This tree is a Neighbor
Joining Phylogenetic <span class=GramE>tree,</span> very similar to the one you
could compute using ClustalW. </p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sproteases_small.fasta -output=clustalw,fasta_aln,msf</span></p>

</div>

<p class=MyNormal>The phylogenetic tree is the file with the ph extension.
Never use the .<span class=SpellE>dnd</span> tree in place of a genuine
phylogenetic tree. The phylogenetic tree output by T-Coffee is only an
indication. You should produce a bootstrapped phylogenetic tree using packages
like <span class=SpellE>Phylip</span> (bioweb.pasteur.fr/seqanal/phylogeny/phylip-uk.html).
You can visualize your tree using online tree drawing programs like <span
class=SpellE>phylodendron</span> (iubio.bio.indiana.edu/treeapp/treeprint-form.html).</p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985533">Using
Several Datasets</a></h3>

<p class=MyNormal>If your sequences are spread across several datasets, you can
give all the files via the -<span class=SpellE>seq</span> flag:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -seq=sprotease1_small.fasta,sprotease2_small.aln
-output=clustalw,fasta_aln,msf</span></p>

</div>

<p class=MyNormal>Note that you can give as many file as you want (the limit is
200) and that the files can be in any format. If you give an alignment, the
gaps will be reset and your alignment will only provide sequences.</p>

<p class=MyNormal>Sequences with the same name between two files are assumed to
be the same sequence. If their sequences differ, they will be aligned and
replaced by the consensus of that alignment. This process is known as sequence
reconciliation.</p>

<div style='mso-element:para-border-div;border:solid windowtext 1.5pt;
padding:0cm 0cm 0cm 0cm;background:#F7CBB7'>

<p class=Warning><span style='mso-ansi-language:EN-US'>You should make sure
that there are no duplicates in your alignment, especially when providing
multiple datasets.<o:p></o:p></span></p>

</div>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985534">How
</a><span class=GramE><span style='mso-bookmark:_Toc235985534'>Good</span></span><span
style='mso-bookmark:_Toc235985534'> is Your Alignment</span></h3>

<p class=MyNormal>Later in this tutorial we show you how to estimate the
accuracy of your alignment. Before we go into details, you should know that the
number that comes on the first line of the header (in ClustalW format) is the
score of your alignment. </p>

<p class=MyNormal>CLUSTAL FORMAT for T-COFFEE Version_4.32
[http://www.tcoffee.org], CPU=19.06 sec, <b style='mso-bidi-font-weight:normal'><u>SCORE=37</u></b>,
<span class=SpellE>Nseq</span>=19, Len=341</p>

<p class=MyNormal>You can use this value to compare alternative alignments of
the same sequences. Alignments with a score higher than 40 are usually pretty
good.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985535">Doing
it over the WWW</a></h3>

<p class=MyNormal>You can run T-Coffee online at www.tcoffee.org. Use the
regular or the advanced form of the T-Coffee server.</p>

<h2><a name="_Toc235985536"><span lang=EN-GB>Aligning Many Sequences</span></a></h2>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985537">Aligning
Very Large Datasets with Muscle</a></h3>

<p class=MyNormal>T-Coffee is not a good choice if you are dealing with very
large datasets, use Mafft or Muscle. To align a large dataset with Muscle, try:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>muscle
-infile sproteases_large.fasta &gt; sproteases_large.muscle</span></p>

</div>

<p class=MyNormal>To use the fastest possible mode (less accurate) run:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>muscle
-in sproteases_large.fasta -maxiters 1 -diags -sv -distance1 kbit20_3 &gt;
sproteases_large.muscle</span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985538">Aligning
Very Large Alignments with Mafft</a></h3>

<p class=MyNormal>The fastest mode with Mafft can be achieved using:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>mafft
--retree 2 input &gt; output</span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985539">Aligning
Very Large Alignments with T-Coffee</a></h3>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MyNormal>T-Coffee is not very well gifted for aligning large datasets,
but you can give it a try using a special option that generates approximate
alignments. These alignments should roughly have the same accuracy as ClustalW.
They are acceptable for sequences more than 40% identical.</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sproteases_large.fasta -mode quickaln</span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985540">Shrinking
Large Alignments With T-Coffee</a></h3>

<p class=MyNormal>Once you have generated your large alignment, you may <span
class=SpellE>nedd</span>/want to shrink it to a smaller one<span class=GramE>,
that will be (<span class=SpellE>hopefuly</span>) as informative and easier to
manipulate</span>. For that purpose, use the trim option (described in detail
in the first section of this document).</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_large.muscle -action +trim _n20
-output &gt; sproteases_large_muscle_trim.aln</span></p>

</div>

<h2><a name="_Toc235985541"><span lang=EN-GB>Modifying the default parameters
of T-Coffee</span></a></h2>

<p class=MsoBodyText><span lang=EN-GB>The main parameters of T-Coffee are similar
to those of ClustalW. They include the substitution matrix and the gap
penalties. In general, T-Coffee's default is adequate. If, however, you are not
satisfied with the default parameters, we encourage you to change the following
parameters. Interestingly, most of what we say here holds reasonably well for
ClustalW.</span></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985542">Changing
the Substitution Matrix</a></h3>

<p class=MyNormal>T-Coffee only uses the substitution matrix to make the
pairwise alignments that go into the library. These are all the global
alignments of every possible pair of sequences, and the ten best local
alignments associated with every pair of sequences.</p>

<p class=MyNormal>By default, these alignments are computed using a Blosum62
matrix, but you can use any matrix you fancy instead, including: pam120mt,
pam160mt, pam250mt, pam350mt, blosum30mt, blosum40mt, blosum45mt, blosum50mt,
blosum55mt, blosum62mt, blosum80mt, or even user-provided matrices in the BLAST
format, as described in the technical manual.</p>

<p class=MyNormal><b style='mso-bidi-font-weight:normal'><i style='mso-bidi-font-style:
normal'>Pam matrices</i></b>: These matrices are allegedly less accurate than
the <span class=SpellE>blosum</span>. The index is correlated to the
evolutionary distances. You should therefore use the pam350mt to align very
distantly related sequences.</p>

<p class=MyNormal><span class=SpellE><b style='mso-bidi-font-weight:normal'><i
style='mso-bidi-font-style:normal'>Blosum</i></b></span><b style='mso-bidi-font-weight:
normal'><i style='mso-bidi-font-style:normal'> matrices</i></b>: These matrices
are allegedly the most accurate. The index is correlated to the maximum percent
identity within the sequences used to estimate the matrix. You should therefore
use the Blosum30mt to align very distantly related sequences. <span
class=SpellE>Blosum</span> matrices are biased toward protein core regions.
This may explain why theses matrices tend to give better alignments, since by
design they can capture the most evolutionary resilient signal contained in
proteins.</p>

<p class=MyNormal>Unless you have some structural information available, the
only way to tell whether your alignment has improved or not is to look at the
score. For instance, if you compute the two following alignments:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sproteases_small.fasta -matrix=blosum30mt -outfile=b30.aln</span></p>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sproteases_small.fasta -matrix=blosum80mt -outfile=b80.aln</span></p>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sproteases_small.fasta -matrix=pam350mt -outfile p350.aln</span></p>

</div>

<p class=MyNormal>You will get two alignments that have roughly the same score
but are different. You can still use these two alternative alignments by
comparing them to identify regions that have been aligned identically by the
two matrices. These regions are usually <span class=GramE>more trustworthy</span>.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985543">Comparing
Two Alternative Alignments</a></h3>

<p class=MyNormal>If you change the parameters, you will end up with
alternative <span class=SpellE>alignemnts</span>. It can be interesting to
compare them quantitatively. T-Coffee comes along with an alignment comparison
module named <span class=SpellE>aln_compare</span>. You can use it to estimate
the amount of difference between your two alignments:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg aln_compare -al1 b30.aln -al2 p350.aln</span></p>

</div>

<p class=MyNormal>This comparison will return the following result:</p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6'>

<p class=examplefileCxSpFirst><span style='mso-ansi-language:EN-US'>*****************************************************<o:p></o:p></span></p>

<p class=examplefileCxSpMiddle><span class=GramE><span style='mso-ansi-language:
EN-US'>seq1</span></span><span style='mso-ansi-language:EN-US'><span
style="mso-spacerun:yes">       </span>seq2<span
style="mso-spacerun:yes">          </span><span class=SpellE>Sim</span><span
style="mso-spacerun:yes">   </span>[ALL]<span
style="mso-spacerun:yes">           </span>Tot<o:p></o:p></span></p>

<p class=examplefileCxSpLast><span class=GramE><span style='mso-ansi-language:
EN-US'>b30</span></span><span style='mso-ansi-language:EN-US'><span
style="mso-spacerun:yes">           </span>19<span
style="mso-spacerun:yes">         </span>32.6<span style="mso-spacerun:yes">   
</span>93.7 [100.0]<span style="mso-spacerun:yes">   </span>[40444]<o:p></o:p></span></p>

</div>

<p class=MyNormal>Where 93.7 is the percentage of similarity (sums of pairs)
between the two alignments. It means that when considering every pair of
aligned residues in b30 (40444), the program found that 93.7% of these pairs
could be found in the alignment p350.aln.</p>

<p class=MyNormal>Of course, this does not tell you where are the good bits,
but you can get this information with the same program:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>t_coffee
-other_pg aln_compare -al1 b30.aln -al2 p350.aln -output_aln
-output_aln_threshold 50</span></p>

</div>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6'>

<p class=examplefileverysmallCxSpFirst><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|O35205|GRAK_MOUSE<span style="mso-spacerun:yes">   </span>M---r----<span
class=SpellE>fssw</span>-------<span class=SpellE>ALvslvagvym</span>----------------SSECFHTEIIGGR<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|Q7YRZ7|GRAA_BOVIN<span style="mso-spacerun:yes">   </span>M--<span
class=SpellE>ni</span>----<span class=SpellE>pfpf</span>--<span class=SpellE>sfppaIClllipgvfp</span>----------------<span
class=SpellE>vs</span>---<span class=SpellE>cEGIIGGN</span><o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P08884|GRAE_MOUSE<span style="mso-spacerun:yes">   </span>M--------<span
class=SpellE>ppv</span>----------<span class=SpellE>lilltlllp</span>----------------l-GAGAEEIIGGH<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|Q06606|GRZ2_RAT<span style="mso-spacerun:yes">     </span>M--------<span
class=SpellE>flf</span>----------<span class=SpellE>lfflvailp</span>----------------v-NTEGGEIIWGT<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P21844|MCPT5_MOUSE<span style="mso-spacerun:yes">  </span>M---h----<span
class=SpellE>llt</span>----------<span class=SpellE>lhllllllg</span>----------------s-STKAGEIIGGT<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P03953|CFAD_MOUSE<span style="mso-spacerun:yes">   </span>M---h----<span
class=SpellE>ssvy</span>-------<span class=SpellE>fvalvilgaav</span>----------------CAAQPRGRILGGQ<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P00773|ELA1_RAT<span style="mso-spacerun:yes">     </span>M---l----<span
class=SpellE>rflv</span>--F----<span class=SpellE>ASlvlyghstq</span>----------------DFPETNARVVGGA<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|Q00871|CTRB1_PENVA<span style="mso-spacerun:yes">  </span><span
class=SpellE>MIgkl</span>----<span class=SpellE>slll</span>--V----<span
class=SpellE>CVavasgnpaagkpwhwKSPKPLVDPRIHVNATPRIVGGV</span><o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P08246|ELNE_HUMAN<span style="mso-spacerun:yes">   </span>M--<span
class=SpellE>tlGR</span>--<span class=SpellE>rlac</span>--L----<span
class=SpellE>FLacvlpalll</span>----------------GGTALASEIVGGR<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P20160|CAP7_HUMAN<span style="mso-spacerun:yes">   </span>M--t-----<span
class=SpellE>rltv</span>--L----<span class=SpellE>ALlagllassr</span>----------------AGSSPLLDIVGGR<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P80015|CAP7_PIG<span style="mso-spacerun:yes">    
</span>-------------------------------------------------------IVGGR<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|Q03238|GRAM_RAT<span style="mso-spacerun:yes">    
</span>l-------------------<span class=SpellE>LLllalktlwa</span>----------------VGNRFEAQIIGGR<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P00757|KLKB4_MOUSE<span style="mso-spacerun:yes"> 
</span>M-----------w-------<span class=SpellE>flilflalslggid</span>-------------AAPP-----<span
class=SpellE>vqsq</span><o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|Q6H321|KLK2_HORSE<span style="mso-spacerun:yes">  
</span>M-----------w-------<span class=SpellE>flvlcldlslgetg</span>-------------ALPPIQSRIIGGW<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|Q91VE3|KLK7_MOUSE<span style="mso-spacerun:yes">   </span>M---------<span
class=SpellE>gvw</span>-------<span class=SpellE>llslitvllslale</span>-------------tag-QGERIIDGY<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|Q9Y5K2|KLK4_HUMAN<span style="mso-spacerun:yes">   </span>M-<span
class=SpellE>ataGN</span>--<span class=SpellE>pwgw</span>-------<span
class=SpellE>flgylilgvag-sl</span>-------------<span class=SpellE>vsg</span>-SCSQIINGE<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P29786|TRY3_AEDAE<span style="mso-spacerun:yes">   </span>M-------<span
class=SpellE>nqflfVSF</span>---------<span class=SpellE>calldsakvsaa</span>------------<span
class=SpellE>tLSSGRIVGGF</span><o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P35037|TRY3_ANOGA<span style="mso-spacerun:yes">   </span>M---<span
class=SpellE>iSNKiaillAVLvvav</span>----<span class=SpellE>acaqarvaqqhrsVQALPRFLPRPKYDVGHRIVGGF</span><o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P07338|CTRB1_RAT<span style="mso-spacerun:yes">    </span>M--a------<span
class=SpellE>flwlvs</span>---------<span class=SpellE>cfalvgatfgcg</span>---<span
class=SpellE>vptiqpv</span>--LTGLSRIVNGE<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span style='mso-ansi-language:EN-US'><span
style="mso-spacerun:yes">                                                                              
</span><span class=GramE>: .</span><o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span style='mso-ansi-language:EN-US'><o:p>&nbsp;</o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|O35205|GRAK_MOUSE<span style="mso-spacerun:yes">  
</span>EVQPHSRPFMASIQYR----<span class=SpellE>SKHICGGVLIHPQWVLTAAHCYSWFprGHSPTVVLGAHSL</span><o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|Q7YRZ7|GRAA_BOVIN<span style="mso-spacerun:yes">  
</span>EVAPHTRRYMALIK------<span class=SpellE>GLKLCAGALIKENWVLTAAHCDlk</span>----GNPQVILGAHST<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P08884|GRAE_MOUSE<span style="mso-spacerun:yes">  
</span>VVKPHSRPYMAFVKSVDIEGNRRYCGGFLVQDDFVLTAAHCRN-----RTMTVTLGAHNI<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|Q06606|GRZ2_RAT<span style="mso-spacerun:yes">    
</span>ESKPHSRPYMAFIKFYDSNSEPHHCGGFLVAKDIVMTAAHCNG-----RNIKVTLGAHNI<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P21844|MCPT5_MOUSE<span style="mso-spacerun:yes"> 
</span>ECIPHSRPYMAYLEIVTSENYLSACSGFLIRRNFVLTAAHCAG-----RSITVLLGAHNK<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P03953|CFAD_MOUSE<span style="mso-spacerun:yes">  
</span>EAAAHARPYMASVQVN----<span class=SpellE>GTHVCGGTLLDEQWVLSAAHCMDGVtdDDSVQVLLGAHSL</span><o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P00773|ELA1_RAT<span style="mso-spacerun:yes">     </span><span
class=SpellE>EARRNSWPSQISLQYLSggswyHTCGGTLIRRNWVMTAAHCVSSQm</span>---TFRVVVGDHNL<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|Q00871|CTRB1_PENVA<span style="mso-spacerun:yes">  </span><span
class=SpellE>EATPHSWPHQAALFId</span>----<span class=SpellE>DMYFCGGSLISSEWVLTAAHCMDGAg</span>---FVEVVLGAHNI<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P08246|ELNE_HUMAN<span style="mso-spacerun:yes">   </span><span
class=SpellE>RARPHAWPFMVSLQLr</span>----GGHFCGATLIAPNFVMSAAHCVANVNV-RAVRVVLGAHNL<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P20160|CAP7_HUMAN<span style="mso-spacerun:yes">   </span><span
class=SpellE>KARPRQFPFLASIQNq</span>----GRHFCGGALIHARFVMTAASCFQSQNP-GVSTVVLGAYDL<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P80015|CAP7_PIG<span style="mso-spacerun:yes">     </span><span
class=SpellE>RAQPQEFPFLASIQKq</span>----GRPFCAGALVHPRFVLTAASCFRGKNS-GSASVVLGAYDL<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|Q03238|GRAM_RAT<span style="mso-spacerun:yes">    
</span>EAVPHSRPYMVSLQNT----KSHMCGGVLVHQKWVLTAAHCLSEP--LQQLKLVFGLHSL<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P00757|KLKB4_MOUSE<span style="mso-spacerun:yes">  </span><span
class=SpellE>vdcENSQPWHVAVYRF</span>----NKYQCGGVLLDRNWVLTAAHCYN-----DKYQVWLGKNNF<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|Q6H321|KLK2_HORSE<span style="mso-spacerun:yes">  
</span>ECEKHSKPWQVAVYHQ----GHFQCGGVLVHPQWVLTAAHCMS-----DDYQIWLGRHNL<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|Q91VE3|KLK7_MOUSE<span style="mso-spacerun:yes">  
</span>KCKEGSHPWQVALLKG----NQLHCGGVLVDKYWVLTAAHCKM-----GQYQVQLGSDKI<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|Q9Y5K2|KLK4_HUMAN<span style="mso-spacerun:yes">  
</span>DCSPHSQPWQAALVME----NELFCSGVLVHPQWVLSAAHCFQ-----NSYTIGLGLHSL<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P29786|TRY3_AEDAE<span style="mso-spacerun:yes">  
</span>QIDIAEVPHQVSLQRS----GRHFCGGSIISPRWVLTRAHCTTNTDP-<span class=SpellE>AAYTIRAGStd</span>-<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P35037|TRY3_ANOGA<span style="mso-spacerun:yes">  
</span>EIDVSETPYQVSLQYF----NSHRCGGSVLNSKWILTAAHCTVNLQP-<span class=SpellE>SSLAVRLGSsr</span>-<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P07338|CTRB1_RAT<span style="mso-spacerun:yes">    </span><span
class=SpellE>DAIPGSWPWQVSLQDKt</span>---<span class=SpellE>gfHFCGGSLISEDWVVTAAHCGVKT</span>----SDVVVAGEFDQ<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpLast><span style='mso-ansi-language:EN-US'><span
style="mso-spacerun:yes">                                   </span>:<span
style="mso-spacerun:yes">           </span>*<span class=GramE>..</span> <span
class=GramE>::</span><span style="mso-spacerun:yes">    </span>::: * *<span
style="mso-spacerun:yes">           </span>:<span style="mso-spacerun:yes"> 
</span>*<o:p></o:p></span></p>

</div>

<p class=MyNormal>This is the alignment al1, but residues that have lost more
than 50% of their pairing partner between the two alignments are now in lower
case. In the section of this tutorial entitled comparing alignments, we show
you more sophisticated ways to do this comparison.</p>

<p class=MyNormal>For an even more drastic display, try:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>t_coffee
-other_pg aln_compare -al1 b30.aln -al2 p350.aln -output_aln
-output_aln_threshold 50 -output_aln_modif x</span></p>

</div>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985544">Changing
Gap Penalties</a></h3>

<p class=MyNormal>Gap penalties are the core of the matter when it comes to
multiple sequence alignments. An interesting feature of T-Coffee is that it
does not really need such penalties when assembling the MSA, because in theory
the penalties have already been applied when computing the library. This is the
theory, as in practice penalties can help improve the quality of the alignment.</p>

<p class=MyNormal>The penalties can be changed via the flags -<span
class=SpellE>gapopen</span> for the <span class=GramE>gap opening</span>
penalty and via -<span class=SpellE>gapext</span> for the gap extension penalty.
The range for <span class=SpellE>gapopen</span> are [-500<span class=GramE>,-</span>-5000],
the range for the extension should rather be [-1, -10]. These values do not
refer to a substitution matrix, but rather to the values range of the <span
class=SpellE>concistensy</span> estimation (i.e. a ratio) normalized to 10000
for a maximum consistency.</p>

<p class=MyNormal>The default values are -<span class=SpellE>gapopen</span>=-50,
-<span class=SpellE>gapext</span>=0. The reasons for these very low values are
that they are meant to be cosmetic only, since a trademark of T-Coffee
(inherited from <span class=SpellE>Dialign</span>) is not to need explicit
penalties. Yet, we know for a fact that alignments with higher gap penalties
often look nicer (for publications) and are sometimes more accurate. For
instance, you can try:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sproteases_small.fasta -gapopen -100 -gapext -5</span></p>

</div>

<p class=MyNormal>This gap penalty is only applied at the alignment level (i.e.
after the library was computed). If you want to change the gap penalties of the
methods used to build the library, you will need to go deeper into the core of
the matter...</p>

<p class=MyNormal>Two methods are used by default to build the library. One does
global pairwise alignments and is named <span class=SpellE>slow_pair</span>,
the other is named <span class=SpellE>lalign_id_pair</span> and <span
class=SpellE>and</span> produces local alignments. These methods are specified
via the -method flag. The default of this flag is:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sproteases_small.fasta -method=lalign_id_pair,slow_pair</span></p>

</div>

<p class=MyNormal>Usually you do not need to write it because it is the
default, but if you want to change the default parameters of the constituting
methods, you will need to do so <span class=SpellE>explicitely</span>. The
default for <span class=SpellE>lalign_id_pair</span> is <span class=SpellE><span
class=GramE>gop</span></span>=-10, GEP=-4, MATRIX=blosum50mt. The default for <span
class=SpellE>slow_pair</span> is: GOP=-10, GEP=-1 and MATRIX=blosum62mt. If you
want to change this, try:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sproteases_small.fasta -method
lalign_id_pair@EP@MATRIX@blosum62mt,slow_pair -outfile sproteases_small.b62_aln
</span></p>

</div>

<p class=MyNormal>This means the library is now computed using the Blosum62mt
with <span class=SpellE>lalign</span>, rather than the Blosum50mt. The good
news is that when using this matrix, the score of our alignment increases from
48(default) to 50. We may assume this new alignment is more accurate than the
previous one.</p>

<div style='mso-element:para-border-div;border:solid windowtext 1.5pt;
padding:0cm 0cm 0cm 0cm;background:#F7CBB7'>

<p class=Warning><span lang=EN-GB>WARNING: It only makes sense to compare the
consistency score of alternative alignments when these alignments have been
computed using the same methods (<span class=SpellE>lalign_id_pair</span> and <span
class=SpellE>slow_pair</span> for instance).</span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985545">Can
You Guess The Optimal Parameters?</a></h3>

<p class=MyNormal>It is a trick question, but the general answer is NO. The
matrix and the gap penalties are simplistic attempts at modeling evolution.
While the matrices do a reasonable job, the penalties are simply inappropriate:
they should have a value that depends on the structure of the protein and a
uniform value cannot be good enough. Yet, since we do not have better we must
use them</p>

<p class=MyNormal>In practice, this means that parameter optimality is a very
add-hoc business. It will change from one dataset to the next and there is no
simple way to predict which matrix and which penalty will do better. The
problem is also that even after your alignment has been computed, it is not
always easy to tell whether your new parameters have improved or degraded your
MSA. There is no systematic way to evaluate an MSA.</p>

<p class=MyNormal>In general, people visually evaluate the alignment, count the
number of identical columns and consider that one more conserved column is good
news. If you are lucky you may know a few functional features that you expect
to see aligned. If you are very lucky, you will have one structure and you can
check the gaps fall in the loops. If you are <b style='mso-bidi-font-weight:
normal'><i style='mso-bidi-font-style:normal'>extremely</i></b> lucky, you will
have two structures and you can assess the quality of your MSA.</p>

<p class=MyNormal>An advantage of T-Coffee is the fact that the overall score
of the alignment (i.e. the consistency with the library) is correlated with the
overall accuracy. In other words, if you <span class=GramE>alignment score</span>
increases, its accuracy probably increases also. All this being said,
consistency is merely an empirical way of estimating the change of parameters
and it does not have the predictive power of a BLAST E-Value.</p>

<h2><a name="_Toc235985546"><span lang=EN-GB>Using Many Methods at once</span></a></h2>

<p class=MsoBodyText><span lang=EN-GB>One of the most common <span class=GramE>situation</span>
when building multiple sequence alignments is to have several alignments
produced by several alternative methods, and not knowing which one to choose.
In this section, we show you that you can use M-Coffee to combine your many alignments
into one single alignment. We show you here that you can either let T-Coffee
compute all the multiple sequence alignments and combine them into one, or you
can specify the methods you want to combine. M-Coffee is not always the best
methods, but extensive benchmarks on BaliBase, <span class=GramE>Prefab<span
style="mso-spacerun:yes">  </span>and</span> Homstrad have shown that it
delivers the best alignment 2 times out of 3. If you do not want to use the
methods provided by M-Coffee, you can also combine pre-computed alignments.</span></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985547">Using
All the Methods at the Same Time: M-Coffee</a></h3>

<p class=MyNormal>In M-Coffee, M stands for Meta. To use M-Coffee, you will
need several packages to be installed (see documentation). The following
command:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sproteases_small.fasta -mode mcoffee -output clustalw, html</span></p>

</div>

<p class=MyNormal>Will compute a Multiple Sequence Alignment with the following
MSA packages:</p>

<p class=MyNormal><span class=SpellE><span class=GramE>clustalw</span></span>, <span
class=SpellE>poa</span>, muscle, <span class=SpellE>probcons</span>, <span
class=SpellE>mafft</span>, dialing-T, <span class=SpellE>pcma</span> and
T-Coffee.</p>

<p class=MyNormal>For those using <span class=SpellE>debian</span>, another
mode is available</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sproteases_small.fasta -mode dmcoffee -output clustalw, html</span></p>

</div>

<p class=MyNormal>Will compute a Multiple Sequence Alignment with the following
MSA packages:</p>

<p class=MyNormal><span class=SpellE><span class=GramE>kalign</span></span>, <span
class=SpellE>poa</span>, muscle, <span class=SpellE>probcons</span>, <span
class=SpellE>mafft</span>, dialing-T, and T-Coffee.</p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6'>

<p class=examplefileCxSpFirst><span lang=EN-GB>Package<span style='mso-tab-count:
2'>              </span>Where From</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>==========================================================</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>ClustalW<span style='mso-tab-count:
2'>             </span>can interact with t_coffee</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span class=SpellE><span lang=EN-GB>Poa</span></span><span
lang=EN-GB><span style='mso-tab-count:2'>           </span><span
style='mso-tab-count:1'>       </span>http://www.bioinformatics.ucla.edu/poa/</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>Muscle<span style='mso-tab-count:
2'>        </span>http://www.bioinformatics.ucla.edu/poa/</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>ProbCons<span style='mso-tab-count:
2'>             </span>http://probcons.stanford.edu/</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>MAFFT<span style='mso-tab-count:
1'>  </span><a href="http://www.biophys.kyoto-/">http://www.biophys.kyoto-</a>u.ac.jp/~katoh/programs/align/mafft/</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span class=SpellE><span lang=EN-GB>Dialign</span></span><span
lang=EN-GB>-T<span style='mso-tab-count:2'>            </span><a
href="http://dialign-t.gobics.de/">http://dialign-t.gobics.de/</a></span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>PCMA<span style='mso-tab-count:
3'>                 </span><a href="file://localhost/pub/PCMA">ftp://iole.swmed.edu/pub/PCMA/</a></span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>T-Coffee<span style='mso-tab-count:
2'>             </span>www.tcoffee.org</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>Kalign <span style='mso-tab-count:
2'>              </span>msa.cgb.ki.se/<span class=SpellE>cgi</span>-bin/<span
class=SpellE>msa.cgi</span></span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span class=SpellE><span class=GramE><span
lang=EN-GB>amap</span></span></span><span lang=EN-GB><span style='mso-tab-count:
3'>                 </span>bio.math.berkeley.edu/<span class=SpellE>amap</span>/</span></p>

<p class=examplefileCxSpLast><span lang=EN-GB>----------------------------------------------------------<br
style='mso-special-character:line-break'>
<![if !supportLineBreakNewLine]><br style='mso-special-character:line-break'>
<![endif]></span></p>

</div>

<p class=MyNormal>When this is done, all the alignments will be combined into
one. If you open the file sproteases_small.html with your favorite web browser,
you will see a colored version of your alignment.</p>

<p class=MyNormal>The alignment is colored according to its consistency with
all the MSA used to compute it. Regions in red have a high consistency and you
can expect them to be fairly accurate. Regions in green/blue have the lowest
consistency and you should not trust them. </p>

<p class=MyNormal>Overall this alignment has a score of 80, which means that it
is 80% consistent with the entire collection. This is a fairly high index,
which means you can probably trust your alignment (at least where it is red).</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><span
style="mso-spacerun:yes"> </span><a name="_Toc235985548">Using Selected Methods
to Compute your MSA</a> </h3>

<p class=MyNormal>Using the 8 Methods of M-Coffee8 can sometimes be a bit
heavy. If you only want to use a subset of your favorite methods, you should
know that each of these methods is available via the -method flag. For
instance, to combine MAFFT, Muscle, t_coffee and ProbCons, you can use:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sproteases_small.fasta -method=t_coffee_msa,mafft_msa,probcons_msa,muscle_msa
-output=html</span></p>

</div>

<p class=MyNormal>This will result in a computation where all the specified
methods are mixed together</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985549">Combining
pre-Computed Alignments</a> </h3>

<p class=MyNormal>You may have a bunch of alignments that you have either
pre-computed, or assembled manually or received from a colleague. You can also
combine these alignments. For instance, let us imagine we generated 4
alignments with ClustalW using different gap penalties:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>clustalw
-infile=sproteases_small.fasta -gapopen=0 -outfile=g0.aln</span></p>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>clustalw
-infile=sproteases_small.fasta -gapopen=-5 -outfile=g5.aln</span></p>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>clustalw
-infile=sproteases_small.fasta -gapopen=-10 -outfile=g10.aln</span></p>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>clustalw
-infile=sproteases_small.fasta -gapopen=-15 -outfile=g15.aln</span></p>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU><o:p>&nbsp;</o:p></span></p>

</div>

<p class=MyNormal>To combine them into ONE single alignment, use the -<span
class=SpellE>aln</span> flag:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sproteases_small.fasta -aln g0.aln g5.aln g10.aln g15.aln -output
clustalw html</span></p>

</div>

<p class=MyNormal>As before, the score indicates a high level of consistency
(91%) between all these alignments. This is an indication that the final
alignment is probably correct.</p>

<h2><a name="_Toc235985550"><span lang=EN-GB>Aligning Profiles</span></a></h2>

<p class=MsoBodyText><span lang=EN-GB>Sometimes, it is better to pre-align a
subset of your sequences, and then to use this small alignment as a master for
adding sequences (sequence to profile alignment) or even to align several
profiles together if your protein family contains distantly related groups. T-Coffee
contains most of the facilities available in ClustalW to deal with profiles,
and the strategy we outline here can be used to deal with large datasets</span></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985551">Using
Profiles as templates</a></h3>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985552">Aligning
One sequence to a Profile</a> </h3>

<p class=MyNormal>Assuming you have a multiple alignment (<span class=SpellE>sproteases_small.aln</span>)
here is a simple strategy to align one sequence to your profile:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sproteases_oneseq.fasta<span style="mso-spacerun:yes"> 
</span>-profile sproteases_small.aln</span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985553">Aligning
Many Sequences to a Profile</a></h3>

<p class=MyNormal>You can align as many sequences as you wish to your profile.
Likewise, you can have as many profiles as you want. For instance, the
following:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sequences.fasta<span style="mso-spacerun:yes"> 
</span>-profile=prf1.aln,prf2.aln,prf3.aln -outfile=combined_profiles.aln</span></p>

</div>

<p class=MyNormal>Will make a multiple alignment of 3 profiles and 5 sequences.
You can mix sequences and profiles in any proportion you like. You can also use
all the methods you want although you should be aware that when using external
methods (see the external method section in this tutorial), the profile is
replaced with its consensus sequence, which will not be quite as accurate.</p>

<p class=MyNormal>Methods supporting full profile information are: <span
class=SpellE>lalign_id_pair</span>, <span class=SpellE>slow_pair</span> and <span
class=SpellE>proba_pair</span>, <span class=SpellE>clustalw_pair</span> and <span
class=SpellE>clustalw_msa</span>. All the other methods (internal or external)
treat the profile as a consensus (less accurate).</p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<h2><a name="_Toc235985554"><span lang=EN-GB>Aligning Other Types of Sequences</span></a></h2>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985555">Splicing
variants</a></h3>

<p class=MyNormal>Splicing variants are especially challenging for most MSA
programs. This is because the splicing variants need very long gaps to be
inserted, while most programs attempt to match as many symbols as possible. </p>

<p class=MyNormal>Standard programs like ClustalW or Muscle are not good at
dealing with this situation and in our experience, the only programs that can
do something with splice variants are those using local information like some
flavors of Mafft and T-<span class=GramE>Coffee .</span></p>

<p class=MyNormal>For instance, if you try muscle on the following dataset:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>muscle
-in sv.fasta -clw</span></p>

</div>

<p class=MyNormal>You will quickly <span class=SpellE>realise</span> that your
alignment is not very good and does not show where the alternative splicing <span
class=SpellE>coocurs</span>. On the other hand, if you use T-Coffee, things
become much clearer</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee<span style="mso-spacerun:yes">  </span>sv.fasta</span></p>

</div>

<p class=MyNormal>The reason why T-Coffee does better than other packages is
mostly because it uses local information (<span class=SpellE>lalign_id_pair</span>)
and is therefore less sensitive to long gaps. If the default mode does not work
for your dataset, you can try to be a bit more aggressive and only use local
information to compute your library:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sv.fasta -method lalign_id_pair</span></p>

</div>

<p class=MyNormal>Of course, the most distantly related your sequences, the
harder the alignment of splicing variants</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985556">Aligning
DNA sequences</a></h3>

<p class=MyNormal>Multiple Sequence Alignment methods are not at their best
when aligning DNA sequences. Whenever you can, try using a local multiple
sequence alignment package like the Gibbs sampler. Yet if you believe your DNA <span
class=GramE>sequence are</span> homologous over their entire length, you can
use T-Coffee.</p>

<p class=MyNormal>In theory, the program automatically recognizes DNA sequences
and uses appropriate methods, yet adding the -type=<span class=SpellE><span
class=GramE>dna</span></span> flag cannot do any harm...</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_dnaseq1.fasta &#8211;type=dna</span></p>

</div>

<p class=MyNormal>The type declaration (or its automatic detection) triggers
the use of the appropriate substitution matrix in most of the methods. In
practice, any time it encounters <span class=SpellE><span class=GramE>dna</span></span>,
the program will try to use 4dna version of the requested methods. These
methods have lower penalties and are better suited for dealing with nucleic
acid sequences.</p>

<p class=MyNormal>However, if you would rather use your own matrix, use:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_dnaseq1.fasta &#8211;in Mlalign_id_pair4dna@EP@MATRIX@idmat</span></p>

</div>

<p class=MyNormal>Where you should replace <span class=SpellE>idmat</span> with
your own matrix, in BLAST format (see the format section of the Reference
Manual).</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985557">Aligning
RNA sequences</a></h3>

<p class=MyNormal>RNA sequences are very important and almost every-where these
days. The main property of RNA sequences is to have a secondary structure that
can be used to guide the alignment. While the default T-Coffee has no special
RNA alignment method incorporated in, smart people have thought about this. If
you are interested in RNA, check: http://www.bio.inf.uni-jena.de/Software/MARNA/.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985558">Noisy
Coding DNA Sequences</a></h3>

<p class=MyNormal>When dealing with coding DNA, the right thing to do is to
translate your DNA sequence and thread the DNA onto the protein alignment if
you really need some DNA. However, sometimes, your <span class=SpellE>cDNA</span>
may not be so clean that you can easily translate it (<span class=SpellE>frameshifts</span>
and so on). Whenever this happens, try (no warranty) the following special
method.</p>

<p class=MyNormal>The test case in <span class=SpellE>three_dna_seq.fasta</span>
contains the DNA sequences of three proteases with a couple of <span
class=SpellE>frameshifts</span> here and there. If you make a regular alignment
of these sequences</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee three_cdna.fasta</span></p>

</div>

<p class=MyNormal>You can immediately see that many gaps have sizes that are
not multiple of 3 (codon size). Most of the information is lost. On the other
hand, when using an appropriate alignment method that takes into account all
the frames at the same time, we get something much more meaningful:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee three_cdna.fasta -method cdna_fast_pair</span></p>

</div>

<p class=MyNormal>And most importantly, the <span class=SpellE>frameshifts</span>
end up at the right place. You can even recover the corrected protein sequence
using a special mode of seq_reformat:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in three_cdna.aln -action +clean_cdna
+translate</span></p>

</div>

<p class=MyNormal>+<span class=GramE>clean</span> <span class=SpellE>cdna</span>
is a small HMM that loops through each sequence and select the frame in order
to maximize the similarity within the alignment.</p>

<h1><span lang=EN-GB>Combining Protein 2D and 3D Structural Information in your
Multiple Sequence Alignment</span></h1>

<p class=MsoBodyText><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<p class=MsoBodyText><span lang=EN-GB>Using structural information when
aligning sequences is very useful. The reason is that structures diverge slower
than sequences. As a consequence, one may still find a discernable homology between
two sequences that have been diverging for so long that their sequences have
evolved beyond recognition. Yet, when assembling the correct structure based
MSA, you will realize that these sequences contain key conserved residues that
a simple alignment procedure was unable to reveal. We show you in this section
how to make the best of T-Coffee tools to incorporate structural information in
your alignment.</span></p>

<h2><a name="_Toc235985566"><span lang=EN-GB>If you are in a Hurry: Expresso</span></a></h2>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985567">What
is Expresso?</a></h3>

<p class=MsoBodyText><span lang=EN-GB>Expresso is the latest T-Coffee mode. It
is not yet available for local installation, but you can run it from the
www.tcoffee.org server. The principle of Expresso is simple: the server runs a
BLAST between every sequence in your query against the PDB database. If it
finds a structure similar enough to a sequence in your dataset (&gt;60%
identity), it will use that structure as a template for your sequence.</span></p>

<p class=MsoBodyText><span lang=EN-GB>Template files look something like:</span></p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmallCxSpFirst><span lang=EN-GB>&gt;<span class=GramE>sp</span>|P08246|ELNE_HUMAN
_P_ 1PPGE</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;<span class=GramE>sp</span>|P20160|CAP7_HUMAN
_P_ 1AE5</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;<span class=GramE>sp</span>|P00757|KLKB4_MOUSE
_P_ 1SGFX</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;<span class=GramE>sp</span>|Q6H321|KLK2_HORSE
_P_ 1GVZA</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;<span class=GramE>sp</span>|P00773|ELA1_RAT
_P_ 2D26C</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;<span class=GramE>sp</span>|Q00871|CTRB1_PENVA
_P_ 1AZZB</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;<span class=GramE>sp</span>|P21844|MCPT5_MOUSE
_P_ 1NN6A</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;<span class=GramE>sp</span>|O35205|GRAK_MOUSE
_P_ 1MZDA</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;<span class=GramE>sp</span>|P07338|CTRB1_RAT
_P_ 2CGAB</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;<span class=GramE>sp</span>|P80015|CAP7_PIG
_P_ 1FY3A</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;<span class=GramE>sp</span>|P03953|CFAD_MOUSE
_P_ 1FDPD</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;<span class=GramE>sp</span>|Q7YRZ7|GRAA_BOVIN
_P_ 1OP8F</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;<span class=GramE>sp</span>|Q06606|GRZ2_RAT
_P_ 1EUFA</span></p>

<p class=examplefilesmallCxSpLast><span lang=EN-GB>&gt;<span class=GramE>sp</span>|P08884|GRAE_MOUSE
_P_ 1FI8B</span></p>

</div>

<p class=MsoBodyText><span lang=EN-GB>In a template file, <span
style="mso-spacerun:yes"> </span><b style='mso-bidi-font-weight:normal'>_P_</b>
indicates that the template is of type structure (P for PDB). <span
class=GramE>Template files can be generated manually or automatically by the
Expresso server</span>. Whenever possible t_coffee will then align your
sequences using the structural information contained in the templates. If it
encounters enough structures (as shown here) it will produce a genuine
structure based sequence alignment. </span></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985568">Using
Expresso</a></h3>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee three_pdb_two_seq.fasta -method sap_pair,slow_pair -template_file PDB</span></p>

</div>

<h2><a name="_Toc235985569"></a><a name="_Toc235985559"><span style='mso-bookmark:
_Toc235985569'><span lang=EN-GB>Using Secondary Structure Predictions:</span></span></a><span
style='mso-bookmark:_Toc235985569'><span lang=EN-GB> </span></span></h2>

<p class=MyNormal><span style='mso-bookmark:_Toc235985569'>T-Coffee can be used
to predict secondary structures and <span class=SpellE>transmembrane</span>
domains. For secondary structure predictions, the current implementation is
only able to run GOR on either single sequences or on a bunch of homologues
found by BLAST. </span></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><span style='mso-bookmark:
_Toc235985569'><a name="_Toc235985560">Single Sequence prediction</a></span></h3>

<p class=MyNormal><span style='mso-bookmark:_Toc235985569'>To make a secondary
structure prediction with GOR, run the following. In this command line SSP is a
hard coded mode. It prompts the computation of predicted secondary structures.</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc235985569'><span lang=ZU>t_coffee sample_aln.fasta
-template_file SSP</span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc235985569'>The predictions are
then displayed in the files:</span></p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 31.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmallCxSpFirst style='margin-right:0cm;mso-add-space:auto;
background:#E6E6E6;border:none;mso-border-alt:solid windowtext .5pt;padding:
0cm;mso-padding-alt:1.0pt 4.0pt 1.0pt 31.0pt'><span style='mso-bookmark:_Toc235985569'><span
lang=EN-GB><o:p>&nbsp;</o:p></span></span></p>

<p class=examplefilesmallCxSpMiddle style='margin-right:0cm;mso-add-space:auto;
background:#E6E6E6;border:none;mso-border-alt:solid windowtext .5pt;padding:
0cm;mso-padding-alt:1.0pt 4.0pt 1.0pt 31.0pt'><span style='mso-bookmark:_Toc235985569'><span
lang=EN-GB>#### File Type= Template Protein Secondary Structure Format<span
class=GramE>=<span style="mso-spacerun:yes">  </span><span class=SpellE>fasta</span></span><span
class=SpellE>_seq</span> Name= <span class=SpellE>hmgb_chite.ssp</span></span></span></p>

<p class=examplefilesmallCxSpMiddle style='margin-right:0cm;mso-add-space:auto;
background:#E6E6E6;border:none;mso-border-alt:solid windowtext .5pt;padding:
0cm;mso-padding-alt:1.0pt 4.0pt 1.0pt 31.0pt'><span style='mso-bookmark:_Toc235985569'><span
lang=EN-GB>#### File Type= Template Protein Secondary Structure Format<span
class=GramE>=<span style="mso-spacerun:yes">  </span><span class=SpellE>fasta</span></span><span
class=SpellE>_seq</span> Name= <span class=SpellE>hmgl_trybr.ssp</span></span></span></p>

<p class=examplefilesmallCxSpMiddle style='margin-right:0cm;mso-add-space:auto;
background:#E6E6E6;border:none;mso-border-alt:solid windowtext .5pt;padding:
0cm;mso-padding-alt:1.0pt 4.0pt 1.0pt 31.0pt'><span style='mso-bookmark:_Toc235985569'><span
lang=EN-GB>#### File Type= Template Protein Secondary Structure Format<span
class=GramE>=<span style="mso-spacerun:yes">  </span><span class=SpellE>fasta</span></span><span
class=SpellE>_seq</span> Name= hmgl_trybr3.ssp</span></span></p>

<p class=examplefilesmallCxSpMiddle style='margin-right:0cm;mso-add-space:auto;
background:#E6E6E6;border:none;mso-border-alt:solid windowtext .5pt;padding:
0cm;mso-padding-alt:1.0pt 4.0pt 1.0pt 31.0pt'><span style='mso-bookmark:_Toc235985569'><span
lang=EN-GB>#### File Type= Template Protein Secondary Structure Format<span
class=GramE>=<span style="mso-spacerun:yes">  </span><span class=SpellE>fasta</span></span><span
class=SpellE>_seq</span> Name= <span class=SpellE>hmgl_wheat.ssp</span></span></span></p>

<p class=examplefilesmallCxSpMiddle style='margin-right:0cm;mso-add-space:auto;
background:#E6E6E6;border:none;mso-border-alt:solid windowtext .5pt;padding:
0cm;mso-padding-alt:1.0pt 4.0pt 1.0pt 31.0pt'><span style='mso-bookmark:_Toc235985569'><span
lang=EN-GB>#### File Type= Template Protein Secondary Structure Format<span
class=GramE>=<span style="mso-spacerun:yes">  </span><span class=SpellE>fasta</span></span><span
class=SpellE>_seq</span> Name= hmgl_wheat2.ssp</span></span></p>

<p class=examplefilesmallCxSpMiddle style='margin-right:0cm;mso-add-space:auto;
background:#E6E6E6;border:none;mso-border-alt:solid windowtext .5pt;padding:
0cm;mso-padding-alt:1.0pt 4.0pt 1.0pt 31.0pt'><span style='mso-bookmark:_Toc235985569'><span
lang=EN-GB>#### File Type= Template Protein Secondary Structure Format<span
class=GramE>=<span style="mso-spacerun:yes">  </span><span class=SpellE>fasta</span></span><span
class=SpellE>_seq</span> Name= <span class=SpellE>hmgt_mouse.ssp</span></span></span></p>

<p class=examplefilesmallCxSpLast style='margin-right:0cm;mso-add-space:auto;
background:#E6E6E6;border:none;mso-border-alt:solid windowtext .5pt;padding:
0cm;mso-padding-alt:1.0pt 4.0pt 1.0pt 31.0pt'><span style='mso-bookmark:_Toc235985569'><span
lang=EN-GB><o:p>&nbsp;</o:p></span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc235985569'><span class=SpellE>Transmembrane</span>
structures can be carried out with:</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc235985569'><span lang=ZU>t_coffee sample_aln.fasta
-template_file TM</span></span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><span style='mso-bookmark:
_Toc235985569'><a name="_Toc235985561">Multiple Sequence Predictions</a></span></h3>

<p class=MyNormal><span style='mso-bookmark:_Toc235985569'>Used this way, the
method will produce for each sequence a secondary prediction file. GOR is a
single sequence with a relatively low accuracy. It is possible to increase the
accuracy by coupling BLAST and <span class=GramE>GOR,</span> this can be
achieved with the following command:</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc235985569'><span lang=ZU>t_coffee sample_aln.fasta
-template_file PSISSP</span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc235985569'>When doing so, the
predictions for each sequence are obtained by averaging the GOR predictions on
every homologue as reported by a BLAST against NR. By default the BLAST is done
remotely at the NCBI using the <span class=SpellE>blastpgp</span> web service
of the EBI.</span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc235985569'>A similar output can
be obtained for <span class=SpellE><span class=GramE>Transmembrane</span></span>
segment predictions:</span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc235985569'><o:p>&nbsp;</o:p></span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc235985569'><span lang=ZU>t_coffee sample_aln.fasta
-template_file PSITM</span></span></p>

</div>

<p class=MsoBodyText><span style='mso-bookmark:_Toc235985569'><span lang=EN-GB><o:p>&nbsp;</o:p></span></span></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><span style='mso-bookmark:
_Toc235985569'><a name="_Toc235985562">Incorporation of the prediction in the
alignment</a></span></h3>

<p class=MyNormal><span style='mso-bookmark:_Toc235985569'>It is possible to
use the secondary prediction in order to reward the alignment of similar
elements</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc235985569'><span lang=ZU>t_coffee sample_aln.fasta
-template_file PSISSP -method_evaluate_mode ssp -method lalign_id_pair
slow_pair</span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc235985569'>Likewise, it is
possible to use this information with trans-membrane domains</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc235985569'><span lang=ZU>t_coffee sample_aln.fasta
-template_file PSITM -method_evaluate_mode tm -method lalign_id_pair slow_pair</span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc235985569'>The overall effect
is very crude and amounts to over-weighting by 30% the score obtained when
matching two residues in a similar secondary structure state. The net
consequence is that residues in similar predicted states tend to be aligned
more easily.</span></p>

<h3 style='mso-list:none'><span style='mso-bookmark:_Toc235985569'><o:p>&nbsp;</o:p></span></h3>

<h3 style='mso-list:none'><span style='mso-bookmark:_Toc235985569'><a
name="_Toc235985563">Using other secondary structure predictions</a></span></h3>

<p class=MyNormal><span style='mso-bookmark:_Toc235985569'>If you have your own
predictions, you can use them. All you need is to produce a template file where
the file containing the secondary structure prediction is declared along with
the sequence:</span></p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 31.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmallCxSpFirst style='margin-right:0cm;mso-add-space:auto;
background:#E6E6E6;border:none;mso-border-alt:solid windowtext .5pt;padding:
0cm;mso-padding-alt:1.0pt 4.0pt 1.0pt 31.0pt'><span style='mso-bookmark:_Toc235985569'><span
lang=EN-GB><o:p>&nbsp;</o:p></span></span></p>

<p class=examplefilesmallCxSpMiddle style='margin-right:0cm;mso-add-space:auto;
background:#E6E6E6;border:none;mso-border-alt:solid windowtext .5pt;padding:
0cm;mso-padding-alt:1.0pt 4.0pt 1.0pt 31.0pt'><span style='mso-bookmark:_Toc235985569'><span
lang=EN-GB>&gt;<span class=SpellE><span class=GramE>hmgl</span>_wheat</span>
_E_ <span class=SpellE>hmgl_wheat.ssp</span></span></span></p>

<p class=examplefilesmallCxSpMiddle style='margin-right:0cm;mso-add-space:auto;
background:#E6E6E6;border:none;mso-border-alt:solid windowtext .5pt;padding:
0cm;mso-padding-alt:1.0pt 4.0pt 1.0pt 31.0pt'><span style='mso-bookmark:_Toc235985569'><span
lang=EN-GB>&gt;<span class=SpellE><span class=GramE>hmgb</span>_chite</span>
_E_ <span class=SpellE>hmgb_chite.ssp</span></span></span></p>

<p class=examplefilesmallCxSpMiddle style='margin-right:0cm;mso-add-space:auto;
background:#E6E6E6;border:none;mso-border-alt:solid windowtext .5pt;padding:
0cm;mso-padding-alt:1.0pt 4.0pt 1.0pt 31.0pt'><span style='mso-bookmark:_Toc235985569'><span
lang=EN-GB>&gt;<span class=GramE>hmgl</span>_trybr3 _E_ hmgl_trybr3.ssp</span></span></p>

<p class=examplefilesmallCxSpMiddle style='margin-right:0cm;mso-add-space:auto;
background:#E6E6E6;border:none;mso-border-alt:solid windowtext .5pt;padding:
0cm;mso-padding-alt:1.0pt 4.0pt 1.0pt 31.0pt'><span style='mso-bookmark:_Toc235985569'><span
lang=EN-GB>&gt;<span class=GramE>hmgl</span>_wheat2 _E_ hmgl_wheat2.ssp</span></span></p>

<p class=examplefilesmallCxSpMiddle style='margin-right:0cm;mso-add-space:auto;
background:#E6E6E6;border:none;mso-border-alt:solid windowtext .5pt;padding:
0cm;mso-padding-alt:1.0pt 4.0pt 1.0pt 31.0pt'><span style='mso-bookmark:_Toc235985569'><span
lang=EN-GB>&gt;<span class=SpellE><span class=GramE>hmgt</span>_mouse</span> _E_
<span class=SpellE>hmgt_mouse.ssp</span></span></span></p>

<p class=examplefilesmallCxSpMiddle style='margin-right:0cm;mso-add-space:auto;
background:#E6E6E6;border:none;mso-border-alt:solid windowtext .5pt;padding:
0cm;mso-padding-alt:1.0pt 4.0pt 1.0pt 31.0pt'><span style='mso-bookmark:_Toc235985569'><span
lang=EN-GB>&gt;<span class=SpellE><span class=GramE>hmgl</span>_trybr</span>
_E_ <span class=SpellE>hmgl_trybr.ssp</span></span></span></p>

<p class=examplefilesmallCxSpLast style='margin-right:0cm;mso-add-space:auto;
background:#E6E6E6;border:none;mso-border-alt:solid windowtext .5pt;padding:
0cm;mso-padding-alt:1.0pt 4.0pt 1.0pt 31.0pt'><span style='mso-bookmark:_Toc235985569'><span
lang=EN-GB><o:p>&nbsp;</o:p></span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc235985569'><span class=GramE>where</span>
each template looks like this:</span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc235985569'><o:p>&nbsp;</o:p></span></p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 31.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmallCxSpFirst style='margin-right:0cm;mso-add-space:auto;
background:#E6E6E6;border:none;mso-border-alt:solid windowtext .5pt;padding:
0cm;mso-padding-alt:1.0pt 4.0pt 1.0pt 31.0pt'><span style='mso-bookmark:_Toc235985569'><span
lang=EN-GB><o:p>&nbsp;</o:p></span></span></p>

<p class=examplefilesmallCxSpMiddle style='margin-right:0cm;mso-add-space:auto;
background:#E6E6E6;border:none;mso-border-alt:solid windowtext .5pt;padding:
0cm;mso-padding-alt:1.0pt 4.0pt 1.0pt 31.0pt'><span style='mso-bookmark:_Toc235985569'><span
lang=EN-GB>&gt;<span class=SpellE><span class=GramE>hmgl</span>_wheat</span></span></span></p>

<p class=examplefilesmallCxSpMiddle style='margin-right:0cm;mso-add-space:auto;
background:#E6E6E6;border:none;mso-border-alt:solid windowtext .5pt;padding:
0cm;mso-padding-alt:1.0pt 4.0pt 1.0pt 31.0pt'><span style='mso-bookmark:_Toc235985569'><span
lang=EN-GB>CCCCCCCCCCCCHHHHHHHCCCCCCCCCHHHHHHHHHHHHHHHCCCCHHHHHHHHHHHHHHHCE</span></span></p>

<p class=examplefilesmallCxSpLast style='margin-right:0cm;mso-add-space:auto;
background:#E6E6E6;border:none;mso-border-alt:solid windowtext .5pt;padding:
0cm;mso-padding-alt:1.0pt 4.0pt 1.0pt 31.0pt'><span style='mso-bookmark:_Toc235985569'><span
lang=EN-GB><o:p>&nbsp;</o:p></span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc235985569'>You can then run
T-Coffee using your own template file</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc235985569'><span lang=ZU>t_coffee sample_aln.fasta
-template_file &lt;template_file&gt; -method_evaluate_mode ssp -method
lalign_id_pair slow_pair</span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc235985569'><o:p>&nbsp;</o:p></span></p>

<h3 style='mso-list:none'><span style='mso-bookmark:_Toc235985569'><span
style='mso-tab-count:1'>    </span><a name="_Toc235985564">Output of the
prediction</a></span></h3>

<p class=MyNormal><span style='mso-bookmark:_Toc235985569'>You can output a <span
class=GramE>color coded</span> version of your alignment using the predicted
structures</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc235985569'><span lang=ZU>t_coffee sample_aln.fasta
-template_file PSISSP -output sec_html</span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc235985569'>A similar result can
be obtained with trans-membrane regions:</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc235985569'><span lang=ZU>t_coffee sample_aln.fasta
-template_file PSITM -output tm_html</span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc235985569'><o:p>&nbsp;</o:p></span></p>

<h2><span style='mso-bookmark:_Toc235985569'><span lang=EN-GB>Aligning
Sequences and Structures</span></span></h2>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985570">Mixing
Sequences and Structures</a></h3>

<p class=MyNormal>Gather your sequences in the same file. Name your structures according
to their PDB identifier. The file <span class=SpellE>three_pdb_two_seq.fasta</span>
contains five sequences, three are the sequences of PDB structures and two are
regular sequences.</p>

<p class=MyNormal>What you want to do is to build a T-Coffee library where
sequences <span class=GramE>with a known structures</span> are aligned with a
structure alignment program (like sap) while the other sequences are aligned
using regular T-Coffee methods. You can achieve this with the following
command:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee three_pdb_two_seq.fasta -method sap_pair,slow_pair -template_file PDB</span></p>

</div>

<p class=MyNormal>The option -<span class=SpellE>template_file</span> is here
to tell the program how to find the PDB. <span class=GramE>In that case.</span>
EXPRESSO means that a remote BLAST (the EBI BLAST) will be used to identify the
best targets. If your sequences are already named according to their PDB name,
you can use:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style="mso-spacerun:yes"> </span><span lang=ZU>PROMPT: t_coffee
three_pdb_two_seq.fasta -method sap_pair,slow_pair -template_file _SELF_P_</span></p>

</div>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MyNormal>_SELF_ means that the PDB identifier is the name of the
sequences, while _P_ is an indication that the template is indeed a PDB. These
indications are necessary for T-Coffee to fetch the relevant structures.</p>

<p class=MyNormal>The good news is that you do not need to have PDB installed
locally as T-Coffee will automatically fetch the structures directly from RCSB
(the home of PDB). Of course, if your dataset only contains structures, your
alignment becomes a structural alignment.</p>

<p class=MyNormal>If you have a fugue license, you can also add the fugue
method to your run. Fugue will align the structures with sequences whose
structure is unknown (this is called threading). </p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee three_pdb_two_seq.fasta -method sap_pair,slow_pair,fugue_pair
-template_file _SELF_P_</span></p>

</div>

<p class=MyNormal>This can be written more concisely, using one of T-Coffee <span
class=SpellE>special_modes</span>:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU><span
style='mso-tab-count:1'>   </span>PROMPT: t_coffee three_pdb_two_seq.fasta -mode
3dcoffee</span></p>

</div>

<p class=MyNormal><span class=GramE>or</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU><span
style='mso-tab-count:1'>   </span>PROMPT: t_coffee three_pdb_two_seq.fasta -mode
expresso</span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985571">Using
Sequences only</a></h3>

<p class=MyNormal>What often happens is that you have already built a dataset
with sequences that are very similar to PDB sequences but not exactly
identical. It may even be the case that the real sequence and the PDB one do
not match exactly because of some genetic engineering on the structure. In this
case, you have no structure whose sequence is exactly similar to the sequences
in your dataset. All you need to do is to declare the equivalence
sequences/structures and run T-Coffee, just like Expresso does.</p>

<p class=MyNormal>The first step is to fill up a template file that contains an
explicit declaration of the structures corresponding to your sequences. The
format is very simple and <span class=SpellE>fasta</span>-like. You can use the
file: <span class=SpellE>sproteases_small.template_file</span></p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6;margin-left:0cm;margin-right:1.0cm'>

<p class=examplefilesmallCxSpFirst><span lang=EN-GB>&gt;<span class=GramE>sp</span>|P08246|ELNE_HUMAN
_P_ 1PPGE</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;<span class=GramE>sp</span>|P20160|CAP7_HUMAN
_P_ 1AE5</span></p>

<p class=examplefilesmallCxSpMiddle><span lang=EN-GB>&gt;<span class=GramE>sp</span>|P00757|KLKB4_MOUSE
_P_ 1SGFX</span></p>

<p class=examplefilesmallCxSpLast><span lang=EN-GB>&gt;<span class=GramE>sp</span>|Q6H321|KLK2_HORSE
_P_ 1GVZA</span></p>

</div>

<p class=MyNormal>In this file, the first line is telling us that sequence <span
class=GramE>sp</span>|P08246|ELNE_HUMAN is associated with the structural
template 1PPGE. The sequence and the structure do not need to be identical
although we recommend using structural templates more than 60% identical with
your actual sequences (i.e. similar enough so that they generate a non
ambiguous alignment). If your template file is ready, all you need to do is run
the following command.</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sproteases_small.fasta -method slow_pair, lalign_id_pair, sap_pair
-template_file sproteases_small.template_file </span></p>

</div>

<p class=MyNormal>When you run this once, T-Coffee goes and fetches the
structures. It will then align them using sap. It takes a lot of time to fetch
structures, and it takes even more time to align them with sap. This is why
T-Coffee saves these important intermediate results in a special location
called the cache. By default, your cache is in ~/<span class=GramE>.t</span>_coffee/cache,
it is a good idea to empty it from time to time</p>

<h2><a name="_Toc235985572"><span lang=EN-GB>Aligning Profile using Structural
Information</span></a></h2>

<p class=MsoBodyText><span lang=EN-GB>If you have two profiles to align, an
ideal situation is when your profiles each contain one or more structures.
These structures will guide the alignment of the profiles, even if they contain
very distally related sequences. We have prepared two such profiles
(prf1_pdb1.aln, prf2_pdb2.aln). You have two choices here. All you need is a
template file that declares which sequences have a known structure. If you only
want to align sequences, you can try:</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
<a name="OLE_LINK2"></a><a name="OLE_LINK1"><span style='mso-bookmark:OLE_LINK2'>t_coffee
-profile=profile1_pdb1.aln, profile2_pdb2.aln -method sap_pair -profile_template_file
two_profiles.template_file </span></a></span></p>

</div>

<h1><a name="_Toc235985573"><span lang=EN-GB>How Good Is Your </span></a><span
class=GramE><span style='mso-bookmark:_Toc235985573'><span lang=EN-GB>Alignment
?</span></span></span><span style='mso-bookmark:_Toc235985573'></span></h1>

<p class=MsoBodyText><span lang=EN-GB>There are three strategies for evaluating
your alignment. Structure is a killer. If you have two structures available for
your protein family, you are in an ideal situation and you can use the iRMSD.
If you don't, you are left with the option of using <span class=GramE>sequence
based</span> methods like the CORE index. These do pretty well in the CORE
regions, but can be limited in the loops. Another killer, less often at hand,
is the use of functional information. If you know some residues MUST be aligned
because they are functionally related, you can easily set up an evaluation
procedure using T-Coffee.</span></p>

<h2><a name="_Toc235985574"><span lang=EN-GB>Evaluating Alignments with The
CORE index</span></a></h2>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985575">Computing
the Local CORE Index</a> </h3>

<p class=MyNormal>The CORE index is an estimation of the consistency between
your alignment and the computed library. The higher the consistency, the better
the alignment. The score reported with every T-Coffee alignment is the <span
class=SpellE>concistency</span> score. <span class=GramE>However, if you want
to go further and estimate the local <span class=SpellE>concistency</span>
(known as the CORE index).</span> Simply request one extra output:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sproteases_small.fasta -output=clustalw,html</span></p>

</div>

<p class=MyNormal>The format html leads to the output of a file named
sproteases_small.html. Open this file. It contains a colorized version of your
alignment. In this colorized version, positions that have no <span
class=SpellE>concistency</span> with the library are in blue, a little in
green, better positions in yellow, then orange, then red. You can expect yellow
positions to be entirely correct.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><span
style="mso-spacerun:yes">  </span><a name="_Toc235985576">Computing the CORE
index of any alignment</a></h3>

<p class=MyNormal>You can evaluate any existing alignment with the CORE index.
All you need to do is provide that alignment with the -<span class=SpellE>infile</span>
flag and specify that you want to evaluate it:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -infile=sproteases_small.g10.cw_aln -output=html -score</span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><span
style="mso-spacerun:yes">  </span><a name="_Toc235985577">Filtering Bad
Residues</a></h3>

<p class=MyNormal>The local consistency score is between 0 and 9. If you need
to build a profile or identify a signature or do some phylogeny, it may be a
good idea to remove portions that are too unreliable. Here is how you can do
it.</p>

<p class=MyNormal>You will first need to produce a score version of your
alignment in <span class=GramE>machine readable</span> format. This format is called
<span class=SpellE>score_ascii</span>:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -infile=sproteases_small.g10.cw_aln -output=score_ascii</span></p>

</div>

<p class=MyNormal>You will then need to use seq_reformat to filter out the
portions you are not interested in, using the file sproteases_small.g10.score_ascii
as a cache for filtering. For instance, use the following command to only keep
the residues whose score is between 5 and 9.</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.g10.cw_aln -struc_in
sproteases_small.g10.score_ascii -struc_in_f number_aln -action<span
style="mso-spacerun:yes">  </span>+keep '[5-9]'</span></p>

</div>

<p class=MyNormal>Not so neat... The reason is that most columns tend to be <span
class=SpellE>heterogenous</span> and contain a couple of unhappy residues. If
you would rather generate nice blocks, filter according to the <span
class=SpellE>consencus</span>. This is easy and simply requires adding the +<span
class=SpellE>use_cons</span> filter</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.g10.cw_aln -struc_in
sproteases_small.g10.score_ascii -struc_in_f number_aln -action<span
style="mso-spacerun:yes">  </span>+use_cons +keep '[5-9]'</span></p>

</div>

<p class=MyNormal>Removing columns of gaps is just as easy. You simply need to
add the switch +<span class=SpellE>rm_gap</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.g10.cw_aln -struc_in
sproteases_small.g10.score_ascii -struc_in_f number_aln -action<span
style="mso-spacerun:yes">  </span>+use_cons +keep '[5-9]' +rm_gap</span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><span
style="mso-spacerun:yes">  </span><a name="_Toc235985578">Filtering Gap Columns</a></h3>

<p class=MyNormal>You may want to remove columns that contain too many gaps. It
is just a small variation around the <span class=SpellE>rm_gap</span> switch:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat -in sproteases_small.g10.cw_aln -struc_in
sproteases_small.g10.score_ascii -struc_in_f number_aln -action +rm_gap 50</span></p>

</div>

<p class=MyNormal>Will remove all the columns containing 40% or more gaps.</p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<h2><a name="_Toc235985579"><span lang=EN-GB>Evaluating an Alignment Using
Structural Information: APDB and iRMSD</span></a></h2>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985580">What
is the iRMSD?</a></h3>

<p class=MyNormal>APDB and the iRMSD are two closely related measures meant to
evaluate the accuracy of a sequence alignment without using a structure based
reference alignment. The iRMSD is a follow up of the APDB measure and we now
recommend using the iRMSD rather than APDB.</p>

<p class=MyNormal>Although it may seem that the iRMSD was an attempt to get
free <span class=SpellE>iPODs</span> from Apple, it is not (or at least we
never got the <span class=SpellE>iPODs</span>). The iRMSD is a special RMSD (It
stands for intra-<span class=SpellE>catener</span>) where the alignments are
evaluated using the structural information of the sequences with known
structures.</p>

<p class=MyNormal>The strength of the iRMSD is its independence from a specific
superposition models.<span style="mso-spacerun:yes">  </span>When using the
iRMSD to evaluate the score of a sequence alignment, one does not need to
superpose the two structures and deduce a sequence alignment that will then be
compared with the target alignment. In practice, we use a Normalized version of
the iRMSD, the <span class=SpellE>NiRMSD</span> that makes it possible to
compare alternative alignments of different length. From a structural point of
view, the iRMSD has a meaning very similar to the iRMSD and it <span
class=SpellE>beahaves</span> in a similar fashion from a numerical point of
view (similar ranges in Angstroms). </p>

<p class=MyNormal>The first step of APDB is to measure the distances between
the C<span style='font-family:Symbol'>a</span> of each residue and its
neighbors. Neighborhood is defined as a sphere of <b style='mso-bidi-font-weight:
normal'>radius &#8211;<span class=SpellE>maximum_distance</span></b> (10Å by
default). However, by setting <b style='mso-bidi-font-weight:normal'>&#8211;<span
class=SpellE>local_mode</span></b> to window, the sphere can be replaced with
a window of 1/2 size '<b style='mso-bidi-font-weight:normal'>-<span
class=SpellE>maximum_distance</span>'</b> residues.</p>

<p class=MyNormal><!--[if gte vml 1]><v:shapetype id="_x0000_t75" coordsize="21600,21600"
 o:spt="75" o:preferrelative="t" path="m@4@5l@4@11@9@11@9@5xe" filled="f"
 stroked="f">
 <v:stroke joinstyle="miter"/>
 <v:formulas>
  <v:f eqn="if lineDrawn pixelLineWidth 0"/>
  <v:f eqn="sum @0 1 0"/>
  <v:f eqn="sum 0 0 @1"/>
  <v:f eqn="prod @2 1 2"/>
  <v:f eqn="prod @3 21600 pixelWidth"/>
  <v:f eqn="prod @3 21600 pixelHeight"/>
  <v:f eqn="sum @0 0 1"/>
  <v:f eqn="prod @6 1 2"/>
  <v:f eqn="prod @7 21600 pixelWidth"/>
  <v:f eqn="sum @8 21600 0"/>
  <v:f eqn="prod @7 21600 pixelHeight"/>
  <v:f eqn="sum @10 21600 0"/>
 </v:formulas>
 <v:path o:extrusionok="f" gradientshapeok="t" o:connecttype="rect"/>
 <o:lock v:ext="edit" aspectratio="t"/>
</v:shapetype><v:shape id="_x0000_i1025" type="#_x0000_t75" style='width:218pt;
 height:191pt'>
 <v:imagedata src="t_coffee_tutorial_files/image001.png" o:title=""/>
</v:shape><![endif]--><![if !vml]><img border=0 width=220 height=193
src="t_coffee_tutorial_files/image002.gif" v:shapes="_x0000_i1025"><![endif]></p>

<p class=MyNormal>Given two aligned residues (X and Y on the Figure) the iRMSD
measure is an attempt to estimate the neighborhood support for the XY
alignment. This is done by measuring the difference of distances between X and
Y and every other pair of aligned residues within the same sphere (W and Z on
Figure 1). The iRMSD is obtained by measuring the average Root Mean Square of
these differences of distances. The lower the iRMSD, the better the alignment.
However, an alignment can obtain a good iRMSD by simply having few aligned
residues. To avoid this, the program also reports the <span class=SpellE>NiRMSD</span>=
<span class=GramE>MIN(</span>L1,L2)*iRMSD/Number Considered columns.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985581">How
to Efficiently Use Structural Information</a></h3>

<p class=MyNormal>When it comes to evaluating Multiple Sequence Alignments,
nothing beats structural information. To use the methods we describe here, you
will need to have at least two structures, similar enough (&gt;60%) to two
sequences in your dataset.</p>

<p class=MyNormal>Here an outline of the best way to proceed:</p>

<p class=MyNormal style='margin-left:108.3pt;text-indent:-37.5pt;mso-list:l13 level1 lfo38;
tab-stops:list 108.3pt'><![if !supportLists]><span style='mso-bidi-font-family:
"Times New Roman"'><span style='mso-list:Ignore'>1-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]>Make sure you include two structures whose
sequences are so distantly related that most of the other sequences are
intermediates.</p>

<p class=MyNormal style='margin-left:108.3pt;text-indent:-37.5pt;mso-list:l13 level1 lfo38;
tab-stops:list 108.3pt'><![if !supportLists]><span style='mso-bidi-font-family:
"Times New Roman"'><span style='mso-list:Ignore'>2-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]>Align your sequences without using the
structural information (i.e. t_coffee, muscle...)</p>

<p class=MyNormal style='margin-left:108.3pt;text-indent:-37.5pt;mso-list:l13 level1 lfo38;
tab-stops:list 108.3pt'><![if !supportLists]><span style='mso-bidi-font-family:
"Times New Roman"'><span style='mso-list:Ignore'>3-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]>Evaluate your alignment with iRMSD (see <span
class=GramE>later<span style="mso-spacerun:yes">  </span>in</span> this
section). The score will be S1</p>

<p class=MyNormal style='margin-left:108.3pt;text-indent:-37.5pt;mso-list:l13 level1 lfo38;
tab-stops:list 108.3pt'><![if !supportLists]><span style='mso-bidi-font-family:
"Times New Roman"'><span style='mso-list:Ignore'>4-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]>Realign your sequences, but this time using
structural information (Expresso)</p>

<p class=MyNormal style='margin-left:108.3pt;text-indent:-37.5pt;mso-list:l13 level1 lfo38;
tab-stops:list 108.3pt'><![if !supportLists]><span style='mso-bidi-font-family:
"Times New Roman"'><span style='mso-list:Ignore'>5-<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></span><![endif]>Measure the score of that alignment (Score=S2)</p>

<p class=MyNormal>If S1 and S2 are almost similar, it means your distantly
related structures were well aligned, and you can expect the intermediate
sequences to be well aligned as well. If S2 is much better than S1, you can
expect the structures to be well aligned in the second alignment, while there
is no guaranty that the alignment of the intermediate sequences has improved as
well, although in practice it often does </p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985582">Evaluating
an Alignment With the iRMSD Package</a></h3>

<p class=MyNormal>Let us evaluate the alignment produced by Expresso, using the
<span class=SpellE>template_file</span> returned by <span class=SpellE>expresso</span>:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg irmsd sproteases_small.expresso -template_file
sproteases_small.template_file</span></p>

</div>

<p class=MyNormal>This will deliver a long output. The most interesting bit is
at the bottom:</p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6'>

<p class=examplefileCxSpFirst><span lang=EN-GB>#TOTAL for the Full MSA</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB><span
style="mso-spacerun:yes">        </span>TOTAL<span
style="mso-spacerun:yes">     </span>EVALUATED:<span style="mso-spacerun:yes"> 
</span>52.90 %</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB><span
style="mso-spacerun:yes">        </span>TOTAL<span
style="mso-spacerun:yes">     </span>APDB<span class=GramE>:<span
style="mso-spacerun:yes">       </span>81.59</span> %</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB><span
style="mso-spacerun:yes">        </span>TOTAL<span
style="mso-spacerun:yes">     </span>iRMSD<span class=GramE>:<span
style="mso-spacerun:yes">       </span>0.71</span> <span class=SpellE>Angs</span></span></p>

<p class=examplefileCxSpLast><span lang=EN-GB><span
style="mso-spacerun:yes">        </span>TOTAL<span
style="mso-spacerun:yes">     </span><span class=SpellE>NiRMSD</span><span
class=GramE>:<span style="mso-spacerun:yes">      </span>1.33</span> <span
class=SpellE>Angs</span></span></p>

</div>

<p class=MyNormal>APDB is an older measure, less robust than the iRMSD and it
is an attempt to estimate the fraction of pairs of residues whose alignment
seems to be correct form a structural point of view. The higher APDB, the
better the alignment, the lower the <span class=SpellE>NiRMSD</span>, the
better the alignment.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985583">Evaluating
Alternative Alignments</a></h3>

<p class=MyNormal>The strength of <span class=GramE>structure based</span>
alignments is that they make it possible to compare alternative alignments. In
this case let us consider:</p>

<table class=MsoNormalTable border=1 cellspacing=0 cellpadding=0
 style='border-collapse:collapse;border:none;mso-border-alt:solid windowtext .5pt;
 mso-yfti-tbllook:480;mso-padding-alt:0cm 5.4pt 0cm 5.4pt;mso-border-insideh:
 .5pt solid windowtext;mso-border-insidev:.5pt solid windowtext'>
 <tr style='mso-yfti-irow:0;mso-yfti-firstrow:yes'>
  <td width=100 valign=top style='width:100.15pt;border:solid windowtext 1.0pt;
  mso-border-alt:solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MyNormal>Method</p>
  </td>
  <td width=127 valign=top style='width:126.6pt;border:solid windowtext 1.0pt;
  border-left:none;mso-border-left-alt:solid windowtext .5pt;mso-border-alt:
  solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MyNormal>File</p>
  </td>
  <td width=100 valign=top style='width:100.35pt;border:solid windowtext 1.0pt;
  border-left:none;mso-border-left-alt:solid windowtext .5pt;mso-border-alt:
  solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MyNormal><span class=SpellE>NiRMSD</span></p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:1'>
  <td width=100 valign=top style='width:100.15pt;border:solid windowtext 1.0pt;
  border-top:none;mso-border-top-alt:solid windowtext .5pt;mso-border-alt:solid windowtext .5pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MyNormal>Expresso</p>
  </td>
  <td width=127 valign=top style='width:126.6pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  mso-border-top-alt:solid windowtext .5pt;mso-border-left-alt:solid windowtext .5pt;
  mso-border-alt:solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MyNormal><span class=SpellE><span class=GramE>sproteases</span>_small.expresso</span></p>
  </td>
  <td width=100 valign=top style='width:100.35pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  mso-border-top-alt:solid windowtext .5pt;mso-border-left-alt:solid windowtext .5pt;
  mso-border-alt:solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MyNormal>1.33 <span style='mso-bidi-font-family:"Times New Roman"'>Å</span></p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:2'>
  <td width=100 valign=top style='width:100.15pt;border:solid windowtext 1.0pt;
  border-top:none;mso-border-top-alt:solid windowtext .5pt;mso-border-alt:solid windowtext .5pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MyNormal>T-Coffee</p>
  </td>
  <td width=127 valign=top style='width:126.6pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  mso-border-top-alt:solid windowtext .5pt;mso-border-left-alt:solid windowtext .5pt;
  mso-border-alt:solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MyNormal><span class=SpellE><span class=GramE>sproteases</span>_small.tc_aln</span></p>
  </td>
  <td width=100 valign=top style='width:100.35pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  mso-border-top-alt:solid windowtext .5pt;mso-border-left-alt:solid windowtext .5pt;
  mso-border-alt:solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MyNormal>1.35 <span style='mso-bidi-font-family:"Times New Roman"'>Å</span></p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:3'>
  <td width=100 valign=top style='width:100.15pt;border:solid windowtext 1.0pt;
  border-top:none;mso-border-top-alt:solid windowtext .5pt;mso-border-alt:solid windowtext .5pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MyNormal>ClustalW</p>
  </td>
  <td width=127 valign=top style='width:126.6pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  mso-border-top-alt:solid windowtext .5pt;mso-border-left-alt:solid windowtext .5pt;
  mso-border-alt:solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MyNormal><span class=SpellE><span class=GramE>sproteases</span>_small.cw_aln</span></p>
  </td>
  <td width=100 valign=top style='width:100.35pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  mso-border-top-alt:solid windowtext .5pt;mso-border-left-alt:solid windowtext .5pt;
  mso-border-alt:solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MyNormal>1.52 <span style='mso-bidi-font-family:"Times New Roman"'>Å</span></p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:4'>
  <td width=100 valign=top style='width:100.15pt;border:solid windowtext 1.0pt;
  border-top:none;mso-border-top-alt:solid windowtext .5pt;mso-border-alt:solid windowtext .5pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MyNormal>Mafft</p>
  </td>
  <td width=127 valign=top style='width:126.6pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  mso-border-top-alt:solid windowtext .5pt;mso-border-left-alt:solid windowtext .5pt;
  mso-border-alt:solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MyNormal><span class=SpellE><span class=GramE>sproteases</span>_small.mafft</span></p>
  </td>
  <td width=100 valign=top style='width:100.35pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  mso-border-top-alt:solid windowtext .5pt;mso-border-left-alt:solid windowtext .5pt;
  mso-border-alt:solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MyNormal>1.36 <span style='mso-bidi-font-family:"Times New Roman"'>Å</span></p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:5;mso-yfti-lastrow:yes'>
  <td width=100 valign=top style='width:100.15pt;border:solid windowtext 1.0pt;
  border-top:none;mso-border-top-alt:solid windowtext .5pt;mso-border-alt:solid windowtext .5pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MyNormal>Muscle</p>
  </td>
  <td width=127 valign=top style='width:126.6pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  mso-border-top-alt:solid windowtext .5pt;mso-border-left-alt:solid windowtext .5pt;
  mso-border-alt:solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MyNormal><span class=SpellE><span class=GramE>sproteases</span>_small.muscle</span></p>
  </td>
  <td width=100 valign=top style='width:100.35pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  mso-border-top-alt:solid windowtext .5pt;mso-border-left-alt:solid windowtext .5pt;
  mso-border-alt:solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MyNormal>1.34 <span style='mso-bidi-font-family:"Times New Roman"'>Å</span></p>
  </td>
 </tr>
</table>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MyNormal>As expected, Expresso delivers the best alignment from a
structural point of view. This makes sense, since Expresso <span class=SpellE>explicitely</span>
USES structural information. The other figures show us that the structural
based alignment is only marginally better than most sequences based alignments.
Muscle seems to have a small edge here although the reality is that all these
figures are impossible to distinguish with the notable exception of ClustalW</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985584">Identifying
the most distantly related sequences in your dataset</a></h3>

<p class=MyNormal>In order to identify the most distantly related sequences in
a dataset, you can use the seq_reformat utility, in order to compare all the
sequences two by two and pick up the two having the lowest level of identity:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -other_pg seq_reformat sproteases_small.fasta -output sim_idscore |
grep TOP |sort -rnk3</span></p>

</div>

<p class=MyNormal>This <span class=SpellE>sim_idscore</span> indicates that
every pair of sequences will need to be aligned when estimating the similarity.
The <span class=SpellE>ouput</span> (below) indicates that the two sequences
having the lowest level of identity are AEDAE and MOUSE. It may not be a bad
idea to choose these sequences (if possible) for evaluating your MSA.</p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6'>

<p class=examplefileverysmallCxSpFirst><span style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span style='mso-ansi-language:EN-US'>TOP<span
style="mso-spacerun:yes">        </span>16<span style="mso-spacerun:yes">  
</span>10<span style="mso-spacerun:yes">       </span><span class=GramE>28.00<span
style="mso-spacerun:yes">  </span>sp</span>|P29786|TRY3_AEDAE<span
style="mso-spacerun:yes">      </span>sp|Q6H321|KLK2_HORSE<span
style="mso-spacerun:yes">   </span>28.00<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span style='mso-ansi-language:EN-US'>TOP<span
style="mso-spacerun:yes">        </span>16<span style="mso-spacerun:yes">   
</span>7<span style="mso-spacerun:yes">       </span><span class=GramE>28.00<span
style="mso-spacerun:yes">  </span>sp</span>|P29786|TRY3_AEDAE<span
style="mso-spacerun:yes">      </span>sp|P08246|ELNE_HUMAN<span
style="mso-spacerun:yes">   </span>28.00<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span style='mso-ansi-language:EN-US'>TOP<span
style="mso-spacerun:yes">        </span>16 <span
style="mso-spacerun:yes">   </span>1<span style="mso-spacerun:yes">      
</span><span class=GramE>28.00<span style="mso-spacerun:yes">  </span>sp</span>|P29786|TRY3_AEDAE<span
style="mso-spacerun:yes">      </span>sp|P08884|GRAE_MOUSE<span
style="mso-spacerun:yes">   </span>28.00<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span style='mso-ansi-language:EN-US'>TOP<span
style="mso-spacerun:yes">        </span>15<span style="mso-spacerun:yes">  
</span>14<span style="mso-spacerun:yes">       </span>27.00<span
style="mso-spacerun:yes">    </span><span class=GramE>sp</span>|P80015|CAP7_PIG<span
style="mso-spacerun:yes">     </span>sp|P00757|KLKB4_MOUSE<span
style="mso-spacerun:yes">   </span>27.00<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span style='mso-ansi-language:EN-US'>TOP<span
style="mso-spacerun:yes">        </span>12<span style="mso-spacerun:yes">   
</span>9<span style="mso-spacerun:yes">       </span><span class=GramE>27.00<span
style="mso-spacerun:yes">  </span>sp</span>|P20160|CAP7_HUMAN<span
style="mso-spacerun:yes">      </span>sp|Q91VE3|KLK7_MOUSE<span
style="mso-spacerun:yes">   </span>27.00<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span style='mso-ansi-language:EN-US'>TOP<span
style="mso-spacerun:yes">      </span><span
style="mso-spacerun:yes">   </span>9<span style="mso-spacerun:yes">   
</span>7<span style="mso-spacerun:yes">       </span><span class=GramE>27.00<span
style="mso-spacerun:yes">  </span>sp</span>|Q91VE3|KLK7_MOUSE<span
style="mso-spacerun:yes">      </span>sp|P08246|ELNE_HUMAN<span
style="mso-spacerun:yes">   </span>27.00<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpLast><span style='mso-ansi-language:EN-US'>TOP<span
style="mso-spacerun:yes">        </span>16<span style="mso-spacerun:yes">   
</span>2<span style="mso-spacerun:yes">       </span><span class=GramE>26.00<span
style="mso-spacerun:yes">  </span>sp</span>|P29786|TRY3_AEDAE<span
style="mso-spacerun:yes">     </span>sp|P21844|MCPT5_MOUSE<span
style="mso-spacerun:yes">   </span>26.00<o:p></o:p></span></p>

</div>

<h2><a name="_Toc235985585"><span lang=EN-GB>Evaluating an Alignment according
to your own Criterion</span></a></h2>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985586">Establishing
Your Own Criterion</a></h3>

<p class=MyNormal>Any kind of Feature can easily be turned into an evaluation
grid. For instance, the protease sequences we have been using here have a <span
class=GramE>well characterized</span> binding site. A possible evaluation can
be made as follows. <span class=GramE>let</span> us consider the <span
class=SpellE>Swissprot</span> annotation of the two most distantly related
sequences. These two sequences contain the electron relay system of the
proteases. We can use it to build an evaluation library: in P29786, the first <span
class=SpellE>Histidine</span> is at position <span class=GramE>68,</span> while
in P21844 this <span class=SpellE>Histidine</span> is on position 66. We can
therefore build a library that will check whether these residues are properly
aligned in any MSA. The library will look like this:</p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6'>

<p class=examplefileverysmallCxSpFirst><span style='mso-ansi-language:EN-US'>!
TC_LIB_FORMAT_01<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span style='mso-ansi-language:EN-US'>2<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P21844|MCPT5_MOUSE 247
MHLLTLHLLLLLLGSSTKAGEIIGGTECIPHSRPYMAYLEIVTSENYLSACSGFLIRRNFVLTAAHCAGRSITVLLGAHNKTSKEDTWQKLEVEKQFLHPKYDENLVVHDIMLLKLKEKAKLTLGVGTLPLSANFNFIPPGRMCRAVGWGRTNV<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span style='mso-ansi-language:EN-US'>NEPASDTLQEVKMRLQEPQACKHFTSFRHNSQLCVGNPKKMQNVYKGDSGGPLLCAGIAQGIASYVHRNAKPPAVFTRISHYRPWINKILREN<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span class=GramE><span
style='mso-ansi-language:EN-US'>sp</span></span><span style='mso-ansi-language:
EN-US'>|P29786|TRY3_AEDAE 254
MNQFLFVSFCALLDSAKVSAATLSSGRIVGGFQIDIAEVPHQVSLQRSGRHFCGGSIISPRWVLTRAHCTTNTDPAAYTIRAGSTDRTNGGIIVKVKSVIPHPQYNGDTYNYDFSLLELDESIGFSRSIEAIALPDASETVADGAMCTVSGWGDT<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span style='mso-ansi-language:EN-US'>KNVFEMNTLLRAVNVPSYNQAECAAALVNVVPVTEQMICAGYAAGGKDSCQGDSGGPLVSGDKLVGVVSWGKGCALPNLPGVYARVSTVRQWIREVSEV<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span style='mso-ansi-language:EN-US'>#1
2<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpMiddle><span style='mso-ansi-language:EN-US'><span
style="mso-spacerun:yes">    </span><span class=GramE>66<span
style="mso-spacerun:yes">  </span>68</span><span style="mso-spacerun:yes">    
</span>100<o:p></o:p></span></p>

<p class=examplefileverysmallCxSpLast><span style='mso-ansi-language:EN-US'>!
SEQ_1_TO_N<o:p></o:p></span></p>

</div>

<p class=MyNormal>You simply need to cut and paste this library in a file and
use this file as a library to measure the <span class=SpellE>concistency</span>
between your alignment and the <span class=SpellE>correspondances</span>
declared in your library. The following command line also makes it possible to <span
class=SpellE>visualy</span> display the agreement between your sequences and
the library. </p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -infile sproteases_small.aln -lib<span style="mso-spacerun:yes"> 
</span>charge_relay_lib.tc_lib<span style="mso-spacerun:yes">  </span>-score
-output html</span></p>

</div>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<h1><a name="_Toc235985587"><span lang=EN-GB>Integrating External Methods In
T-Coffee</span></a></h1>

<p class=MsoBodyText><a name="_Toc135066776"></a><a name="_Toc102494687"></a><a
name="_Toc102494304"><span style='mso-bookmark:_Toc102494687'><span
style='mso-bookmark:_Toc135066776'><span lang=EN-GB>The real power of T-Coffee
is its ability to seamlessly combine many methods into one. While we try to
integrate as many methods as we can in the default distribution, we do not have
the means to be exhaustive and if you desperately need your favourite method to
be integrated, you will need to bite the bullet </span></span></span></a></p>

<span style='mso-bookmark:_Toc135066776'></span><span style='mso-bookmark:_Toc102494687'></span><span
style='mso-bookmark:_Toc102494304'></span>

<h2><a name="_Toc235985588"><span lang=EN-GB>What Are The Methods Already
Integrated in T-Coffee</span></a></h2>

<p class=MyNormal>Although, it does not necessarily do so explicitly, T-Coffee
always <span class=GramE>end</span> up combining libraries. Libraries are
collections of pairs of residues. Given a set of libraries, T-Coffee makes an
attempt to assemble the alignment with the highest level of consistence. You
can think of the alignment as a timetable. Each library pair would be a request
from students or teachers, and the job of T-Coffee would be to assemble the <span
class=GramE>time table</span> that makes as many people as possible happy</p>

<p class=MyNormal>In T-Coffee, methods replace the students/professors as
constraints generators. These methods can be any standard/non standard
alignment methods that can be used to generate alignments (pairwise, most of
the time). These alignments can be viewed as collections of constraints that
must be fit within the final alignment. Of course, the constraints do not have
to agree with one another</p>

<p class=MyNormal>This section shows you what <span class=GramE>are the <span
class=SpellE>vailable</span> method</span> in T-Coffee, and how you can add
your own methods, either through direct parameterization or via a <span
class=SpellE>perl</span> script. There are two kinds of methods: the internal
and the external. For the internal methods, you simply need to have T-Coffee up
and running. The external methods will require you to <span class=SpellE>instal</span>
a package.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985589">List
of INTERNAL Methods</a></h3>

<p class=MyNormal>Built in methods <span class=SpellE>methods</span> can be
requested using the following names. To </p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>proba</b></span><b style='mso-bidi-font-weight:normal'>_pair</b></span><b
style='mso-bidi-font-weight:normal'><span style='mso-tab-count:1'>      </span></b><span
class=SpellE>Adpated</span> from <span class=SpellE>Probcons</span>, this
method <b>[the current default]</b> uses a pair HMM to compute a pairwise
alignment with a bi-phasic gap penalty. <b style='mso-bidi-font-weight:normal'><o:p></o:p></b></p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>fast</b></span><b style='mso-bidi-font-weight:normal'>_pair</b></span><b
style='mso-bidi-font-weight:normal'><span style='mso-tab-count:1'>          </span></b>Makes
a global <span class=SpellE>fasta</span> style pairwise alignment. For
proteins, matrix=blosum62mt, <span class=SpellE>gep</span>=-1, <span
class=SpellE><span class=GramE>gop</span></span>=-10, <span class=SpellE>ktup</span>=2.
For DNA, matrix=<span class=SpellE>idmat</span> (id=10), <span class=SpellE>gep</span>=-1,
<span class=SpellE><span class=GramE>gop</span></span>=-20, <span class=SpellE>ktup</span>=5.
Each pair of residue is given a score function of the weighting mode defined by
-weight. </p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>slow</b></span><b style='mso-bidi-font-weight:normal'>_pair</b></span><b
style='mso-bidi-font-weight:normal'><span style='mso-tab-count:1'>         </span></b>Identical
to fast pair, but does a full dynamic programming, using the <span
class=SpellE>myers</span> and miller algorithm. This method is recommended if
your sequences are distantly related.</p>

<p class=Mylist><span class=SpellE><span class=GramE>ifast</span>_pair</span></p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>islow</b></span><b style='mso-bidi-font-weight:normal'>_pair</b></span><b
style='mso-bidi-font-weight:normal'><span style='mso-tab-count:1'>        </span><br>
</b>Makes a global <span class=SpellE>fasta</span> <span class=SpellE>alignmnet</span>
using the previously computed pairs as a library. `<span class=SpellE><span
class=GramE>i</span></span>` stands for iterative. Each pair of residue is
given a score function of the weighting mode defined by -weight. The Library
used for the computation is the one computed before the method is used. The <span
class=SpellE>resullt</span> is therefore dependant on the order in methods and <span
class=GramE>library are</span> set via the &#8211;in flag.</p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>align</b></span><b style='mso-bidi-font-weight:normal'>_pdb_pair</b></span><span
style='mso-tab-count:1'>    </span>Uses the align_pdb routine to align two
structures. The pairwise scores are those <span class=SpellE>returnes</span> by
the align_pdb program. If a structure is missing, <span class=SpellE>fast_pair</span>
is used instead. Each pair of residue is given a score function defined by
align_pdb. [UNSUPORTED]</p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>lalign</b></span><b style='mso-bidi-font-weight:normal'>_id_pair</b></span><b
style='mso-bidi-font-weight:normal'><span style='mso-tab-count:1'>  </span></b>Uses
the ten top non intersecting local alignments, as delivered by <span
class=SpellE>lalign</span>. Each <span class=SpellE>alignement</span> is
weighted with its average percent identity.</p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>lalign</b></span><b style='mso-bidi-font-weight:normal'>_rs_s_pair</b></span><span
style='mso-tab-count:1'>   </span>Same as above but does also does self
comparison and uses the <span class=SpellE>lalign</span> <span class=SpellE>raw_score</span>
(s stands for self). This is needed when extracting repeats. </p>

<p class=Mylist><b style='mso-bidi-font-weight:normal'>Matrix<span
style='mso-tab-count:1'>             </span></b>Amy matrix can be requested.
Simply indicate as a method the name of the matrix preceded with an X (i.e.
Xpam250mt). If you indicate such a matrix, all the other methods will simply be
ignored, and a standard fast progressive alignment will be computed. If you
want to change the substitution matrix used by the methods, use the
&#8211;matrix flag.</p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>cdna</b></span><b style='mso-bidi-font-weight:normal'>_fast_pair</b></span><span
style='mso-tab-count:1'> </span>This method computes the pairwise alignment of
two <span class=SpellE>cDNA</span> sequences. It is a <span class=SpellE>fast_pair</span>
alignment that only takes into account the amino-acid similarity and uses
different penalties for amino-acid insertions and <span class=SpellE>frameshifts</span>.
This alignment is turned into a library where matched nucleotides receive a score
<span class=SpellE>equql</span> to the average level of identity at the
amino-acid level. This mode is intended to clean <span class=SpellE>cDNA</span>
obtained from ESTs, or to align pseudo-genes.</p>

<p class=Mylist>WARNING: This method is currently <span class=SpellE>unsuported</span>.</p>

<h4><span class=SpellE>PLUG-INs<span class=GramE>:List</span></span> OF
EXTERNAL METHODS</h4>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc101065342"></a><a
name="_Toc235985590"></a><a name="_Toc135066777"></a><a name="_Toc102494688"></a><a
name="_Toc102494305"><span style='mso-bookmark:_Toc102494688'><span
style='mso-bookmark:_Toc135066777'><span style='mso-bookmark:_Toc235985590'><span
style='mso-bookmark:_Toc101065342'>Plug-In: Using Methods Integrated in
T-Coffee</span></span></span></span></a><span style='mso-bookmark:_Toc101065342'>
</span></h3>

<p class=MsoNormal><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<p class=MsoNormal><span lang=AF style='mso-ansi-language:AF'><o:p>&nbsp;</o:p></span></p>

<p class=MyNormal>The following methods are external. They correspond to
packages <span class=SpellE>developped</span> by other groups that you may want
to run within T-Coffee. We are very open to extending these options and we
welcome any request to ad an extra interface. The following table lists the
methods that can be used as plug-ins:</p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6'>

<p class=examplefileCxSpFirst><span lang=EN-GB>Package<span style='mso-tab-count:
2'>              </span>Where From</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>==========================================================</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>ClustalW<span style='mso-tab-count:
2'>             </span>can interact with t_coffee</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span class=SpellE><span lang=EN-GB>Poa</span></span><span
lang=EN-GB><span style='mso-tab-count:2'>           </span><span
style='mso-tab-count:1'>       </span>http://www.bioinformatics.ucla.edu/poa/</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>Muscle<span style='mso-tab-count:
2'>        </span>http://www.bioinformatics.ucla.edu/poa/</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>ProbCons<span style='mso-tab-count:
2'>             </span>http://probcons.stanford.edu/</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>MAFFT<span style='mso-tab-count:
1'>  </span><a href="http://www.biophys.kyoto-/">http://www.biophys.kyoto-</a>u.ac.jp/~katoh/programs/align/mafft/</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span class=SpellE><span lang=EN-GB>Dialign</span></span><span
lang=EN-GB>-T<span style='mso-tab-count:2'>            </span><a
href="http://dialign-t.gobics.de/">http://dialign-t.gobics.de/</a></span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>PCMA<span style='mso-tab-count:
3'>                 </span><a href="file://localhost/pub/PCMA">ftp://iole.swmed.edu/pub/PCMA/</a></span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>----------------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span class=GramE><span lang=EN-GB>sap</span></span><span
lang=EN-GB><span style='mso-tab-count:2'>           </span><span
style='mso-tab-count:1'>       </span>structure/structure comparisons </span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>(<span class=GramE>obtain</span>
it from W. Taylor, NIMR-MRC).</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>---------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>Blast<span style='mso-tab-count:
2'>         </span><span style='mso-tab-count:1'>       </span><a
href="http://www.ncbi.nih.nlm.gov/"><span lang=EN-US style='mso-ansi-language:
EN-US'>www.ncbi.nih.nlm.gov</span></a></span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>---------------------------------------------------</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>Fugue<span style='mso-tab-count:
3'>                </span>protein to structure alignment program</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB><span style='mso-tab-count:
3'>                     </span>http://www-cryst.bioc.cam.ac.uk/fugue/download.html</span></p>

<p class=examplefileCxSpLast><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

</div>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MyNormal>Once installed, most of these methods can be used as either
pairwise or multiple alignment methods. Note that all these methods use
Blosum62 as a default.</p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>clustalw</b></span><b style='mso-bidi-font-weight:normal'>_pair</b></span><b
style='mso-bidi-font-weight:normal'><span style='mso-tab-count:1'>   </span></b>Uses
<span class=SpellE>clustalw</span> (default parameters) to align two sequences.
Each pair of residue is given a score function of the weighting mode defined by
-weight.</p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>clustalw</b></span><b style='mso-bidi-font-weight:normal'>_msa</b></span><b
style='mso-bidi-font-weight:normal'><span style='mso-tab-count:1'>   </span></b>Makes
a multiple alignment using ClustalW and adds it to the library. Each pair of
residue is given a score function of the weighting mode defined by -weight.</p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>probcons</b></span><b style='mso-bidi-font-weight:normal'>_pair</b></span><b
style='mso-bidi-font-weight:normal'><span style='mso-tab-count:1'> </span></b><span
class=SpellE>Probcons</span> package: probcons.stanford.edu/.</p>

<p class=Mylist><span class=SpellE><span class=GramE>probcons</span>_msa</span><span
style='mso-tab-count:1'> </span>idem.</p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>muscle</b></span><b style='mso-bidi-font-weight:normal'>_pair</b></span><b
style='mso-bidi-font-weight:normal'><span style='mso-tab-count:1'>     </span></b>Muscle
package www.drive5.com/muscle/ .</p>

<p class=Mylist><span class=SpellE><span class=GramE>muscle</span>_msa</span><span
style='mso-tab-count:1'>     </span>idem.</p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>mafft</b></span><b style='mso-bidi-font-weight:normal'>_pair</b></span><b
style='mso-bidi-font-weight:normal'><span style='mso-tab-count:1'>        </span></b>www.biophys.kyoto-u.ac.jp/~katoh/programs/align/mafft/
.</p>

<p class=Mylist><span class=SpellE><span class=GramE>mafft</span>_msa</span><span
style='mso-tab-count:1'>        </span>idem.</p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>pcma</b></span><b style='mso-bidi-font-weight:normal'>_msa</b></span><b
style='mso-bidi-font-weight:normal'><span style='mso-tab-count:1'>       </span></b><span
class=SpellE>pcma</span> package<b style='mso-bidi-font-weight:normal'><o:p></o:p></b></p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>pcma</b></span><b style='mso-bidi-font-weight:normal'>_pair</b></span><b
style='mso-bidi-font-weight:normal'><span style='mso-tab-count:1'>       </span></b><span
class=SpellE>pcma</span> package<b style='mso-bidi-font-weight:normal'><o:p></o:p></b></p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>poa</b></span><b style='mso-bidi-font-weight:normal'>_msa</b></span><b
style='mso-bidi-font-weight:normal'><span style='mso-tab-count:1'>          </span></b><span
class=SpellE>poa</span> package<b style='mso-bidi-font-weight:normal'><o:p></o:p></b></p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>poa</b></span><b style='mso-bidi-font-weight:normal'>_pair</b></span><b
style='mso-bidi-font-weight:normal'><span style='mso-tab-count:1'>          </span></b><span
class=SpellE>poa</span> package<b style='mso-bidi-font-weight:normal'><o:p></o:p></b></p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>dialignt</b></span><b style='mso-bidi-font-weight:normal'>_pair</b></span><b
style='mso-bidi-font-weight:normal'><span style='mso-tab-count:1'>    </span></b><span
class=SpellE>dialignt</span> <span style="mso-spacerun:yes"> </span>package<b
style='mso-bidi-font-weight:normal'><o:p></o:p></b></p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>dialignt</b></span><b style='mso-bidi-font-weight:normal'>_msa</b></span><span
style='mso-tab-count:1'>    </span><span class=SpellE>pcma</span> package</p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>sap</b></span><b style='mso-bidi-font-weight:normal'>_pair</b></span><b
style='mso-bidi-font-weight:normal'><span style='mso-tab-count:1'>          </span></b>Uses
sap to align two structures. Each pair of residue is given a score function
defined by sap. You must have sap installed on your system to use this method.</p>

<p class=Mylist><span class=SpellE><span class=GramE><b style='mso-bidi-font-weight:
normal'>fugue</b></span><b style='mso-bidi-font-weight:normal'>_pair</b></span><b
style='mso-bidi-font-weight:normal'><span style='mso-tab-count:1'>       </span></b>Uses
a standard fugue installation to make a sequence /structure alignment. Fugue
installation must be standard. It does not have to include all the fugue
packages but only:</p>

<p class=Mylist>1- joy, melody, <span class=SpellE>fugueali</span>, <span
class=SpellE>sstruc</span>, <span class=SpellE>hbond</span></p>

<p class=Mylist>2-copy fugue/classdef.dat<span style='mso-tab-count:1'>            </span>/data/fugue/SUBST/classdef.dat</p>

<p class=Mylist>OR</p>

<p class=Mylist><span class=SpellE>Setenv</span>
MELODY_CLASSDEF=&lt;location&gt;</p>

<p class=Mylist><span class=SpellE>Setenv</span> MELODY_SUBST=fugue/allmat.dat</p>

<p class=Mylist><span style="mso-spacerun:yes"> </span>All the configuration
files must be in the right location. </p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MyNormal>To request a method, see the -in or the -method flag. For
instance, if you wish to request the use of <span class=SpellE>fast_pair</span>
and <span class=SpellE>lalign_id_pair</span> (the current default):</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -seq sample_seq1.fasta -method fast_pair,lalign_id_pair </span></p>

</div>

<h2><a name="_Toc101065343"></a><a name="_Toc102494689"></a><a
name="_Toc102494306"></a><a name="_Toc235985591"><span style='mso-bookmark:
_Toc102494306'><span style='mso-bookmark:_Toc102494689'><span style='mso-bookmark:
_Toc101065343'><span lang=EN-GB>Modifying the parameters of Internal and
External Methods</span></span></span></span></a><span style='mso-bookmark:_Toc102494306'><span
style='mso-bookmark:_Toc102494689'><span style='mso-bookmark:_Toc101065343'></span></span></span></h2>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><span style='mso-bookmark:
_Toc102494306'><span style='mso-bookmark:_Toc102494689'><span style='mso-bookmark:
_Toc101065343'><a name="_Toc235985592">Internal Methods</a></span></span></span></h3>

<p class=MyNormal><span style='mso-bookmark:_Toc102494306'><span
style='mso-bookmark:_Toc102494689'><span style='mso-bookmark:_Toc101065343'>It
is possible to modify on the fly the parameters of hard coded methods:</span></span></span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc102494306'><span style='mso-bookmark:_Toc102494689'><span
style='mso-bookmark:_Toc101065343'><span lang=ZU>PROMPT: t_coffee
sample_seq1.fasta &#8211;method slow_pair@EP@MATRIX@pam250mt@GOP@-10@GEP@-1</span></span></span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc102494306'><span
style='mso-bookmark:_Toc102494689'><span style='mso-bookmark:_Toc101065343'>EP
stands for Extra parameters. These parameters will <span class=SpellE>superseed</span>
any other parameters.</span></span></span></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><span style='mso-bookmark:
_Toc102494306'><span style='mso-bookmark:_Toc102494689'><span style='mso-bookmark:
_Toc101065343'><a name="_Toc235985593">External Methods</a></span></span></span></h3>

<p class=MyNormal><span style='mso-bookmark:_Toc102494306'><span
style='mso-bookmark:_Toc102494689'><span style='mso-bookmark:_Toc101065343'>External
methods receive a command line built with the information provided via the
parameter file (see next heading). It is possible to produce such a parameter
file and to modify it in order to modify the commands passed to the methods. </span></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc102494306'><span
style='mso-bookmark:_Toc102494689'><span style='mso-bookmark:_Toc101065343'>The
passed command is built as follows:</span></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc102494306'><span
style='mso-bookmark:_Toc102494689'><span style='mso-bookmark:_Toc101065343'>&lt;EXECUTABLE&gt;&lt;PARAM1&gt;&lt;IN_FLAG&gt;&lt;seq_file&gt;&lt;PARAM2&gt;&lt;OUT_FLAG&gt;&lt;outname&gt;&lt;PARAM&gt;</span></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc102494306'><span
style='mso-bookmark:_Toc102494689'><span style='mso-bookmark:_Toc101065343'><o:p>&nbsp;</o:p></span></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc102494306'><span
style='mso-bookmark:_Toc102494689'><span style='mso-bookmark:_Toc101065343'>You
should know what is the best place for <span class=SpellE>squizing</span> your
extra parameters. It will depend on the application, although PARAM2 is usually
a good guess. Now if you want, for instance to modify the gap penalty of <span
class=SpellE>clustalw</span>, you can try the following:</span></span></span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc102494306'><span style='mso-bookmark:_Toc102494689'><span
style='mso-bookmark:_Toc101065343'><span lang=ZU><span
style="mso-spacerun:yes"> </span>PROMPT: t_coffee sample_seq1.fasta -method
clustalw_msa@EP@PARAM2@-GAPOPEN%e100%s-GAPEXT%e10</span></span></span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc102494306'><span
style='mso-bookmark:_Toc102494689'><span style='mso-bookmark:_Toc101065343'>@EP
is here to indicate that you will pass an extra parameter</span></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc102494306'><span
style='mso-bookmark:_Toc102494689'><span style='mso-bookmark:_Toc101065343'>@PARAM1
is the name of this parameter</span></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc102494306'><span
style='mso-bookmark:_Toc102494689'><span style='mso-bookmark:_Toc101065343'><span
class=GramE>the</span> next filed is the parameter itself, where:</span></span></span></p>

<p class=MyNormal style='text-indent:28.25pt'><span style='mso-bookmark:_Toc102494306'><span
style='mso-bookmark:_Toc102494689'><span style='mso-bookmark:_Toc101065343'>%<span
class=GramE>s</span> replace spaces </span></span></span></p>

<p class=MyNormal style='text-indent:28.25pt'><span style='mso-bookmark:_Toc102494306'><span
style='mso-bookmark:_Toc102494689'><span style='mso-bookmark:_Toc101065343'>%<span
class=GramE>e</span> replaces the equal sign</span></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc102494306'><span
style='mso-bookmark:_Toc102494689'><span style='mso-bookmark:_Toc101065343'>Of
course, you must know the command line of the program you are trying to modify
(<span class=SpellE>clustalw</span> in this case).</span></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc102494306'><span
style='mso-bookmark:_Toc102494689'><span style='mso-bookmark:_Toc101065343'><o:p>&nbsp;</o:p></span></span></span></p>

<h2><span style='mso-bookmark:_Toc102494306'><span style='mso-bookmark:_Toc102494689'><span
style='mso-bookmark:_Toc101065343'><a name="_Toc235985594"></a><a
name="_Toc135066778"><span style='mso-bookmark:_Toc235985594'><span lang=EN-GB>Integrating
External Methods</span></span></a></span></span></span><span style='mso-bookmark:
_Toc101065343'></span></h2>

<p class=MsoBodyText><span style='mso-bookmark:_Toc101065343'><span lang=EN-GB>If
the method you need is not already included in T-Coffee, you will need to
integrate it yourself. We give you here some guidelines on how to do so.</span></span></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><span style='mso-bookmark:
_Toc101065343'><a name="_Toc235985595">Direct access to external methods</a></span></h3>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'>A special method
exists in T-Coffee that can be used to invoke any existing program:</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc101065343'><span lang=ZU>PROMPT: t_coffee
sample_seq1.fasta &#8211;method=em@clustalw@pairwise </span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'>In this context,
Clustalw is a method that can be ran with the following command line:</span></p>

<p class=example2><span style='mso-bookmark:_Toc101065343'><span class=GramE>method</span>
&#8211;<span class=SpellE>infile</span>=&lt;<span class=SpellE>infile</span>&gt;
-<span class=SpellE>outfile</span>=&lt;<span class=SpellE>outfile</span>&gt;</span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'>Clustalw can be
replaced with any method using a similar syntax. If the program you want to use
cannot be run this way, you can either write a <span class=SpellE>perl</span>
wrapper that fits the bill or write a <span class=SpellE>tc_method</span> file
adapted to your program (<span class=SpellE>cf</span> next section).</span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'>This special method
(<span class=SpellE>em</span>, external method) uses the following syntax:</span></p>

<p class=example2><span style='mso-bookmark:_Toc101065343'><span lang=AF
style='mso-ansi-language:AF'>em@&lt;method&gt;@&lt;aln_mode:pairwise¦s_pairwise|multiple&gt;<o:p></o:p></span></span></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><span style='mso-bookmark:
_Toc101065343'><a name="_Toc235985596">Customizing an external method (with
parameters) for T-Coffee</a></span></h3>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'>T-Coffee can run
external methods, using a <span class=SpellE><i style='mso-bidi-font-style:
normal'>tc_method</i></span> file that can be used in place of an established
method. Two such files are incorporated in T-Coffee. You can dump them and
customize them according to your needs:</span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'>For instance if you
have ClustalW installed, you can use the following file to run the </span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc101065343'><span lang=ZU>PROMPT: t_coffee
&#8211;other_pg unpack_clustalw_method.tc_method </span></span></p>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc101065343'><span lang=ZU>PROMPT: t_coffee
&#8211;other_pg unpack_generic_method.tc_method </span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'>The second file (<span
class=SpellE>generic_method.tc_method</span>) contains many hints on how to
customize your new method. The first file is a very straightforward example on
how to have t_coffee to run Clustalw with a set of parameters you may be
interested in:</span></p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6'>

<p class=examplefileCxSpFirst><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB><o:p>&nbsp;</o:p></span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*TC_METHOD_FORMAT_01</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>***************<span class=SpellE><span class=GramE>clustalw</span>_method.tc_method</span>*********</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>EXECUTABLE<span style='mso-tab-count:1'>    </span><span
class=SpellE>clustalw</span></span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>ALN_MODE<span style='mso-tab-count:2'>             </span>pairwise</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>IN_FLAG<span style='mso-tab-count:2'>              </span>-INFILE=</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>OUT_FLAG<span style='mso-tab-count:2'>             </span>-OUTFILE=</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>OUT_MODE<span style='mso-tab-count:2'>             </span><span
class=SpellE>aln</span></span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>PARAM<span style='mso-tab-count:2'>         </span>-<span
class=SpellE>gapopen</span>=-10</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>SEQ_TYPE<span style='mso-tab-count:2'>             </span>S</span></span></p>

<p class=examplefileCxSpLast><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*************************************************</span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'>This configuration
file will cause T-Coffee to emit the following system call:</span></p>

<p class=example2><span style='mso-bookmark:_Toc101065343'><span class=SpellE><span
class=GramE>clustalw</span></span> &#8211;INFILE=tmpfile1
&#8211;OUTFILE=tmpfile2 &#8211;<span class=SpellE>gapopen</span>=-10</span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'>Note that ALN_MODE
instructs t_coffee to run <span class=SpellE>clustalw</span> on every pair of
sequences (<span class=SpellE>cf</span> <span class=SpellE>generic_method.tc_method</span>
for more details). </span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'>The <span
class=SpellE>tc_method</span> files are treated like any standard established
method in T-Coffee. For instance, if the file <span class=SpellE><i
style='mso-bidi-font-style:normal'>clustalw_method.tc_method</i></span> is in
your current directory, run:</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc101065343'><span lang=ZU>PROMPT: t_coffee
sample_seq1.fasta &#8211;method clustalw_method.tc_method </span></span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><span style='mso-bookmark:
_Toc101065343'><a name="_Toc235985597"></a><a name="_Toc135066779"></a><a
name="_Toc102494690"></a><a name="_Toc102494307"><span style='mso-bookmark:
_Toc102494690'><span style='mso-bookmark:_Toc135066779'><span style='mso-bookmark:
_Toc235985597'>Managing a collection of method files</span></span></span></a></span></h3>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'>It may be convenient
to store all the method files in a single location on your system. By default,
t_coffee will go looking into the directory ~/<span class=GramE>.t</span>_coffee/methods/.
You can change this by either modifying the METHODS_4_TCOFFEE in <span
class=SpellE>define_headers.h</span> (and recompile) or by modifying the <span
class=SpellE>envoronement</span> variable METHODS_4_TCOFFEE.</span></p>

<h2><span style='mso-bookmark:_Toc101065343'><a name="_Toc235985598"></a><a
name="_Toc135066780"></a><a name="_Toc102494691"></a><a name="_Toc102494308"><span
style='mso-bookmark:_Toc102494691'><span style='mso-bookmark:_Toc135066780'><span
style='mso-bookmark:_Toc235985598'><span lang=EN-GB>Advanced Method Integration</span></span></span></span></a></span></h2>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'>It may sometimes be
difficult to customize the program you want to use through a <span
class=SpellE>tc_method</span> file. In that case, you may rather use an
external <span class=SpellE>perl_script</span> to run your external
application. This can easily be achieved using the <span class=SpellE>generic_method.tc_method</span>
file.</span></p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6'>

<p class=examplefileCxSpFirst><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*TC_METHOD_FORMAT_01</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>***************<span class=SpellE><span class=GramE>generic</span>_method.tc_method</span>*********</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>EXECUTABLE<span style='mso-tab-count:1'>    </span>tc_generic_method.pl</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>ALN_MODE<span style='mso-tab-count:2'>             </span>pairwise</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>IN_FLAG<span style='mso-tab-count:2'>              </span>-<span
class=SpellE>infile</span>=</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>OUT_FLAG<span style='mso-tab-count:2'>             </span>-<span
class=SpellE>outfile</span>=</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>OUT_MODE<span style='mso-tab-count:2'>             </span><span
class=SpellE>aln</span></span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>PARAM<span style='mso-tab-count:2'>         </span>-method <span
class=SpellE>clustalw</span></span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>PARAM<span style='mso-tab-count:2'>         </span>-<span
class=SpellE>gapopen</span>=-10</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>SEQ_TYPE<span style='mso-tab-count:2'>             </span>S</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*************************************************</span></span></p>

<p class=examplefileCxSpLast><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>* Note: &amp;<span class=SpellE>bsnp</span> can be used to <span
class=GramE>for<span style="mso-spacerun:yes">  </span>white</span> spaces</span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'>When you run this
method:</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc101065343'><span lang=ZU>PROMPT: t_coffee
&#8211;other_pg unpack_generic_method.tc_method</span></span></p>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc101065343'><span lang=ZU>PROMPT: t_coffee
sample_seq1.fasta &#8211;method generic_method.tc_method </span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'>T-Coffee runs the
script tc_generic_method.pl on your data. It also provides the script with
parameters. In this case &#8211;method <span class=SpellE>clustalw</span>
indicates that the script should run <span class=SpellE>clustalw</span> on your
data. The script tc_generic_method.pl is incorporated in t_coffee. Over the
time, this script will be the place where novel methods will be integrated </span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'><span
style="mso-spacerun:yes"> </span><span class=GramE>will</span> be used to run
the script <i style='mso-bidi-font-style:normal'>tc_generic_method.pl</i>. The
file tc_generic_method.pl is a <span class=SpellE>perl</span> file,
automatically generated by t_coffee. Over the time this file will make it
possible to run all available methods. You can dump the script using the
following command:</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc101065343'><span lang=ZU>PROMPT: t_coffee
&#8211;other_pg=unpack_tc_generic_method.pl</span></span></p>

</div>

<p class=Note><span style='mso-bookmark:_Toc101065343'><span style='mso-ansi-language:
EN-US'>Note: If there is a copy of that script in your local directory, that
copy will be used in place of the internal copy of T-Coffee.<o:p></o:p></span></span></p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><span style='mso-bookmark:
_Toc101065343'><a name="_Toc235985599"></a><a name="_Toc135066781"></a><a
name="_Toc102494692"></a><a name="_Toc102494309"><span style='mso-bookmark:
_Toc102494692'><span style='mso-bookmark:_Toc135066781'><span style='mso-bookmark:
_Toc235985599'>The Mother of All method file</span></span></span></a><span
style='mso-bookmark:_Toc135066781'><span style='mso-bookmark:_Toc235985599'>s</span></span></span></h3>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6'>

<p class=examplefileCxSpFirst><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*TC_METHOD_FORMAT_01</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>******************<span class=SpellE><span class=GramE>generic</span>_method.tc_method</span>*************</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span style="mso-spacerun:yes">       </span>Incorporating new
methods in T-Coffee</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span style="mso-spacerun:yes">       </span>Cedric Notredame
17/04/05</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*******************************************************</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*This file is a method file</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*Copy it and adapt it to your need so that the method </span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span class=GramE>you</span> want to use can be incorporated within
T-Coffee</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*******************************************************</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span style="mso-spacerun:yes">                  </span>USAGE<span
style="mso-spacerun:yes">                              </span>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*******************************************************</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*This file is passed to t_coffee via &#8211;in:</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span style='mso-tab-count:1'>      </span><span class=GramE>t</span>_coffee
&#8211;in <span class=SpellE>Mgeneric_method.method</span></span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span style='mso-tab-count:1'>      </span>The method is passed to
the shell using the following</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span class=GramE>call</span>:</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*&lt;EXECUTABLE&gt;&lt;IN_FLAG&gt;&lt;<span class=SpellE>seq_file</span>&gt;&lt;OUT_FLAG&gt;&lt;<span
class=SpellE>outname</span>&gt;&lt;PARAM&gt;</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*Conventions:</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*&lt;FLAG_NAME&gt; <span style='mso-tab-count:1'> </span>&lt;TYPE&gt;<span
style='mso-tab-count:2'>        </span>&lt;VALUE&gt;</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*&lt;VALUE&gt;:<span style='mso-tab-count:1'>     </span><span
class=SpellE>no_name</span> <span style='mso-tab-count:1'>      </span>&lt;=&gt;
Replaced with a space</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*&lt;VALUE&gt;:<span style='mso-tab-count:1'>     </span>&amp;<span
class=SpellE>nbsp</span><span style='mso-tab-count:1'>  </span>&lt;=&gt;
Replaced with a space</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*******************************************************</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span style="mso-spacerun:yes">                 
</span>EXECUTABLE<span style="mso-spacerun:yes">                        
</span>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*******************************************************</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span class=GramE>name</span> of the executable</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span class=GramE>passed</span> to the shell: executable</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span style='mso-tab-count:1'>      </span></span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>EXECUTABLE<span style='mso-tab-count:1'>    </span>tc_generic_method.pl</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*******************************************************</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span style="mso-spacerun:yes">                 
</span>ALN_MODE<span style="mso-spacerun:yes">                          
</span>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*******************************************************</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span class=GramE>pairwise</span><span style="mso-spacerun:yes">  
</span>-&gt;all Vs all (no self )[(n2-n)/2aln]</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span class=SpellE><span class=GramE>m</span>_pairwise</span>
-&gt;all Vs all (no self)[n^2-n]^2</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span class=SpellE><span class=GramE>s</span>_pairwise</span>
-&gt;all Vs all (self): [n^2-n]/2 + n</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span class=GramE>multiple</span><span style="mso-spacerun:yes">  
</span>-&gt;All the sequences in one go</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=FR style='mso-ansi-language:FR'>*<o:p></o:p></span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=FR style='mso-ansi-language:FR'>ALN_MODE<span style='mso-tab-count:2'>             </span>pairwise<o:p></o:p></span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=FR style='mso-ansi-language:FR'>*<o:p></o:p></span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=FR style='mso-ansi-language:FR'>*******************************************************<o:p></o:p></span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=FR style='mso-ansi-language:FR'>*<span
style="mso-spacerun:yes">               </span><span
style="mso-spacerun:yes">   </span>OUT_MODE<span
style="mso-spacerun:yes">                           </span>*<o:p></o:p></span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*******************************************************</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>* <span class=GramE>mode</span> for the output:</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*External methods: </span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>* <span class=SpellE><span class=GramE>aln</span></span> -&gt; <span
class=SpellE>alignmnent</span> File (<span class=SpellE>Fasta</span> or
ClustalW Format)</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>* <span class=GramE>lib</span>-&gt; Library file (TC_LIB_FORMAT_01)</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*Internal Methods:</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>* <span class=SpellE><span class=GramE>fL</span></span> -&gt; Internal
Function returning a Lib (<span class=SpellE>Librairie</span>)</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>* <span class=SpellE><span class=GramE>fA</span></span> -&gt;
Internal Function returning an <span class=SpellE>Alignmnent</span></span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>OUT_MODE<span style='mso-tab-count:2'>             </span><span
class=SpellE>aln</span></span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*******************************************************</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span style="mso-spacerun:yes">                 
</span>IN_FLAG<span style="mso-spacerun:yes">                            
</span>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*******************************************************</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*IN_FLAG</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span class=GramE>flag</span> indicating the name of the in coming
sequences</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*IN_FLAG S <span class=SpellE>no_name</span> -&gt;no flag</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*IN_FLAG S &amp;<span class=SpellE>nbsp</span>&#8211;<span
class=SpellE>in&amp;nbsp</span> -&gt;  &#8211;in </span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>IN_FLAG<span style='mso-tab-count:2'>              </span>-<span
class=SpellE>infile</span>=</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*******************************************************</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span style="mso-spacerun:yes">                  </span>OUT_FLAG<span
style="mso-spacerun:yes">                           </span>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*******************************************************</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*OUT_FLAG</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span class=GramE>flag</span> indicating the name of the out-coming
data</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span class=GramE>same</span> conventions as IN_FLAG</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*OUT_FLAG<span style='mso-tab-count:1'>     </span>S <span
class=SpellE>no_name</span> -&gt;no flag</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>OUT_FLAG<span style='mso-tab-count:2'>             </span>-<span
class=SpellE>outfile</span>=</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*******************************************************</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span style="mso-spacerun:yes">                 
</span>SEQ_TYPE<span style="mso-spacerun:yes">                          
</span>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*******************************************************</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*G: Genomic, S: Sequence, P: PDB, R: Profile</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*Examples:</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*SEQTYPE<span style='mso-tab-count:1'>      </span>S<span
style='mso-tab-count:1'>      </span>sequences against sequences (default)</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*SEQTYPE<span style='mso-tab-count:1'>      </span>S_P<span
style='mso-tab-count:1'>    </span>sequence against structure</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*SEQTYPE<span style='mso-tab-count:1'>      </span>P_P<span
style='mso-tab-count:1'>    </span>structure against structure</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*SEQTYPE<span style='mso-tab-count:1'>      </span>PS<span
style='mso-tab-count:1'>     </span><span class=GramE>mix</span> of sequences
and structure<span style='mso-tab-count:1'>     </span></span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>SEQ_TYPE<span style='mso-tab-count:1'>      </span>S</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*******************************************************</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span style="mso-spacerun:yes">                  </span>PARAM<span
style="mso-spacerun:yes">                              </span>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*******************************************************</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*Parameters sent to the EXECUTABLE </span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*If there is more than 1 PARAM line, the lines are</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span class=GramE>concatenated</span></span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>PARAM<span style='mso-tab-count:1'>  </span>-method <span
class=SpellE>clustalw</span></span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>PARAM<span style="mso-spacerun:yes">   </span>-OUTORDER=INPUT
-NEWTREE=core -align -<span class=SpellE>gapopen</span>=-15</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*******************************************************</span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*<span style="mso-spacerun:yes">                  </span>END<span
style="mso-spacerun:yes">                                </span>*</span></span></p>

<p class=examplefileCxSpLast><span style='mso-bookmark:_Toc101065343'><span
lang=EN-GB>*******************************************************<a
name="_Toc102494693"></a><a name="_Toc102494310"><span style='mso-bookmark:
_Toc102494693'></span></a></span></span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><span style='mso-bookmark:
_Toc101065343'><span style='mso-bookmark:_Toc102494310'><span style='mso-bookmark:
_Toc102494693'><a name="_Toc135066782"></a><a name="_Toc235985600"><span
style='mso-bookmark:_Toc135066782'>Weighting your Method</span></a><span
style='mso-bookmark:_Toc135066782'></span></span></span></span></h3>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'><span
style='mso-bookmark:_Toc102494310'><span style='mso-bookmark:_Toc102494693'><span
style='mso-bookmark:_Toc135066782'>By default, the alignment produced by your
method will be weighted according to its percent identity. However, this can be
customized via the WEIGHT parameter.</span></span></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'><span
style='mso-bookmark:_Toc102494310'><span style='mso-bookmark:_Toc102494693'><span
style='mso-bookmark:_Toc135066782'>The WEIGHT parameter supports all the values
of the &#8211;weight flag. The only difference is that the &#8211;weight value
thus declared will only be applied onto your method.</span></span></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'><span
style='mso-bookmark:_Toc102494310'><span style='mso-bookmark:_Toc102494693'><span
style='mso-bookmark:_Toc135066782'>If needed you can also modify on the fly the
WEIGHT value of your method:</span></span></span></span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc101065343'><span style='mso-bookmark:_Toc102494310'><span
style='mso-bookmark:_Toc102494693'><span style='mso-bookmark:_Toc135066782'><span
lang=ZU>PROMPT: t_coffee sample_seq1.fasta &#8211;method slow_pair@WEIGHT@OW2</span></span></span></span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'><span
style='mso-bookmark:_Toc102494310'><span style='mso-bookmark:_Toc102494693'><span
style='mso-bookmark:_Toc135066782'>Will overweight by a factor 2 the weight of <span
class=SpellE>slow_pair</span>.</span></span></span></span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc101065343'><span style='mso-bookmark:_Toc102494310'><span
style='mso-bookmark:_Toc102494693'><span style='mso-bookmark:_Toc135066782'><span
lang=ZU>PROMPT: t_coffee sample_seq1.fasta &#8211;method slow_pair@WEIGHT@250</span></span></span></span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'><span
style='mso-bookmark:_Toc102494310'><span style='mso-bookmark:_Toc102494693'><span
style='mso-bookmark:_Toc135066782'>Will cause every pair of <span class=SpellE>slow_pair</span>
to have a weight equal to 250</span></span></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'><span
style='mso-bookmark:_Toc102494310'><span style='mso-bookmark:_Toc102494693'><span
style='mso-bookmark:_Toc135066782'><o:p>&nbsp;</o:p></span></span></span></span></p>

<h2><span style='mso-bookmark:_Toc101065343'><span style='mso-bookmark:_Toc102494310'><span
style='mso-bookmark:_Toc102494693'><span style='mso-bookmark:_Toc135066782'><a
name="_Toc235985601"><span lang=EN-GB>Plug-Out: Using T-Coffee as a Plug-In</span></a></span></span></span></span></h2>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'><span
style='mso-bookmark:_Toc102494310'><span style='mso-bookmark:_Toc102494693'>Just
because it enjoys enslaving other methods as plug-ins, does not mean that
T-Coffee does not enjoy being incorporated within other packages. We try to
give as much support as possible to anyone who wishes to <span class=SpellE>incorportae</span>
T-Coffee in an alignment pipeline.</span></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'><span
style='mso-bookmark:_Toc102494310'><span style='mso-bookmark:_Toc102494693'>If
you want to do so, please work out some way to incorporate T-Coffee in your <span
class=GramE>script .</span> If you need some help along the ways, do not
hesitate to ask, as we will always be happy to either give assistance, or even
modify the package so that it <span class=SpellE>accomodates</span> as many
needs as possible.</span></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'><span
style='mso-bookmark:_Toc102494310'><span style='mso-bookmark:_Toc102494693'>Once
that procedure is over, set aside a couple of input files with the correct <span
class=SpellE>parameterisation</span> and send them to us. These will be
included as a distribution test, to insure that any further distribution
remains compliant with your application.</span></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065343'><span
style='mso-bookmark:_Toc102494310'><span style='mso-bookmark:_Toc102494693'>We
currently support:</span></span></span></p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6'>

<p class=examplefileCxSpFirst><span style='mso-bookmark:_Toc101065343'><span
style='mso-bookmark:_Toc102494310'><span style='mso-bookmark:_Toc102494693'><span
lang=EN-GB>Package<span style='mso-tab-count:2'>              </span>Where From</span></span></span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
style='mso-bookmark:_Toc102494310'><span style='mso-bookmark:_Toc102494693'><span
lang=EN-GB>==========================================================</span></span></span></span></p>

<p class=examplefileCxSpMiddle><span style='mso-bookmark:_Toc101065343'><span
style='mso-bookmark:_Toc102494310'><span style='mso-bookmark:_Toc102494693'><span
class=SpellE><span lang=EN-GB>Marna</span></span><span lang=EN-GB><span
style='mso-tab-count:1'>  </span><span style='mso-tab-count:2'>              </span>www.bio.inf.unijena.de/Software/MARNA/download</span></span></span></span></p>

<p class=examplefileCxSpLast><span style='mso-bookmark:_Toc101065343'><span
style='mso-bookmark:_Toc102494310'><span style='mso-bookmark:_Toc102494693'><span
lang=EN-GB>----------------------------------------------------------</span></span></span></span></p>

</div>

<h2><span style='mso-bookmark:_Toc101065343'><span style='mso-bookmark:_Toc102494310'><span
style='mso-bookmark:_Toc102494693'><a name="_Toc235985602"></a><a
name="_Toc135066783"><span style='mso-bookmark:_Toc235985602'><span lang=EN-GB>Creating
Your Own T-Coffee Libraries</span></span></a></span></span></span></h2>

<span style='mso-bookmark:_Toc101065343'></span>

<p class=MyNormal>If the method you want to use is not integrated, or
impossible to integrate, you can generate your own libraries, either directly
or by turning existing alignments into libraries. You may also want to <span
class=SpellE>precompute</span> your libraries, in order to combine them at your
convenience.</p>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985603"></a><a
name="_Toc135066784"></a><a name="_Toc102494694"></a><a name="_Toc102494311"><span
style='mso-bookmark:_Toc102494694'><span style='mso-bookmark:_Toc135066784'><span
style='mso-bookmark:_Toc235985603'>Using Pre-Computed Alignments</span></span></span></a></h3>

<p class=MyNormal>If the method you wish to use is not supported, or if you
simply have the alignments, the simplest thing to do is to generate yourself
the pairwise/multiple alignments, in FASTA, ClustalW, <span class=SpellE>msf</span>
or <span class=SpellE>Pir</span> format and feed them into t_coffee using the <i
style='mso-bidi-font-style:normal'>-in</i> flag:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee &#8211;aln=sample_aln1_1.aln,sample_aln1_2.aln
&#8211;outfile=combined_aln.aln </span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985604"></a><a
name="_Toc135066785"></a><a name="_Toc102494695"></a><a name="_Toc102494312"><span
style='mso-bookmark:_Toc102494695'><span style='mso-bookmark:_Toc135066785'><span
style='mso-bookmark:_Toc235985604'>Customizing the Weighting Scheme</span></span></span></a></h3>

<p class=MyNormal>The previous integration method forces you to use the same
weighting scheme for each alignment and the rest of the libraries generated on
the fly. This weighting scheme is based on global pairwise sequence identity.
If you want to use a more specific weighting scheme with a given method, you
should either:</p>

<p class=MyNormal><span class=GramE>generate</span> your own library (<span
class=SpellE>cf</span> next section)</p>

<p class=MyNormal><span class=GramE>convert</span> your <span class=SpellE>aln</span>
into a lib, using the &#8211;weight flag:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee &#8211;aln sample_aln1.aln &#8211;out_lib=test_lib.tc_lib
&#8211;lib_only &#8211;weight=sim_pam250mt </span></p>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee &#8211;aln sample_aln1.aln -lib test_lib.tc_lib &#8211;outfile=outaln </span></p>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee &#8211;aln=sample_aln1_1.aln,sample_aln1_2.aln -method= fast_pair,lalign_id_pair
&#8211;outfile=out_aln </span></p>

</div>

<h3 style='text-indent:0cm;mso-list:l10 level3 lfo3'><a name="_Toc235985605"></a><a
name="_Toc135066786"></a><a name="_Toc102494696"></a><a name="_Toc102494313"></a><a
name="_Toc101065344"><span style='mso-bookmark:_Toc102494313'><span
style='mso-bookmark:_Toc102494696'><span style='mso-bookmark:_Toc135066786'><span
style='mso-bookmark:_Toc235985605'>Generating Your Own Libraries</span></span></span></span></a></h3>

<p class=MyNormal>This is suitable if you have local alignments, or very
detailed information about your potential residue pairs, or if you want to use
a very specific weighting scheme. You will need to generate your own libraries,
using the format described in the last section.</p>

<p class=MyNormal>You may also want to pre-compute your libraries in order to
save them for further use. For instance, in the following example, we generate
the local and the global libraries and later re-use them for combination into a
multiple alignment.</p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_seq1.fasta &#8211;method slow_pair &#8211;out_lib
slow_pair_seq1.tc_lib &#8211;lib_only</span></p>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_seq1.fasta &#8211;method lalign_id_pair &#8211;out_lib
lalign_id_pair_seq1.tc_lib &#8211;lib_only</span></p>

</div>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MyNormal>Once these libraries have been computed, you can then combine
tem at your convenience in a single MSA. Of course you can decide to only use
the local or the global library</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_seq1.fasta &#8211;lib lalign_id_pair_seq1.tc_lib, slow_pair_seq1.tc_lib</span></p>

</div>

<p class=MsoBodyText><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<h1><a name="_Toc235985606"></a><a name="_Toc135066788"></a><a
name="_Toc102494697"><span style='mso-bookmark:_Toc135066788'><span
style='mso-bookmark:_Toc235985606'><span class=ChapterTitleChar><span lang=AF
style='font-size:36.0pt;mso-bidi-font-size:48.0pt;mso-ansi-language:AF'>Frequently
Asked Questions</span></span></span></span></a><span class=ChapterTitleChar><span
lang=AF style='font-size:36.0pt;mso-bidi-font-size:48.0pt;mso-ansi-language:
AF'><o:p></o:p></span></span></h1>

<div style='mso-element:para-border-div;border:solid windowtext 1.5pt;
padding:0cm 0cm 0cm 0cm;background:#F7CBB7'>

<p class=Warning><a name="_Toc102494698"></a><a name="_Toc102494314"></a><a
name="_Toc101065347"><span style='mso-bookmark:_Toc102494314'><span
style='mso-bookmark:_Toc102494698'><span lang=EN-GB>IMPORTANT: All the files </span></span></span></a><span
class=SpellE><span style='mso-bookmark:_Toc101065347'><span style='mso-bookmark:
_Toc102494314'><span style='mso-bookmark:_Toc102494698'><span lang=EN-GB>mentionned</span></span></span></span></span><span
style='mso-bookmark:_Toc101065347'><span style='mso-bookmark:_Toc102494314'><span
style='mso-bookmark:_Toc102494698'><span lang=EN-GB> here (<span class=SpellE>sample_seq</span><span
class=GramE>..</span>.) can be found in the example directory of the
distribution.</span></span></span></span></p>

</div>

<h2><span style='mso-bookmark:_Toc101065347'><span style='mso-bookmark:_Toc102494314'><span
style='mso-bookmark:_Toc102494698'><a name="_Toc235985607"></a><a
name="_Toc135066789"><span style='mso-bookmark:_Toc235985607'><span lang=EN-GB>Abnormal
Terminations and Wrong Results</span></span></a></span></span></span></h2>

<p class=FAQ><a name="_Toc235985608"></a><a name="_Toc135066790"></a><a
name="_Toc102494699"></a><a name="_Toc102494315"><span style='mso-bookmark:
_Toc102494699'><span style='mso-bookmark:_Toc135066790'><span style='mso-bookmark:
_Toc235985608'><span style='mso-ansi-language:EN-US'>Q: The program keeps
crashing when I give my sequences</span></span></span></span></a><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: This may be a format problem. Try to reformat your
sequences using any utility (<span class=SpellE>readseq</span>...). We
recommend the <span class=SpellE>Fasta</span> format. If the problem persists,
contact us.</p>

<p class=MyNormal>A: Your sequences may not be recognized for what they really
are. Normally T-Coffee recognizes the type of your sequences automatically, but
if it fails, use:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_seq1.fasta -type=PROTEIN </span></p>

</div>

<p class=MyNormal>A: Costly computation or data gathered over the net is stored
by T-Coffee in a cache directory. Sometimes, some of these files can be
corrupted and cause an abnormal termination. You can either empty the cache <span
class=GramE>( ~</span>/.t_coffee/cache/) or request T-Coffee to run without
using it:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee &#8211;pdb=struc1.pdb,struc2.pdb,struc3.pdb -method sap_pair
&#8211;cache=no</span></p>

</div>

<p class=MyNormal>If you do not want to empty your cache, you may also use
&#8211;cache=update that will only update the files corresponding to your data</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee &#8211;pdb=struc1.pdb,struc2.pdb,struc3.pdb -method sap_pair
&#8211;cache=update</span></p>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU><o:p>&nbsp;</o:p></span></p>

</div>

<p class=FAQ><a name="_Toc235985609"></a><a name="_Toc135066791"></a><a
name="_Toc102494700"></a><a name="_Toc102494316"><span style='mso-bookmark:
_Toc102494700'><span style='mso-bookmark:_Toc135066791'><span style='mso-bookmark:
_Toc235985609'><span style='mso-ansi-language:EN-US'>Q: The default alignment
is not good enough</span></span></span></span></a><span style='mso-ansi-language:
EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: see next question</p>

<p class=FAQ><a name="_Toc235985610"></a><a name="_Toc135066792"></a><a
name="_Toc102494701"></a><a name="_Toc102494317"><span style='mso-bookmark:
_Toc102494701'><span style='mso-bookmark:_Toc135066792'><span style='mso-bookmark:
_Toc235985610'><span style='mso-ansi-language:EN-US'>Q: The alignment contains
obvious mistakes</span></span></span></span></a><span style='mso-ansi-language:
EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: This happens with most multiple alignment procedures.
However, <span class=GramE>wrong alignments are sometimes caused by bugs or an
implementation mistake</span>. Please report the most unexpected results to the
authors.</p>

<p class=FAQ><a name="_Toc235985611"></a><a name="_Toc135066793"></a><a
name="_Toc102494702"></a><a name="_Toc102494318"><span style='mso-bookmark:
_Toc102494702'><span style='mso-bookmark:_Toc135066793'><span style='mso-bookmark:
_Toc235985611'><span style='mso-ansi-language:EN-US'>Q: The program is crashing</span></span></span></span></a><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: If you get the message:</p>

<p class=output><span lang=EN-GB>FAILED TO ALLOCATE REQUIRED MEMORY</span></p>

<p class=MyNormal>See the next question.</p>

<p class=MyNormal>If the program crashes for some other reason, please check
whether you are using the right syntax and if the problem persists get in touch
with the authors.</p>

<p class=FAQ><a name="_Toc235985612"></a><a name="_Toc135066794"></a><a
name="_Toc102494703"></a><a name="_Toc102494319"><span style='mso-bookmark:
_Toc102494703'><span style='mso-bookmark:_Toc135066794'><span style='mso-bookmark:
_Toc235985612'><span style='mso-ansi-language:EN-US'>Q: I am running out of
memory</span></span></span></span></a><span style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: You can use a more accurate, slower and less memory hungry
dynamic programming mode called <span class=SpellE>myers_miller_pair_wise</span>.
Simply indicate the flag:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_seq1.fasta &#8211;special_mode low_memory </span><span
class=MyNormalChar><span style='font-size:11.0pt;mso-bidi-font-size:9.0pt'><o:p></o:p></span></span></p>

</div>

<p class=MyNormal><span class=GramE>Note that this mode will be much less time
efficient than the default, although it may be slightly more accurate.</span>
In practice the parameterization associate with special mode turns off every
memory expensive heuristic within T-Coffee. For version 2.11 this amounts to</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee<span style="mso-spacerun:yes">  </span>sample_seq1.fasta -method=slow_pair,lalign_id_pair
&#8211;distance_matrix_mode=idscore -dp_mode=myers_miller_pair_wise </span></p>

</div>

<p class=MyNormal>If you keep running out of memory, you may also want to lower
&#8211;<span class=SpellE>maxnseq</span>, to ensure that <span class=SpellE>t_coffee_dpa</span>
will be used. </p>

<h2><a name="_Toc235985613"></a><a name="_Toc135066795"></a><a
name="_Toc102494704"></a><a name="_Toc102494320"></a><a name="_Toc101065348"><span
style='mso-bookmark:_Toc102494320'><span style='mso-bookmark:_Toc102494704'><span
style='mso-bookmark:_Toc135066795'><span style='mso-bookmark:_Toc235985613'><span
lang=EN-GB>Input/Output Control</span></span></span></span></span></a></h2>

<p class=FAQ><a name="_Toc235985614"></a><a name="_Toc135066796"></a><a
name="_Toc102494705"></a><a name="_Toc102494321"><span style='mso-bookmark:
_Toc102494705'><span style='mso-bookmark:_Toc135066796'><span style='mso-bookmark:
_Toc235985614'><span style='mso-ansi-language:EN-US'>Q: How many Sequences can
t_coffee </span></span></span></span></a><span class=GramE><span
style='mso-bookmark:_Toc102494321'><span style='mso-bookmark:_Toc102494705'><span
style='mso-bookmark:_Toc135066796'><span style='mso-bookmark:_Toc235985614'><span
style='mso-ansi-language:EN-US'>handle</span></span></span></span></span></span><span
style='mso-bookmark:_Toc235985614'></span><span style='mso-bookmark:_Toc135066796'></span><span
style='mso-bookmark:_Toc102494705'></span><span style='mso-bookmark:_Toc102494321'></span><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: T-Coffee is limited to a maximum of 50 sequences. Above
this number, the program automatically switches to a heuristic mode, named DPA,
where DPA stands for Double Progressive Alignment.</p>

<p class=MyNormal>DPA is still in development and the version currently shipped
with T-Coffee is only a beta version.</p>

<p class=FAQ><a name="_Toc102494706"></a><a name="_Toc102494322"></a><a
name="_Toc235985615"></a><a name="_Toc135066797"><span style='mso-bookmark:
_Toc235985615'><span style='mso-bookmark:_Toc102494322'><span style='mso-bookmark:
_Toc102494706'><span style='mso-ansi-language:EN-US'>Q: Can I prevent the
Output of all the warnings?</span></span></span></span></a><span
style='mso-bookmark:_Toc102494322'><span style='mso-bookmark:_Toc102494706'><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc102494322'><span
style='mso-bookmark:_Toc102494706'>A: Yes, by setting<span
style="mso-spacerun:yes">  </span><b style='mso-bidi-font-weight:normal'><i
style='mso-bidi-font-style:normal'>&#8211;<span class=SpellE>no_warning</span></i></b></span></span></p>

<p class=FAQ><span style='mso-bookmark:_Toc102494322'><span style='mso-bookmark:
_Toc102494706'><a name="_Toc235985616"></a><a name="_Toc135066798"><span
style='mso-bookmark:_Toc235985616'><span style='mso-ansi-language:EN-US'>Q: How
many ways to pass parameters to t_coffee?</span></span></a></span></span><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: See the section well behaved parameters</p>

<p class=FAQ><a name="_Toc235985617"></a><a name="_Toc135066799"></a><a
name="_Toc102494707"></a><a name="_Toc102494323"><span style='mso-bookmark:
_Toc102494707'><span style='mso-bookmark:_Toc135066799'><span style='mso-bookmark:
_Toc235985617'><span style='mso-ansi-language:EN-US'>Q: How can I change the
default output format?</span></span></span></span></a><span style='mso-ansi-language:
EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: See the -output option, common output formats are:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_seq1.fasta -output=msf,fasta_aln </span></p>

</div>

<p class=FAQ><a name="_Toc235985618"></a><a name="_Toc135066800"></a><a
name="_Toc102494708"></a><a name="_Toc102494324"><span style='mso-bookmark:
_Toc102494708'><span style='mso-bookmark:_Toc135066800'><span style='mso-bookmark:
_Toc235985618'><span style='mso-ansi-language:EN-US'>Q: My sequences are
slightly different between all the alignments.</span></span></span></span></a><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: It does not matter. T-Coffee will reconstruct a set of
sequences that incorporates all the residues potentially missing in some of the
sequences <span class=GramE>( see</span> flag -in).</p>

<p class=FAQ><a name="_Toc235985619"></a><a name="_Toc135066801"></a><a
name="_Toc102494709"></a><a name="_Toc102494325"><span style='mso-bookmark:
_Toc102494709'><span style='mso-bookmark:_Toc135066801'><span style='mso-bookmark:
_Toc235985619'><span style='mso-ansi-language:EN-US'>Q: Is it possible to pipe
stuff OUT of t_coffee?</span></span></span></span></a><span style='mso-ansi-language:
EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: Specify <span class=SpellE>stderr</span> or <span
class=SpellE>stdout</span> as output <span class=GramE>filename,</span> the
output will be redirected accordingly. For instance</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_seq1.fasta -outfile=stdout -out_lib=stdout </span></p>

</div>

<p class=MyNormal>This instruction will output the tree (in new <span
class=SpellE>hampshire</span> format) and the alignment to <span class=SpellE>stdout</span>.</p>

<p class=FAQ><a name="_Toc235985620"></a><a name="_Toc135066802"></a><a
name="_Toc102494710"></a><a name="_Toc102494326"><span style='mso-bookmark:
_Toc102494710'><span style='mso-bookmark:_Toc135066802'><span style='mso-bookmark:
_Toc235985620'><span style='mso-ansi-language:EN-US'>Q: Is it possible to pipe
stuff INTO t_coffee?</span></span></span></span></a><span style='mso-ansi-language:
EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: If as a file name, you specify stdin, the content of this
file will be expected <span class=SpellE>throught</span> pipe:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
cat sample_seq1.fasta | t_coffee -infile=stdin </span></p>

</div>

<p class=MyNormal><span class=GramE>will</span> be equivalent to</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_seq1.fasta </span></p>

</div>

<p class=MyNormal>If you do not give any argument to t_coffee, they will be
expected to come from pipe:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
cat sample_param_file.param<span style="mso-spacerun:yes">  </span>| t_coffee
-parameters=stdin </span></p>

</div>

<p class=MyNormal>For instance:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
echo &#8211;seq=sample_seq1.fasta -method=clustalw_pair | t_coffee
&#8211;parameters=stdin </span></p>

</div>

<p class=FAQ><a name="_Toc235985621"></a><a name="_Toc135066803"></a><a
name="_Toc102494711"></a><a name="_Toc102494327"><span style='mso-bookmark:
_Toc102494711'><span style='mso-bookmark:_Toc135066803'><span style='mso-bookmark:
_Toc235985621'><span style='mso-ansi-language:EN-US'>Q: Can I read my
parameters from a file?</span></span></span></span></a><span style='mso-ansi-language:
EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: See the <span class=GramE>well behaved</span> parameters
section. </p>

<p class=FAQ><a name="_Toc235985622"></a><a name="_Toc135066804"></a><a
name="_Toc102494712"></a><a name="_Toc102494328"><span style='mso-bookmark:
_Toc102494712'><span style='mso-bookmark:_Toc135066804'><span style='mso-bookmark:
_Toc235985622'><span style='mso-ansi-language:EN-US'>Q: I want </span></span></span></span></a><span
class=GramE><span style='mso-bookmark:_Toc102494328'><span style='mso-bookmark:
_Toc102494712'><span style='mso-bookmark:_Toc135066804'><span style='mso-bookmark:
_Toc235985622'><span style='mso-ansi-language:EN-US'>to<span
style="mso-spacerun:yes">  </span>decide</span></span></span></span></span></span><span
style='mso-bookmark:_Toc102494328'><span style='mso-bookmark:_Toc102494712'><span
style='mso-bookmark:_Toc135066804'><span style='mso-bookmark:_Toc235985622'><span
style='mso-ansi-language:EN-US'> myself on the name of the output files!!!</span></span></span></span></span><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: Use the <i style='mso-bidi-font-style:normal'>-<span
class=SpellE>run_name</span></i> flag.</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_seq1.fasta &#8211;run_name=guacamole </span></p>

</div>

<p class=FAQ><a name="_Toc235985623"></a><a name="_Toc135066805"></a><a
name="_Toc102494713"></a><a name="_Toc102494329"><span style='mso-bookmark:
_Toc102494713'><span style='mso-bookmark:_Toc135066805'><span style='mso-bookmark:
_Toc235985623'><span style='mso-ansi-language:EN-US'>Q: I want to use the
sequences in an alignment file</span></span></span></span></a><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: Simply fed your alignment, any way you like, but do not
forget to append the prefix S for sequence:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=FR
style='mso-ansi-language:FR'>PROMPT: </span><span lang=ZU>t_coffee
Ssample_aln1.aln </span></p>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=FR
style='mso-ansi-language:FR'>PROMPT: </span><span lang=ZU>t_coffee
-infile=Ssample_aln1.aln </span></p>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee &#8211;seq=sample_aln1.aln -method=slow_pair,lalign_id_pair
&#8211;outfile=outaln</span></p>

</div>

<p class=MyNormal>This means that the gaps will be reset and that the alignment
you provide will not be considered as an alignment, but as a set of sequences.</p>

<p class=FAQ><a name="_Toc235985624"></a><a name="_Toc135066806"></a><a
name="_Toc102494714"></a><a name="_Toc102494330"><span style='mso-bookmark:
_Toc102494714'><span style='mso-bookmark:_Toc135066806'><span style='mso-bookmark:
_Toc235985624'><span style='mso-ansi-language:EN-US'>Q: I only want to produce
a library</span></span></span></span></a><span style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: use the &#8211;<span class=SpellE>lib_only</span> flag</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_seq1.fasta -out_lib=sample_lib1.tc_lib -lib_only </span></p>

</div>

<p class=MyNormal>Please, note that the previous usage supersedes the use of
the &#8211;convert flag. Its main advantage is to restrict computation time to
the actual library computation.</p>

<p class=FAQ><a name="_Toc235985625"></a><a name="_Toc135066807"></a><a
name="_Toc102494715"></a><a name="_Toc102494331"><span style='mso-bookmark:
_Toc102494715'><span style='mso-bookmark:_Toc135066807'><span style='mso-bookmark:
_Toc235985625'><span style='mso-ansi-language:EN-US'>Q: I want to turn an
alignment into a library</span></span></span></span></a><span style='mso-ansi-language:
EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: use the &#8211;<span class=SpellE>lib_only</span> flag</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee &#8211;in=Asample_aln1.aln -out_lib=sample_lib1.tc_lib -lib_only </span></p>

</div>

<p class=MyNormal>It is also possible to control the weight associated with
this alignment (see the &#8211;weight section).</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee &#8211;aln=sample_aln1.aln -out_lib=sample_lib1.tc_lib -lib_only
&#8211;weight=1000 </span></p>

</div>

<p class=FAQ><a name="_Toc235985626"></a><a name="_Toc135066808"></a><a
name="_Toc102494716"></a><a name="_Toc102494332"><span style='mso-bookmark:
_Toc102494716'><span style='mso-bookmark:_Toc135066808'><span style='mso-bookmark:
_Toc235985626'><span style='mso-ansi-language:EN-US'>Q: I want to concatenate
two libraries</span></span></span></span></a><span style='mso-ansi-language:
EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: You cannot concatenate these files <span class=GramE>on
their own</span>. You will have to use t_coffee. Assume you want to combine
tc_lib1.tc_lib and tc_lib2.tc_lib.</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee -lib=sample_lib1.tc_lib,sample_lib2.tc_lib &#8211;lib_only
-out_lib=sample_lib3.tc_lib </span></p>

</div>

<p class=FAQ><a name="_Toc235985627"></a><a name="_Toc135066809"></a><a
name="_Toc102494717"></a><a name="_Toc102494333"><span style='mso-bookmark:
_Toc102494717'><span style='mso-bookmark:_Toc135066809'><span style='mso-bookmark:
_Toc235985627'><span style='mso-ansi-language:EN-US'>Q: What happens to the
gaps when an alignment is fed to T-Coffee</span></span></span></span></a><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: An alignment is ALWAYS considered as a library AND a set
of sequences. If you want your alignment to be considered as a library only,
use the S identifier.</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee Ssample_aln1.aln &#8211;outfile=outaln</span></p>

</div>

<p class=MyNormal>It will be seen as a sequence file, even if it has an
alignment format (gaps will be removed).</p>

<p class=FAQ><a name="_Toc235985628"></a><a name="_Toc135066810"></a><a
name="_Toc102494718"></a><a name="_Toc102494334"><span style='mso-bookmark:
_Toc102494718'><span style='mso-bookmark:_Toc135066810'><span style='mso-bookmark:
_Toc235985628'><span style='mso-ansi-language:EN-US'>Q: I cannot print the html
graphic display!!!</span></span></span></span></a><span style='mso-ansi-language:
EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: This is a problem that has to do with your browser.
Instead of requesting the <span class=SpellE>score_html</span> output, request
the <span class=SpellE>score_ps</span> output that <span class=GramE>can</span>
be read using <span class=SpellE>ghostview</span>:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_seq1.fasta -output=score_ps </span></p>

</div>

<p class=MyNormal><span class=GramE>or</span><span style='mso-tab-count:1'>      </span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_seq2.fasta -output=score_pdf </span></p>

</div>

<p class=FAQ><a name="_Toc235985629"></a><a name="_Toc135066811"></a><a
name="_Toc102494719"></a><a name="_Toc102494335"><span style='mso-bookmark:
_Toc102494719'><span style='mso-bookmark:_Toc135066811'><span style='mso-bookmark:
_Toc235985629'><span style='mso-ansi-language:EN-US'>Q: I want to output an
html file and a regular file</span></span></span></span></a><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: see the next question</p>

<p class=FAQ><a name="_Toc235985630"></a><a name="_Toc135066812"></a><a
name="_Toc102494720"></a><a name="_Toc102494336"><span style='mso-bookmark:
_Toc102494720'><span style='mso-bookmark:_Toc135066812'><span style='mso-bookmark:
_Toc235985630'><span style='mso-ansi-language:EN-US'>Q: I would like to output
more than one alignment format at the same time</span></span></span></span></a><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: The flag -output accepts more than one parameter. For
instance,</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_seq1.fasta -output=clustalw,html,score_ps,msf </span></p>

</div>

<p class=MyNormal>This will output <span class=SpellE>founr</span> alignment
files in the corresponding formats. Alignments' names will have the format name
as an extension. </p>

<p class=Note><span style='mso-ansi-language:EN-US'>Note: you need to have the
converter ps2pdf installed on your system (standard under Linux and <span
class=SpellE>cygwin</span>). The latest versions of Internet Explorer and
Netscape now allow the user to print the HTML display <i style='mso-bidi-font-style:
normal'>Do not forget to request Background printing.</i><o:p></o:p></span></p>

<h2><a name="_Toc235985631"></a><a name="_Toc135066813"></a><a
name="_Toc102494721"></a><a name="_Toc102494337"></a><a name="_Toc101065349"><span
style='mso-bookmark:_Toc102494337'><span style='mso-bookmark:_Toc102494721'><span
style='mso-bookmark:_Toc135066813'><span style='mso-bookmark:_Toc235985631'><span
lang=EN-GB>Alignment Computation</span></span></span></span></span></a><span
lang=EN-GB> </span></h2>

<p class=FAQ><a name="_Toc102494722"></a><a name="_Toc102494338"></a><a
name="_Toc235985632"></a><a name="_Toc135066814"><span style='mso-bookmark:
_Toc235985632'><span style='mso-bookmark:_Toc102494338'><span style='mso-bookmark:
_Toc102494722'><span style='mso-ansi-language:EN-US'>Q: Is T-Coffee the best? </span></span></span></span></a><span
class=GramE><span style='mso-bookmark:_Toc135066814'><span style='mso-bookmark:
_Toc235985632'><span style='mso-bookmark:_Toc102494338'><span style='mso-bookmark:
_Toc102494722'><span style='mso-ansi-language:EN-US'>Why Not Using Muscle, or
Mafft, or ProbCons???</span></span></span></span></span></span><span
style='mso-bookmark:_Toc235985632'></span><span style='mso-bookmark:_Toc135066814'></span><span
style='mso-bookmark:_Toc102494338'><span style='mso-bookmark:_Toc102494722'><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc102494338'><span
style='mso-bookmark:_Toc102494722'>A: All these packages are good packages and
they sometimes outperform T-Coffee. They also claim to outperform one
another... If you have them installed locally, you can have T-Coffee to
generate a consensus alignment:</span></span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc102494338'><span style='mso-bookmark:_Toc102494722'><span
lang=ZU>PROMPT: t_coffee sample_seq1.fasta &#8211;method
muscle_msa,probcons_msa, mafft_msa, lalign_id_pair,slow_pair</span></span></span></p>

</div>

<p class=FAQ><span style='mso-bookmark:_Toc102494338'><span style='mso-bookmark:
_Toc102494722'><a name="_Toc235985633"></a><a name="_Toc135066815"><span
style='mso-bookmark:_Toc235985633'><span style='mso-ansi-language:EN-US'>Q: Can
t_coffee align Nucleic </span></span></a><span class=GramE><span
style='mso-bookmark:_Toc135066815'><span style='mso-bookmark:_Toc235985633'><span
style='mso-ansi-language:EN-US'>Acids ?</span></span></span></span></span></span><span
style='mso-bookmark:_Toc102494338'><span style='mso-bookmark:_Toc102494722'><span
style='mso-bookmark:_Toc135066815'><span style='mso-bookmark:_Toc235985633'><span
style='mso-ansi-language:EN-US'>??</span></span></span></span></span><span
style='mso-bookmark:_Toc102494338'><span style='mso-bookmark:_Toc102494722'><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc102494338'><span
style='mso-bookmark:_Toc102494722'>A: Normally it can, but check in the log
that the program <span class=SpellE>recognises</span> the right type <span
class=GramE>( In</span> the INPUT SEQ section, Type: <span class=SpellE>xxxx</span>).
If this fails, you will need to manually set the type:</span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc102494338'><span
style='mso-bookmark:_Toc102494722'><o:p>&nbsp;</o:p></span></span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc102494338'><span style='mso-bookmark:_Toc102494722'><span
lang=ZU>PROMPT: t_coffee sample_dnaseq1.fasta &#8211;type dna</span></span></span></p>

</div>

<p class=FAQ><span style='mso-bookmark:_Toc102494338'><span style='mso-bookmark:
_Toc102494722'><a name="_Toc235985634"></a><a name="_Toc135066816"><span
style='mso-bookmark:_Toc235985634'><span style='mso-ansi-language:EN-US'>Q: I
do not want to compute the alignment.</span></span></a></span></span><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: use the -convert flag</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_aln1.aln -convert -output=gcg </span></p>

</div>

<p class=MyNormal>This command will read the .<span class=SpellE>aln</span>
file and turn it into an .<span class=SpellE>msf</span> alignment.</p>

<p class=FAQ><a name="_Toc235985635"></a><a name="_Toc135066817"></a><a
name="_Toc102494723"></a><a name="_Toc102494339"><span style='mso-bookmark:
_Toc102494723'><span style='mso-bookmark:_Toc135066817'><span style='mso-bookmark:
_Toc235985635'><span style='mso-ansi-language:EN-US'>Q: I would like to force
some residues to be aligned.</span></span></span></span></a><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>If you want to brutally force some residues to be aligned,
you may use as a post processing, the <span class=SpellE>force_aln</span>
function of seq_reformat:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee &#8211;other_pg seq_reformat &#8211;in sample_aln4.aln &#8211;action
+force_aln seq1 10 seq2 15 </span></p>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee &#8211;other_pg seq_reformat &#8211;in sample_aln4.aln &#8211;action
+force_aln sample_lib4.tc_lib02 </span></p>

</div>

<p class=MyNormal><span class=GramE>sample</span>_lib4.tc_lib02 is a T-Coffee
library using the tc_lib02 format:</p>

<p class=example2CxSpFirst><span lang=AF style='mso-ansi-language:AF'>*TC_LIB_FORMAT_02<o:p></o:p></span></p>

<p class=example2CxSpLast><span lang=AF style='mso-ansi-language:AF'>SeqX resY
ResY_index<span style='mso-tab-count:1'>      </span> SeqZ ResZ ResZ_index<o:p></o:p></span></p>

<div style='mso-element:para-border-div;border:solid windowtext 1.5pt;
padding:0cm 0cm 0cm 0cm;background:#F7CBB7'>

<p class=Warning><span lang=EN-GB>The TC_LIB_FORMAT_02 is still experimental
and unsupported. It can only be used in the context of the <span class=SpellE>force_aln</span>
function described here.</span></p>

</div>

<p class=MyNormal>Given more than one constraint, these will be applied one
after the other, in the order they are provided. This greedy procedure means
that the Nth constraint may disrupt the (N-<span class=GramE>1)<span
class=SpellE>th</span></span> previously imposed constraint, hence the
importance of forcing the constraints in the right order, with the most important
coming last.</p>

<p class=MyNormal>We do not recommend imposing hard constraints on an
alignment, and it is much more advisable to use the soft constraints provided
by standard t_coffee libraries (cf. building your own libraries section) </p>

<p class=FAQ><a name="_Toc235985636"></a><a name="_Toc135066818"></a><a
name="_Toc102494724"></a><a name="_Toc102494340"><span style='mso-bookmark:
_Toc102494724'><span style='mso-bookmark:_Toc135066818'><span style='mso-bookmark:
_Toc235985636'><span style='mso-ansi-language:EN-US'>Q: I would like to use
structural alignments.</span></span></span></span></a><span style='mso-ansi-language:
EN-US'><o:p></o:p></span></p>

<p class=MyNormal>See the section Using structures in Multiple Sequence
Alignments, or see the question<i style='mso-bidi-font-style:normal'> I want to
build my own libraries.<o:p></o:p></i></p>

<p class=FAQ><a name="_Toc235985637"></a><a name="_Toc135066819"></a><a
name="_Toc102494725"></a><a name="_Toc102494341"><span style='mso-bookmark:
_Toc102494725'><span style='mso-bookmark:_Toc135066819'><span style='mso-bookmark:
_Toc235985637'><span style='mso-ansi-language:EN-US'>Q: I want to build my own
libraries.</span></span></span></span></a><span style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: Turn your alignment into a library, forcing the residues
to have a very good weight, using structure:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee &#8211;aln=sample_seq1.aln -weight=1000 -out_lib=sample_seq1.tc_lib
&#8211;lib_only </span></p>

</div>

<p class=MyNormal>The value 1000 is simply a high value that should make it
more likely for the substitution found in your alignment to reoccur in the
final alignment. This will produce the library <i style='mso-bidi-font-style:
normal'>sample_aln1.tc_lib</i> that you can later use when aligning all the
sequences:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee &#8211;seq=sample_seq1.fasta -lib=sample_seq1.tc_lib &#8211;outfile
sample_seq1.aln </span></p>

</div>

<p class=MyNormal>If you only want some of these residues to be aligned, or want
to give them individual weights, you will have to edit the library file
yourself or use the &#8211;<span class=SpellE>force_aln</span> option (<span
class=SpellE>cf</span> FAQ: I would like to force some residues to be aligned).
A value of N*N * 1000 (N being the number of sequences) usually ensure the
respect of a constraint.</p>

<p class=FAQ><a name="_Toc235985638"></a><a name="_Toc135066820"></a><a
name="_Toc102494726"></a><a name="_Toc102494342"><span style='mso-bookmark:
_Toc102494726'><span style='mso-bookmark:_Toc135066820'><span style='mso-bookmark:
_Toc235985638'><span style='mso-ansi-language:EN-US'>Q: I want to use my own
tree</span></span></span></span></a><span style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: Use the -<span class=SpellE>usetree</span>=&lt;your own
tree&gt; flag.</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_seq1.fasta &#8211;usetree=sample_tree.dnd </span></p>

</div>

<p class=FAQ><a name="_Toc235985639"></a><a name="_Toc135066821"></a><a
name="_Toc102494727"></a><a name="_Toc102494343"><span style='mso-bookmark:
_Toc102494727'><span style='mso-bookmark:_Toc135066821'><span style='mso-bookmark:
_Toc235985639'><span style='mso-ansi-language:EN-US'>Q: I want to align coding
DNA</span></span></span></span></a><span style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: use the <span class=SpellE>fasta_cdna_pair</span> method
that compares two <span class=SpellE>cDNA</span> using the best reading frame
and taking <span class=SpellE>frameshifts</span> into account. </p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee three_cdna.fasta &#8211;method=cdna_fast_pair </span></p>

</div>

<p class=MyNormal>Notice that in the resulting alignments, all the gaps are of
modulo3, except one small gap in the first line of sequence <span class=SpellE>hmgl_trybr</span>.
This is a <span class=SpellE>framshift</span>, made on purpose. You can realign
the same sequences while ignoring their coding potential and treating them like
standard DNA:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee three_cdna.fasta </span></p>

</div>

<div style='mso-element:para-border-div;border:solid windowtext 1.5pt;
padding:0cm 0cm 0cm 0cm;background:#F7CBB7'>

<p class=Warning><span lang=EN-GB>Note: This method has not yet been fully
tested and is only provided as-is with no warranty. Any feedback will be much
appreciated.</span><span class=MyNormalChar><span style='font-size:11.0pt;
mso-bidi-font-size:10.0pt;font-family:Garamond'><o:p></o:p></span></span></p>

</div>

<p class=FAQ><a name="_Toc235985640"></a><a name="_Toc135066822"></a><a
name="_Toc102494728"></a><a name="_Toc102494344"><span style='mso-bookmark:
_Toc102494728'><span style='mso-bookmark:_Toc135066822'><span style='mso-bookmark:
_Toc235985640'><span style='mso-ansi-language:EN-US'>Q: I do not want to use
all the possible pairs when computing the library</span></span></span></span></a><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=FAQ><a name="_Toc235985641"></a><a name="_Toc135066823"></a><a
name="_Toc102494729"></a><a name="_Toc102494345"><span style='mso-bookmark:
_Toc102494729'><span style='mso-bookmark:_Toc135066823'><span style='mso-bookmark:
_Toc235985641'><span style='mso-ansi-language:EN-US'>Q: I only want to use
specific pairs to compute the library</span></span></span></span></a><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: Simply write in a file the list of sequence groups you
want to use:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_seq1.fasta &#8211;method=clustalw_pair,clustalw_msa
&#8211;lib_list=sample_list1.lib_list &#8211;outfile=test</span></p>

</div>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6'>

<p class=examplefileCxSpFirst><span lang=EN-GB>***************<span
class=GramE>sample</span>_list1.lib_list****</span></p>

<p class=examplefileCxSpMiddle><span class=GramE><span lang=EN-GB>2 <span
class=SpellE>hmgl_</span></span></span><span class=SpellE><span lang=EN-GB>trybr</span></span><span
lang=EN-GB> <span class=SpellE>hmgt_mouse</span></span></p>

<p class=examplefileCxSpMiddle><span class=GramE><span lang=EN-GB>2 <span
class=SpellE>hmgl_</span></span></span><span class=SpellE><span lang=EN-GB>trybr</span></span><span
lang=EN-GB> <span class=SpellE>hmgb_chite</span></span></p>

<p class=examplefileCxSpMiddle><span class=GramE><span lang=EN-GB>2 <span
class=SpellE>hmgl_</span></span></span><span class=SpellE><span lang=EN-GB>trybr</span></span><span
lang=EN-GB> <span class=SpellE>hmgl_wheat</span></span></p>

<p class=examplefileCxSpMiddle><span class=GramE><span lang=EN-GB>3 <span
class=SpellE>hmgl_</span></span></span><span class=SpellE><span lang=EN-GB>trybr</span></span><span
lang=EN-GB> <span class=SpellE>hmgl_wheat</span> <span class=SpellE>hmgl_mouse</span></span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>***************<span
class=GramE>sample</span>_list1.lib_list****</span></p>

<p class=examplefileCxSpLast><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

</div>

<p class=Note><span style='mso-ansi-language:EN-US'>Note: Pairwise methods (<span
class=SpellE>slow_pair</span>) will only be applied to list of pairs of
sequences, while multiple methods (<span class=SpellE>clustalw_aln</span>) will
be applied to any dataset having more than two sequences. <o:p></o:p></span></p>

<p class=FAQ><a name="_Toc235985642"></a><a name="_Toc135066824"></a><a
name="_Toc102494730"></a><a name="_Toc102494346"><span style='mso-bookmark:
_Toc102494730'><span style='mso-bookmark:_Toc135066824'><span style='mso-bookmark:
_Toc235985642'><span style='mso-ansi-language:EN-US'>Q: There are duplicates or
quasi-duplicates in my set</span></span></span></span></a><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: If you can remove them, this will make the program run
faster, otherwise, the t<span class=GramE>_coffee scoring</span> scheme should
be able to avoid over-weighting of over-represented sequences.<span
style='mso-tab-count:2'>             </span></p>

<h2><a name="_Toc101065350"></a><a name="_Toc235985643"></a><a
name="_Toc135066825"></a><a name="_Toc102494731"></a><a name="_Toc102494347"><span
style='mso-bookmark:_Toc102494731'><span style='mso-bookmark:_Toc135066825'><span
style='mso-bookmark:_Toc235985643'><span style='mso-bookmark:_Toc101065350'><span
lang=EN-GB>Using Structures and Profiles</span></span></span></span></span></a><span
style='mso-bookmark:_Toc101065350'></span></h2>

<p class=FAQ><span style='mso-bookmark:_Toc101065350'><a name="_Toc235985644"></a><a
name="_Toc135066826"></a><a name="_Toc102494732"></a><a name="_Toc102494348"><span
style='mso-bookmark:_Toc102494732'><span style='mso-bookmark:_Toc135066826'><span
style='mso-bookmark:_Toc235985644'><span style='mso-ansi-language:EN-US'>Q: Can
I align sequences to a profile with T-Coffee?</span></span></span></span></a></span><span
style='mso-bookmark:_Toc101065350'><span style='mso-ansi-language:EN-US'><o:p></o:p></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065350'>A: Yes, you simply
need to indicate that your alignment is a profile with the R tag<span
class=GramE>..</span></span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc101065350'><span lang=ZU>PROMPT: t_coffee
sample_seq1.fasta -profile=sample_aln2.aln &#8211;outfile tacos</span></span></p>

</div>

<p class=FAQ><span style='mso-bookmark:_Toc101065350'><a name="_Toc235985645"></a><a
name="_Toc135066827"></a><a name="_Toc102494733"></a><a name="_Toc102494349"><span
style='mso-bookmark:_Toc102494733'><span style='mso-bookmark:_Toc135066827'><span
style='mso-bookmark:_Toc235985645'><span style='mso-ansi-language:EN-US'>Q: Can
I align sequences Two or More Profiles?</span></span></span></span></a></span><span
style='mso-bookmark:_Toc101065350'><span style='mso-ansi-language:EN-US'><o:p></o:p></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065350'>A: Yes, <span
class=GramE>you,</span> simply tag your profiles with the letter R and the
program will treat them like standard sequences.</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc101065350'><span lang=ZU>PROMPT: t_coffee
-profile=sample_aln1.fasta,sample_aln2.aln &#8211;outfile tacos</span></span></p>

</div>

<p class=FAQ><span style='mso-bookmark:_Toc101065350'><a name="_Toc235985646"></a><a
name="_Toc135066828"></a><a name="_Toc102494734"></a><a name="_Toc102494350"><span
style='mso-bookmark:_Toc102494734'><span style='mso-bookmark:_Toc135066828'><span
style='mso-bookmark:_Toc235985646'><span style='mso-ansi-language:EN-US'>Q: Can
I align two profiles according to the structures they contain?</span></span></span></span></a></span><span
style='mso-bookmark:_Toc101065350'><span style='mso-ansi-language:EN-US'><o:p></o:p></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065350'>A: Yes. As long as
the structure sequences are named according to their PDB identifier</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc101065350'><span lang=ZU>PROMPT: t_coffee<span
style="mso-spacerun:yes">  </span>-profile=sample_profile1.aln,sample_profile2.aln
&#8211;special_mode=3dcoffee &#8211;outfile=aligne_prf.aln</span></span></p>

</div>

<p class=FAQ><span style='mso-bookmark:_Toc101065350'><a name="_Toc235985647"></a><a
name="_Toc135066829"><span style='mso-bookmark:_Toc235985647'><span
style='mso-ansi-language:EN-US'>Q: T-Coffee becomes very slow when combining
sequences and structures</span></span></a></span><span style='mso-bookmark:
_Toc101065350'><span style='mso-ansi-language:EN-US'><o:p></o:p></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065350'>A: This is true. By
default the structures are <span class=SpellE>feteched</span> on the net, using
RCSB. The problem arises when T-Coffee looks for the structure of sequences
WITHOUT structures. One solution is to install PDB locally. In that case you
will need to set two environment variables:</span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065350'><span
style='mso-tab-count:1'>         </span><span style="mso-spacerun:yes"> </span></span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc101065350'><span lang=ZU>setenv (or export)<span
style="mso-spacerun:yes">  </span>PDB_DIR=directory containing the pdb
structures<br>
setenv (or export)<span style="mso-spacerun:yes">  </span>NO_REMOTE_PDB_DIR=1</span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc101065350'>Interestingly, the
observation that sequences without structures are those that take the most time
to be checked is a reminder of the strongest rational argument that I know of
against torture: any innocent would require the maximum amount of torture to
establish his/her innocence, which sounds...ahem...strange<span class=GramE>.,</span>
and at least <span class=SpellE>inneficient</span>. Then again I was never
struck by the efficiency of the Bush administration.</span></p>

<p class=FAQ><span style='mso-bookmark:_Toc101065350'><a name="_Toc235985648"></a><a
name="_Toc135066830"><span style='mso-bookmark:_Toc235985648'><span
style='mso-ansi-language:EN-US'>Q: Can I use a local installation of PDB?</span></span></a></span><span
style='mso-bookmark:_Toc101065350'><span style='mso-ansi-language:EN-US'><o:p></o:p></span></span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065350'>A: Yes, T-<span
class=SpellE>Coffe</span> supports three types of installations: </span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065350'><span
style='mso-tab-count:1'>         </span>-<span class=GramE>an</span> add-hoc
installation where all your structures are in a directory, under the form
pdbid.pdb or <span class=SpellE>pdbid.id.Z</span> or pdbid.pdb.gz. In that
case, all you need to do is set the <span class=SpellE>environement</span>
variables correctly:</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc101065350'><span lang=ZU>setenv (or export)<span
style="mso-spacerun:yes">  </span>PDB_DIR=directory containing the pdb
structures<br>
setenv (or export)<span style="mso-spacerun:yes">  </span>NO_REMOTE_PDB_DIR=1</span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc101065350'>-A standard <span
class=SpellE>pdb</span> installation using the all section of <span
class=SpellE>pdb</span>. In that case, you must set the variables to:</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc101065350'><span lang=ZU>setenv (or export)<span
style="mso-spacerun:yes">  </span>PDB_DIR=&lt;some absolute
path&gt;/data/structures/all/pdb/<br>
setenv (or export)<span style="mso-spacerun:yes">  </span>NO_REMOTE_PDB_DIR=1</span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc101065350'>-A standard <span
class=SpellE>pdb</span> installation using the divided section of <span
class=SpellE>pdb</span>:</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc101065350'><span lang=ZU>setenv (or export)<span
style="mso-spacerun:yes">  </span>PDB_DIR=&lt;some absolute
path&gt;/data/structures/divided/pdb/<br>
setenv (or export)<span style="mso-spacerun:yes">  </span>NO_REMOTE_PDB_DIR=1</span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc101065350'>If you need to do
more clever things, you should know that all the PDB manipulation is made in
T-Coffee by a <span class=SpellE>perl</span> script named <span class=SpellE>extract_from_pdb</span>.
You can extract this script from T-Coffee:</span></p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span
style='mso-bookmark:_Toc101065350'><span lang=ZU>t_coffee &#8211;other_pg
unpack_extract_from_pdb <br>
chmod u+x extract_from_pdb</span></span></p>

</div>

<p class=MyNormal><span style='mso-bookmark:_Toc101065350'>You can then edit
the script to suit your needs. T-Coffee will use your edited version if it is
in the current directory. It will issue a warning that it used a local version.</span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065350'>If you make
extensive modifications, I would appreciate you send me the corrected file so
that I can incorporate it in the next distribution.</span></p>

<p class=MyNormal><span style='mso-bookmark:_Toc101065350'><o:p>&nbsp;</o:p></span></p>

<h2><span style='mso-bookmark:_Toc101065350'><a name="_Toc235985649"></a><a
name="_Toc135066831"></a><a name="_Toc102494735"></a><a name="_Toc102494351"><span
style='mso-bookmark:_Toc102494735'><span style='mso-bookmark:_Toc135066831'><span
style='mso-bookmark:_Toc235985649'><span lang=EN-GB>Alignment Evaluation</span></span></span></span></a></span></h2>

<p class=FAQ><a name="_Toc235985650"></a><a name="_Toc135066832"></a><a
name="_Toc102494736"></a><a name="_Toc102494352"><span style='mso-bookmark:
_Toc102494736'><span style='mso-bookmark:_Toc135066832'><span style='mso-bookmark:
_Toc235985650'><span style='mso-ansi-language:EN-US'>Q: How good is my
alignment?</span></span></span></span></a><span style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: see what is the color index?</p>

<p class=FAQ><a name="_Toc235985651"></a><a name="_Toc135066833"></a><a
name="_Toc102494737"></a><a name="_Toc102494353"><span style='mso-bookmark:
_Toc102494737'><span style='mso-bookmark:_Toc135066833'><span style='mso-bookmark:
_Toc235985651'><span style='mso-ansi-language:EN-US'>Q: What is that color
index?</span></span></span></span></a><span style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: T-Coffee can provide you with a measure of consistency
among all the methods used. You can produce such an output using:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_seq1.fasta -output=html </span></p>

</div>

<p class=MyNormal>This will compute <span class=SpellE>your_seq.score_html</span>
that you can view using <span class=SpellE>netscape</span>. An alternative is
to use <span class=SpellE>score_ps</span> or <span class=SpellE>score_pdf</span>
that can be viewed using <span class=SpellE>ghostview</span> or <span
class=SpellE>acroread</span>, <span class=SpellE>score_ascii</span> will give
you an alignment that can be parsed as a text file.</p>

<p class=MyNormal>A book chapter describing the CORE index is available on:</p>

<p class=MyNormal><a href="http://www.tcoffee.org/Publications/Pdf/core.pp.pdf">http://www.tcoffee.org/Publications/Pdf/core.pp.pdf</a></p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=FAQ><a name="_Toc235985652"></a><a name="_Toc135066834"></a><a
name="_Toc102494738"></a><a name="_Toc102494354"><span style='mso-bookmark:
_Toc102494738'><span style='mso-bookmark:_Toc135066834'><span style='mso-bookmark:
_Toc235985652'><span style='mso-ansi-language:EN-US'>Q: Can I evaluate
alignments NOT produced with T-Coffee?</span></span></span></span></a><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: Yes. You may have an alignment produced from any source
you like. To evaluate it do:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee &#8211;infile=sample_aln1.aln -lib=sample_aln1.tc_lib
&#8211;special_mode=evaluate </span><span class=MyNormalChar><span
style='font-size:11.0pt;mso-bidi-font-size:9.0pt'><o:p></o:p></span></span></p>

</div>

<p class=MyNormal>If you have no library available, the library will be
computed on the fly using the following command. This can take some time,
depending on your sample size. To monitor the progress in a situation where the
default library is being built, use:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee &#8211;infile=sample_aln1.aln &#8211;special_mode evaluate </span></p>

</div>

<p class=FAQ><a name="_Toc235985653"></a><a name="_Toc135066835"></a><a
name="_Toc102494739"></a><a name="_Toc102494355"><span style='mso-bookmark:
_Toc102494739'><span style='mso-bookmark:_Toc135066835'><span style='mso-bookmark:
_Toc235985653'><span style='mso-ansi-language:EN-US'>Q: Can I Compare Two
Alignments?</span></span></span></span></a><span style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: Yes. You can treat one of your alignments as a library and
compare it with the second alignment:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee &#8211;infile=sample_aln1_1.aln -aln=sample_aln1_2.aln
&#8211;special_mode=evaluate </span></p>

</div>

<p class=MyNormal>If you have no library available, the library will be
computed on the fly using the following command. This can take some time,
depending on your sample size. To monitor the progress in a situation where the
default library is being built, use:</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee &#8211;infile=sample_aln1.aln &#8211;special_mode evaluate </span></p>

</div>

<p class=FAQ><a name="_Toc235985654"></a><a name="_Toc135066836"></a><a
name="_Toc102494740"></a><a name="_Toc102494356"><span style='mso-bookmark:
_Toc102494740'><span style='mso-bookmark:_Toc135066836'><span style='mso-bookmark:
_Toc235985654'><span style='mso-ansi-language:EN-US'>Q: I am aligning sequences
with long regions of very good overlap</span></span></span></span></a><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: Increase the <span class=SpellE>ktuple</span> size <span
class=GramE>( up</span> to 4 or 5 for DNA) and up to 3 for proteins.</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee sample_seq1.fasta -ktuple=3</span></p>

</div>

<p class=MyNormal>This will speed up the program. It can be very useful,
especially when aligning ESTs. </p>

<p class=FAQ><a name="_Toc235985655"></a><a name="_Toc135066837"></a><a
name="_Toc102494741"></a><a name="_Toc102494357"><span style='mso-bookmark:
_Toc102494741'><span style='mso-bookmark:_Toc135066837'><span style='mso-bookmark:
_Toc235985655'><span style='mso-ansi-language:EN-US'>Q: Why is T-Coffee
changing the names of my sequences!!!!</span></span></span></span></a><span
style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: If there is no duplicated name in your sequence set,
T-Coffee's handling of names is consistent with Clustalw, (<span class=SpellE>Cf</span>
Sequence Name Handling in the Format section). If your dataset contains
sequences with identical names, these will automatically be renamed to:</p>

<div style='mso-element:para-border-div;border:solid windowtext 1.0pt;
mso-border-alt:solid windowtext .5pt;padding:1.0pt 4.0pt 1.0pt 4.0pt;
background:#E6E6E6'>

<p class=examplefileCxSpFirst><span lang=EN-GB>************************</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>&gt;<span class=GramE>seq1</span></span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>&gt;<span class=GramE>seq1</span></span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>************************</span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>&gt;<span class=GramE>seq1</span></span></p>

<p class=examplefileCxSpMiddle><span lang=EN-GB>&gt;<span class=GramE>seq1</span>_1</span></p>

<p class=examplefileCxSpLast><span lang=EN-GB>************************ </span></p>

</div>

<div style='mso-element:para-border-div;border:solid windowtext 1.5pt;
padding:0cm 0cm 0cm 0cm;background:#F7CBB7'>

<p class=Warning><span lang=EN-GB>Warning: The behaviour is undefined when this
creates two <span class=GramE>sequence</span> with a similar names.</span></p>

</div>

<h2><a name="_Toc235985656"><span lang=EN-GB>Improving Your Alignment</span></a></h2>

<p class=FAQ><a name="_Toc235985657"><span style='mso-ansi-language:EN-US'>Q: How
Can I Edit my Alignment Manually?</span></a><span style='mso-ansi-language:
EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: Use <span class=SpellE>jalview</span>, a Java online MSA
editor: www.jalview.org</p>

<p class=FAQ><a name="_Toc235985658"><span style='mso-ansi-language:EN-US'>Q: Have
I Improved or Not my Alignment?</span></a><span style='mso-ansi-language:EN-US'><o:p></o:p></span></p>

<p class=MyNormal>A: Using structural information is the only way to establish
whether you have improved or not your alignment. The CORE index can also give
you some information.</p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MsoBodyText><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<p class=MsoBodyText><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

<p class=MyNormal><a name="_Toc102494653"></a><a name="_Toc101065186"><span
style='mso-bookmark:_Toc102494653'><o:p>&nbsp;</o:p></span></a></p>

<h1><span style='mso-bookmark:_Toc101065186'><span style='mso-bookmark:_Toc102494653'><a
name="_Toc235985659"><span lang=EN-GB>Addresses and Contacts</span></a></span></span><span
style='mso-bookmark:_Toc235985659'></span><span style='mso-bookmark:_Toc102494653'></span><span
style='mso-bookmark:_Toc101065186'></span><span lang=EN-GB style='font-size:
98.0pt;mso-bidi-font-size:10.0pt'><o:p></o:p></span></h1>

<h2><a name="_Toc235985660"><span lang=EN-GB>Contributors</span></a></h2>

<p class=MsoBodyText><span lang=EN-GB>T-coffee is developed, maintained,
monitored, used and debugged by a dedicated team that include:</span></p>

<p class=MsoBodyText><span lang=EN-GB><span style='mso-tab-count:1'>            </span>Cédric
Notredame</span></p>

<p class=MsoBodyText style='text-indent:36.0pt'><span class=SpellE><span
lang=EN-GB>Fabrice</span></span><span lang=EN-GB> <span class=SpellE>Armougom</span></span></p>

<p class=MsoBodyText style='text-indent:36.0pt'><span lang=EN-GB>Des Higgins</span></p>

<p class=MsoBodyText style='text-indent:36.0pt'><span class=SpellE><span
lang=EN-GB>Sebastien</span></span><span lang=EN-GB> <span class=SpellE>Moretti</span></span></p>

<p class=MsoBodyText style='text-indent:36.0pt'><span class=SpellE><span
lang=EN-GB>Orla</span></span><span lang=EN-GB> OSullivan</span></p>

<p class=MsoBodyText style='text-indent:36.0pt'><span class=SpellE><span
lang=EN-GB>Eamon</span></span><span lang=EN-GB> OToole</span></p>

<p class=MsoBodyText style='text-indent:36.0pt'><span lang=EN-GB>Olivier <span
class=SpellE>Poirot</span></span></p>

<p class=MsoBodyText style='text-indent:36.0pt'><span class=SpellE><span
lang=EN-GB>Karsten</span></span><span lang=EN-GB> <span class=SpellE>Suhre</span></span></p>

<p class=MsoBodyText style='text-indent:36.0pt'><span lang=EN-GB>Vladimir <span
class=SpellE>Keduas</span></span></p>

<p class=MsoBodyText style='text-indent:36.0pt'><span lang=EN-GB>Iain Wallace</span></p>

<h2><a name="_Toc235985661"><span lang=EN-GB>Addresses</span></a></h2>

<p class=MyNormal>We are always very eager to get some user feedback. Please do
not hesitate to drop us a <span class=GramE>line<span
style="mso-spacerun:yes">  </span>at</span>: <a
href="mailto:cedric.notredame@europe.com"><span lang=EN-GB style='mso-ansi-language:
EN-GB'>cedric.notredame@europe.com</span></a> The latest updates of T-Coffee
are always available<span style="mso-spacerun:yes">  </span>on: www.tcoffee.org
. On this address you will also find a link to some of the online T-Coffee
servers, including <span class=SpellE>Tcoffee@igs</span> </p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MyNormal>T-Coffee can be used to automatically check if an updated
version is available, however the program will not update automatically, as this
can cause endless reproducibility problems.</p>

<div style='mso-element:para-border-div;border-top:solid black 1.0pt;
border-left:none;border-bottom:solid black 1.0pt;border-right:none;mso-border-top-alt:
solid black .25pt;mso-border-bottom-alt:solid black .25pt;padding:7.0pt 0cm 10.0pt 0cm;
background:#FFFFCC;margin-left:2.0cm;margin-right:0cm'>

<p class=MyExample style='margin-left:0cm;background:#FFFFCC'><span lang=ZU>PROMPT:
t_coffee &#8211;update</span></p>

</div>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<div style='mso-element:para-border-div;border:solid windowtext 3.0pt;
padding:1.0pt 4.0pt 1.0pt 4.0pt;background:#CCCCCC;margin-left:0cm;margin-right:
4.25pt'>

<p class=ChapterTitle><a name="_Toc235985662"><span lang=EN-GB>References</span></a></p>

</div>

<p class=MyNormal>It is important that you cite T-Coffee when you use it.
Citing us is (almost) like giving us money: it helps us convincing our
institutions that what we do is useful and that they should keep paying our
salaries and deliver Donuts to our offices from time to time (Not that they
ever did it, but it would be nice anyway). </p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MyNormal>Cite the server if you used it, otherwise, cite the original
paper from 2000 (No, it was never named &quot;T-Coffee 2000&quot;).</p>

<table class=MsoNormalTable border=0 cellspacing=4 cellpadding=0 width="81%"
 style='width:81.78%;mso-cellspacing:3.7pt;mso-padding-alt:0cm 0cm 0cm 0cm'>
 <tr style='mso-yfti-irow:0;mso-yfti-firstrow:yes'>
  <td width="53%" style='width:53.48%;padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified><a
  href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=10964570"><span
  lang=EN-GB style='mso-ansi-language:EN-GB'>Notredame C, Higgins DG, Heringa
  J.</span></a><span style='mso-bidi-font-size:12.0pt;color:black'><o:p></o:p></span></p>
  </td>
  <td width="43%" nowrap valign=top style='width:43.3%;padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified><a
  href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=pubmed&amp;cmd=Display&amp;dopt=pubmed_pubmed&amp;from_uid=10964570"><span
  lang=EN-GB style='mso-bidi-font-size:9.0pt;color:#336699;mso-ansi-language:
  EN-GB'>Related Articles,</span></a> <!--[if gte vml 1]><v:shape id="_x0000_i1026"
   type="#_x0000_t75" style='width:15pt;height:11pt;visibility:hidden' print="f">
   <v:imagedata src="t_coffee_tutorial_files/image003.png" o:title="&#37008;&#28494;&#28018;&#27745;&#36864;&#24902;x0&#37008;1"/>
   <o:lock v:ext="edit" grouping="t"/>
  </v:shape><![endif]--><!--[if gte vml 1]><v:shape id="_x0000_i1027" type="#_x0000_t75"
   style='width:15pt;height:11pt;visibility:hidden' print="f">
   <v:imagedata src="t_coffee_tutorial_files/image004.png" o:title="&#37008;&#28494;&#28018;&#27745;&#36864;&#24902;x0&#37008;1"/>
   <o:lock v:ext="edit" grouping="t"/>
  </v:shape><![endif]--><span style='mso-field-code:" HYPERLINK \0022javascript\:PopUpMenu2_Set\(Menu10964570\)\;\0022 \\t \0022_self\0022 "'><span
  class=MsoHyperlink><span lang=EN-GB style='mso-bidi-font-size:9.0pt;
  color:#336699;mso-ansi-language:EN-GB'>Links</span></span></span> <span
  style='mso-bidi-font-size:12.0pt;color:black'><o:p></o:p></span></p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:1;mso-yfti-lastrow:yes'>
  <td width="97%" colspan=2 style='width:97.86%;padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified>T-Coffee: A novel method for fast and accurate
  multiple sequence alignment.<br>
  J Mol Biol. 2000 Sep 8<span class=GramE>;302</span>(1):205-17. <br>
  PMID: 10964570 [PubMed - indexed for MEDLINE]<span style='mso-bidi-font-size:
  12.0pt;color:black'><o:p></o:p></span></p>
  </td>
 </tr>
</table>

<p class=MyNormal>Other useful publications include:</p>

<h2><a name="_Toc235985663"></a><a name="_Toc102494655"></a><a
name="_Toc102494277"></a><a name="_Toc101065188"><span style='mso-bookmark:
_Toc102494277'><span style='mso-bookmark:_Toc102494655'><span style='mso-bookmark:
_Toc235985663'><span lang=EN-GB>T-Coffee</span></span></span></span></a></h2>

<table class=MsoNormalTable border=0 cellspacing=4 cellpadding=0 width="81%"
 style='width:81.78%;mso-cellspacing:3.7pt;mso-padding-alt:0cm 0cm 0cm 0cm'>
 <tr style='mso-yfti-irow:0;mso-yfti-firstrow:yes'>
  <td width="53%" style='width:53.48%;padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified><a
  href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=15215460"><span
  lang=EN-GB style='mso-ansi-language:EN-GB'>Claude JB, <span class=SpellE>Suhre</span>
  K, Notredame C, Claverie JM, <span class=SpellE>Abergel</span> C.</span></a><span
  style='mso-bidi-font-size:12.0pt;color:black'><o:p></o:p></span></p>
  </td>
  <td width="43%" nowrap valign=top style='width:43.3%;padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified><a
  href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=pubmed&amp;cmd=Display&amp;dopt=pubmed_pubmed&amp;from_uid=15215460"><span
  lang=EN-GB style='mso-bidi-font-size:9.0pt;color:#336699;mso-ansi-language:
  EN-GB'>Related Articles,</span></a> <!--[if gte vml 1]><v:shape id="_x0000_i1028"
   type="#_x0000_t75" style='width:15pt;height:11pt;visibility:hidden' print="f">
   <v:imagedata src="t_coffee_tutorial_files/image005.png" o:title="&#37008;&#28494;&#28018;&#27745;&#36864;&#24902;x0&#37008;1"/>
   <o:lock v:ext="edit" grouping="t"/>
  </v:shape><![endif]--><!--[if gte vml 1]><v:shape id="_x0000_i1029" type="#_x0000_t75"
   style='width:15pt;height:11pt;visibility:hidden' print="f">
   <v:imagedata src="t_coffee_tutorial_files/image006.png" o:title="&#37008;&#28494;&#28018;&#27745;&#36864;&#24902;x0&#37008;1"/>
   <o:lock v:ext="edit" grouping="t"/>
  </v:shape><![endif]--><span style='mso-field-code:" HYPERLINK \0022javascript\:PopUpMenu2_Set\(Menu15215460\)\;\0022 \\t \0022_self\0022 "'><span
  class=MsoHyperlink><span lang=EN-GB style='mso-bidi-font-size:9.0pt;
  color:#336699;mso-ansi-language:EN-GB'>Links</span></span></span> <span
  style='mso-bidi-font-size:12.0pt;color:black'><o:p></o:p></span></p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:1;mso-yfti-lastrow:yes'>
  <td width="97%" colspan=2 style='width:97.86%;padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified><span class=SpellE>CaspR</span>: a web server for
  automated molecular replacement using homology <span class=SpellE>modelling</span>.<br>
  Nucleic Acids Res. 2004 Jul 1<span class=GramE>;32</span>(Web Server
  issue):W606-9. <br>
  PMID: 15215460 [PubMed - indexed for MEDLINE]<span style='mso-bidi-font-size:
  12.0pt;color:black'><o:p></o:p></span></p>
  </td>
 </tr>
</table>

<p class=MyNormalJustified><o:p>&nbsp;</o:p></p>

<table class=MsoNormalTable border=0 cellspacing=4 cellpadding=0 width="81%"
 style='width:81.98%;mso-cellspacing:3.7pt;mso-padding-alt:0cm 0cm 0cm 0cm'>
 <tr style='mso-yfti-irow:0;mso-yfti-firstrow:yes'>
  <td width="53%" style='width:53.5%;padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified><a
  href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=15215345"><span
  class=SpellE><span lang=EN-GB style='mso-ansi-language:EN-GB'>Poirot</span></span><span
  lang=EN-GB style='mso-ansi-language:EN-GB'> O, <span class=SpellE>Suhre</span>
  K, <span class=SpellE>Abergel</span> C, O'Toole E, Notredame C.</span></a><span
  style='mso-bidi-font-size:12.0pt;color:black'><o:p></o:p></span></p>
  </td>
  <td nowrap valign=top style='padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified><a
  href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=pubmed&amp;cmd=Display&amp;dopt=pubmed_pubmed&amp;from_uid=15215345"><span
  lang=EN-GB style='mso-bidi-font-size:9.0pt;color:#336699;mso-ansi-language:
  EN-GB'>Related Articles,</span></a> <!--[if gte vml 1]><v:shape id="_x0000_i1030"
   type="#_x0000_t75" style='width:15pt;height:11pt;visibility:hidden' print="f">
   <v:imagedata src="t_coffee_tutorial_files/image007.png" o:title="&#37008;&#28494;&#28018;&#27745;&#36864;&#24902;x0&#37008;1"/>
   <o:lock v:ext="edit" grouping="t"/>
  </v:shape><![endif]--><span style='mso-field-code:" HYPERLINK \0022javascript\:PopUpMenu2_Set\(Menu15215345\)\;\0022 \\t \0022_self\0022 "'><span
  class=MsoHyperlink><span lang=EN-GB style='mso-bidi-font-size:9.0pt;
  color:#336699;mso-ansi-language:EN-GB'>Links</span></span></span> <span
  style='mso-bidi-font-size:12.0pt;color:black'><o:p></o:p></span></p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:1;mso-yfti-lastrow:yes'>
  <td colspan=2 style='padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified>3DCoffee@igs: a web server for combining sequences
  and structures into a multiple sequence alignment.<br>
  Nucleic Acids Res. 2004 Jul 1<span class=GramE>;32</span>(Web Server
  issue):W37-40. <br>
  PMID: 15215345 [PubMed - indexed for MEDLINE]<span style='mso-bidi-font-size:
  12.0pt;color:black'><o:p></o:p></span></p>
  </td>
 </tr>
</table>

<p class=MyNormalJustified><o:p>&nbsp;</o:p></p>

<table class=MsoNormalTable border=0 cellspacing=4 cellpadding=0 width="81%"
 style='width:81.98%;mso-cellspacing:3.7pt;mso-padding-alt:0cm 0cm 0cm 0cm'>
 <tr style='mso-yfti-irow:0;mso-yfti-firstrow:yes'>
  <td width="53%" style='width:53.5%;padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified><a
  href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=15201059"><span
  lang=EN-GB style='mso-ansi-language:EN-GB'>O'Sullivan O, <span class=SpellE>Suhre</span>
  K, <span class=SpellE>Abergel</span> C, Higgins DG, Notredame C.</span></a><span
  style='mso-bidi-font-size:12.0pt;color:black'><o:p></o:p></span></p>
  </td>
  <td nowrap valign=top style='padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified><a
  href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=pubmed&amp;cmd=Display&amp;dopt=pubmed_pubmed&amp;from_uid=15201059"><span
  lang=EN-GB style='mso-bidi-font-size:9.0pt;color:#336699;mso-ansi-language:
  EN-GB'>Related Articles,</span></a> <!--[if gte vml 1]><v:shape id="_x0000_i1031"
   type="#_x0000_t75" style='width:15pt;height:11pt;visibility:hidden' print="f">
   <v:imagedata src="t_coffee_tutorial_files/image008.png" o:title="&#37008;&#28494;&#28018;&#27745;&#36864;&#24902;x0&#37008;1"/>
   <o:lock v:ext="edit" grouping="t"/>
  </v:shape><![endif]--><span style='mso-field-code:" HYPERLINK \0022javascript\:PopUpMenu2_Set\(Menu15201059\)\;\0022 \\t \0022_self\0022 "'><span
  class=MsoHyperlink><span lang=EN-GB style='mso-bidi-font-size:9.0pt;
  color:#336699;mso-ansi-language:EN-GB'>Links</span></span></span> <span
  style='mso-bidi-font-size:12.0pt;color:black'><o:p></o:p></span></p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:1;mso-yfti-lastrow:yes'>
  <td colspan=2 style='padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified>3DCoffee: combining protein sequences and
  structures within multiple sequence alignments.<br>
  J Mol Biol. 2004 Jul 2<span class=GramE>;340</span>(2):385-95. <br>
  PMID: 15201059 [PubMed - indexed for MEDLINE]<span style='mso-bidi-font-size:
  12.0pt;color:black'><o:p></o:p></span></p>
  </td>
 </tr>
</table>

<p class=MyNormalJustified><o:p>&nbsp;</o:p></p>

<table class=MsoNormalTable border=0 cellspacing=4 cellpadding=0 width="81%"
 style='width:81.98%;mso-cellspacing:3.7pt;mso-padding-alt:0cm 0cm 0cm 0cm'>
 <tr style='mso-yfti-irow:0;mso-yfti-firstrow:yes'>
  <td width="53%" style='width:53.5%;padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified><a
  href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=12824354"><span
  class=SpellE><span lang=EN-GB style='mso-ansi-language:EN-GB'>Poirot</span></span><span
  lang=EN-GB style='mso-ansi-language:EN-GB'> O, O'Toole E, Notredame C.</span></a><span
  style='mso-bidi-font-size:12.0pt;color:black'><o:p></o:p></span></p>
  </td>
  <td nowrap valign=top style='padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified><a
  href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=pubmed&amp;cmd=Display&amp;dopt=pubmed_pubmed&amp;from_uid=12824354"><span
  lang=EN-GB style='mso-bidi-font-size:9.0pt;color:#336699;mso-ansi-language:
  EN-GB'>Related Articles,</span></a> <!--[if gte vml 1]><v:shape id="_x0000_i1032"
   type="#_x0000_t75" style='width:15pt;height:11pt;visibility:hidden' print="f">
   <v:imagedata src="t_coffee_tutorial_files/image009.png" o:title="&#37008;&#28494;&#28018;&#27745;&#36864;&#24902;x0&#37008;1"/>
   <o:lock v:ext="edit" grouping="t"/>
  </v:shape><![endif]--><span style='mso-field-code:" HYPERLINK \0022javascript\:PopUpMenu2_Set\(Menu12824354\)\;\0022 \\t \0022_self\0022 "'><span
  class=MsoHyperlink><span lang=EN-GB style='mso-bidi-font-size:9.0pt;
  color:#336699;mso-ansi-language:EN-GB'>Links</span></span></span> <span
  style='mso-bidi-font-size:12.0pt;color:black'><o:p></o:p></span></p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:1;mso-yfti-lastrow:yes'>
  <td colspan=2 style='padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified><span class=SpellE>Tcoffee@igs</span>: A web
  server for computing, evaluating and combining multiple sequence alignments.<br>
  Nucleic Acids Res. 2003 Jul 1<span class=GramE>;31</span>(13):3503-6. <br>
  PMID: 12824354 [PubMed - indexed for MEDLINE]<span style='mso-bidi-font-size:
  12.0pt;color:black'><o:p></o:p></span></p>
  </td>
 </tr>
</table>

<p class=MyNormalJustified><o:p>&nbsp;</o:p></p>

<table class=MsoNormalTable border=0 cellspacing=4 cellpadding=0 width="81%"
 style='width:81.98%;mso-cellspacing:3.7pt;mso-padding-alt:0cm 0cm 0cm 0cm'>
 <tr style='mso-yfti-irow:0;mso-yfti-firstrow:yes'>
  <td width="53%" style='width:53.5%;padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified><a
  href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=11301309"><span
  lang=EN-GB style='mso-ansi-language:EN-GB'>Notredame C.</span></a><span
  style='mso-bidi-font-size:12.0pt;color:black'><o:p></o:p></span></p>
  </td>
  <td nowrap valign=top style='padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified><a
  href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=pubmed&amp;cmd=Display&amp;dopt=pubmed_pubmed&amp;from_uid=11301309"><span
  lang=EN-GB style='mso-bidi-font-size:9.0pt;color:#336699;mso-ansi-language:
  EN-GB'>Related Articles,</span></a> <!--[if gte vml 1]><v:shape id="_x0000_i1033"
   type="#_x0000_t75" style='width:15pt;height:11pt;visibility:hidden' print="f">
   <v:imagedata src="t_coffee_tutorial_files/image010.png" o:title="&#37008;&#28494;&#28018;&#27745;&#36864;&#24902;x0&#37008;1"/>
   <o:lock v:ext="edit" grouping="t"/>
  </v:shape><![endif]--><span style='mso-field-code:" HYPERLINK \0022javascript\:PopUpMenu2_Set\(Menu11301309\)\;\0022 \\t \0022_self\0022 "'><span
  class=MsoHyperlink><span lang=EN-GB style='mso-bidi-font-size:9.0pt;
  color:#336699;mso-ansi-language:EN-GB'>Links</span></span></span> <span
  style='mso-bidi-font-size:12.0pt;color:black'><o:p></o:p></span></p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:1;mso-yfti-lastrow:yes'>
  <td colspan=2 style='padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified><span class=SpellE>Mocca</span>: semi-automatic
  method for domain hunting.<br>
  Bioinformatics. 2001 Apr<span class=GramE>;17</span>(4):373-4. <br>
  PMID: 11301309 [PubMed - indexed for MEDLINE]<span style='mso-bidi-font-size:
  12.0pt;color:black'><o:p></o:p></span></p>
  </td>
 </tr>
</table>

<p class=MyNormalJustified><o:p>&nbsp;</o:p></p>

<table class=MsoNormalTable border=0 cellspacing=4 cellpadding=0 width="81%"
 style='width:81.98%;mso-cellspacing:3.7pt;mso-padding-alt:0cm 0cm 0cm 0cm'>
 <tr style='mso-yfti-irow:0;mso-yfti-firstrow:yes'>
  <td width="53%" style='width:53.5%;padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified><a
  href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=10964570"><span
  lang=EN-GB style='mso-ansi-language:EN-GB'>Notredame C, Higgins DG, Heringa
  J.</span></a><span style='mso-bidi-font-size:12.0pt;color:black'><o:p></o:p></span></p>
  </td>
  <td nowrap valign=top style='padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified><a
  href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=pubmed&amp;cmd=Display&amp;dopt=pubmed_pubmed&amp;from_uid=10964570"><span
  lang=EN-GB style='mso-bidi-font-size:9.0pt;color:#336699;mso-ansi-language:
  EN-GB'>Related Articles,</span></a> <!--[if gte vml 1]><v:shape id="_x0000_i1034"
   type="#_x0000_t75" style='width:15pt;height:11pt;visibility:hidden' print="f">
   <v:imagedata src="t_coffee_tutorial_files/image011.png" o:title="&#37008;&#28494;&#28018;&#27745;&#36864;&#24902;x0&#37008;1"/>
   <o:lock v:ext="edit" grouping="t"/>
  </v:shape><![endif]--><span style='mso-field-code:" HYPERLINK \0022javascript\:PopUpMenu2_Set\(Menu10964570\)\;\0022 \\t \0022_self\0022 "'><span
  class=MsoHyperlink><span lang=EN-GB style='mso-bidi-font-size:9.0pt;
  color:#336699;mso-ansi-language:EN-GB'>Links</span></span></span> <span
  style='mso-bidi-font-size:12.0pt;color:black'><o:p></o:p></span></p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:1;mso-yfti-lastrow:yes'>
  <td colspan=2 style='padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified>T-Coffee: A novel method for fast and accurate
  multiple sequence alignment.<br>
  J Mol Biol. 2000 Sep 8<span class=GramE>;302</span>(1):205-17. <br>
  PMID: 10964570 [PubMed - indexed for MEDLINE]<span style='mso-bidi-font-size:
  12.0pt;color:black'><o:p></o:p></span></p>
  </td>
 </tr>
</table>

<p class=MyNormalJustified><o:p>&nbsp;</o:p></p>

<table class=MsoNormalTable border=0 cellspacing=4 cellpadding=0 width="81%"
 style='width:81.98%;mso-cellspacing:3.7pt;mso-padding-alt:0cm 0cm 0cm 0cm'>
 <tr style='mso-yfti-irow:0;mso-yfti-firstrow:yes'>
  <td width="53%" style='width:53.5%;padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified><a
  href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=9682054"><span
  lang=EN-GB style='mso-ansi-language:EN-GB'>Notredame C, Holm L, Higgins DG.</span></a><span
  style='mso-bidi-font-size:12.0pt;color:black'><o:p></o:p></span></p>
  </td>
  <td nowrap valign=top style='padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified><a
  href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=pubmed&amp;cmd=Display&amp;dopt=pubmed_pubmed&amp;from_uid=9682054"><span
  lang=EN-GB style='mso-bidi-font-size:9.0pt;color:#336699;mso-ansi-language:
  EN-GB'>Related Articles,</span></a> <!--[if gte vml 1]><v:shape id="_x0000_i1035"
   type="#_x0000_t75" style='width:15pt;height:11pt;visibility:hidden' print="f">
   <v:imagedata src="t_coffee_tutorial_files/image012.png" o:title="&#37008;&#28494;&#28018;&#27745;&#36864;&#24902;x0&#37008;1"/>
   <o:lock v:ext="edit" grouping="t"/>
  </v:shape><![endif]--><span style='mso-field-code:" HYPERLINK \0022javascript\:PopUpMenu2_Set\(Menu9682054\)\;\0022 \\t \0022_self\0022 "'><span
  class=MsoHyperlink><span lang=EN-GB style='mso-bidi-font-size:9.0pt;
  color:#336699;mso-ansi-language:EN-GB'>Links</span></span></span> <span
  style='mso-bidi-font-size:12.0pt;color:black'><o:p></o:p></span></p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:1;mso-yfti-lastrow:yes'>
  <td colspan=2 style='padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormalJustified>COFFEE: an objective function for multiple
  sequence alignments.<br>
  Bioinformatics. 1998 Jun<span class=GramE>;14</span>(5):407-22. <br>
  PMID: 9682054 [PubMed - indexed for MEDLINE]<span style='mso-bidi-font-size:
  12.0pt;color:black'><o:p></o:p></span></p>
  </td>
 </tr>
</table>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<h2><a name="_Toc235985664"></a><a name="_Toc102494656"></a><a
name="_Toc102494278"></a><a name="_Toc101065189"><span style='mso-bookmark:
_Toc102494278'><span style='mso-bookmark:_Toc102494656'><span style='mso-bookmark:
_Toc235985664'></span></span></span></a><span class=SpellE><span
style='mso-bookmark:_Toc101065189'><span style='mso-bookmark:_Toc102494278'><span
style='mso-bookmark:_Toc102494656'><span style='mso-bookmark:_Toc235985664'><span
lang=EN-GB>Mocca</span></span></span></span></span></span><span
style='mso-bookmark:_Toc235985664'></span><span style='mso-bookmark:_Toc102494656'></span><span
style='mso-bookmark:_Toc102494278'></span><span style='mso-bookmark:_Toc101065189'></span></h2>

<table class=MsoNormalTable border=0 cellspacing=4 cellpadding=0 width="81%"
 style='width:81.98%;mso-cellspacing:3.7pt;mso-padding-alt:0cm 0cm 0cm 0cm'>
 <tr style='mso-yfti-irow:0;mso-yfti-firstrow:yes'>
  <td width="53%" style='width:53.5%;padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormal><a
  href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=11301309"><span
  lang=EN-GB style='mso-ansi-language:EN-GB'>Notredame C.</span></a><span
  style='mso-bidi-font-size:12.0pt;color:black'><o:p></o:p></span></p>
  </td>
  <td nowrap valign=top style='padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormal><a
  href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=pubmed&amp;cmd=Display&amp;dopt=pubmed_pubmed&amp;from_uid=11301309&amp;tool=ExternalSearch"><span
  lang=EN-GB style='mso-bidi-font-size:9.0pt;color:#336699;mso-ansi-language:
  EN-GB'>Related Articles,</span></a> <!--[if gte vml 1]><v:shape id="_x0000_i1036"
   type="#_x0000_t75" style='width:15pt;height:11pt;visibility:hidden' print="f">
   <v:imagedata src="t_coffee_tutorial_files/image013.png" o:title="&#37008;&#28494;&#28018;&#27745;&#36864;&#24902;x0&#37008;1"/>
   <o:lock v:ext="edit" grouping="t"/>
  </v:shape><![endif]--><span style='mso-field-code:" HYPERLINK \0022javascript\:PopUpMenu2_Set\(Menu11301309\)\;\0022 \\t \0022_self\0022 "'><span
  class=MsoHyperlink><span lang=EN-GB style='mso-bidi-font-size:9.0pt;
  color:#336699;mso-ansi-language:EN-GB'>Links</span></span></span> <span
  style='mso-bidi-font-size:12.0pt;color:black'><o:p></o:p></span></p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:1;mso-yfti-lastrow:yes'>
  <td colspan=2 style='padding:0cm 0cm 0cm 0cm'>
  <p class=MyNormal><span class=SpellE>Mocca</span>: semi-automatic method for
  domain hunting.<br>
  Bioinformatics. 2001 Apr<span class=GramE>;17</span>(4):373-4. <br>
  PMID: 11301309 [PubMed - indexed for MEDLINE]<span style='mso-bidi-font-size:
  12.0pt;color:black'><o:p></o:p></span></p>
  </td>
 </tr>
</table>

<h2><a name="_Toc235985665"></a><a name="_Toc102494657"></a><a
name="_Toc102494279"><span style='mso-bookmark:_Toc102494657'><span
style='mso-bookmark:_Toc235985665'><span lang=EN-GB>CORE</span></span></span></a></h2>

<p class=MyNormal><a name="_Toc102494658"></a><a name="_Toc102494280"></a><a
name="_Toc101065367"><span style='mso-bookmark:_Toc102494280'><span
style='mso-bookmark:_Toc102494658'></span></span></a><a
href="http://www.tcoffee.org/Publications/Pdf/core.pp.pdf"><span
style='mso-bookmark:_Toc101065367'><span style='mso-bookmark:_Toc102494280'><span
style='mso-bookmark:_Toc102494658'>http://www.tcoffee.org/Publications/Pdf/core.pp.pdf</span></span></span><span
style='mso-bookmark:_Toc101065367'><span style='mso-bookmark:_Toc102494280'><span
style='mso-bookmark:_Toc102494658'></span></span></span></a><span
style='mso-bookmark:_Toc101065367'><span style='mso-bookmark:_Toc102494280'><span
style='mso-bookmark:_Toc102494658'></span></span></span></p>

<h2><span style='mso-bookmark:_Toc101065367'><span style='mso-bookmark:_Toc102494280'><span
style='mso-bookmark:_Toc102494658'><a name="_Toc235985666"><span lang=EN-GB>Other
Contributions</span></a></span></span></span></h2>

<p class=MyNormal>We do not mean to steal code, but we will always try to
re-use pre-existing code whenever that code exists, free of copyright, just
like we expect people to do with our code. However, whenever this happens, we
make a point at properly citing the source of the original contribution. If
ever you recognize a piece of your code improperly cited, please drop us a note
and we will be happy to correct that.</p>

<p class=MyNormal>In the mean time, here are some important pieces of code from
other packages that have been incorporated within the T-Coffee package. These
include: </p>

<p class=MyNormal>-TM-align package from Zhang, Jeffrey and <span class=SpellE>Skolnik</span>
(NAR, 2005, 33:2303)</p>

<p class=MyNormal>-The <span class=SpellE>Sim</span> algorithm of Huang and
Miller that given two sequences computes the N best scoring local alignments.</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span>-The tree
reading/computing routines are taken from the ClustalW Package, courtesy of
Julie Thompson, Des Higgins and Toby Gibson (Thompson, Higgins, Gibson, 1994,
4673-4680,vol. 22, Nucleic Acid Research).</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span>-The
implementation of the algorithm for aligning two sequences in linear space was
adapted from Myers and Miller, in CABIOS, 1988, 11-17, vol. 1)</p>

<p class=MyNormal><span style="mso-spacerun:yes"> </span><span
style='mso-tab-count:1'>        </span>-Various techniques and algorithms have
been implemented. Whenever relevant, the source of the code/algorithm/idea is
indicated in the corresponding function.</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span>-64 Bits
compliance was implemented by Benjamin <span class=SpellE>Sohn</span>,
Performance Computing Center Stuttgart (HLRS), <span class=GramE>Germany</span>.</p>

<p class=MyNormal><b>An enormous thanks to these people who believe in free
open source code<span class=GramE>..</span><o:p></o:p></b></p>

<h2><a name="_Toc235985667"><span lang=EN-GB>Bug Reports and Feedback</span></a></h2>

<p class=MyNormal><span style='mso-tab-count:1'>         </span>-Prof David
Jones (UCL) reported and corrected the PDB1K bug (now t_coffee/sap can align
PDB sequences longer than 1000 AA).</p>

<p class=MyNormal><span style='mso-tab-count:1'>         </span>-Johan <span
class=SpellE>Leckner</span> reported several bugs related to the treatment of
PDB structures, insuring a consistent behavior between version 1.37 and current
ones.</p>

<p class=MyNormal><span style="mso-spacerun:yes"> </span></p>

<p class=MyNormal><o:p>&nbsp;</o:p></p>

<p class=MsoBodyText><span lang=EN-GB><o:p>&nbsp;</o:p></span></p>

</div>

</body>

</html>
